// @google/adk/dist/cjs/context/summarizers/llm_summarizer.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var llm_summarizer_exports = {};
__export(llm_summarizer_exports, {
  LlmSummarizer: () => LlmSummarizer
});
module.exports = __toCommonJS(llm_summarizer_exports);
var import_compacted_event = require("../../events/compacted_event.js");
var import_event = require("../../events/event.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const DEFAULT_PROMPT = "The following is a conversation history between a user and an AI agent. Please summarize the conversation, focusing on key information and decisions made, as well as any unresolved questions or tasks. The summary should be concise and capture the essence of the interaction.";
class LlmSummarizer {
  constructor(options) {
    this.llm = options.llm;
    this.prompt = options.prompt || DEFAULT_PROMPT;
  }
  async summarize(events) {
    var _a, _b, _c, _d, _e, _f;
    if (events.length === 0) {
      throw new Error("Cannot summarize an empty list of events.");
    }
    const startTime = events[0].timestamp;
    const endTime = events[events.length - 1].timestamp;
    let formattedEvents = "";
    for (let i = 0; i < events.length; i++) {
      formattedEvents += `[Event ${i + 1} - Author: ${events[i].author}]
`;
      formattedEvents += `${(0, import_event.stringifyContent)(events[i])}

`;
    }
    const fullPrompt = `${this.prompt}

${formattedEvents}`;
    const request = {
      contents: [{ role: "user", parts: [{ text: fullPrompt }] }],
      toolsDict: {},
      liveConnectConfig: {}
    };
    const responseGen = this.llm.generateContentAsync(request, false);
    let compactedContent = "";
    const firstResponse = await responseGen.next();
    if (firstResponse.done || !((_c = (_b = (_a = firstResponse.value.content) == null ? void 0 : _a.parts) == null ? void 0 : _b[0]) == null ? void 0 : _c.text)) {
      throw new Error("LLM failed to return a valid summary.");
    }
    compactedContent += firstResponse.value.content.parts[0].text;
    for await (const chunk of responseGen) {
      if ((_f = (_e = (_d = chunk.content) == null ? void 0 : _d.parts) == null ? void 0 : _e[0]) == null ? void 0 : _f.text) {
        compactedContent += chunk.content.parts[0].text;
      }
    }
    return (0, import_compacted_event.createCompactedEvent)({
      author: "system",
      content: {
        role: "model",
        parts: [{ text: compactedContent }]
      },
      startTime,
      endTime,
      compactedContent
    });
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LlmSummarizer
});
