// @google/adk/dist/esm/events/structured_events.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { isEmpty } from "lodash-es";
import { isFinalResponse } from "./event.js";
var EventType = /* @__PURE__ */ ((EventType2) => {
  EventType2["THOUGHT"] = "thought";
  EventType2["CONTENT"] = "content";
  EventType2["TOOL_CALL"] = "tool_call";
  EventType2["TOOL_RESULT"] = "tool_result";
  EventType2["CALL_CODE"] = "call_code";
  EventType2["CODE_RESULT"] = "code_result";
  EventType2["ERROR"] = "error";
  EventType2["ACTIVITY"] = "activity";
  EventType2["TOOL_CONFIRMATION"] = "tool_confirmation";
  EventType2["FINISHED"] = "finished";
  return EventType2;
})(EventType || {});
function toStructuredEvents(event) {
  var _a, _b;
  const structuredEvents = [];
  if (event.errorCode) {
    structuredEvents.push({
      type: "error" /* ERROR */,
      error: new Error(event.errorMessage || event.errorCode)
    });
    return structuredEvents;
  }
  for (const part of (_b = (_a = event.content) == null ? void 0 : _a.parts) != null ? _b : []) {
    if (part.functionCall && !isEmpty(part.functionCall)) {
      structuredEvents.push({
        type: "tool_call" /* TOOL_CALL */,
        call: part.functionCall
      });
    } else if (part.functionResponse && !isEmpty(part.functionResponse)) {
      structuredEvents.push({
        type: "tool_result" /* TOOL_RESULT */,
        result: part.functionResponse
      });
    } else if (part.executableCode && !isEmpty(part.executableCode)) {
      structuredEvents.push({
        type: "call_code" /* CALL_CODE */,
        code: part.executableCode
      });
    } else if (part.codeExecutionResult && !isEmpty(part.codeExecutionResult)) {
      structuredEvents.push({
        type: "code_result" /* CODE_RESULT */,
        result: part.codeExecutionResult
      });
    } else if (part.text) {
      if (part.thought) {
        structuredEvents.push({ type: "thought" /* THOUGHT */, content: part.text });
      } else {
        structuredEvents.push({ type: "content" /* CONTENT */, content: part.text });
      }
    }
  }
  if (event.actions.requestedToolConfirmations && !isEmpty(event.actions.requestedToolConfirmations)) {
    structuredEvents.push({
      type: "tool_confirmation" /* TOOL_CONFIRMATION */,
      confirmations: event.actions.requestedToolConfirmations
    });
  }
  if (isFinalResponse(event)) {
    structuredEvents.push({ type: "finished" /* FINISHED */, output: void 0 });
  }
  return structuredEvents;
}
export {
  EventType,
  toStructuredEvents
};
