// @google/adk-devtools/dist/cjs/utils/agent_loader.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var agent_loader_exports = {};
__export(agent_loader_exports, {
  AgentFile: () => AgentFile,
  AgentLoader: () => AgentLoader,
  FileModuleType: () => FileModuleType,
  replaceDirnamePlugin: () => replaceDirnamePlugin
});
module.exports = __toCommonJS(agent_loader_exports);
var import_adk = require("@google/adk");
var import_esbuild = __toESM(require("esbuild"), 1);
var import_esbuild_shim_plugin = require("esbuild-shim-plugin");
var fsPromises = __toESM(require("node:fs/promises"), 1);
var path = __toESM(require("node:path"), 1);
var import_node_url = require("node:url");
var import_file_utils = require("./file_utils.js");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const JS_FILES_EXTENSIONS = [".js", ".cjs", ".mjs", ".ts", ".mts", ".cts"];
var FileModuleType = /* @__PURE__ */ ((FileModuleType2) => {
  FileModuleType2["CJS"] = "cjs";
  FileModuleType2["ESM"] = "esm";
  return FileModuleType2;
})(FileModuleType || {});
const FILE_MODULE_TYPE_EXTENSION_MAP = {
  ["cjs" /* CJS */]: ".cjs",
  ["esm" /* ESM */]: ".mjs"
};
class AgentFileLoadingError extends Error {
}
const DEFAULT_AGENT_FILE_OPTIONS = {
  compile: true,
  bundle: true
};
function replaceDirnamePlugin(filePath, originalDir) {
  return {
    name: "replace-dirname",
    setup(build) {
      build.onLoad({ filter: /.*/ }, async (args) => {
        if (args.path === filePath) {
          const content = await fsPromises.readFile(args.path, "utf8");
          const fileUrl = (0, import_node_url.pathToFileURL)(filePath).href;
          const loader = [".ts", ".mts", ".cts"].includes(
            path.extname(filePath)
          ) ? "ts" : "js";
          const transformResult = await import_esbuild.default.transform(content, {
            loader,
            define: {
              "__dirname": JSON.stringify(originalDir),
              "__filename": JSON.stringify(filePath),
              "import.meta.url": JSON.stringify(fileUrl)
            }
          });
          return {
            contents: transformResult.code,
            loader: "js"
          };
        }
        return void 0;
      });
    }
  };
}
class AgentFile {
  constructor(filePath, options = DEFAULT_AGENT_FILE_OPTIONS) {
    this.filePath = filePath;
    this.options = options;
    this.disposed = false;
  }
  async load() {
    if (this.agent) {
      return this.agent;
    }
    try {
      await fsPromises.stat(this.filePath);
    } catch (e) {
      if (e.code === "ENOENT") {
        throw new AgentFileLoadingError(
          `Agent file ${this.filePath} does not exists`
        );
      }
    }
    let filePath = this.filePath;
    const shouldCompile = this.options.compile || this.options.bundle;
    if (shouldCompile) {
      const moduleType = this.options.moduleType || await getFileModuleType(filePath);
      const parsedPath = path.parse(filePath);
      const outputDir = (0, import_file_utils.getTempDir)("adk_agent_loader");
      const compiledFilePath = path.join(
        outputDir,
        parsedPath.name + FILE_MODULE_TYPE_EXTENSION_MAP[moduleType]
      );
      const originalDir = path.dirname(filePath);
      await fsPromises.mkdir(outputDir, { recursive: true });
      await linkProjectNodeModules(outputDir, parsedPath.dir);
      await import_esbuild.default.build({
        entryPoints: [filePath],
        outfile: compiledFilePath,
        target: "node16",
        platform: "node",
        format: moduleType,
        packages: "bundle",
        bundle: this.options.bundle,
        minify: this.options.bundle,
        allowOverwrite: true,
        plugins: [replaceDirnamePlugin(filePath, originalDir), (0, import_esbuild_shim_plugin.shimPlugin)()],
        // See http://mikro-orm.io/docs/deployment#deploy-a-bundle-of-entities-and-dependencies-with-esbuild for more details
        external: [
          "sqlite3",
          "better-sqlite3",
          "mysql",
          "mysql2",
          // Native addons must remain external so Node can resolve their
          // platform-specific assets at runtime.
          "onnxruntime-node",
          "oracledb",
          "pg-native",
          "pg-query-stream",
          "tedious",
          "libsql"
        ]
      });
      this.cleanupDirPath = outputDir;
      this.cleanupFilePath = compiledFilePath;
      filePath = compiledFilePath;
    }
    const jsModule = await import((0, import_node_url.pathToFileURL)(filePath).href);
    if (jsModule) {
      if ((0, import_adk.isBaseAgent)(jsModule.rootAgent)) {
        return this.agent = jsModule.rootAgent;
      }
      if ((0, import_adk.isBaseAgent)(jsModule.default)) {
        return this.agent = jsModule.default;
      }
      const rootAgents = Object.values(jsModule).filter(
        (exportValue) => (0, import_adk.isBaseAgent)(exportValue)
      );
      if (rootAgents.length > 1) {
        console.warn(
          `Multiple agents found in ${filePath}. Using the ${rootAgents[0].name} as a root agent.`
        );
      }
      if (rootAgents.length > 0) {
        return this.agent = rootAgents[0];
      }
    }
    await this.dispose();
    throw new AgentFileLoadingError(
      `Failed to load agent ${filePath}: No @google/adk BaseAgent class instance found. Please check that file is not empty and it has export of @google/adk BaseAgent class (e.g. LlmAgent) instance.`
    );
  }
  getFilePath() {
    if (!this.agent) {
      throw new Error("Agent is not loaded yet");
    }
    if (this.disposed) {
      throw new Error("Agent is disposed and can not be used");
    }
    return this.cleanupFilePath || this.filePath;
  }
  async [Symbol.asyncDispose]() {
    return this.dispose();
  }
  async dispose() {
    if (this.disposed) {
      return;
    }
    if (this.cleanupFilePath) {
      this.disposed = true;
      await fsPromises.unlink(this.cleanupFilePath);
      if (this.cleanupDirPath) {
        await (0, import_file_utils.removeFolder)(this.cleanupDirPath);
      }
    }
  }
}
class AgentLoader {
  constructor(agentsDirPath = process.cwd(), options = DEFAULT_AGENT_FILE_OPTIONS) {
    this.agentsDirPath = agentsDirPath;
    this.options = options;
    this.agentsAlreadyPreloaded = false;
    this.preloadedAgents = {};
    const exitHandler = async ({
      exit,
      cleanup
    }) => {
      if (cleanup) {
        await this.disposeAll();
      }
      if (exit) {
        process.exit();
      }
    };
    process.on("exit", () => exitHandler({ cleanup: true }));
    process.on("SIGINT", () => exitHandler({ exit: true }));
    process.on("SIGUSR1", () => exitHandler({ exit: true }));
    process.on("SIGUSR2", () => exitHandler({ exit: true }));
    process.on("uncaughtException", () => exitHandler({ exit: true }));
  }
  async listAgents() {
    await this.preloadAgents();
    return Object.keys(this.preloadedAgents).sort();
  }
  async getAgentFile(agentName) {
    await this.preloadAgents();
    return this.preloadedAgents[agentName];
  }
  async disposeAll() {
    await Promise.all(
      Object.values(this.preloadedAgents).map((f) => f.dispose())
    );
  }
  async preloadAgents() {
    if (this.agentsAlreadyPreloaded) {
      return;
    }
    const files = await (0, import_file_utils.isFile)(this.agentsDirPath) ? [await getFileMetadata(this.agentsDirPath)] : await getDirFiles(this.agentsDirPath);
    await Promise.all(
      files.map(async (fileOrDir) => {
        if (fileOrDir.isFile && isJsFile(fileOrDir.ext)) {
          return this.loadAgentFromFile(fileOrDir);
        }
        if (fileOrDir.isDirectory) {
          return this.loadAgentFromDirectory(fileOrDir);
        }
      })
    );
    this.agentsAlreadyPreloaded = true;
    return;
  }
  async loadAgentFromFile(file) {
    try {
      const agentFile = new AgentFile(file.path, this.options);
      await agentFile.load();
      this.preloadedAgents[file.name] = agentFile;
    } catch (e) {
      if (e instanceof AgentFileLoadingError) {
        return;
      }
      throw e;
    }
  }
  async loadAgentFromDirectory(dir) {
    const subFiles = await getDirFiles(dir.path);
    const possibleAgentJsFile = subFiles.find(
      (f) => f.isFile && f.name === "agent" && isJsFile(f.ext)
    );
    if (!possibleAgentJsFile) {
      return;
    }
    try {
      const agentFile = new AgentFile(possibleAgentJsFile.path, this.options);
      await agentFile.load();
      this.preloadedAgents[dir.name] = agentFile;
    } catch (e) {
      if (e instanceof AgentFileLoadingError) {
        return;
      }
      throw e;
    }
  }
}
function isJsFile(fileExt) {
  return !!fileExt && JS_FILES_EXTENSIONS.includes(fileExt);
}
async function getDirFiles(dir) {
  const files = await fsPromises.readdir(dir);
  return await Promise.all(
    files.map((filePath) => getFileMetadata(path.join(dir, filePath)))
  );
}
async function getFileMetadata(filePath) {
  const fileStats = await fsPromises.stat(filePath);
  const isFile2 = fileStats.isFile();
  const baseName = path.basename(filePath);
  const ext = path.extname(filePath);
  return {
    path: filePath,
    name: isFile2 ? baseName.slice(0, baseName.length - ext.length) : baseName,
    ext: isFile2 ? path.extname(filePath) : void 0,
    isFile: isFile2,
    isDirectory: fileStats.isDirectory()
  };
}
async function getFileModuleType(filePath) {
  const { ext } = path.parse(filePath);
  if ([".cjs", ".cts"].includes(ext)) {
    return "cjs" /* CJS */;
  }
  if ([".mts", ".mjs"].includes(ext)) {
    return "esm" /* ESM */;
  }
  if ([".js", ".ts"].includes(ext)) {
    return getTypeFromPackageJson(path.dirname(filePath));
  }
  return "cjs" /* CJS */;
}
async function getTypeFromPackageJson(dir) {
  const packagePath = path.join(dir, "package.json");
  if (await (0, import_file_utils.isFileExists)(packagePath)) {
    try {
      const packageJson = await (0, import_file_utils.loadFileData)(packagePath);
      return packageJson.type === "module" ? "esm" /* ESM */ : "cjs" /* CJS */;
    } catch {
      return "cjs" /* CJS */;
    }
  }
  const parentDir = path.dirname(dir);
  if (parentDir === dir) {
    return "cjs" /* CJS */;
  }
  return getTypeFromPackageJson(parentDir);
}
async function linkProjectNodeModules(outputDir, sourceDir) {
  const nodeModulesDir = await getProjectNodeModulesDir(sourceDir);
  if (!nodeModulesDir) {
    return;
  }
  const linkPath = path.join(outputDir, "node_modules");
  if (await (0, import_file_utils.isFolderExists)(linkPath)) {
    return;
  }
  try {
    await fsPromises.symlink(
      path.resolve(nodeModulesDir),
      linkPath,
      process.platform === "win32" ? "junction" : "dir"
    );
  } catch (error) {
    if (error.code !== "EEXIST") {
      throw error;
    }
  }
}
async function getProjectNodeModulesDir(sourceDir) {
  try {
    const packageJsonPath = await (0, import_file_utils.tryToFindFileRecursively)(
      sourceDir,
      "package.json",
      10
    );
    const nodeModulesDir = path.join(
      path.dirname(packageJsonPath),
      "node_modules"
    );
    return await (0, import_file_utils.isFolderExists)(nodeModulesDir) ? nodeModulesDir : void 0;
  } catch {
    return void 0;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AgentFile,
  AgentLoader,
  FileModuleType,
  replaceDirnamePlugin
});
