// @google/adk-devtools/dist/cjs/utils/file_utils.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var file_utils_exports = {};
__export(file_utils_exports, {
  createFolder: () => createFolder,
  getTempDir: () => getTempDir,
  isFile: () => isFile,
  isFileExists: () => isFileExists,
  isFolderExists: () => isFolderExists,
  listFiles: () => listFiles,
  loadFileData: () => loadFileData,
  removeFolder: () => removeFolder,
  saveToFile: () => saveToFile,
  tryToFindFileRecursively: () => tryToFindFileRecursively
});
module.exports = __toCommonJS(file_utils_exports);
var fs = __toESM(require("node:fs/promises"), 1);
var os = __toESM(require("node:os"), 1);
var path = __toESM(require("node:path"), 1);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
async function isFolderExists(folderPath) {
  try {
    await fs.access(folderPath);
    const stat = await fs.stat(folderPath);
    return stat.isDirectory();
  } catch (_e) {
    return false;
  }
}
async function isFileExists(folderPath) {
  try {
    const stat = await fs.stat(folderPath);
    return stat.isFile();
  } catch (_e) {
    return false;
  }
}
async function createFolder(folderPath) {
  try {
    await fs.mkdir(folderPath);
  } catch (e) {
    console.error(`Failed to create folder ${folderPath}`, e);
  }
}
async function removeFolder(folderPath) {
  try {
    await fs.rm(folderPath, { recursive: true });
  } catch (e) {
    console.error(`Failed to remove folder ${folderPath}`, e);
  }
}
async function listFiles(folderPath) {
  try {
    return await fs.readdir(folderPath);
  } catch (e) {
    console.error(`Failed to list files in folder ${folderPath}`, e);
    return [];
  }
}
async function isFile(filePath) {
  try {
    const stat = await fs.stat(filePath);
    return stat.isFile();
  } catch (_e) {
    return false;
  }
}
async function loadFileData(filePath) {
  try {
    return JSON.parse(await fs.readFile(filePath, { encoding: "utf-8" }));
  } catch (e) {
    console.error(`Failed to read or parse file ${filePath}:`, e);
    throw e;
  }
}
async function saveToFile(filePath, data) {
  try {
    await fs.writeFile(
      filePath,
      typeof data === "string" ? data : JSON.stringify(data, null, 2),
      { encoding: "utf-8" }
    );
  } catch (e) {
    console.error(`Failed to write file ${filePath}:`, e);
    throw e;
  }
}
function getTempDir(prefix) {
  const pathParts = [os.tmpdir()];
  if (prefix) {
    pathParts.push(prefix);
  }
  pathParts.push(Date.now().toString());
  return path.join(...pathParts);
}
async function tryToFindFileRecursively(sourceFolder, fileName, maxIterations) {
  let currentFolder = sourceFolder;
  for (let i = 0; i < maxIterations; i++) {
    const filePath = path.join(currentFolder, fileName);
    if (await isFileExists(filePath)) {
      return filePath;
    }
    currentFolder = path.dirname(currentFolder);
  }
  throw new Error(
    `No ${fileName} found in ${sourceFolder} or its parent folders up to ${maxIterations} levels.`
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createFolder,
  getTempDir,
  isFile,
  isFileExists,
  isFolderExists,
  listFiles,
  loadFileData,
  removeFolder,
  saveToFile,
  tryToFindFileRecursively
});
