// @google/adk-devtools/dist/cjs/utils/logger.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var logger_exports = {};
__export(logger_exports, {
  AdkLogger: () => AdkLogger
});
module.exports = __toCommonJS(logger_exports);
var import_adk = require("@google/adk");
var winston = __toESM(require("winston"), 1);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class AdkLogger {
  constructor(options) {
    this.logLevel = import_adk.LogLevel.INFO;
    var _a;
    const formats = [
      winston.format.label({
        label: options.label,
        message: (_a = options.colorize) == null ? void 0 : _a.all
      }),
      winston.format((info) => {
        info.level = info.level.toUpperCase();
        return info;
      })()
    ];
    if (options.colorize) {
      formats.push(winston.format.colorize(options.colorize));
    }
    if (options.timestamp) {
      formats.push(winston.format.timestamp());
    }
    if (options.printFormat) {
      formats.push(winston.format.printf(options.printFormat));
    } else {
      formats.push(winston.format.printf((info) => info.message));
    }
    this.logger = winston.createLogger({
      levels: {
        "debug": import_adk.LogLevel.DEBUG,
        "info": import_adk.LogLevel.INFO,
        "warn": import_adk.LogLevel.WARN,
        "error": import_adk.LogLevel.ERROR
      },
      level: "error",
      format: winston.format.combine(...formats),
      transports: [new winston.transports.Console()]
    });
  }
  setLogLevel(level) {
    this.logLevel = level;
  }
  log(level, ...messages) {
    if (this.logLevel > level) {
      return;
    }
    this.logger.log(level.toString(), messages.join(" "));
  }
  debug(...messages) {
    if (this.logLevel > import_adk.LogLevel.DEBUG) {
      return;
    }
    this.logger.debug(messages.join(" "));
  }
  info(...messages) {
    if (this.logLevel > import_adk.LogLevel.INFO) {
      return;
    }
    this.logger.info(messages.join(" "));
  }
  warn(...messages) {
    if (this.logLevel > import_adk.LogLevel.WARN) {
      return;
    }
    this.logger.warn(messages.join(" "));
  }
  error(...messages) {
    if (this.logLevel > import_adk.LogLevel.ERROR) {
      return;
    }
    this.logger.error(messages.join(" "));
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AdkLogger
});
