// @google/adk-devtools/dist/cjs/utils/telemetry_utils.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var telemetry_utils_exports = {};
__export(telemetry_utils_exports, {
  ApiServerSpanExporter: () => ApiServerSpanExporter,
  InMemoryExporter: () => InMemoryExporter,
  hrTimeToNanoseconds: () => hrTimeToNanoseconds,
  setupTelemetry: () => setupTelemetry
});
module.exports = __toCommonJS(telemetry_utils_exports);
var import_adk = require("@google/adk");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function hrTimeToNanoseconds(hrTime) {
  if (!hrTime || !Array.isArray(hrTime) || hrTime.length !== 2) {
    return 0;
  }
  const [seconds, nanoseconds] = hrTime;
  return seconds * 1e9 + nanoseconds;
}
class ApiServerSpanExporter {
  constructor(traceDict) {
    this.traceDict = traceDict;
  }
  export(spans, resultCallback) {
    for (const span of spans) {
      if (span.name === "call_llm" || span.name === "send_data" || span.name.startsWith("execute_tool")) {
        const attributes = { ...span.attributes };
        attributes["trace_id"] = span.spanContext().traceId;
        attributes["span_id"] = span.spanContext().spanId;
        const eventId = attributes["gcp.vertex.agent.event_id"];
        if (eventId) {
          this.traceDict[eventId] = attributes;
        }
      }
    }
    resultCallback({ code: 0 });
  }
  forceFlush() {
    return Promise.resolve();
  }
  shutdown() {
    return Promise.resolve();
  }
}
class InMemoryExporter {
  constructor(traceDict) {
    this.spans = [];
    this.traceDict = traceDict;
  }
  export(spans, resultCallback) {
    for (const span of spans) {
      const traceId = span.spanContext().traceId;
      if (span.name === "call_llm") {
        const attributes = { ...span.attributes };
        const sessionId = attributes["gcp.vertex.agent.session_id"];
        if (sessionId) {
          if (!this.traceDict[sessionId]) {
            this.traceDict[sessionId] = [traceId];
          } else {
            this.traceDict[sessionId].push(traceId);
          }
        }
      }
    }
    this.spans.push(...spans);
    resultCallback({ code: 0 });
  }
  forceFlush() {
    return Promise.resolve();
  }
  shutdown() {
    return Promise.resolve();
  }
  getFinishedSpans(sessionId) {
    const traceIds = this.traceDict[sessionId];
    if (!traceIds || traceIds.length === 0) {
      return [];
    }
    return this.spans.filter(
      (span) => traceIds.includes(span.spanContext().traceId)
    );
  }
  clear() {
    this.spans = [];
  }
}
function otelEnvVarsEnabled() {
  const endpointVars = [
    "OTEL_EXPORTER_OTLP_ENDPOINT",
    "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT",
    "OTEL_EXPORTER_OTLP_METRICS_ENDPOINT",
    "OTEL_EXPORTER_OTLP_LOGS_ENDPOINT"
  ];
  return endpointVars.some((varName) => process.env[varName]);
}
async function setupTelemetry(otelToCloud = false, internalExporters = []) {
  if (otelToCloud) {
    await setupGcpTelemetryExperimental(internalExporters);
  } else if (otelEnvVarsEnabled()) {
    await setupTelemetryFromEnvExperimental(internalExporters);
  } else {
    const otelHooks = {
      spanProcessors: internalExporters
    };
    (0, import_adk.maybeSetOtelProviders)([otelHooks]);
  }
}
async function setupGcpTelemetryExperimental(internalExporters = []) {
  const otelHooksToAdd = [];
  if (internalExporters.length > 0) {
    otelHooksToAdd.push({
      spanProcessors: internalExporters
    });
  }
  const gcpExporters = await (0, import_adk.getGcpExporters)({
    enableTracing: true,
    enableLogging: false,
    enableMetrics: true
  });
  otelHooksToAdd.push(gcpExporters);
  const otelResource = (0, import_adk.getGcpResource)();
  (0, import_adk.maybeSetOtelProviders)(otelHooksToAdd, otelResource);
}
async function setupTelemetryFromEnvExperimental(internalExporters = []) {
  const otelHooksToAdd = [];
  if (internalExporters.length > 0) {
    otelHooksToAdd.push({
      spanProcessors: internalExporters
    });
  }
  (0, import_adk.maybeSetOtelProviders)(otelHooksToAdd);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApiServerSpanExporter,
  InMemoryExporter,
  hrTimeToNanoseconds,
  setupTelemetry
});
