// @google/adk/dist/cjs/tools/load_memory_tool.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var load_memory_tool_exports = {};
__export(load_memory_tool_exports, {
  LOAD_MEMORY: () => LOAD_MEMORY,
  LoadMemoryTool: () => LoadMemoryTool
});
module.exports = __toCommonJS(load_memory_tool_exports);
var import_genai = require("@google/genai");
var import_llm_request = require("../models/llm_request.js");
var import_base_tool = require("./base_tool.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class LoadMemoryTool extends import_base_tool.BaseTool {
  constructor() {
    super({
      name: "load_memory",
      description: "Loads the memory for the current user.\n\nNOTE: Currently this tool only uses text part from the memory."
    });
  }
  _getDeclaration() {
    return {
      name: this.name,
      description: this.description,
      parameters: {
        type: import_genai.Type.OBJECT,
        properties: {
          query: {
            type: import_genai.Type.STRING,
            description: "The query to load the memory for."
          }
        },
        required: ["query"]
      }
    };
  }
  async runAsync({
    args,
    toolContext
  }) {
    try {
      const query = args["query"];
      if (!toolContext.invocationContext.memoryService) {
        throw new Error("Memory service is not initialized.");
      }
      const searchMemoryResponse = await toolContext.searchMemory(query);
      return {
        memories: searchMemoryResponse.memories.map((m) => {
          var _a, _b;
          return {
            // Join all text parts by a space, or empty string if no text parts
            content: (_b = (_a = m.content.parts) == null ? void 0 : _a.map((p) => {
              var _a2;
              return (_a2 = p.text) != null ? _a2 : "";
            }).join(" ")) != null ? _b : "",
            author: m.author,
            timestamp: m.timestamp
          };
        })
      };
    } catch (e) {
      console.error("ERROR in LoadMemoryTool runAsync:", e);
      throw e;
    }
  }
  async processLlmRequest(request) {
    await super.processLlmRequest(request);
    if (!request.toolContext.invocationContext.memoryService) {
      return;
    }
    (0, import_llm_request.appendInstructions)(request.llmRequest, [
      `You have memory. You can use it to answer questions. If any questions need
you to look up the memory, you should call load_memory function with a query.`
    ]);
  }
}
const LOAD_MEMORY = new LoadMemoryTool();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LOAD_MEMORY,
  LoadMemoryTool
});
