// @google/adk/dist/cjs/tools/load_artifacts_tool.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var load_artifacts_tool_exports = {};
__export(load_artifacts_tool_exports, {
  LOAD_ARTIFACTS: () => LOAD_ARTIFACTS,
  LoadArtifactsTool: () => LoadArtifactsTool
});
module.exports = __toCommonJS(load_artifacts_tool_exports);
var import_genai = require("@google/genai");
var import_llm_request = require("../models/llm_request.js");
var import_logger = require("../utils/logger.js");
var import_base_tool = require("./base_tool.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = (0, import_logger.getLogger)();
const GEMINI_SUPPORTED_INLINE_MIME_PREFIXES = ["image/", "audio/", "video/"];
const GEMINI_SUPPORTED_INLINE_MIME_TYPES = /* @__PURE__ */ new Set(["application/pdf"]);
const TEXT_LIKE_MIME_TYPES = /* @__PURE__ */ new Set([
  "application/csv",
  "application/json",
  "application/xml"
]);
function normalizeMimeType(mimeType) {
  if (!mimeType) {
    return void 0;
  }
  return mimeType.split(";")[0].trim();
}
function isInlineMimeTypeSupported(mimeType) {
  const normalized = normalizeMimeType(mimeType);
  if (!normalized) {
    return false;
  }
  return GEMINI_SUPPORTED_INLINE_MIME_PREFIXES.some(
    (prefix) => normalized.startsWith(prefix)
  ) || GEMINI_SUPPORTED_INLINE_MIME_TYPES.has(normalized);
}
function asSafePartForLlm(artifact, artifactName) {
  const inlineData = artifact.inlineData;
  if (!inlineData) {
    return artifact;
  }
  if (isInlineMimeTypeSupported(inlineData.mimeType)) {
    return artifact;
  }
  const mimeType = normalizeMimeType(inlineData.mimeType) || "application/octet-stream";
  const data = inlineData.data;
  if (!data) {
    return {
      text: `[Artifact: ${artifactName}, type: ${mimeType}. No inline data was provided.]`
    };
  }
  const isTextLike = mimeType.startsWith("text/") || TEXT_LIKE_MIME_TYPES.has(mimeType);
  const decodedBuffer = Buffer.from(data, "base64");
  if (isTextLike) {
    try {
      const decoded = decodedBuffer.toString("utf8");
      return { text: decoded };
    } catch {
    }
  }
  const sizeKb = decodedBuffer.length / 1024;
  return {
    text: `[Binary artifact: ${artifactName}, type: ${mimeType}, size: ${sizeKb.toFixed(1)} KB. Content cannot be displayed inline.]`
  };
}
class LoadArtifactsTool extends import_base_tool.BaseTool {
  constructor() {
    super({
      name: "load_artifacts",
      description: `Loads artifacts into the session for this request.

NOTE: Call when you need access to artifacts (for example, uploads saved by the web UI).`
    });
  }
  _getDeclaration() {
    return {
      name: this.name,
      description: this.description,
      parameters: {
        type: import_genai.Type.OBJECT,
        properties: {
          artifact_names: {
            type: import_genai.Type.ARRAY,
            items: {
              type: import_genai.Type.STRING
            },
            description: "The names of the artifacts to load."
          }
        }
      }
    };
  }
  async runAsync({ args }) {
    const artifactNames = args["artifact_names"] || [];
    return {
      artifact_names: artifactNames,
      status: "artifact contents temporarily inserted and removed. to access these artifacts, call load_artifacts tool again."
    };
  }
  async processLlmRequest(request) {
    await super.processLlmRequest(request);
    await this.appendArtifactsToLlmRequest(
      request.toolContext,
      request.llmRequest
    );
  }
  async appendArtifactsToLlmRequest(toolContext, llmRequest) {
    if (!toolContext.invocationContext.artifactService) {
      return;
    }
    const artifactNames = await toolContext.listArtifacts();
    if (!artifactNames || artifactNames.length === 0) {
      return;
    }
    (0, import_llm_request.appendInstructions)(llmRequest, [
      `You have a list of artifacts:
  ${JSON.stringify(
        artifactNames
      )}

  When the user asks questions about any of the artifacts, you should call the
  \`load_artifacts\` function to load the artifact. Always call load_artifacts
  before answering questions related to the artifacts, regardless of whether the
  artifacts have been loaded before. Do not depend on prior answers about the
  artifacts.`
    ]);
    const contents = llmRequest.contents;
    if (contents && contents.length > 0) {
      const lastContent = contents[contents.length - 1];
      if (lastContent.role === "user" && lastContent.parts && lastContent.parts.length > 0) {
        const functionResponsePart = lastContent.parts[0];
        const functionResponse = functionResponsePart.functionResponse;
        if (functionResponse && functionResponse.name === "load_artifacts") {
          const response = functionResponse.response || {};
          const namesToLoad = response["artifact_names"] || [];
          for (const artifactName of namesToLoad) {
            let artifact = await toolContext.loadArtifact(artifactName);
            if (!artifact && !artifactName.startsWith("user:")) {
              const prefixedName = `user:${artifactName}`;
              artifact = await toolContext.loadArtifact(prefixedName);
            }
            if (!artifact) {
              logger.warn(`Artifact "${artifactName}" not found, skipping`);
              continue;
            }
            const artifactPart = asSafePartForLlm(artifact, artifactName);
            llmRequest.contents.push({
              role: "user",
              parts: [{ text: `Artifact ${artifactName} is:` }, artifactPart]
            });
          }
        }
      }
    }
  }
}
const LOAD_ARTIFACTS = new LoadArtifactsTool();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LOAD_ARTIFACTS,
  LoadArtifactsTool
});
