// @google/adk/dist/cjs/tools/vertex_ai_search_tool.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var vertex_ai_search_tool_exports = {};
__export(vertex_ai_search_tool_exports, {
  VertexAiSearchTool: () => VertexAiSearchTool
});
module.exports = __toCommonJS(vertex_ai_search_tool_exports);
var import_logger = require("../utils/logger.js");
var import_model_name = require("../utils/model_name.js");
var import_base_tool = require("./base_tool.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = (0, import_logger.getLogger)();
class VertexAiSearchTool extends import_base_tool.BaseTool {
  constructor(params) {
    super({ name: "vertex_ai_search", description: "vertex_ai_search" });
    const {
      dataStoreId,
      dataStoreSpecs,
      searchEngineId,
      filter,
      maxResults,
      bypassMultiToolsLimit = false
    } = params;
    if (dataStoreId === void 0 && searchEngineId === void 0 || dataStoreId !== void 0 && searchEngineId !== void 0) {
      throw new Error(
        "Either dataStoreId or searchEngineId must be specified."
      );
    }
    if (dataStoreSpecs !== void 0 && searchEngineId === void 0) {
      throw new Error(
        "searchEngineId must be specified if dataStoreSpecs is specified."
      );
    }
    this.dataStoreId = dataStoreId;
    this.dataStoreSpecs = dataStoreSpecs;
    this.searchEngineId = searchEngineId;
    this.filter = filter;
    this.maxResults = maxResults;
    this.bypassMultiToolsLimit = bypassMultiToolsLimit;
  }
  runAsync() {
    return Promise.resolve();
  }
  /**
   * Builds the VertexAISearch configuration.
   *
   * Override this method in a subclass to dynamically customize the search
   * configuration based on the context (e.g., set filter based on session
   * state).
   */
  buildVertexAiSearchConfig(_readonlyContext) {
    return {
      datastore: this.dataStoreId,
      dataStoreSpecs: this.dataStoreSpecs,
      engine: this.searchEngineId,
      filter: this.filter,
      maxResults: this.maxResults
    };
  }
  async processLlmRequest({
    toolContext,
    llmRequest
  }) {
    if (!llmRequest.model) {
      return;
    }
    const modelCheckDisabled = (0, import_model_name.isGeminiModelIdCheckDisabled)();
    llmRequest.config = llmRequest.config || {};
    llmRequest.config.tools = llmRequest.config.tools || [];
    if (!(0, import_model_name.isGeminiModel)(llmRequest.model) && !modelCheckDisabled) {
      throw new Error(
        `Vertex AI search tool is not supported for model ${llmRequest.model}`
      );
    }
    if ((0, import_model_name.isGemini1Model)(llmRequest.model) && llmRequest.config.tools.length > 0 && !this.bypassMultiToolsLimit) {
      throw new Error(
        "Vertex AI search tool cannot be used with other tools in Gemini 1.x."
      );
    }
    const vertexAiSearchConfig = this.buildVertexAiSearchConfig(toolContext);
    let specsInfo;
    if (vertexAiSearchConfig.dataStoreSpecs) {
      const specIds = vertexAiSearchConfig.dataStoreSpecs.map(
        (spec) => spec.dataStore ? spec.dataStore.split("/").pop() : "unnamed"
      );
      specsInfo = `${vertexAiSearchConfig.dataStoreSpecs.length} spec(s): [${specIds.join(", ")}]`;
    }
    logger.debug(
      `Adding Vertex AI Search tool config to LLM request: datastore=${vertexAiSearchConfig.datastore}, engine=${vertexAiSearchConfig.engine}, filter=${vertexAiSearchConfig.filter}, maxResults=${vertexAiSearchConfig.maxResults}, dataStoreSpecs=${specsInfo}`
    );
    llmRequest.config.tools.push({
      retrieval: {
        vertexAiSearch: vertexAiSearchConfig
      }
    });
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VertexAiSearchTool
});
