// @google/adk-devtools/dist/cjs/server/adk_api_server.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
var adk_api_server_exports = {};
__export(adk_api_server_exports, {
  AdkApiServer: () => AdkApiServer
});
module.exports = __toCommonJS(adk_api_server_exports);
var import_adk = require("@google/adk");
var import_api = require("@opentelemetry/api");
var import_sdk_trace_base = require("@opentelemetry/sdk-trace-base");
var import_cors = __toESM(require("cors"), 1);
var import_express = __toESM(require("express"), 1);
var path = __toESM(require("node:path"), 1);
var import_agent_loader = require("../utils/agent_loader.js");
var import_logger = require("../utils/logger.js");
var import_telemetry_utils = require("../utils/telemetry_utils.js");
var import_agent_graph = require("./agent_graph.js");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class AdkApiServer {
  constructor(options) {
    this.runnerCache = {};
    this.traceDict = {};
    this.sessionTraceDict = {};
    this.host = options.host ?? "localhost";
    this.port = options.port ?? 0;
    this.sessionService = options.sessionService ?? new import_adk.InMemorySessionService();
    this.memoryService = options.memoryService ?? new import_adk.InMemoryMemoryService();
    this.artifactService = options.artifactService ?? new import_adk.InMemoryArtifactService();
    this.agentLoader = options.agentLoader ?? new import_agent_loader.AgentLoader(options.agentsDir, options.agentFileLoadOptions);
    this.serveDebugUI = options.serveDebugUI ?? false;
    this.allowOrigins = options.allowOrigins;
    this.otelToCloud = options.otelToCloud ?? false;
    this.registerProcessors = options.registerProcessors;
    this.memoryExporter = new import_telemetry_utils.InMemoryExporter(this.sessionTraceDict);
    this.logger = options.logger ?? new import_logger.AdkLogger({
      label: "ADK API Server",
      timestamp: true,
      colorize: { level: true },
      printFormat: (info) => {
        return `${info.level}: [${info.label}] ${info.timestamp} ${info.message}`;
      }
    });
    this.logger.setLogLevel(options.logLevel ?? import_adk.LogLevel.INFO);
    this.a2a = options.a2a ?? false;
    this.app = (0, import_express.default)();
  }
  get url() {
    if (this.server) {
      const address = this.server.address();
      if (address && typeof address !== "string") {
        return `http://${this.host}:${address.port}`;
      }
    }
    return `http://${this.host}:${this.port}`;
  }
  async setupTelemetry() {
    const internalExporters = [
      new import_sdk_trace_base.SimpleSpanProcessor(new import_telemetry_utils.ApiServerSpanExporter(this.traceDict)),
      new import_sdk_trace_base.SimpleSpanProcessor(this.memoryExporter)
    ];
    await (0, import_telemetry_utils.setupTelemetry)(this.otelToCloud, internalExporters);
    if (this.registerProcessors) {
      const tracerProvider = import_api.trace.getTracerProvider();
      this.registerProcessors(tracerProvider);
    }
  }
  async initA2A() {
    const appNames = await this.agentLoader.listAgents();
    for (const appName of appNames) {
      const agentFile = await this.agentLoader.getAgentFile(appName);
      const agent = await agentFile.load();
      await (0, import_adk.toA2a)(agent, {
        protocol: "http",
        host: this.host,
        port: this.port,
        basePath: `/a2a/${appName}`,
        sessionService: this.sessionService,
        memoryService: this.memoryService,
        artifactService: this.artifactService,
        app: this.app
      });
    }
  }
  async init() {
    const app = this.app;
    await this.setupTelemetry();
    if (this.serveDebugUI) {
      app.get("/", (req, res) => {
        res.redirect("/dev-ui");
      });
      app.use(
        "/dev-ui",
        import_express.default.static(path.join(__dirname, "../browser"), {
          setHeaders: (res, path2) => {
            if (path2.endsWith(".js")) {
              res.setHeader("Content-Type", "text/javascript");
            }
          }
        })
      );
    }
    if (this.allowOrigins) {
      app.use(
        (0, import_cors.default)({
          origin: this.allowOrigins
        })
      );
    }
    app.use(import_express.default.urlencoded({ limit: "50mb", extended: true }));
    app.use(
      import_express.default.json({
        limit: "50mb"
      })
    );
    app.use((req, res, next) => {
      this.logger.info(`${req.method} ${req.originalUrl}`);
      next();
    });
    app.get("/list-apps", async (req, res) => {
      try {
        const apps = await this.agentLoader.listAgents();
        res.json(apps);
      } catch (e) {
        const error = `Failed to list apps: ${e}`;
        res.status(500).json({ error });
        this.logger.error(error);
        return;
      }
    });
    app.get("/debug/trace/:eventId", (req, res) => {
      try {
        const eventId = req.params["eventId"];
        const eventDict = this.traceDict[eventId];
        if (!eventDict) {
          return res.status(404).json({ error: "Trace not found" });
        }
        return res.json(eventDict);
      } catch (e) {
        const error = `Failed to get trace: ${e}`;
        res.status(500).json({ error });
        this.logger.error(error);
        return;
      }
    });
    app.get(
      "/debug/trace/session/:sessionId",
      (req, res) => {
        try {
          const sessionId = req.params["sessionId"];
          const spans = this.memoryExporter.getFinishedSpans(sessionId);
          if (spans.length === 0) {
            return res.json([]);
          }
          const spanData = spans.map((span) => {
            var _a;
            return {
              name: span.name,
              span_id: span.spanContext().spanId,
              trace_id: span.spanContext().traceId,
              start_time: (0, import_telemetry_utils.hrTimeToNanoseconds)(span.startTime),
              end_time: (0, import_telemetry_utils.hrTimeToNanoseconds)(span.endTime),
              attributes: { ...span.attributes },
              parent_span_id: ((_a = span.parentSpanContext) == null ? void 0 : _a.spanId) || null
            };
          });
          return res.json(spanData);
        } catch (e) {
          const error = `Failed to get trace: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
          return;
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/events/:eventId/graph",
      async (req, res) => {
        try {
          var _stack = [];
          try {
            const appName = req.params["appName"];
            const userId = req.params["userId"];
            const sessionId = req.params["sessionId"];
            const eventId = req.params["eventId"];
            const session = await this.sessionService.getSession({
              appName,
              userId,
              sessionId
            });
            if (!session) {
              res.status(404).json({ error: `Session not found: ${sessionId}` });
              return;
            }
            const sessionEvents = session.events || [];
            const event = sessionEvents.find((e) => e.id === eventId);
            if (!event) {
              res.status(404).json({ error: `Event not found: ${eventId}` });
              return;
            }
            const functionCalls = (0, import_adk.getFunctionCalls)(event);
            const functionResponses = (0, import_adk.getFunctionResponses)(event);
            const agentFile = __using(_stack, await this.agentLoader.getAgentFile(appName), true);
            const rootAgent = await agentFile.load();
            if (functionCalls.length > 0) {
              const functionCallHighlights = [];
              for (const functionCall of functionCalls) {
                functionCallHighlights.push([event.author, functionCall.name]);
              }
              return res.send({
                dotSrc: await (0, import_agent_graph.getAgentGraphAsDot)(
                  rootAgent,
                  functionCallHighlights
                )
              });
            }
            if (functionResponses.length > 0) {
              const functionCallHighlights = [];
              for (const functionResponse of functionResponses) {
                functionCallHighlights.push([
                  functionResponse.name,
                  event.author
                ]);
              }
              return res.send({
                dotSrc: await (0, import_agent_graph.getAgentGraphAsDot)(
                  rootAgent,
                  functionCallHighlights
                )
              });
            }
            return res.send({
              dotSrc: await (0, import_agent_graph.getAgentGraphAsDot)(rootAgent, [[event.author, ""]])
            });
          } catch (_) {
            var _error = _, _hasError = true;
          } finally {
            var _promise = __callDispose(_stack, _error, _hasError);
            _promise && await _promise;
          }
        } catch (e) {
          const error = `Failed to get agent graph: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
          return;
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const session = await this.sessionService.getSession({
            appName,
            userId,
            sessionId
          });
          if (!session) {
            res.status(404).json({ error: `Session not found: ${sessionId}` });
            return;
          }
          res.json(session);
        } catch (e) {
          const error = `Failed to get session: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessions = await this.sessionService.listSessions({
            appName,
            userId
          });
          res.json(sessions);
        } catch (e) {
          const error = `Failed to list sessions: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.post(
      "/apps/:appName/users/:userId/sessions/:sessionId",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const state = req.body["state"] || {};
          const existingSession = await this.sessionService.getSession({
            appName,
            userId,
            sessionId
          });
          if (existingSession) {
            res.status(400).json({ error: `Session already exists: ${sessionId}` });
            return;
          }
          const createdSession = await this.sessionService.createSession({
            appName,
            userId,
            state,
            sessionId
          });
          res.json(createdSession);
        } catch (e) {
          const error = `Failed to create session: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.post(
      "/apps/:appName/users/:userId/sessions",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const state = req.body["state"] || {};
          const createdSession = await this.sessionService.createSession({
            appName,
            userId,
            state
          });
          res.json(createdSession);
        } catch (e) {
          const error = `Failed to create session: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.delete(
      "/apps/:appName/users/:userId/sessions/:sessionId",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const session = await this.sessionService.getSession({
            appName,
            userId,
            sessionId
          });
          if (!session) {
            res.status(404).json({ error: `Session not found: ${sessionId}` });
            return;
          }
          await this.sessionService.deleteSession({
            appName,
            userId,
            sessionId
          });
          res.status(204).json({});
        } catch (e) {
          const error = `Failed to delete session: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts/:artifactName",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactName = req.params["artifactName"];
          const artifact = await this.artifactService.loadArtifact({
            appName,
            userId,
            sessionId,
            filename: artifactName
          });
          if (!artifact) {
            res.status(404).json({ error: `Artifact not found: ${artifactName}` });
            return;
          }
          res.json(artifact);
        } catch (e) {
          const error = `Failed to load artifact: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts/:artifactName/versions/:versionId",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactName = req.params["artifactName"];
          const versionId = req.params["versionId"];
          const artifact = await this.artifactService.loadArtifact({
            appName,
            userId,
            sessionId,
            filename: artifactName,
            version: parseInt(versionId, 10)
          });
          if (!artifact) {
            res.status(404).json({ error: `Artifact not found: ${artifactName}` });
            return;
          }
          res.json(artifact);
        } catch (e) {
          const error = `Failed to load artifact version: ${e.message}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactKeys = await this.artifactService.listArtifactKeys({
            appName,
            userId,
            sessionId
          });
          res.json(artifactKeys);
        } catch (e) {
          const error = `Failed to list artifacts: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts/:artifactName/versions",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactName = req.params["artifactName"];
          const artifactVersions = await this.artifactService.listVersions({
            appName,
            userId,
            sessionId,
            filename: artifactName
          });
          res.json(artifactVersions);
        } catch (e) {
          const error = `Failed to list artifact versions: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.delete(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts/:artifactName",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactName = req.params["artifactName"];
          await this.artifactService.deleteArtifact({
            appName,
            userId,
            sessionId,
            filename: artifactName
          });
          res.status(204).json({});
        } catch (e) {
          const error = `Failed to delete artifact: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.post(
      "/apps/:appName/eval_sets/:evalSetId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get("/apps/:appName/eval_sets", (req, res) => {
      return res.status(501).json({ error: "Not implemented" });
    });
    app.post(
      "/apps/:appName/eval_sets/:evalSetId/add_session",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get(
      "/apps/:appName/eval_sets/:evalSetId/evals",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get(
      "/apps/:appName/eval_sets/:evalSetId/evals/:evalCaseId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.put(
      "/apps/:appName/eval_sets/:evalSetId/evals/:evalCaseId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.delete(
      "/apps/:appName/eval_sets/:evalSetId/evals/:evalCaseId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.post(
      "/apps/:appName/eval_sets/:evalSetId/run_eval",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get(
      "/apps/:appName/eval_results/:evalResultId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get("/apps/:appName/eval_results", (req, res) => {
      return res.status(501).json({ error: "Not implemented" });
    });
    app.get("/apps/:appName/eval_metrics", (req, res) => {
      return res.status(501).json({ error: "Not implemented" });
    });
    app.post("/run", async (req, res) => {
      const { appName, userId, sessionId, newMessage, stateDelta } = req.body;
      const session = await this.sessionService.getSession({
        appName,
        userId,
        sessionId
      });
      if (!session) {
        res.status(404).json({ error: `Session not found: ${sessionId}` });
        return;
      }
      try {
        var _stack = [];
        try {
          const agentFile = __using(_stack, await this.agentLoader.getAgentFile(appName), true);
          const agent = await agentFile.load();
          const runner = await this.getRunner(agent, appName);
          const events = [];
          for await (const e of runner.runAsync({
            userId,
            sessionId,
            newMessage,
            stateDelta
          })) {
            events.push(e);
          }
          res.json(events);
        } catch (_) {
          var _error = _, _hasError = true;
        } finally {
          var _promise = __callDispose(_stack, _error, _hasError);
          _promise && await _promise;
        }
      } catch (e) {
        const error = `Failed to run agent: ${e}`;
        res.status(500).json({ error });
        this.logger.error(error);
      }
    });
    app.post("/run_sse", async (req, res) => {
      const { appName, userId, sessionId, newMessage, streaming, stateDelta } = req.body;
      const session = await this.sessionService.getSession({
        appName,
        userId,
        sessionId
      });
      if (!session) {
        const error = `Session not found: ${sessionId}`;
        res.status(404).json({ error });
        this.logger.error(error);
        return;
      }
      try {
        var _stack = [];
        try {
          const agentFile = __using(_stack, await this.agentLoader.getAgentFile(appName), true);
          const agent = await agentFile.load();
          const runner = await this.getRunner(agent, appName);
          res.setHeader("Cache-Control", "no-cache");
          res.setHeader("Content-Type", "text/event-stream");
          res.setHeader("Access-Control-Allow-Origin", "*");
          res.setHeader("Connection", "keep-alive");
          res.flushHeaders();
          for await (const event of runner.runAsync({
            userId,
            sessionId,
            newMessage,
            runConfig: {
              streamingMode: streaming ? import_adk.StreamingMode.SSE : import_adk.StreamingMode.NONE
            },
            stateDelta
          })) {
            res.write(`data: ${JSON.stringify(event)}

`);
          }
          res.end();
        } catch (_) {
          var _error = _, _hasError = true;
        } finally {
          var _promise = __callDispose(_stack, _error, _hasError);
          _promise && await _promise;
        }
      } catch (e) {
        if (res.headersSent) {
          const error = e.message;
          res.end(`data: ${JSON.stringify({ error })}

`);
          this.logger.error(error);
        } else {
          const error = `Failed to run agent: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    });
  }
  async start() {
    await this.init();
    return new Promise((resolve, reject) => {
      this.server = this.app.listen(this.port, async () => {
        try {
          if (this.a2a) {
            await this.initA2A();
          }
          console.log(`
+-----------------------------------------------------------------------------+
| ADK API Server started                                                      |
|                                                                             |
| For local testing, access at ${this.url}.${"".padStart(39 - this.url.length)}       |
+-----------------------------------------------------------------------------+`);
          resolve();
        } catch (error) {
          this.logger.error("Error during AdkApiServer startup:", error);
          reject(error);
        }
      });
      this.server.on("error", (err) => {
        if (err.code === "EADDRINUSE") {
          const error = new Error();
          error.cause = err;
          error.message = `Port ${this.port} is already in use`;
          reject(error);
        } else {
          reject(err);
        }
      });
    });
  }
  stop() {
    if (!this.server) {
      return Promise.resolve();
    }
    return new Promise((resolve, reject) => {
      this.server.close((err) => {
        if (err) {
          reject(err);
          return;
        }
        console.log(`
+-----------------------------------------------------------------------------+
| ADK API Server stopped                                                      |
+-----------------------------------------------------------------------------+`);
        resolve();
      });
    });
  }
  async getRunner(agent, appName) {
    if (!(appName in this.runnerCache)) {
      this.runnerCache[appName] = new import_adk.Runner({
        appName,
        agent,
        memoryService: this.memoryService,
        sessionService: this.sessionService,
        artifactService: this.artifactService
      });
    }
    return this.runnerCache[appName];
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AdkApiServer
});
