// @google/adk-devtools/dist/cjs/server/adk_api_client.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var adk_api_client_exports = {};
__export(adk_api_client_exports, {
  AdkApiClient: () => AdkApiClient
});
module.exports = __toCommonJS(adk_api_client_exports);
var import_genai = require("@google/genai");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class AdkApiClient {
  constructor(config) {
    this.backendUrl = config.backendUrl;
  }
  async listApps() {
    return this.fetch(`${this.backendUrl}/list-apps`);
  }
  async getSession(params) {
    return this.fetch(
      `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json"
        }
      }
    );
  }
  async createSession(params) {
    let url = `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions`;
    if (params.sessionId) {
      url += `/${params.sessionId}`;
    }
    return this.fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        state: params.state
      })
    });
  }
  async deleteSession(params) {
    return this.fetch(
      `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}`,
      {
        method: "DELETE",
        headers: {
          "Content-Type": "application/json"
        }
      }
    );
  }
  async listAllSessions(params) {
    const apps = await this.listApps();
    return Promise.all(
      apps.map(
        (appName) => this.listSessions({ appName, userId: params.userId })
      )
    ).then((sessions) => sessions.flat());
  }
  async listSessions(params) {
    const sessions = await this.fetch(
      `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json"
        }
      }
    );
    if ("sessions" in sessions) {
      return sessions.sessions;
    }
    return sessions;
  }
  async *runAsync(params) {
    const response = await fetch(`${this.backendUrl}/run_sse`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        appName: params.appName,
        userId: params.userId,
        sessionId: params.sessionId,
        streaming: params.streaming,
        stateDelta: params.stateDelta,
        newMessage: typeof params.newMessage === "string" ? (0, import_genai.createUserContent)(params.newMessage) : params.newMessage
      })
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(
        error.error || `Request failed with status ${response.status}`
      );
    }
    if (!response.body) {
      return;
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";
    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n\n");
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        if (line.startsWith("data: ")) {
          const data = JSON.parse(line.substring("data: ".length));
          if (data.error) {
            throw new Error(data.error);
          }
          yield data;
        }
      }
    }
  }
  async listArtifacts(params) {
    return this.fetch(
      `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}/artifacts`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json"
        }
      }
    );
  }
  async loadArtifact(params) {
    let url = `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}/artifacts/${params.artifactName}`;
    if (params.version !== void 0) {
      url += `/versions/${params.version}`;
    }
    return this.fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json"
      }
    });
  }
  async listArtifactVersions(params) {
    const url = `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}/artifacts/${params.artifactName}/versions`;
    return this.fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json"
      }
    });
  }
  async deleteArtifact(params) {
    const url = `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}/artifacts/${params.artifactName}`;
    return this.fetch(url, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json"
      }
    });
  }
  async fetch(url, options) {
    const response = await fetch(url, options);
    if (!response.ok) {
      let error;
      try {
        error = await response.json();
      } catch (_e) {
      }
      throw new Error(
        (error == null ? void 0 : error.error) ?? `Request failed with status ${response.status}`
      );
    }
    return response.json();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AdkApiClient
});
