// @google/adk-devtools/dist/cjs/server/agent_graph.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var agent_graph_exports = {};
__export(agent_graph_exports, {
  buildGraph: () => buildGraph,
  getAgentGraph: () => getAgentGraph,
  getAgentGraphAsDot: () => getAgentGraphAsDot
});
module.exports = __toCommonJS(agent_graph_exports);
var import_adk = require("@google/adk");
var import_ts_graphviz = require("ts-graphviz");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const DARK_GREEN = "#0F5223";
const LIGHT_GREEN = "#69CB87";
const LIGHT_GRAY = "#cccccc";
const WHITE = "#ffffff";
async function buildGraph(graph, rootAgent, highlightsPairs, parentAgent) {
  async function buildCluster(subgraph, agent) {
    if ((0, import_adk.isLoopAgent)(agent)) {
      if (parentAgent) {
        drawEdge(parentAgent.name, agent.subAgents[0].name);
      }
      const length = agent.subAgents.length;
      let currLength = 0;
      for (const subAgent of agent.subAgents) {
        await buildGraph(subgraph, subAgent, highlightsPairs, agent);
        const adjAgent = currLength === length - 1 ? agent.subAgents[0] : agent.subAgents[currLength + 1];
        drawEdge(agent.subAgents[currLength].name, adjAgent.name);
        currLength++;
      }
    } else if ((0, import_adk.isSequentialAgent)(agent)) {
      if (parentAgent) {
        drawEdge(parentAgent.name, agent.subAgents[0].name);
      }
      const length = agent.subAgents.length;
      let currLength = 0;
      for (const subAgent of agent.subAgents) {
        await buildGraph(subgraph, subAgent, highlightsPairs, agent);
        if (currLength !== length - 1) {
          drawEdge(
            agent.subAgents[currLength].name,
            agent.subAgents[currLength + 1].name
          );
        }
        currLength++;
      }
    } else if ((0, import_adk.isParallelAgent)(agent)) {
      for (const subAgent of agent.subAgents) {
        await buildGraph(subgraph, subAgent, highlightsPairs, agent);
        if (parentAgent) {
          drawEdge(parentAgent.name, subAgent.name);
        }
      }
    } else {
      for (const subAgent of agent.subAgents) {
        await buildGraph(subgraph, subAgent, highlightsPairs, agent);
        drawEdge(agent.name, subAgent.name);
      }
    }
    return subgraph;
  }
  async function drawNode(toolOrAgent) {
    const name = getNodeName(toolOrAgent);
    const shape = getNodeShape(toolOrAgent);
    const caption = getNodeCaption(toolOrAgent);
    const asCluster = shouldBuildAgentCluster(toolOrAgent);
    if (highlightsPairs) {
      for (const highlightsPair of highlightsPairs) {
        if (highlightsPair.includes(name)) {
          if (asCluster) {
            const cluster = new import_ts_graphviz.Subgraph(`cluster_${name}`, {
              label: `cluster_${name}`,
              style: "rounded",
              bgcolor: WHITE,
              fontcolor: LIGHT_GRAY
            });
            graph.addSubgraph(cluster);
            await buildCluster(cluster, rootAgent);
          } else {
            graph.addNode(
              new import_ts_graphviz.Node(name, {
                label: caption,
                style: "filled,rounded",
                fillcolor: DARK_GREEN,
                color: DARK_GREEN,
                shape,
                fontcolor: LIGHT_GRAY
              })
            );
          }
          return;
        }
      }
    }
    if (asCluster) {
      const cluster = new import_ts_graphviz.Subgraph(`cluster_${name}`, {
        label: `cluster_${name}`,
        style: "rounded",
        bgcolor: WHITE,
        fontcolor: LIGHT_GRAY
      });
      graph.addSubgraph(cluster);
      await buildCluster(cluster, rootAgent);
      return;
    }
    graph.addNode(
      new import_ts_graphviz.Node(name, {
        label: caption,
        style: "rounded",
        fillcolor: WHITE,
        color: LIGHT_GRAY,
        shape,
        fontcolor: LIGHT_GRAY
      })
    );
  }
  function drawEdge(fromName, toName) {
    if (highlightsPairs) {
      for (const [highlightFrom, highlightTo] of highlightsPairs) {
        if (fromName === highlightFrom && toName === highlightTo) {
          graph.addEdge(
            new import_ts_graphviz.Edge([graph.node(fromName), graph.node(toName)], {
              color: LIGHT_GREEN
            })
          );
          return;
        }
        if (fromName === highlightTo && toName === highlightFrom) {
          graph.addEdge(
            new import_ts_graphviz.Edge([graph.node(fromName), graph.node(toName)], {
              color: LIGHT_GREEN,
              dir: "back"
            })
          );
          return;
        }
      }
    }
    if (shouldBuildAgentCluster(rootAgent)) {
      graph.addEdge(
        new import_ts_graphviz.Edge([new import_ts_graphviz.Node(fromName), new import_ts_graphviz.Node(toName)], {
          color: LIGHT_GREEN
        })
      );
      return;
    }
    graph.addEdge(
      new import_ts_graphviz.Edge([new import_ts_graphviz.Node(fromName), new import_ts_graphviz.Node(toName)], {
        arrowhead: "none",
        color: LIGHT_GRAY
      })
    );
  }
  await drawNode(rootAgent);
  for (const subAgent of rootAgent.subAgents) {
    await buildGraph(graph, subAgent, highlightsPairs, rootAgent);
    if (!shouldBuildAgentCluster(subAgent) && !shouldBuildAgentCluster(rootAgent)) {
      drawEdge(rootAgent.name, subAgent.name);
    }
  }
  if ((0, import_adk.isLlmAgent)(rootAgent)) {
    for (const tool of await rootAgent.canonicalTools()) {
      await drawNode(tool);
      drawEdge(rootAgent.name, getNodeName(tool));
    }
  }
}
function getNodeName(toolOrAgent) {
  if ((0, import_adk.isBaseAgent)(toolOrAgent)) {
    if ((0, import_adk.isSequentialAgent)(toolOrAgent)) {
      return `${toolOrAgent.name} (Sequential Agent)`;
    }
    if ((0, import_adk.isLoopAgent)(toolOrAgent)) {
      return `${toolOrAgent.name} (Loop Agent)`;
    }
    if ((0, import_adk.isParallelAgent)(toolOrAgent)) {
      return `${toolOrAgent.name} (Parallel Agent)`;
    }
    return toolOrAgent.name;
  }
  if ((0, import_adk.isBaseTool)(toolOrAgent)) {
    return toolOrAgent.name;
  }
  throw new Error(`Unsupported tool type: ${toolOrAgent}`);
}
function getNodeCaption(toolOrAgent) {
  if ((0, import_adk.isBaseAgent)(toolOrAgent)) {
    return `\u{1F916} ${toolOrAgent.name}`;
  }
  if ((0, import_adk.isFunctionTool)(toolOrAgent)) {
    return `\u{1F527} ${toolOrAgent.name}`;
  }
  if ((0, import_adk.isAgentTool)(toolOrAgent)) {
    return `\u{1F916} ${toolOrAgent.name}`;
  }
  if ((0, import_adk.isBaseTool)(toolOrAgent)) {
    return `\u{1F527} ${toolOrAgent.name}`;
  }
  console.warn(`Unsupported tool type: ${typeof toolOrAgent}`);
  return `\u2753 Unsupported tool type: ${typeof toolOrAgent}`;
}
function getNodeShape(toolOrAgent) {
  if ((0, import_adk.isBaseAgent)(toolOrAgent)) {
    return "ellipse";
  }
  if ((0, import_adk.isFunctionTool)(toolOrAgent)) {
    return "box";
  }
  if ((0, import_adk.isBaseTool)(toolOrAgent)) {
    return "box";
  }
  console.warn(`Unsupported tool type: ${typeof toolOrAgent}`);
  return "cylinder";
}
function shouldBuildAgentCluster(toolOrAgent) {
  if ((0, import_adk.isSequentialAgent)(toolOrAgent)) {
    return true;
  }
  if ((0, import_adk.isLoopAgent)(toolOrAgent)) {
    return true;
  }
  if ((0, import_adk.isParallelAgent)(toolOrAgent)) {
    return true;
  }
  return false;
}
async function getAgentGraph(rootAgent, highlightsPairs) {
  const graph = new import_ts_graphviz.Digraph(
    rootAgent.name,
    /* strict= */
    true,
    {
      rankdir: "LR",
      bgcolor: "#333537"
    }
  );
  await buildGraph(graph, rootAgent, highlightsPairs);
  return graph;
}
async function getAgentGraphAsDot(rootAgent, highlightsPairs) {
  const graph = await getAgentGraph(rootAgent, highlightsPairs);
  return (0, import_ts_graphviz.toDot)(graph);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  buildGraph,
  getAgentGraph,
  getAgentGraphAsDot
});
