// @google/adk/dist/esm/a2a/a2a_event.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { randomUUID } from "../utils/env_aware_utils.js";
var MessageRole = /* @__PURE__ */ ((MessageRole2) => {
  MessageRole2["USER"] = "user";
  MessageRole2["AGENT"] = "agent";
  return MessageRole2;
})(MessageRole || {});
var TaskState = /* @__PURE__ */ ((TaskState2) => {
  TaskState2["SUBMITTED"] = "submitted";
  TaskState2["WORKING"] = "working";
  TaskState2["COMPLETED"] = "completed";
  TaskState2["FAILED"] = "failed";
  TaskState2["CANCELED"] = "canceled";
  TaskState2["REJECTED"] = "rejected";
  TaskState2["INPUT_REQUIRED"] = "input-required";
  return TaskState2;
})(TaskState || {});
function isTaskStatusUpdateEvent(event) {
  return (event == null ? void 0 : event.kind) === "status-update";
}
function isTaskArtifactUpdateEvent(event) {
  return (event == null ? void 0 : event.kind) === "artifact-update";
}
function isMessage(event) {
  return (event == null ? void 0 : event.kind) === "message";
}
function isTask(event) {
  return (event == null ? void 0 : event.kind) === "task";
}
function getEventMetadata(event) {
  if (isTaskArtifactUpdateEvent(event)) {
    return event.artifact.metadata || {};
  }
  if (isTaskStatusUpdateEvent(event) || isTask(event) || isMessage(event)) {
    return event.metadata || {};
  }
  return {};
}
function isFailedTaskStatusUpdateEvent(event) {
  return (isTaskStatusUpdateEvent(event) || isTask(event)) && event.status.state === "failed" /* FAILED */;
}
function isTerminalTaskStatusUpdateEvent(event) {
  return (isTaskStatusUpdateEvent(event) || isTask(event)) && [
    "completed" /* COMPLETED */,
    "failed" /* FAILED */,
    "canceled" /* CANCELED */,
    "rejected" /* REJECTED */
  ].includes(event.status.state);
}
function isInputRequiredTaskStatusUpdateEvent(event) {
  return (isTaskStatusUpdateEvent(event) || isTask(event)) && event.status.state === "input-required" /* INPUT_REQUIRED */;
}
function getFailedTaskStatusUpdateEventError(event) {
  var _a;
  if (!isFailedTaskStatusUpdateEvent(event)) {
    return void 0;
  }
  const parts = ((_a = event.status.message) == null ? void 0 : _a.parts) || [];
  if (parts.length === 0) {
    return void 0;
  }
  if (parts[0].kind !== "text") {
    return void 0;
  }
  return parts[0].text;
}
function createTaskSubmittedEvent({
  taskId,
  contextId,
  message,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: false,
    status: {
      state: "submitted" /* SUBMITTED */,
      message,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTask({
  contextId,
  message,
  taskId,
  metadata
}) {
  return {
    kind: "task",
    id: taskId || randomUUID(),
    contextId,
    history: [message],
    status: {
      state: "submitted" /* SUBMITTED */,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTaskWorkingEvent({
  taskId,
  contextId,
  message,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: false,
    status: {
      state: "working" /* WORKING */,
      message,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTaskCompletedEvent({
  taskId,
  contextId,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: true,
    status: {
      state: "completed" /* COMPLETED */,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTaskArtifactUpdateEvent({
  taskId,
  contextId,
  artifactId,
  parts = [],
  metadata,
  append,
  lastChunk
}) {
  return {
    kind: "artifact-update",
    taskId,
    contextId,
    append,
    lastChunk,
    artifact: {
      artifactId: artifactId || randomUUID(),
      parts
    },
    metadata
  };
}
function createTaskFailedEvent({
  taskId,
  contextId,
  error,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: true,
    status: {
      state: "failed" /* FAILED */,
      message: {
        kind: "message",
        messageId: randomUUID(),
        role: "agent",
        taskId,
        contextId,
        parts: [
          {
            kind: "text",
            text: error.message
          }
        ]
      },
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTaskInputRequiredEvent({
  taskId,
  contextId,
  parts,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: true,
    status: {
      state: "input-required" /* INPUT_REQUIRED */,
      message: {
        kind: "message",
        messageId: randomUUID(),
        role: "agent",
        taskId,
        contextId,
        parts
      },
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createInputMissingErrorEvent({
  taskId,
  contextId,
  parts,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: true,
    status: {
      state: "input-required" /* INPUT_REQUIRED */,
      message: {
        kind: "message",
        messageId: randomUUID(),
        role: "agent",
        taskId,
        contextId,
        parts
      },
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
export {
  MessageRole,
  TaskState,
  createInputMissingErrorEvent,
  createTask,
  createTaskArtifactUpdateEvent,
  createTaskCompletedEvent,
  createTaskFailedEvent,
  createTaskInputRequiredEvent,
  createTaskSubmittedEvent,
  createTaskWorkingEvent,
  getEventMetadata,
  getFailedTaskStatusUpdateEventError,
  isFailedTaskStatusUpdateEvent,
  isInputRequiredTaskStatusUpdateEvent,
  isMessage,
  isTask,
  isTaskArtifactUpdateEvent,
  isTaskStatusUpdateEvent,
  isTerminalTaskStatusUpdateEvent
};
