// @google/adk/dist/esm/a2a/event_converter_utils.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  createModelContent,
  createUserContent
} from "@google/genai";
import { createEvent } from "../events/event.js";
import { createEventActions } from "../events/event_actions.js";
import { randomUUID } from "../utils/env_aware_utils.js";
import {
  getEventMetadata,
  getFailedTaskStatusUpdateEventError,
  isFailedTaskStatusUpdateEvent,
  isInputRequiredTaskStatusUpdateEvent,
  isMessage,
  isTask,
  isTaskArtifactUpdateEvent,
  isTaskStatusUpdateEvent,
  isTerminalTaskStatusUpdateEvent,
  MessageRole
} from "./a2a_event.js";
import {
  A2AMetadataKeys,
  getA2AEventMetadata
} from "./metadata_converter_utils.js";
import { toA2AParts, toGenAIPart, toGenAIParts } from "./part_converter_utils.js";
function toA2AMessage(event, {
  appName,
  userId,
  sessionId
}) {
  var _a;
  return {
    kind: "message",
    messageId: randomUUID(),
    role: event.author === MessageRole.USER ? MessageRole.USER : MessageRole.AGENT,
    parts: toA2AParts(((_a = event.content) == null ? void 0 : _a.parts) || [], event.longRunningToolIds),
    metadata: getA2AEventMetadata(event, { appName, userId, sessionId })
  };
}
function toAdkEvent(event, invocationId, agentName) {
  if (isMessage(event)) {
    return messageToAdkEvent(event, invocationId, agentName);
  }
  if (isTask(event)) {
    return taskToAdkEvent(event, invocationId, agentName);
  }
  if (isTaskArtifactUpdateEvent(event)) {
    return artifactUpdateToAdkEvent(event, invocationId, agentName);
  }
  if (isTaskStatusUpdateEvent(event)) {
    return event.final ? finalTaskStatusUpdateToAdkEvent(event, invocationId, agentName) : taskStatusUpdateToAdkEvent(event, invocationId, agentName);
  }
  return void 0;
}
function messageToAdkEvent(msg, invocationId, agentName, parentEvent) {
  const parts = toGenAIParts(msg.parts);
  return {
    ...createAdkEventFromMetadata(parentEvent || msg),
    invocationId,
    author: msg.role === MessageRole.USER ? MessageRole.USER : agentName,
    content: msg.role === MessageRole.USER ? createUserContent(parts) : createModelContent(parts),
    turnComplete: true,
    partial: false
  };
}
function artifactUpdateToAdkEvent(a2aEvent, invocationId, agentName) {
  var _a;
  const partsToConvert = ((_a = a2aEvent.artifact) == null ? void 0 : _a.parts) || [];
  if (partsToConvert.length === 0) {
    return void 0;
  }
  const partial = !!getEventMetadata(a2aEvent)[A2AMetadataKeys.PARTIAL] || a2aEvent.append || !a2aEvent.lastChunk;
  return {
    ...createAdkEventFromMetadata(a2aEvent),
    invocationId,
    author: agentName,
    content: createModelContent(toGenAIParts(partsToConvert)),
    longRunningToolIds: getLongRunningToolIDs(partsToConvert),
    partial
  };
}
function finalTaskStatusUpdateToAdkEvent(a2aEvent, invocationId, agentName) {
  var _a;
  const partsToConvert = ((_a = a2aEvent.status.message) == null ? void 0 : _a.parts) || [];
  if (partsToConvert.length === 0) {
    return void 0;
  }
  const parts = toGenAIParts(partsToConvert);
  const isFailedTask = isFailedTaskStatusUpdateEvent(a2aEvent);
  const hasContent = !isFailedTask && parts.length > 0;
  return {
    ...createAdkEventFromMetadata(a2aEvent),
    invocationId,
    author: agentName,
    errorMessage: isFailedTask ? getFailedTaskStatusUpdateEventError(a2aEvent) : void 0,
    content: hasContent ? createModelContent(parts) : void 0,
    longRunningToolIds: getLongRunningToolIDs(partsToConvert),
    turnComplete: true
  };
}
function taskStatusUpdateToAdkEvent(a2aEvent, invocationId, agentName) {
  const msg = a2aEvent.status.message;
  if (!msg) {
    return void 0;
  }
  const parts = toGenAIParts(msg.parts);
  return {
    ...createAdkEventFromMetadata(a2aEvent),
    invocationId,
    author: agentName,
    content: createModelContent(parts),
    turnComplete: false,
    partial: true
  };
}
function taskToAdkEvent(a2aTask, invocationId, agentName) {
  var _a, _b;
  const parts = [];
  const longRunningToolIds = [];
  if (a2aTask.artifacts) {
    for (const artifact of a2aTask.artifacts) {
      if (((_a = artifact.parts) == null ? void 0 : _a.length) > 0) {
        const artifactParts = toGenAIParts(artifact.parts);
        parts.push(...artifactParts);
        longRunningToolIds.push(...getLongRunningToolIDs(artifact.parts));
      }
    }
  }
  if ((_b = a2aTask.status) == null ? void 0 : _b.message) {
    const a2aParts = a2aTask.status.message.parts;
    const genAIParts = toGenAIParts(a2aParts);
    parts.push(...genAIParts);
    longRunningToolIds.push(...getLongRunningToolIDs(a2aParts));
  }
  const isTerminal = isTerminalTaskStatusUpdateEvent(a2aTask) || isInputRequiredTaskStatusUpdateEvent(a2aTask);
  const isFailed = isFailedTaskStatusUpdateEvent(a2aTask);
  if (parts.length === 0 && !isTerminal) {
    return void 0;
  }
  return {
    ...createAdkEventFromMetadata(a2aTask),
    invocationId,
    author: agentName,
    content: isFailed ? void 0 : createModelContent(parts),
    errorMessage: isFailed ? getFailedTaskStatusUpdateEventError(a2aTask) : void 0,
    longRunningToolIds,
    turnComplete: isTerminal
  };
}
function createAdkEventFromMetadata(a2aEvent) {
  const metadata = a2aEvent.metadata || {};
  return createEvent({
    branch: metadata[A2AMetadataKeys.BRANCH],
    author: metadata[A2AMetadataKeys.AUTHOR],
    partial: metadata[A2AMetadataKeys.PARTIAL],
    errorCode: metadata[A2AMetadataKeys.ERROR_CODE],
    errorMessage: metadata[A2AMetadataKeys.ERROR_MESSAGE],
    citationMetadata: metadata[A2AMetadataKeys.CITATION_METADATA],
    groundingMetadata: metadata[A2AMetadataKeys.GROUNDING_METADATA],
    usageMetadata: metadata[A2AMetadataKeys.USAGE_METADATA],
    customMetadata: metadata[A2AMetadataKeys.CUSTOM_METADATA],
    actions: createEventActions({
      escalate: !!metadata[A2AMetadataKeys.ESCALATE],
      transferToAgent: metadata[A2AMetadataKeys.TRANSFER_TO_AGENT]
    })
  });
}
function getLongRunningToolIDs(parts) {
  const ids = [];
  for (const a2aPart of parts) {
    if (a2aPart.metadata && a2aPart.metadata[A2AMetadataKeys.IS_LONG_RUNNING]) {
      const genAIPart = toGenAIPart(a2aPart);
      if (genAIPart.functionCall && genAIPart.functionCall.id) {
        ids.push(genAIPart.functionCall.id);
      }
    }
  }
  return ids;
}
export {
  toA2AMessage,
  toAdkEvent
};
