// @google/adk/dist/cjs/events/compacted_event.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var compacted_event_exports = {};
__export(compacted_event_exports, {
  createCompactedEvent: () => createCompactedEvent,
  isCompactedEvent: () => isCompactedEvent
});
module.exports = __toCommonJS(compacted_event_exports);
var import_event = require("./event.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function isCompactedEvent(event) {
  return "isCompacted" in event && event.isCompacted === true;
}
function createCompactedEvent(params = {}) {
  return {
    ...(0, import_event.createEvent)(params),
    isCompacted: params.isCompacted || true,
    startTime: params.startTime,
    endTime: params.endTime,
    compactedContent: params.compactedContent
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createCompactedEvent,
  isCompactedEvent
});
