// @google/adk/dist/cjs/events/structured_events.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var structured_events_exports = {};
__export(structured_events_exports, {
  EventType: () => EventType,
  toStructuredEvents: () => toStructuredEvents
});
module.exports = __toCommonJS(structured_events_exports);
var import_lodash_es = require("lodash-es");
var import_event = require("./event.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var EventType = /* @__PURE__ */ ((EventType2) => {
  EventType2["THOUGHT"] = "thought";
  EventType2["CONTENT"] = "content";
  EventType2["TOOL_CALL"] = "tool_call";
  EventType2["TOOL_RESULT"] = "tool_result";
  EventType2["CALL_CODE"] = "call_code";
  EventType2["CODE_RESULT"] = "code_result";
  EventType2["ERROR"] = "error";
  EventType2["ACTIVITY"] = "activity";
  EventType2["TOOL_CONFIRMATION"] = "tool_confirmation";
  EventType2["FINISHED"] = "finished";
  return EventType2;
})(EventType || {});
function toStructuredEvents(event) {
  var _a, _b;
  const structuredEvents = [];
  if (event.errorCode) {
    structuredEvents.push({
      type: "error" /* ERROR */,
      error: new Error(event.errorMessage || event.errorCode)
    });
    return structuredEvents;
  }
  for (const part of (_b = (_a = event.content) == null ? void 0 : _a.parts) != null ? _b : []) {
    if (part.functionCall && !(0, import_lodash_es.isEmpty)(part.functionCall)) {
      structuredEvents.push({
        type: "tool_call" /* TOOL_CALL */,
        call: part.functionCall
      });
    } else if (part.functionResponse && !(0, import_lodash_es.isEmpty)(part.functionResponse)) {
      structuredEvents.push({
        type: "tool_result" /* TOOL_RESULT */,
        result: part.functionResponse
      });
    } else if (part.executableCode && !(0, import_lodash_es.isEmpty)(part.executableCode)) {
      structuredEvents.push({
        type: "call_code" /* CALL_CODE */,
        code: part.executableCode
      });
    } else if (part.codeExecutionResult && !(0, import_lodash_es.isEmpty)(part.codeExecutionResult)) {
      structuredEvents.push({
        type: "code_result" /* CODE_RESULT */,
        result: part.codeExecutionResult
      });
    } else if (part.text) {
      if (part.thought) {
        structuredEvents.push({ type: "thought" /* THOUGHT */, content: part.text });
      } else {
        structuredEvents.push({ type: "content" /* CONTENT */, content: part.text });
      }
    }
  }
  if (event.actions.requestedToolConfirmations && !(0, import_lodash_es.isEmpty)(event.actions.requestedToolConfirmations)) {
    structuredEvents.push({
      type: "tool_confirmation" /* TOOL_CONFIRMATION */,
      confirmations: event.actions.requestedToolConfirmations
    });
  }
  if ((0, import_event.isFinalResponse)(event)) {
    structuredEvents.push({ type: "finished" /* FINISHED */, output: void 0 });
  }
  return structuredEvents;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EventType,
  toStructuredEvents
});
