// @google/adk/dist/web/models/apigee_llm.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __await = function(promise, isYieldStar) {
  this[0] = promise;
  this[1] = isYieldStar;
};
var __asyncGenerator = (__this, __arguments, generator) => {
  var resume = (k, v, yes, no) => {
    try {
      var x = generator[k](v), isAwait = (v = x.value) instanceof __await, done = x.done;
      Promise.resolve(isAwait ? v[0] : v).then((y) => isAwait ? resume(k === "return" ? k : "next", v[1] ? { done: y.done, value: y.value } : y, yes, no) : yes({ value: y, done })).catch((e) => resume("throw", e, yes, no));
    } catch (e) {
      no(e);
    }
  }, method = (k) => it[k] = (x) => new Promise((yes, no) => resume(k, x, yes, no)), it = {};
  return generator = generator.apply(__this, __arguments), it[__knownSymbol("asyncIterator")] = () => it, method("next"), method("throw"), method("return"), it;
};
var __yieldStar = (value) => {
  var obj = value[__knownSymbol("asyncIterator")], isAwait = false, method, it = {};
  if (obj == null) {
    obj = value[__knownSymbol("iterator")]();
    method = (k) => it[k] = (x) => obj[k](x);
  } else {
    obj = obj.call(value);
    method = (k) => it[k] = (v) => {
      if (isAwait) {
        isAwait = false;
        if (k === "throw") throw v;
        return v;
      }
      isAwait = true;
      return {
        done: false,
        value: new __await(new Promise((resolve) => {
          var x = obj[k](v);
          if (!(x instanceof Object)) __typeError("Object expected");
          resolve(x);
        }), 1)
      };
    };
  }
  return it[__knownSymbol("iterator")] = () => it, method("next"), "throw" in obj ? method("throw") : it.throw = (x) => {
    throw x;
  }, "return" in obj && method("return"), it;
};
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { isBrowser } from "../utils/env_aware_utils.js";
import { logger } from "../utils/logger.js";
import { Gemini, geminiInitParams } from "./google_llm.js";
const APIGEE_PROXY_URL_ENV_VARIABLE_NAME = "APIGEE_PROXY_URL";
class ApigeeLlm extends Gemini {
  constructor({
    model,
    proxyUrl,
    apiKey,
    vertexai,
    location,
    project,
    headers
  }) {
    var _a;
    if (!validateModel(model)) {
      throw new Error(
        "Model ".concat(model, " is not a valid Apigee model, expected apigee/[<provider>/][<version>/]<model_id>")
      );
    }
    super(__spreadProps(__spreadValues({}, apigeeToGeminiInitParams({ model, vertexai, project, location, apiKey })), {
      headers
    }));
    this.proxyUrl = proxyUrl != null ? proxyUrl : "";
    if (!isBrowser() && !this.proxyUrl) {
      this.proxyUrl = (_a = process.env[APIGEE_PROXY_URL_ENV_VARIABLE_NAME]) != null ? _a : "";
    }
    if (!this.proxyUrl) {
      throw new Error(
        "Proxy URL must be provided via the constructor or ".concat(APIGEE_PROXY_URL_ENV_VARIABLE_NAME, " environment variable.")
      );
    }
  }
  getHttpOptions() {
    const opts = super.getHttpOptions();
    opts.baseUrl = this.proxyUrl;
    return opts;
  }
  getLiveHttpOptions() {
    const opts = super.getLiveHttpOptions();
    opts.baseUrl = this.proxyUrl;
    return opts;
  }
  identifyApiVersion() {
    const modelTrimmed = this.model.startsWith("apigee/") ? this.model.substring("apigee/".length) : this.model;
    const components = modelTrimmed.split("/");
    if (components.length === 3) {
      return components[1];
    }
    if (components.length === 2) {
      if (components[0] != "vertex_ai" && components[0] != "gemini" && components[0].startsWith("v")) {
        return components[0];
      }
    }
    return this.vertexai ? "v1beta1" : "v1alpha";
  }
  get liveApiVersion() {
    if (!this._apigeeLiveApiVersion) {
      this._apigeeLiveApiVersion = this.identifyApiVersion();
    }
    return this._apigeeLiveApiVersion;
  }
  generateContentAsync(llmRequest, stream = false, abortSignal) {
    return __asyncGenerator(this, null, function* () {
      var _a;
      const modelToUse = (_a = llmRequest.model) != null ? _a : this.model;
      llmRequest.model = getModelId(modelToUse);
      yield* __yieldStar(super.generateContentAsync(llmRequest, stream, abortSignal));
    });
  }
  async connect(llmRequest) {
    var _a;
    const modelToUse = (_a = llmRequest.model) != null ? _a : this.model;
    llmRequest.model = getModelId(modelToUse);
    return super.connect(llmRequest);
  }
}
/**
 * A list of model name patterns that are supported by this LLM.
 *
 * @returns A list of supported models.
 */
ApigeeLlm.supportedModels = [
  /apigee\/.*/
];
function apigeeToGeminiInitParams({
  model,
  vertexai,
  project,
  location,
  apiKey
}) {
  var _a;
  const params = geminiInitParams({ model, vertexai, project, location, apiKey });
  params.vertexai = params.vertexai || ((_a = params.model) == null ? void 0 : _a.startsWith("apigee/vertex_ai/"));
  if (params.vertexai) {
    return params;
  }
  if (!params.apiKey) {
    logger.warn(
      'No API key provided when using a Gemini model, using a fake key "-".'
    );
    params.apiKey = "-";
  }
  return params;
}
function getModelId(model) {
  if (!validateModel(model)) {
    throw new Error(
      "Model ".concat(model, " is not a valid Apigee model, expected apigee/[<provider>/][<version>/]<model_id>")
    );
  }
  const components = model.split("/");
  return components[components.length - 1];
}
function validateModel(model) {
  const validProviders = ["vertex_ai", "gemini"];
  if (!model.startsWith("apigee/")) {
    return false;
  }
  const modelPart = model.substring("apigee/".length);
  if (modelPart.length === 0) {
    return false;
  }
  const components = modelPart.split("/", -1);
  if (components[components.length - 1].length === 0) {
    return false;
  }
  if (components.length == 1) {
    return true;
  }
  if (components.length == 2) {
    if (validProviders.includes(components[0])) {
      return true;
    }
    return components[0].startsWith("v");
  }
  if (components.length == 3) {
    if (!validProviders.includes(components[0])) {
      return false;
    }
    return components[1].startsWith("v");
  }
  return false;
}
export {
  ApigeeLlm
};
