// @google/adk/dist/web/models/routed_llm.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};
var __await = function(promise, isYieldStar) {
  this[0] = promise;
  this[1] = isYieldStar;
};
var __asyncGenerator = (__this, __arguments, generator) => {
  var resume = (k, v, yes, no) => {
    try {
      var x = generator[k](v), isAwait = (v = x.value) instanceof __await, done = x.done;
      Promise.resolve(isAwait ? v[0] : v).then((y) => isAwait ? resume(k === "return" ? k : "next", v[1] ? { done: y.done, value: y.value } : y, yes, no) : yes({ value: y, done })).catch((e) => resume("throw", e, yes, no));
    } catch (e) {
      no(e);
    }
  }, method = (k) => it[k] = (x) => new Promise((yes, no) => resume(k, x, yes, no)), it = {};
  return generator = generator.apply(__this, __arguments), it[__knownSymbol("asyncIterator")] = () => it, method("next"), method("throw"), method("return"), it;
};
var __yieldStar = (value) => {
  var obj = value[__knownSymbol("asyncIterator")], isAwait = false, method, it = {};
  if (obj == null) {
    obj = value[__knownSymbol("iterator")]();
    method = (k) => it[k] = (x) => obj[k](x);
  } else {
    obj = obj.call(value);
    method = (k) => it[k] = (v) => {
      if (isAwait) {
        isAwait = false;
        if (k === "throw") throw v;
        return v;
      }
      isAwait = true;
      return {
        done: false,
        value: new __await(new Promise((resolve) => {
          var x = obj[k](v);
          if (!(x instanceof Object)) __typeError("Object expected");
          resolve(x);
        }), 1)
      };
    };
  }
  return it[__knownSymbol("iterator")] = () => it, method("next"), "throw" in obj ? method("throw") : it.throw = (x) => {
    throw x;
  }, "return" in obj && method("return"), it;
};
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { BaseLlm } from "./base_llm.js";
import { experimental } from "../utils/experimental.js";
import { runWithRouting } from "../utils/failover_utils.js";
import { logger } from "../utils/logger.js";
let RoutedLlm = class extends BaseLlm {
  constructor({
    models,
    router,
    modelName = "routed-llm"
  }) {
    const modelsMap = Array.isArray(models) ? Object.fromEntries(models.map((m) => [m.model, m])) : models;
    const modelNames = Object.entries(modelsMap).map(
      ([name, model]) => "".concat(name, " (").concat(model.model, ")")
    );
    const computedName = "RoutedLlm[".concat(modelNames.join(", "), "]");
    super({ model: modelName === "routed-llm" ? computedName : modelName });
    this.models = modelsMap;
    this.router = router;
  }
  /**
   * Generates content by delegating to the selected model.
   */
  generateContentAsync(llmRequest, stream) {
    return __asyncGenerator(this, null, function* () {
      logger.info("Routing request via ".concat(this.model));
      yield* __yieldStar(runWithRouting(
        this.models,
        llmRequest,
        this.router,
        (model) => model.generateContentAsync(llmRequest, stream)
      ));
    });
  }
  /**
   * Creates a live connection to the LLM by delegating to the selected model.
   * This live connection cannot be switched mid-stream, it is tied to the model
   * selected at the time of connection.
   */
  async connect(llmRequest) {
    const generator = runWithRouting(
      this.models,
      llmRequest,
      this.router,
      (model) => model.connect(llmRequest)
    );
    const result = await generator.next();
    if (result.done || result.value === void 0) {
      throw new Error("Failed to establish connection: No connection yielded.");
    }
    return result.value;
  }
};
RoutedLlm = __decorateClass([
  experimental
], RoutedLlm);
export {
  RoutedLlm
};
