// @google/adk-devtools/dist/esm/server/adk_api_client.js

const Module = await import('node:module'); 
const require = /* @__PURE__ */ (Module?.default?.createRequire ?? Module?.createRequire)(import.meta.url);
const __filename = /* @__PURE__ */ (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = /* @__PURE__ */ (await import("node:path")).dirname(__filename);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { createUserContent } from "@google/genai";
class AdkApiClient {
  constructor(config) {
    this.backendUrl = config.backendUrl;
  }
  async listApps() {
    return this.fetch(`${this.backendUrl}/list-apps`);
  }
  async getSession(params) {
    return this.fetch(
      `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json"
        }
      }
    );
  }
  async createSession(params) {
    let url = `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions`;
    if (params.sessionId) {
      url += `/${params.sessionId}`;
    }
    return this.fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        state: params.state
      })
    });
  }
  async deleteSession(params) {
    return this.fetch(
      `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}`,
      {
        method: "DELETE",
        headers: {
          "Content-Type": "application/json"
        }
      }
    );
  }
  async listAllSessions(params) {
    const apps = await this.listApps();
    return Promise.all(
      apps.map(
        (appName) => this.listSessions({ appName, userId: params.userId })
      )
    ).then((sessions) => sessions.flat());
  }
  async listSessions(params) {
    const sessions = await this.fetch(
      `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json"
        }
      }
    );
    if ("sessions" in sessions) {
      return sessions.sessions;
    }
    return sessions;
  }
  async *runAsync(params) {
    const response = await fetch(`${this.backendUrl}/run_sse`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        appName: params.appName,
        userId: params.userId,
        sessionId: params.sessionId,
        streaming: params.streaming,
        stateDelta: params.stateDelta,
        newMessage: typeof params.newMessage === "string" ? createUserContent(params.newMessage) : params.newMessage
      })
    });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(
        error.error || `Request failed with status ${response.status}`
      );
    }
    if (!response.body) {
      return;
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";
    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n\n");
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        if (line.startsWith("data: ")) {
          const data = JSON.parse(line.substring("data: ".length));
          if (data.error) {
            throw new Error(data.error);
          }
          yield data;
        }
      }
    }
  }
  async listArtifacts(params) {
    return this.fetch(
      `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}/artifacts`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json"
        }
      }
    );
  }
  async loadArtifact(params) {
    let url = `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}/artifacts/${params.artifactName}`;
    if (params.version !== void 0) {
      url += `/versions/${params.version}`;
    }
    return this.fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json"
      }
    });
  }
  async listArtifactVersions(params) {
    const url = `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}/artifacts/${params.artifactName}/versions`;
    return this.fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json"
      }
    });
  }
  async deleteArtifact(params) {
    const url = `${this.backendUrl}/apps/${params.appName}/users/${params.userId}/sessions/${params.sessionId}/artifacts/${params.artifactName}`;
    return this.fetch(url, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json"
      }
    });
  }
  async fetch(url, options) {
    const response = await fetch(url, options);
    if (!response.ok) {
      let error;
      try {
        error = await response.json();
      } catch (_e) {
      }
      throw new Error(
        (error == null ? void 0 : error.error) ?? `Request failed with status ${response.status}`
      );
    }
    return response.json();
  }
}
export {
  AdkApiClient
};
