// @google/adk-devtools/dist/esm/server/adk_api_server.js

const Module = await import('node:module'); 
const require = /* @__PURE__ */ (Module?.default?.createRequire ?? Module?.createRequire)(import.meta.url);
const __filename = /* @__PURE__ */ (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = /* @__PURE__ */ (await import("node:path")).dirname(__filename);
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  getFunctionCalls,
  getFunctionResponses,
  InMemoryArtifactService,
  InMemoryMemoryService,
  InMemorySessionService,
  LogLevel,
  Runner,
  StreamingMode,
  toA2a
} from "@google/adk";
import { trace } from "@opentelemetry/api";
import { SimpleSpanProcessor } from "@opentelemetry/sdk-trace-base";
import cors from "cors";
import express from "express";
import * as path from "node:path";
import { AgentLoader } from "../utils/agent_loader.js";
import { AdkLogger } from "../utils/logger.js";
import {
  ApiServerSpanExporter,
  hrTimeToNanoseconds,
  InMemoryExporter,
  setupTelemetry
} from "../utils/telemetry_utils.js";
import { getAgentGraphAsDot } from "./agent_graph.js";
class AdkApiServer {
  constructor(options) {
    this.runnerCache = {};
    this.traceDict = {};
    this.sessionTraceDict = {};
    this.host = options.host ?? "localhost";
    this.port = options.port ?? 0;
    this.sessionService = options.sessionService ?? new InMemorySessionService();
    this.memoryService = options.memoryService ?? new InMemoryMemoryService();
    this.artifactService = options.artifactService ?? new InMemoryArtifactService();
    this.agentLoader = options.agentLoader ?? new AgentLoader(options.agentsDir, options.agentFileLoadOptions);
    this.serveDebugUI = options.serveDebugUI ?? false;
    this.allowOrigins = options.allowOrigins;
    this.otelToCloud = options.otelToCloud ?? false;
    this.registerProcessors = options.registerProcessors;
    this.memoryExporter = new InMemoryExporter(this.sessionTraceDict);
    this.logger = options.logger ?? new AdkLogger({
      label: "ADK API Server",
      timestamp: true,
      colorize: { level: true },
      printFormat: (info) => {
        return `${info.level}: [${info.label}] ${info.timestamp} ${info.message}`;
      }
    });
    this.logger.setLogLevel(options.logLevel ?? LogLevel.INFO);
    this.a2a = options.a2a ?? false;
    this.app = express();
  }
  get url() {
    if (this.server) {
      const address = this.server.address();
      if (address && typeof address !== "string") {
        return `http://${this.host}:${address.port}`;
      }
    }
    return `http://${this.host}:${this.port}`;
  }
  async setupTelemetry() {
    const internalExporters = [
      new SimpleSpanProcessor(new ApiServerSpanExporter(this.traceDict)),
      new SimpleSpanProcessor(this.memoryExporter)
    ];
    await setupTelemetry(this.otelToCloud, internalExporters);
    if (this.registerProcessors) {
      const tracerProvider = trace.getTracerProvider();
      this.registerProcessors(tracerProvider);
    }
  }
  async initA2A() {
    const appNames = await this.agentLoader.listAgents();
    for (const appName of appNames) {
      const agentFile = await this.agentLoader.getAgentFile(appName);
      const agent = await agentFile.load();
      await toA2a(agent, {
        protocol: "http",
        host: this.host,
        port: this.port,
        basePath: `/a2a/${appName}`,
        sessionService: this.sessionService,
        memoryService: this.memoryService,
        artifactService: this.artifactService,
        app: this.app
      });
    }
  }
  async init() {
    const app = this.app;
    await this.setupTelemetry();
    if (this.serveDebugUI) {
      app.get("/", (req, res) => {
        res.redirect("/dev-ui");
      });
      app.use(
        "/dev-ui",
        express.static(path.join(__dirname, "../browser"), {
          setHeaders: (res, path2) => {
            if (path2.endsWith(".js")) {
              res.setHeader("Content-Type", "text/javascript");
            }
          }
        })
      );
    }
    if (this.allowOrigins) {
      app.use(
        cors({
          origin: this.allowOrigins
        })
      );
    }
    app.use(express.urlencoded({ limit: "50mb", extended: true }));
    app.use(
      express.json({
        limit: "50mb"
      })
    );
    app.use((req, res, next) => {
      this.logger.info(`${req.method} ${req.originalUrl}`);
      next();
    });
    app.get("/list-apps", async (req, res) => {
      try {
        const apps = await this.agentLoader.listAgents();
        res.json(apps);
      } catch (e) {
        const error = `Failed to list apps: ${e}`;
        res.status(500).json({ error });
        this.logger.error(error);
        return;
      }
    });
    app.get("/debug/trace/:eventId", (req, res) => {
      try {
        const eventId = req.params["eventId"];
        const eventDict = this.traceDict[eventId];
        if (!eventDict) {
          return res.status(404).json({ error: "Trace not found" });
        }
        return res.json(eventDict);
      } catch (e) {
        const error = `Failed to get trace: ${e}`;
        res.status(500).json({ error });
        this.logger.error(error);
        return;
      }
    });
    app.get(
      "/debug/trace/session/:sessionId",
      (req, res) => {
        try {
          const sessionId = req.params["sessionId"];
          const spans = this.memoryExporter.getFinishedSpans(sessionId);
          if (spans.length === 0) {
            return res.json([]);
          }
          const spanData = spans.map((span) => {
            var _a;
            return {
              name: span.name,
              span_id: span.spanContext().spanId,
              trace_id: span.spanContext().traceId,
              start_time: hrTimeToNanoseconds(span.startTime),
              end_time: hrTimeToNanoseconds(span.endTime),
              attributes: { ...span.attributes },
              parent_span_id: ((_a = span.parentSpanContext) == null ? void 0 : _a.spanId) || null
            };
          });
          return res.json(spanData);
        } catch (e) {
          const error = `Failed to get trace: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
          return;
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/events/:eventId/graph",
      async (req, res) => {
        try {
          var _stack = [];
          try {
            const appName = req.params["appName"];
            const userId = req.params["userId"];
            const sessionId = req.params["sessionId"];
            const eventId = req.params["eventId"];
            const session = await this.sessionService.getSession({
              appName,
              userId,
              sessionId
            });
            if (!session) {
              res.status(404).json({ error: `Session not found: ${sessionId}` });
              return;
            }
            const sessionEvents = session.events || [];
            const event = sessionEvents.find((e) => e.id === eventId);
            if (!event) {
              res.status(404).json({ error: `Event not found: ${eventId}` });
              return;
            }
            const functionCalls = getFunctionCalls(event);
            const functionResponses = getFunctionResponses(event);
            const agentFile = __using(_stack, await this.agentLoader.getAgentFile(appName), true);
            const rootAgent = await agentFile.load();
            if (functionCalls.length > 0) {
              const functionCallHighlights = [];
              for (const functionCall of functionCalls) {
                functionCallHighlights.push([event.author, functionCall.name]);
              }
              return res.send({
                dotSrc: await getAgentGraphAsDot(
                  rootAgent,
                  functionCallHighlights
                )
              });
            }
            if (functionResponses.length > 0) {
              const functionCallHighlights = [];
              for (const functionResponse of functionResponses) {
                functionCallHighlights.push([
                  functionResponse.name,
                  event.author
                ]);
              }
              return res.send({
                dotSrc: await getAgentGraphAsDot(
                  rootAgent,
                  functionCallHighlights
                )
              });
            }
            return res.send({
              dotSrc: await getAgentGraphAsDot(rootAgent, [[event.author, ""]])
            });
          } catch (_) {
            var _error = _, _hasError = true;
          } finally {
            var _promise = __callDispose(_stack, _error, _hasError);
            _promise && await _promise;
          }
        } catch (e) {
          const error = `Failed to get agent graph: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
          return;
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const session = await this.sessionService.getSession({
            appName,
            userId,
            sessionId
          });
          if (!session) {
            res.status(404).json({ error: `Session not found: ${sessionId}` });
            return;
          }
          res.json(session);
        } catch (e) {
          const error = `Failed to get session: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessions = await this.sessionService.listSessions({
            appName,
            userId
          });
          res.json(sessions);
        } catch (e) {
          const error = `Failed to list sessions: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.post(
      "/apps/:appName/users/:userId/sessions/:sessionId",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const state = req.body["state"] || {};
          const existingSession = await this.sessionService.getSession({
            appName,
            userId,
            sessionId
          });
          if (existingSession) {
            res.status(400).json({ error: `Session already exists: ${sessionId}` });
            return;
          }
          const createdSession = await this.sessionService.createSession({
            appName,
            userId,
            state,
            sessionId
          });
          res.json(createdSession);
        } catch (e) {
          const error = `Failed to create session: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.post(
      "/apps/:appName/users/:userId/sessions",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const state = req.body["state"] || {};
          const createdSession = await this.sessionService.createSession({
            appName,
            userId,
            state
          });
          res.json(createdSession);
        } catch (e) {
          const error = `Failed to create session: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.delete(
      "/apps/:appName/users/:userId/sessions/:sessionId",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const session = await this.sessionService.getSession({
            appName,
            userId,
            sessionId
          });
          if (!session) {
            res.status(404).json({ error: `Session not found: ${sessionId}` });
            return;
          }
          await this.sessionService.deleteSession({
            appName,
            userId,
            sessionId
          });
          res.status(204).json({});
        } catch (e) {
          const error = `Failed to delete session: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts/:artifactName",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactName = req.params["artifactName"];
          const artifact = await this.artifactService.loadArtifact({
            appName,
            userId,
            sessionId,
            filename: artifactName
          });
          if (!artifact) {
            res.status(404).json({ error: `Artifact not found: ${artifactName}` });
            return;
          }
          res.json(artifact);
        } catch (e) {
          const error = `Failed to load artifact: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts/:artifactName/versions/:versionId",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactName = req.params["artifactName"];
          const versionId = req.params["versionId"];
          const artifact = await this.artifactService.loadArtifact({
            appName,
            userId,
            sessionId,
            filename: artifactName,
            version: parseInt(versionId, 10)
          });
          if (!artifact) {
            res.status(404).json({ error: `Artifact not found: ${artifactName}` });
            return;
          }
          res.json(artifact);
        } catch (e) {
          const error = `Failed to load artifact version: ${e.message}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactKeys = await this.artifactService.listArtifactKeys({
            appName,
            userId,
            sessionId
          });
          res.json(artifactKeys);
        } catch (e) {
          const error = `Failed to list artifacts: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.get(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts/:artifactName/versions",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactName = req.params["artifactName"];
          const artifactVersions = await this.artifactService.listVersions({
            appName,
            userId,
            sessionId,
            filename: artifactName
          });
          res.json(artifactVersions);
        } catch (e) {
          const error = `Failed to list artifact versions: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.delete(
      "/apps/:appName/users/:userId/sessions/:sessionId/artifacts/:artifactName",
      async (req, res) => {
        try {
          const appName = req.params["appName"];
          const userId = req.params["userId"];
          const sessionId = req.params["sessionId"];
          const artifactName = req.params["artifactName"];
          await this.artifactService.deleteArtifact({
            appName,
            userId,
            sessionId,
            filename: artifactName
          });
          res.status(204).json({});
        } catch (e) {
          const error = `Failed to delete artifact: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    );
    app.post(
      "/apps/:appName/eval_sets/:evalSetId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get("/apps/:appName/eval_sets", (req, res) => {
      return res.status(501).json({ error: "Not implemented" });
    });
    app.post(
      "/apps/:appName/eval_sets/:evalSetId/add_session",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get(
      "/apps/:appName/eval_sets/:evalSetId/evals",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get(
      "/apps/:appName/eval_sets/:evalSetId/evals/:evalCaseId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.put(
      "/apps/:appName/eval_sets/:evalSetId/evals/:evalCaseId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.delete(
      "/apps/:appName/eval_sets/:evalSetId/evals/:evalCaseId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.post(
      "/apps/:appName/eval_sets/:evalSetId/run_eval",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get(
      "/apps/:appName/eval_results/:evalResultId",
      (req, res) => {
        return res.status(501).json({ error: "Not implemented" });
      }
    );
    app.get("/apps/:appName/eval_results", (req, res) => {
      return res.status(501).json({ error: "Not implemented" });
    });
    app.get("/apps/:appName/eval_metrics", (req, res) => {
      return res.status(501).json({ error: "Not implemented" });
    });
    app.post("/run", async (req, res) => {
      const { appName, userId, sessionId, newMessage, stateDelta } = req.body;
      const session = await this.sessionService.getSession({
        appName,
        userId,
        sessionId
      });
      if (!session) {
        res.status(404).json({ error: `Session not found: ${sessionId}` });
        return;
      }
      try {
        var _stack = [];
        try {
          const agentFile = __using(_stack, await this.agentLoader.getAgentFile(appName), true);
          const agent = await agentFile.load();
          const runner = await this.getRunner(agent, appName);
          const events = [];
          for await (const e of runner.runAsync({
            userId,
            sessionId,
            newMessage,
            stateDelta
          })) {
            events.push(e);
          }
          res.json(events);
        } catch (_) {
          var _error = _, _hasError = true;
        } finally {
          var _promise = __callDispose(_stack, _error, _hasError);
          _promise && await _promise;
        }
      } catch (e) {
        const error = `Failed to run agent: ${e}`;
        res.status(500).json({ error });
        this.logger.error(error);
      }
    });
    app.post("/run_sse", async (req, res) => {
      const { appName, userId, sessionId, newMessage, streaming, stateDelta } = req.body;
      const session = await this.sessionService.getSession({
        appName,
        userId,
        sessionId
      });
      if (!session) {
        const error = `Session not found: ${sessionId}`;
        res.status(404).json({ error });
        this.logger.error(error);
        return;
      }
      try {
        var _stack = [];
        try {
          const agentFile = __using(_stack, await this.agentLoader.getAgentFile(appName), true);
          const agent = await agentFile.load();
          const runner = await this.getRunner(agent, appName);
          res.setHeader("Cache-Control", "no-cache");
          res.setHeader("Content-Type", "text/event-stream");
          res.setHeader("Access-Control-Allow-Origin", "*");
          res.setHeader("Connection", "keep-alive");
          res.flushHeaders();
          for await (const event of runner.runAsync({
            userId,
            sessionId,
            newMessage,
            runConfig: {
              streamingMode: streaming ? StreamingMode.SSE : StreamingMode.NONE
            },
            stateDelta
          })) {
            res.write(`data: ${JSON.stringify(event)}

`);
          }
          res.end();
        } catch (_) {
          var _error = _, _hasError = true;
        } finally {
          var _promise = __callDispose(_stack, _error, _hasError);
          _promise && await _promise;
        }
      } catch (e) {
        if (res.headersSent) {
          const error = e.message;
          res.end(`data: ${JSON.stringify({ error })}

`);
          this.logger.error(error);
        } else {
          const error = `Failed to run agent: ${e}`;
          res.status(500).json({ error });
          this.logger.error(error);
        }
      }
    });
  }
  async start() {
    await this.init();
    return new Promise((resolve, reject) => {
      this.server = this.app.listen(this.port, async () => {
        try {
          if (this.a2a) {
            await this.initA2A();
          }
          console.log(`
+-----------------------------------------------------------------------------+
| ADK API Server started                                                      |
|                                                                             |
| For local testing, access at ${this.url}.${"".padStart(39 - this.url.length)}       |
+-----------------------------------------------------------------------------+`);
          resolve();
        } catch (error) {
          this.logger.error("Error during AdkApiServer startup:", error);
          reject(error);
        }
      });
      this.server.on("error", (err) => {
        if (err.code === "EADDRINUSE") {
          const error = new Error();
          error.cause = err;
          error.message = `Port ${this.port} is already in use`;
          reject(error);
        } else {
          reject(err);
        }
      });
    });
  }
  stop() {
    if (!this.server) {
      return Promise.resolve();
    }
    return new Promise((resolve, reject) => {
      this.server.close((err) => {
        if (err) {
          reject(err);
          return;
        }
        console.log(`
+-----------------------------------------------------------------------------+
| ADK API Server stopped                                                      |
+-----------------------------------------------------------------------------+`);
        resolve();
      });
    });
  }
  async getRunner(agent, appName) {
    if (!(appName in this.runnerCache)) {
      this.runnerCache[appName] = new Runner({
        appName,
        agent,
        memoryService: this.memoryService,
        sessionService: this.sessionService,
        artifactService: this.artifactService
      });
    }
    return this.runnerCache[appName];
  }
}
export {
  AdkApiServer
};
