// @google/adk/dist/cjs/auth/auth_preprocessor.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var auth_preprocessor_exports = {};
__export(auth_preprocessor_exports, {
  AUTH_PREPROCESSOR: () => AUTH_PREPROCESSOR,
  AuthPreprocessor: () => AuthPreprocessor
});
module.exports = __toCommonJS(auth_preprocessor_exports);
var import_functions = require("../agents/functions.js");
var import_llm_agent = require("../agents/llm_agent.js");
var import_base_llm_processor = require("../agents/processors/base_llm_processor.js");
var import_readonly_context = require("../agents/readonly_context.js");
var import_event = require("../events/event.js");
var import_state = require("../sessions/state.js");
var import_auth_handler = require("./auth_handler.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const TOOLSET_AUTH_CREDENTIAL_ID_PREFIX = "_adk_toolset_auth_";
async function storeAuthAndCollectResumeTargets(events, authFcIds, authResponses, state) {
  const requestedAuthConfigById = {};
  for (const event of events) {
    const eventFunctionCalls = (0, import_event.getFunctionCalls)(event);
    for (const functionCall of eventFunctionCalls) {
      if (functionCall.id && authFcIds.has(functionCall.id) && functionCall.name === import_functions.REQUEST_EUC_FUNCTION_CALL_NAME) {
        const args = functionCall.args;
        if (args && args.authConfig) {
          requestedAuthConfigById[functionCall.id] = args.authConfig;
        }
      }
    }
  }
  for (const fcId of authFcIds) {
    if (!(fcId in authResponses)) {
      continue;
    }
    const authConfig = authResponses[fcId];
    const requestedAuthConfig = requestedAuthConfigById[fcId];
    if (requestedAuthConfig && requestedAuthConfig.credentialKey) {
      authConfig.credentialKey = requestedAuthConfig.credentialKey;
    }
    await new import_auth_handler.AuthHandler(authConfig).parseAndStoreAuthResponse(state);
  }
  const toolsToResume = /* @__PURE__ */ new Set();
  for (const fcId of authFcIds) {
    const requestedAuthConfig = requestedAuthConfigById[fcId];
    if (!requestedAuthConfig) {
      continue;
    }
    for (const event of events) {
      const eventFunctionCalls = (0, import_event.getFunctionCalls)(event);
      for (const functionCall of eventFunctionCalls) {
        if (functionCall.id === fcId && functionCall.name === import_functions.REQUEST_EUC_FUNCTION_CALL_NAME) {
          const args = functionCall.args;
          if (args && args.functionCallId) {
            if (args.functionCallId.startsWith(TOOLSET_AUTH_CREDENTIAL_ID_PREFIX)) {
              continue;
            }
            toolsToResume.add(args.functionCallId);
          }
        }
      }
    }
  }
  return toolsToResume;
}
class AuthPreprocessor extends import_base_llm_processor.BaseLlmRequestProcessor {
  async *runAsync(invocationContext) {
    const agent = invocationContext.agent;
    if (!(0, import_llm_agent.isLlmAgent)(agent)) {
      return;
    }
    const events = invocationContext.session.events;
    if (!events || events.length === 0) {
      return;
    }
    let lastEventWithContent = null;
    for (let i = events.length - 1; i >= 0; i--) {
      const event = events[i];
      if (event.content !== void 0) {
        lastEventWithContent = event;
        break;
      }
    }
    if (!lastEventWithContent || lastEventWithContent.author !== "user") {
      return;
    }
    const responses = (0, import_event.getFunctionResponses)(lastEventWithContent);
    if (!responses || responses.length === 0) {
      return;
    }
    const authFcIds = /* @__PURE__ */ new Set();
    const authResponses = {};
    for (const functionCallResponse of responses) {
      if (functionCallResponse.name !== import_functions.REQUEST_EUC_FUNCTION_CALL_NAME) {
        continue;
      }
      if (functionCallResponse.id) {
        authFcIds.add(functionCallResponse.id);
        authResponses[functionCallResponse.id] = functionCallResponse.response;
      }
    }
    if (authFcIds.size === 0) {
      return;
    }
    const state = new import_state.State(invocationContext.session.state);
    const toolsToResume = await storeAuthAndCollectResumeTargets(
      events,
      authFcIds,
      authResponses,
      state
    );
    if (toolsToResume.size === 0) {
      return;
    }
    for (let i = events.length - 2; i >= 0; i--) {
      const event = events[i];
      const functionCalls = (0, import_event.getFunctionCalls)(event);
      if (!functionCalls || functionCalls.length === 0) {
        continue;
      }
      const hasMatchingCall = functionCalls.some(
        (call) => call.id ? toolsToResume.has(call.id) : false
      );
      if (hasMatchingCall) {
        const canonicalTools = await agent.canonicalTools(
          new import_readonly_context.ReadonlyContext(invocationContext)
        );
        const toolsDict = {};
        for (const tool of canonicalTools) {
          toolsDict[tool.name] = tool;
        }
        const functionResponseEvent = await (0, import_functions.handleFunctionCallsAsync)({
          invocationContext,
          functionCallEvent: event,
          toolsDict,
          beforeToolCallbacks: agent.canonicalBeforeToolCallbacks,
          afterToolCallbacks: agent.canonicalAfterToolCallbacks,
          filters: toolsToResume
        });
        if (functionResponseEvent) {
          yield functionResponseEvent;
        }
        return;
      }
    }
  }
}
const AUTH_PREPROCESSOR = new AuthPreprocessor();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AUTH_PREPROCESSOR,
  AuthPreprocessor
});
