// @google/adk-devtools/dist/cjs/integration/agent_registry.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var agent_registry_exports = {};
__export(agent_registry_exports, {
  AgentRegistry: () => AgentRegistry
});
module.exports = __toCommonJS(agent_registry_exports);
var import_adk = require("@google/adk");
var path = __toESM(require("node:path"), 1);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const BUILTIN_TOOLS = [
  "exit_loop",
  "google_search",
  "url_context",
  "google_maps_grounding"
];
class AgentRegistry {
  constructor(integrationRegistry) {
    this.agents = /* @__PURE__ */ new Map();
    this.configs = /* @__PURE__ */ new Map();
    this.instantiating = /* @__PURE__ */ new Set();
    this.integrationRegistry = integrationRegistry;
  }
  registerAgent(name, agent) {
    this.agents.set(name, agent);
  }
  getAgent(name) {
    if (this.agents.has(name)) {
      return this.agents.get(name);
    }
    if (this.configs.has(name)) {
      return this.instantiateAgent(name, this.configs.get(name));
    }
    return void 0;
  }
  // Returns the root agent based on the last directory name.
  // if there are somehow more than one root agent in a directory, an arbitrary one is returned
  getRootAgentByShortName(shortName) {
    for (const [name, config] of this.configs) {
      if (path.basename(path.dirname(name)) === shortName && config.isRootAgent) {
        return this.getAgent(name);
      }
    }
    return void 0;
  }
  registerAgentConfig(name, config) {
    this.configs.set(name, config);
  }
  findToolOrThrow(name) {
    const tool = this.integrationRegistry.getTool(name);
    if (!tool) {
      console.log("Tool not found in registry", name);
      throw new Error(`Tool ${name} not found in registry`);
    }
    return tool;
  }
  instantiateAgent(name, config) {
    var _a, _b, _c, _d, _e;
    console.log(
      "Instantiating",
      name,
      "of class",
      config.agentClass ?? "LlmAgent"
    );
    if (this.instantiating.has(name)) {
      throw new Error(`Circular dependency detected for agent ${name}`);
    }
    this.instantiating.add(name);
    try {
      const beforeAgentCallbacks = (_a = config.beforeAgentCallbacks) == null ? void 0 : _a.map(
        (callbackInfo) => {
          const callback = this.integrationRegistry.getBeforeAgentCallback(
            callbackInfo.name
          );
          if (!callback) {
            throw new Error(
              `BeforeAgentCallback ${callbackInfo.name} not found in registry`
            );
          }
          return callback;
        }
      );
      const afterAgentCallbacks = (_b = config.afterAgentCallbacks) == null ? void 0 : _b.map(
        (callbackInfo) => {
          const callback = this.integrationRegistry.getAfterAgentCallback(
            callbackInfo.name
          );
          if (!callback) {
            throw new Error(
              `AfterAgentCallback ${callbackInfo.name} not found in registry`
            );
          }
          return callback;
        }
      );
      const subAgents = (_c = config.subAgents) == null ? void 0 : _c.map((ref) => {
        const subAgent = this.getAgent(ref.configPath);
        if (!subAgent) {
          throw new Error(
            `SubAgent ${ref.configPath} not found in registry (referenced by ${name})`
          );
        }
        return subAgent;
      });
      const tools = (_d = config.tools) == null ? void 0 : _d.map((toolConfig) => {
        if (BUILTIN_TOOLS.includes(toolConfig.name)) {
          return void 0;
        }
        if (toolConfig.name == "LongRunningFunctionTool") {
          const args = toolConfig.args;
          const subTool = this.findToolOrThrow(args.func);
          return subTool;
        }
        if (toolConfig.name === "AgentTool") {
          const args = toolConfig.args;
          const subAgent = this.getAgent(args.agent.configPath);
          if (!subAgent) {
            throw new Error(
              `Agent ${args.agent.configPath} not found in registry (referenced by AgentTool in ${name})`
            );
          }
          return new import_adk.AgentTool({ agent: subAgent });
        }
        if (toolConfig.name === "McpToolset") {
          const args = toolConfig.args;
          if (args.stdioConnectionParams.type !== "StdioConnectionParams") {
            args.stdioConnectionParams.type = "StdioConnectionParams";
          }
          const toolset = new import_adk.MCPToolset(
            args.stdioConnectionParams,
            args.toolFilter
          );
          return toolset;
        }
        return this.findToolOrThrow(toolConfig.name);
      }).filter((tool) => tool !== void 0);
      const options = {
        name: config.name,
        model: config.model,
        description: config.description,
        instruction: config.instruction,
        beforeAgentCallback: beforeAgentCallbacks ?? [],
        afterAgentCallback: afterAgentCallbacks ?? [],
        subAgents: subAgents ?? [],
        tools: tools ?? [],
        disallowTransferToParent: config.disallowTransferToParent === "true",
        disallowTransferToPeers: config.disallowTransferToPeers === "true",
        temperature: (_e = config.generateContentConfig) == null ? void 0 : _e.temperature
      };
      let agent;
      switch (config.agentClass) {
        case "LoopAgent":
          agent = new import_adk.LoopAgent({
            ...options,
            maxIterations: config.maxIterations ? parseInt(config.maxIterations, 10) : void 0
          });
          break;
        case "ParallelAgent":
          agent = new import_adk.ParallelAgent(options);
          break;
        case "SequentialAgent":
          agent = new import_adk.SequentialAgent(options);
          break;
        case "LlmAgent":
        default:
          agent = new import_adk.LlmAgent(options);
          break;
      }
      this.registerAgent(name, agent);
      return agent;
    } finally {
      this.instantiating.delete(name);
    }
  }
  summary() {
    return `${this.configs.size} configs, ${this.agents.size} instantiated agents`;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AgentRegistry
});
