// @google/adk-devtools/dist/cjs/integration/integration_registry.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var integration_registry_exports = {};
__export(integration_registry_exports, {
  IntegrationRegistry: () => IntegrationRegistry
});
module.exports = __toCommonJS(integration_registry_exports);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class IntegrationRegistry {
  constructor() {
    this.tools = /* @__PURE__ */ new Map();
    this.beforeAgentCallbacks = /* @__PURE__ */ new Map();
    this.afterAgentCallbacks = /* @__PURE__ */ new Map();
    this.plugins = /* @__PURE__ */ new Map();
  }
  summary() {
    return `${this.tools.size} tools, ${this.beforeAgentCallbacks.size} before agent callbacks, ${this.afterAgentCallbacks.size} after agent callbacks, and ${this.plugins.size} plugins.`;
  }
  registerTool(name, tool) {
    this.tools.set(name, tool);
  }
  getTool(name) {
    return this.tools.get(name);
  }
  registerBeforeAgentCallback(name, callback) {
    this.beforeAgentCallbacks.set(name, callback);
  }
  getBeforeAgentCallback(name) {
    return this.beforeAgentCallbacks.get(name);
  }
  registerAfterAgentCallback(name, callback) {
    this.afterAgentCallbacks.set(name, callback);
  }
  getAfterAgentCallback(name) {
    return this.afterAgentCallbacks.get(name);
  }
  registerPlugin(name, plugin) {
    this.plugins.set(name, plugin);
  }
  getPlugin(name) {
    return this.plugins.get(name);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  IntegrationRegistry
});
