// @google/adk-devtools/dist/cjs/integration/test_runner.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var test_runner_exports = {};
__export(test_runner_exports, {
  TestRunner: () => TestRunner
});
module.exports = __toCommonJS(test_runner_exports);
var import_adk = require("@google/adk");
var import_lodash_es = require("lodash-es");
var assert = __toESM(require("node:assert"), 1);
var import_dummy_llm = require("./dummy_llm.js");
var import_replay_plugin = require("./replay_plugin.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const SKIPPED_TESTS = [
  {
    name: "tool/example_tool_001",
    reason: "ExampleTool is not implemented yet."
  },
  { name: "workflow/loop_001", reason: "ExitLoopTool is not implemented yet." },
  {
    name: "core/multi_005",
    reason: "Suspected broken test. Need to re-evaluate."
  },
  {
    name: "tool/long_running_tool_001",
    reason: "Suspected broken test. Need to re-evaluate."
  }
];
class TestRunner {
  constructor(agentRegistry) {
    this.agentRegistry = agentRegistry;
  }
  async run(testInfo, force) {
    if (!force) {
      for (const skip of SKIPPED_TESTS) {
        if (skip.name == testInfo.name) {
          console.log("Skipping test", testInfo.name, "because:", skip.reason);
          return true;
        }
      }
    }
    const agentName = testInfo.spec.agent;
    const agent = this.agentRegistry.getRootAgentByShortName(agentName);
    if (!agent) {
      throw new Error(`Agent ${agentName} not found in registry`);
    }
    const recordings = (0, import_lodash_es.cloneDeep)(testInfo.recordings.recordings);
    const context = { userMessageIndex: 0 };
    injectDummyLlm(agent);
    const replayPlugin = new import_replay_plugin.ReplayPlugin(recordings, context);
    const sessionService = new import_adk.InMemorySessionService();
    const runner = new import_adk.Runner({
      agent,
      sessionService,
      plugins: [replayPlugin],
      appName: "test-runner"
    });
    const userId = "test-user";
    const sessionId = "test-session";
    await sessionService.createSession({
      appName: "test-runner",
      userId,
      sessionId
    });
    const userMessages = testInfo.spec.userMessages;
    for (let i = 0; i < userMessages.length; i++) {
      context.userMessageIndex = i;
      const userMsg = userMessages[i];
      const content = userMessageToContent(userMsg);
      const iterator = runner.runAsync({
        userId,
        sessionId,
        newMessage: content,
        stateDelta: i === 0 ? testInfo.spec.initialState : void 0
      });
      for await (const _ of iterator) {
      }
    }
    const session = await sessionService.getSession({
      appName: "test-runner",
      userId,
      sessionId
    });
    if (!session) {
      throw new Error("Session not found after execution");
    }
    validateSession(session, testInfo.session);
    return false;
  }
}
function injectDummyLlm(agent) {
  if ((0, import_adk.isLlmAgent)(agent)) {
    agent.model = new import_dummy_llm.DummyLlm();
  }
  const subAgents = agent.subAgents;
  if (subAgents && Array.isArray(subAgents)) {
    for (const sub of subAgents) {
      injectDummyLlm(sub);
    }
  }
}
function userMessageToContent(msg) {
  if (msg.content) {
    const content = msg.content;
    content.role = "user";
    return content;
  }
  if (msg.text) {
    return { role: "user", parts: [{ text: msg.text }] };
  }
  throw new Error("Either Content text or content field is required");
}
function validateSession(actual, expected) {
  const actualEvents = actual.events.map(normalizeEvent);
  const expectedEvents = expected.events.map(normalizeEvent);
  assert.deepStrictEqual(actualEvents, expectedEvents);
}
function normalizeEvent(event) {
  const filteredEvent = event;
  filterEventFields(filteredEvent);
  removeEmptyAndUndefinedFields(
    filteredEvent
  );
  return filteredEvent;
}
function removeEmptyAndUndefinedFields(obj) {
  for (const key in obj) {
    if (Object.hasOwn(obj, key)) {
      if (obj[key] === void 0 || obj[key] === null) {
        delete obj[key];
      } else if (Array.isArray(obj[key])) {
        for (let i = 0; i < obj[key].length; i++) {
          removeEmptyAndUndefinedFields(obj[key][i]);
        }
        if (obj[key].length === 0) {
          delete obj[key];
          continue;
        }
      } else if (typeof obj[key] === "object") {
        removeEmptyAndUndefinedFields(obj[key]);
        if (Object.keys(obj[key]).length === 0) {
          delete obj[key];
          continue;
        }
      }
    }
  }
}
function filterEventActionsStateDelta(actions) {
  if (!(actions == null ? void 0 : actions.stateDelta)) {
    return;
  }
  delete actions.stateDelta["_adk_recordings_config"];
  delete actions.stateDelta["_adk_replay_config"];
}
function filterPartFields(part) {
  delete part.thoughtSignature;
  delete part.functionCall;
  delete part.functionResponse;
}
function filterEventFields(event) {
  var _a;
  delete event.id;
  delete event.timestamp;
  delete event.invocationId;
  delete event.longRunningToolIds;
  filterEventActionsStateDelta(event.actions);
  if (event.content) {
    (_a = event.content.parts) == null ? void 0 : _a.forEach(filterPartFields);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TestRunner
});
