// @google/adk/dist/esm/tools/base_toolset.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var _a;
const BASE_TOOLSET_SIGNATURE_SYMBOL = Symbol.for("google.adk.baseToolset");
function isBaseToolset(obj) {
  return typeof obj === "object" && obj !== null && BASE_TOOLSET_SIGNATURE_SYMBOL in obj && obj[BASE_TOOLSET_SIGNATURE_SYMBOL] === true;
}
_a = BASE_TOOLSET_SIGNATURE_SYMBOL;
class BaseToolset {
  constructor(toolFilter, prefix) {
    this.toolFilter = toolFilter;
    this.prefix = prefix;
    this[_a] = true;
  }
  /**
   * Returns whether the tool should be exposed to LLM.
   *
   * @param tool The tool to check.
   * @param context Context used to filter tools available to the agent.
   * @return Whether the tool should be exposed to LLM.
   */
  isToolSelected(tool, context) {
    if (!this.toolFilter) {
      return true;
    }
    if (typeof this.toolFilter === "function") {
      return this.toolFilter(tool, context);
    }
    if (Array.isArray(this.toolFilter)) {
      return this.toolFilter.includes(tool.name);
    }
    return false;
  }
  /**
   * Processes the outgoing LLM request for this toolset. This method will be
   * called before each tool processes the llm request.
   *
   * Use cases:
   * - Instead of let each tool process the llm request, we can let the toolset
   *   process the llm request. e.g. ComputerUseToolset can add computer use
   *   tool to the llm request.
   *
   * @param toolContext The context of the tool.
   * @param llmRequest The outgoing LLM request, mutable this method.
   */
  async processLlmRequest(toolContext, llmRequest) {
  }
}
export {
  BaseToolset,
  isBaseToolset
};
