// @google/adk/dist/web/auth/oauth2/oauth2_credential_refresher.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { logger } from "../../utils/logger.js";
import {
  fetchOAuth2Tokens,
  getTokenEndpoint,
  isTokenExpired
} from "./oauth2_utils.js";
class OAuth2CredentialRefresher {
  /**
   * Check if the OAuth2 credential needs to be refreshed.
   *
   * @param authCredential The OAuth2 credential to check.
   * @param authScheme The OAuth2 authentication scheme (optional).
   * @returns True if the credential needs to be refreshed, False otherwise.
   */
  async isRefreshNeeded(authCredential) {
    if (!authCredential.oauth2) {
      return false;
    }
    if (authCredential.oauth2 && authCredential.oauth2.expiresAt) {
      return isTokenExpired(authCredential.oauth2);
    }
    return false;
  }
  /**
   * Refresh the OAuth2 credential.
   *
   * @param authCredential The OAuth2 credential to refresh.
   * @param authScheme The OAuth2 authentication scheme.
   * @returns The refreshed credential.
   */
  async refresh(authCredential, authScheme) {
    if (!authCredential.oauth2 || !authScheme) {
      return authCredential;
    }
    if (!authCredential.oauth2.refreshToken) {
      logger.warn("No refresh token available to refresh credential");
      return authCredential;
    }
    const isNeeded = await this.isRefreshNeeded(authCredential);
    if (!isNeeded) {
      return authCredential;
    }
    const tokenEndpoint = getTokenEndpoint(authScheme);
    if (!tokenEndpoint) {
      logger.warn("Token endpoint not found in auth scheme.");
      return authCredential;
    }
    if (!authCredential.oauth2.clientId || !authCredential.oauth2.clientSecret) {
      logger.warn("clientId and clientSecret are required for token refresh.");
      return authCredential;
    }
    const body = new URLSearchParams();
    body.set("grant_type", "refresh_token");
    body.set("refresh_token", authCredential.oauth2.refreshToken);
    body.set("client_id", authCredential.oauth2.clientId);
    body.set("client_secret", authCredential.oauth2.clientSecret);
    try {
      const data = await fetchOAuth2Tokens(tokenEndpoint, body);
      const updatedOAuth2 = __spreadProps(__spreadValues({}, authCredential.oauth2), {
        accessToken: data.accessToken || authCredential.oauth2.accessToken,
        refreshToken: data.refreshToken || authCredential.oauth2.refreshToken,
        expiresIn: data.expiresIn,
        expiresAt: data.expiresAt || authCredential.oauth2.expiresAt
      });
      return __spreadProps(__spreadValues({}, authCredential), {
        oauth2: updatedOAuth2
      });
    } catch (error) {
      logger.error("Failed to refresh tokens:", error);
      return authCredential;
    }
  }
}
export {
  OAuth2CredentialRefresher
};
