// @google/adk/dist/web/auth/oauth2/oauth2_discovery.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { z } from "zod";
import { logger } from "../../utils/logger.js";
const AuthorizationServerMetadataSchema = z.object({
  issuer: z.string(),
  authorization_endpoint: z.string(),
  token_endpoint: z.string(),
  scopes_supported: z.array(z.string()).optional(),
  registration_endpoint: z.string().optional()
});
const ProtectedResourceMetadataSchema = z.object({
  resource: z.string(),
  authorization_servers: z.array(z.string()).default([])
});
class OAuth2DiscoveryManager {
  /**
   * Discovers the OAuth2 authorization server metadata.
   */
  async discoverAuthServerMetadata(issuerUrl) {
    if (!validateDiscoveryUrl(issuerUrl)) {
      return void 0;
    }
    let baseUrl;
    let path;
    try {
      const url = new URL(issuerUrl);
      baseUrl = "".concat(url.protocol, "//").concat(url.host);
      path = url.pathname;
    } catch (e) {
      logger.warn("Failed to parse issuerUrl ".concat(issuerUrl, ": ").concat(e));
      return void 0;
    }
    const endpointsToTry = [];
    if (path && path !== "/") {
      endpointsToTry.push(
        "".concat(baseUrl, "/.well-known/oauth-authorization-server").concat(path),
        "".concat(baseUrl, "/.well-known/openid-configuration").concat(path),
        "".concat(baseUrl).concat(path, "/.well-known/openid-configuration")
      );
    } else {
      endpointsToTry.push(
        "".concat(baseUrl, "/.well-known/oauth-authorization-server"),
        "".concat(baseUrl, "/.well-known/openid-configuration")
      );
    }
    for (const endpoint of endpointsToTry) {
      try {
        const response = await fetch(endpoint, {
          method: "GET",
          headers: {
            Accept: "application/json"
          }
        });
        if (!response.ok) {
          continue;
        }
        const data = await response.json();
        const metadata = AuthorizationServerMetadataSchema.parse(data);
        if (metadata.issuer.replace(/\/$/, "") === issuerUrl.replace(/\/$/, "")) {
          return metadata;
        } else {
          logger.warn(
            "Issuer in metadata ".concat(metadata.issuer, " does not match issuerUrl ").concat(issuerUrl)
          );
        }
      } catch (e) {
        logger.debug("Failed to fetch metadata from ".concat(endpoint, ": ").concat(e));
      }
    }
    return void 0;
  }
  /**
   * Discovers the OAuth2 protected resource metadata.
   */
  async discoverResourceMetadata(resourceUrl) {
    if (!validateDiscoveryUrl(resourceUrl)) {
      return void 0;
    }
    let baseUrl;
    let path;
    try {
      const url = new URL(resourceUrl);
      baseUrl = "".concat(url.protocol, "//").concat(url.host);
      path = url.pathname;
    } catch (e) {
      logger.warn("Failed to parse resourceUrl ".concat(resourceUrl, ": ").concat(e));
      return void 0;
    }
    let wellKnownEndpoint;
    if (path && path !== "/") {
      wellKnownEndpoint = "".concat(baseUrl, "/.well-known/oauth-protected-resource").concat(path);
    } else {
      wellKnownEndpoint = "".concat(baseUrl, "/.well-known/oauth-protected-resource");
    }
    try {
      const response = await fetch(wellKnownEndpoint, {
        method: "GET",
        headers: {
          Accept: "application/json"
        }
      });
      if (!response.ok) {
        return void 0;
      }
      const data = await response.json();
      const metadata = ProtectedResourceMetadataSchema.parse(data);
      if (metadata.resource.replace(/\/$/, "") === resourceUrl.replace(/\/$/, "")) {
        return metadata;
      } else {
        logger.warn(
          "Resource in metadata ".concat(metadata.resource, " does not match resourceUrl ").concat(resourceUrl)
        );
      }
    } catch (e) {
      logger.debug("Failed to fetch metadata from ".concat(wellKnownEndpoint, ": ").concat(e));
    }
    return void 0;
  }
}
function validateDiscoveryUrl(urlStr) {
  try {
    const url = new URL(urlStr);
    if (url.protocol !== "https:") {
      logger.warn("Unsafe protocol for discovery URL: ".concat(url.protocol));
      return false;
    }
    const host = url.hostname.toLowerCase();
    if (host === "localhost" || host === "127.0.0.1" || host === "[::1]" || host.startsWith("10.") || host.startsWith("192.168.") || host.startsWith("169.254.")) {
      logger.warn("Unsafe host for discovery URL: ".concat(host));
      return false;
    }
    const match = host.match(/^172\.(\d+)\./);
    if (match) {
      const secondOctet = parseInt(match[1], 10);
      if (secondOctet >= 16 && secondOctet <= 31) {
        logger.warn("Unsafe host for discovery URL: ".concat(host));
        return false;
      }
    }
    return true;
  } catch (e) {
    logger.warn("Failed to parse URL for validation ".concat(urlStr, ": ").concat(e));
    return false;
  }
}
export {
  AuthorizationServerMetadataSchema,
  OAuth2DiscoveryManager,
  ProtectedResourceMetadataSchema
};
