// @google/adk/dist/web/tools/load_memory_tool.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { Type } from "@google/genai";
import { appendInstructions } from "../models/llm_request.js";
import {
  BaseTool
} from "./base_tool.js";
class LoadMemoryTool extends BaseTool {
  constructor() {
    super({
      name: "load_memory",
      description: "Loads the memory for the current user.\n\nNOTE: Currently this tool only uses text part from the memory."
    });
  }
  _getDeclaration() {
    return {
      name: this.name,
      description: this.description,
      parameters: {
        type: Type.OBJECT,
        properties: {
          query: {
            type: Type.STRING,
            description: "The query to load the memory for."
          }
        },
        required: ["query"]
      }
    };
  }
  async runAsync({
    args,
    toolContext
  }) {
    try {
      const query = args["query"];
      if (!toolContext.invocationContext.memoryService) {
        throw new Error("Memory service is not initialized.");
      }
      const searchMemoryResponse = await toolContext.searchMemory(query);
      return {
        memories: searchMemoryResponse.memories.map((m) => {
          var _a, _b;
          return {
            // Join all text parts by a space, or empty string if no text parts
            content: (_b = (_a = m.content.parts) == null ? void 0 : _a.map((p) => {
              var _a2;
              return (_a2 = p.text) != null ? _a2 : "";
            }).join(" ")) != null ? _b : "",
            author: m.author,
            timestamp: m.timestamp
          };
        })
      };
    } catch (e) {
      console.error("ERROR in LoadMemoryTool runAsync:", e);
      throw e;
    }
  }
  async processLlmRequest(request) {
    await super.processLlmRequest(request);
    if (!request.toolContext.invocationContext.memoryService) {
      return;
    }
    appendInstructions(request.llmRequest, [
      "You have memory. You can use it to answer questions. If any questions need\nyou to look up the memory, you should call load_memory function with a query."
    ]);
  }
}
const LOAD_MEMORY = new LoadMemoryTool();
export {
  LOAD_MEMORY,
  LoadMemoryTool
};
