// @google/adk/dist/web/tools/vertex_ai_search_tool.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { getLogger } from "../utils/logger.js";
import {
  isGemini1Model,
  isGeminiModel,
  isGeminiModelIdCheckDisabled
} from "../utils/model_name.js";
import { BaseTool } from "./base_tool.js";
const logger = getLogger();
class VertexAiSearchTool extends BaseTool {
  constructor(params) {
    super({ name: "vertex_ai_search", description: "vertex_ai_search" });
    const {
      dataStoreId,
      dataStoreSpecs,
      searchEngineId,
      filter,
      maxResults,
      bypassMultiToolsLimit = false
    } = params;
    if (dataStoreId === void 0 && searchEngineId === void 0 || dataStoreId !== void 0 && searchEngineId !== void 0) {
      throw new Error(
        "Either dataStoreId or searchEngineId must be specified."
      );
    }
    if (dataStoreSpecs !== void 0 && searchEngineId === void 0) {
      throw new Error(
        "searchEngineId must be specified if dataStoreSpecs is specified."
      );
    }
    this.dataStoreId = dataStoreId;
    this.dataStoreSpecs = dataStoreSpecs;
    this.searchEngineId = searchEngineId;
    this.filter = filter;
    this.maxResults = maxResults;
    this.bypassMultiToolsLimit = bypassMultiToolsLimit;
  }
  runAsync() {
    return Promise.resolve();
  }
  /**
   * Builds the VertexAISearch configuration.
   *
   * Override this method in a subclass to dynamically customize the search
   * configuration based on the context (e.g., set filter based on session
   * state).
   */
  buildVertexAiSearchConfig(_readonlyContext) {
    return {
      datastore: this.dataStoreId,
      dataStoreSpecs: this.dataStoreSpecs,
      engine: this.searchEngineId,
      filter: this.filter,
      maxResults: this.maxResults
    };
  }
  async processLlmRequest({
    toolContext,
    llmRequest
  }) {
    if (!llmRequest.model) {
      return;
    }
    const modelCheckDisabled = isGeminiModelIdCheckDisabled();
    llmRequest.config = llmRequest.config || {};
    llmRequest.config.tools = llmRequest.config.tools || [];
    if (!isGeminiModel(llmRequest.model) && !modelCheckDisabled) {
      throw new Error(
        "Vertex AI search tool is not supported for model ".concat(llmRequest.model)
      );
    }
    if (isGemini1Model(llmRequest.model) && llmRequest.config.tools.length > 0 && !this.bypassMultiToolsLimit) {
      throw new Error(
        "Vertex AI search tool cannot be used with other tools in Gemini 1.x."
      );
    }
    const vertexAiSearchConfig = this.buildVertexAiSearchConfig(toolContext);
    let specsInfo;
    if (vertexAiSearchConfig.dataStoreSpecs) {
      const specIds = vertexAiSearchConfig.dataStoreSpecs.map(
        (spec) => spec.dataStore ? spec.dataStore.split("/").pop() : "unnamed"
      );
      specsInfo = "".concat(vertexAiSearchConfig.dataStoreSpecs.length, " spec(s): [").concat(specIds.join(", "), "]");
    }
    logger.debug(
      "Adding Vertex AI Search tool config to LLM request: " + "datastore=".concat(vertexAiSearchConfig.datastore, ", ") + "engine=".concat(vertexAiSearchConfig.engine, ", ") + "filter=".concat(vertexAiSearchConfig.filter, ", ") + "maxResults=".concat(vertexAiSearchConfig.maxResults, ", ") + "dataStoreSpecs=".concat(specsInfo)
    );
    llmRequest.config.tools.push({
      retrieval: {
        vertexAiSearch: vertexAiSearchConfig
      }
    });
  }
}
export {
  VertexAiSearchTool
};
