// @google/adk/dist/web/events/event.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { toCamelCase, toSnakeCase } from "../utils/object_notation_utils.js";
import { createEventActions } from "./event_actions.js";
function createEvent(params = {}) {
  return __spreadProps(__spreadValues({}, params), {
    id: params.id || createNewEventId(),
    invocationId: params.invocationId || "",
    author: params.author,
    actions: params.actions || createEventActions(),
    longRunningToolIds: params.longRunningToolIds || [],
    branch: params.branch,
    timestamp: params.timestamp || Date.now()
  });
}
function isFinalResponse(event) {
  if (event.actions.skipSummarization || event.longRunningToolIds && event.longRunningToolIds.length > 0) {
    return true;
  }
  return getFunctionCalls(event).length === 0 && getFunctionResponses(event).length === 0 && !event.partial && !hasTrailingCodeExecutionResult(event);
}
function getFunctionCalls(event) {
  const funcCalls = [];
  if (event.content && event.content.parts) {
    for (const part of event.content.parts) {
      if (part.functionCall) {
        funcCalls.push(part.functionCall);
      }
    }
  }
  return funcCalls;
}
function getFunctionResponses(event) {
  const funcResponses = [];
  if (event.content && event.content.parts) {
    for (const part of event.content.parts) {
      if (part.functionResponse) {
        funcResponses.push(part.functionResponse);
      }
    }
  }
  return funcResponses;
}
function hasTrailingCodeExecutionResult(event) {
  var _a;
  if (event.content && ((_a = event.content.parts) == null ? void 0 : _a.length)) {
    const lastPart = event.content.parts[event.content.parts.length - 1];
    return lastPart.codeExecutionResult !== void 0;
  }
  return false;
}
function stringifyContent(event) {
  var _a;
  if (!((_a = event.content) == null ? void 0 : _a.parts)) {
    return "";
  }
  return event.content.parts.map((part) => {
    var _a2;
    return (_a2 = part.text) != null ? _a2 : "";
  }).join("");
}
const ASCII_LETTERS_AND_NUMBERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
function createNewEventId() {
  let id = "";
  for (let i = 0; i < 8; i++) {
    id += ASCII_LETTERS_AND_NUMBERS[Math.floor(Math.random() * ASCII_LETTERS_AND_NUMBERS.length)];
  }
  return id;
}
const PRESERVE_KEYS_CAMEL_CASE = [
  "actions.stateDelta",
  "actions.artifactDelta",
  "actions.requestedAuthConfigs",
  "actions.requestedToolConfirmations",
  "actions.customMetadata",
  "content.parts.functionCall.args",
  "content.parts.functionResponse.response"
];
const PRESERVE_KEYS_SNAKE_CASE = [
  "actions.state_delta",
  "actions.artifact_delta",
  "actions.requested_auth_configs",
  "actions.requested_tool_confirmations",
  "actions.custom_metadata",
  "content.parts.function_call.args",
  "content.parts.function_response.response"
];
function transformToCamelCaseEvent(event) {
  return toCamelCase(event, PRESERVE_KEYS_SNAKE_CASE);
}
function transformToSnakeCaseEvent(event) {
  return toSnakeCase(event, PRESERVE_KEYS_CAMEL_CASE);
}
export {
  createEvent,
  createNewEventId,
  getFunctionCalls,
  getFunctionResponses,
  hasTrailingCodeExecutionResult,
  isFinalResponse,
  stringifyContent,
  transformToCamelCaseEvent,
  transformToSnakeCaseEvent
};
