// @google/adk/dist/cjs/code_executors/unsafe_local_code_executor.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var unsafe_local_code_executor_exports = {};
__export(unsafe_local_code_executor_exports, {
  UnsafeLocalCodeExecutor: () => UnsafeLocalCodeExecutor
});
module.exports = __toCommonJS(unsafe_local_code_executor_exports);
var import_child_process = require("child_process");
var fs = __toESM(require("node:fs/promises"), 1);
var os = __toESM(require("node:os"), 1);
var path = __toESM(require("node:path"), 1);
var import_file_extension_utils = require("../utils/file_extension_utils.js");
var import_file_utils = require("../utils/file_utils.js");
var import_logger = require("../utils/logger.js");
var import_base_code_executor = require("./base_code_executor.js");
var import_code_execution_utils = require("./code_execution_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const IS_WINDOWS = os.platform() === "win32";
async function createTempScriptFile(code, language, shellCommandPath) {
  const tempDir = path.join(
    os.tmpdir(),
    "adk_js_unsafe_code_executor",
    Date.now().toString() + "_" + Math.random().toString(36).slice(2)
  );
  await fs.mkdir(tempDir, { recursive: true });
  const ext = getExtensionForLanguage(language, shellCommandPath) || ".js";
  const filePath = path.join(tempDir, `script${ext}`);
  await fs.writeFile(filePath, code);
  return { filePath, tempDir };
}
function getExtensionForLanguage(language, shellCommandPath) {
  if (language === import_code_execution_utils.CodeExecutionLanguage.JAVASCRIPT) {
    return ".js";
  }
  if (language === import_code_execution_utils.CodeExecutionLanguage.PYTHON) {
    return ".py";
  }
  if (language === import_code_execution_utils.CodeExecutionLanguage.POWERSHELL) {
    return ".ps1";
  }
  if (language === import_code_execution_utils.CodeExecutionLanguage.WINDOWS_CMD) {
    return ".bat";
  }
  if (language === import_code_execution_utils.CodeExecutionLanguage.SHELL) {
    if (IS_WINDOWS) {
      if (shellCommandPath && shellCommandPath.toLowerCase().includes("cmd")) {
        return ".bat";
      }
      return ".ps1";
    }
    return ".sh";
  }
  return void 0;
}
class UnsafeLocalCodeExecutor extends import_base_code_executor.BaseCodeExecutor {
  constructor(options = {}) {
    var _a, _b, _c, _d;
    super();
    this.timeoutSeconds = (_a = options.timeoutSeconds) != null ? _a : 30;
    this.nodeCommandPath = (_b = options.commandPath) != null ? _b : process.execPath;
    this.pythonCommandPath = (_c = options.pythonCommandPath) != null ? _c : IS_WINDOWS ? "python" : "python3";
    this.shellCommandPath = (_d = options.shellCommandPath) != null ? _d : IS_WINDOWS ? "powershell" : "bash";
    this.stateful = false;
    this.optimizeDataFile = false;
  }
  async executeCode(params) {
    var _a;
    const { code, language } = params.codeExecutionInput;
    if (![
      import_code_execution_utils.CodeExecutionLanguage.JAVASCRIPT,
      import_code_execution_utils.CodeExecutionLanguage.PYTHON,
      import_code_execution_utils.CodeExecutionLanguage.SHELL,
      import_code_execution_utils.CodeExecutionLanguage.WINDOWS_CMD,
      import_code_execution_utils.CodeExecutionLanguage.POWERSHELL
    ].includes(language)) {
      return {
        stdout: "",
        stderr: `Unsupported language: ${language}`,
        outputFiles: []
      };
    }
    import_logger.logger.warn(
      "\n====================================================================================\n\u26A0\uFE0F DANGER: UnsafeLocalCodeExecutor is executing code locally on this host machine!\nThis component provides NO sandboxing or container isolation. Arbitrary shell/script\ncommands generated by AI or untrusted sources could lead to serious security risks.\n====================================================================================\n"
    );
    let tempDir;
    try {
      const res = await createTempScriptFile(
        code,
        language,
        this.shellCommandPath
      );
      const filePath = res.filePath;
      tempDir = res.tempDir;
      if (params.codeExecutionInput.inputFiles) {
        await (0, import_file_utils.materializeFiles)(params.codeExecutionInput.inputFiles, tempDir);
      }
      let command = this.nodeCommandPath;
      let args = [filePath];
      if (language === import_code_execution_utils.CodeExecutionLanguage.PYTHON) {
        command = this.pythonCommandPath;
      } else if (language === import_code_execution_utils.CodeExecutionLanguage.SHELL) {
        command = this.shellCommandPath;
        if (this.shellCommandPath.toLowerCase().includes("powershell")) {
          args = ["-NoLogo", "-ExecutionPolicy", "Bypass", "-File", filePath];
        } else if (this.shellCommandPath.toLowerCase().includes("cmd")) {
          args = ["/c", filePath];
        }
      } else if (language === import_code_execution_utils.CodeExecutionLanguage.POWERSHELL) {
        command = IS_WINDOWS ? "powershell" : "pwsh";
        args = ["-NoLogo", "-ExecutionPolicy", "Bypass", "-File", filePath];
      } else if (language === import_code_execution_utils.CodeExecutionLanguage.WINDOWS_CMD) {
        command = "cmd.exe";
        args = ["/c", filePath];
      }
      if (params.codeExecutionInput.args) {
        if (Array.isArray(params.codeExecutionInput.args)) {
          args.push(...params.codeExecutionInput.args);
        } else {
          for (const [k, v] of Object.entries(params.codeExecutionInput.args)) {
            args.push(`--${k}`, String(v));
          }
        }
      }
      const executionResult = await new Promise((resolve) => {
        const child = (0, import_child_process.spawn)(command, args, {
          timeout: this.timeoutSeconds * 1e3,
          killSignal: "SIGKILL",
          cwd: tempDir
        });
        let stdout = "";
        let stderr = "";
        if (child.stdout) {
          child.stdout.on("data", (data) => {
            stdout += data.toString();
          });
        }
        if (child.stderr) {
          child.stderr.on("data", (data) => {
            stderr += data.toString();
          });
        }
        child.on("error", (err) => {
          stderr += `Process error: ${err.message}
`;
        });
        child.on("close", (exitCode, signal) => {
          if (signal === "SIGKILL" || signal === "SIGTERM") {
            stderr += `
Code execution timed out after ${this.timeoutSeconds} seconds.`;
          } else if (exitCode !== 0 && exitCode !== null) {
            if (!stderr) {
              stderr = `Exit code ${exitCode}`;
            }
          }
          resolve({ stdout, stderr, exitCode });
        });
      });
      const outputFiles = [];
      try {
        const allFiles = await fs.readdir(tempDir, { recursive: true });
        for (const relativeFilePath of allFiles) {
          const fullPath = path.join(tempDir, relativeFilePath);
          const stat = await fs.lstat(fullPath);
          if (!stat.isFile()) {
            continue;
          }
          if (relativeFilePath === path.basename(filePath)) {
            continue;
          }
          const isInputFile = (_a = params.codeExecutionInput.inputFiles) == null ? void 0 : _a.some(
            (f) => f.name === relativeFilePath
          );
          if (isInputFile) {
            continue;
          }
          const fileContent = await fs.readFile(fullPath);
          const { mimeType, encoding } = (0, import_file_extension_utils.getMimeTypeAndEncoding)(
            path.extname(relativeFilePath)
          );
          outputFiles.push({
            name: relativeFilePath,
            content: fileContent.toString(encoding),
            contentEncoding: encoding,
            mimeType
          });
        }
      } catch (e) {
        import_logger.logger.error(`Error scanning output files: ${e}`);
      }
      return {
        stdout: executionResult.stdout,
        stderr: executionResult.stderr,
        outputFiles
      };
    } finally {
      if (tempDir) {
        await fs.rm(tempDir, { recursive: true, force: true });
      }
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  UnsafeLocalCodeExecutor
});
