// @google/adk/dist/web/auth/auth_preprocessor.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __await = function(promise, isYieldStar) {
  this[0] = promise;
  this[1] = isYieldStar;
};
var __asyncGenerator = (__this, __arguments, generator) => {
  var resume = (k, v, yes, no) => {
    try {
      var x = generator[k](v), isAwait = (v = x.value) instanceof __await, done = x.done;
      Promise.resolve(isAwait ? v[0] : v).then((y) => isAwait ? resume(k === "return" ? k : "next", v[1] ? { done: y.done, value: y.value } : y, yes, no) : yes({ value: y, done })).catch((e) => resume("throw", e, yes, no));
    } catch (e) {
      no(e);
    }
  }, method = (k) => it[k] = (x) => new Promise((yes, no) => resume(k, x, yes, no)), it = {};
  return generator = generator.apply(__this, __arguments), it[__knownSymbol("asyncIterator")] = () => it, method("next"), method("throw"), method("return"), it;
};
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  REQUEST_EUC_FUNCTION_CALL_NAME,
  handleFunctionCallsAsync
} from "../agents/functions.js";
import { isLlmAgent } from "../agents/llm_agent.js";
import { BaseLlmRequestProcessor } from "../agents/processors/base_llm_processor.js";
import { ReadonlyContext } from "../agents/readonly_context.js";
import {
  getFunctionCalls,
  getFunctionResponses
} from "../events/event.js";
import { State } from "../sessions/state.js";
import { AuthHandler } from "./auth_handler.js";
const TOOLSET_AUTH_CREDENTIAL_ID_PREFIX = "_adk_toolset_auth_";
async function storeAuthAndCollectResumeTargets(events, authFcIds, authResponses, state) {
  const requestedAuthConfigById = {};
  for (const event of events) {
    const eventFunctionCalls = getFunctionCalls(event);
    for (const functionCall of eventFunctionCalls) {
      if (functionCall.id && authFcIds.has(functionCall.id) && functionCall.name === REQUEST_EUC_FUNCTION_CALL_NAME) {
        const args = functionCall.args;
        if (args && args.authConfig) {
          requestedAuthConfigById[functionCall.id] = args.authConfig;
        }
      }
    }
  }
  for (const fcId of authFcIds) {
    if (!(fcId in authResponses)) {
      continue;
    }
    const authConfig = authResponses[fcId];
    const requestedAuthConfig = requestedAuthConfigById[fcId];
    if (requestedAuthConfig && requestedAuthConfig.credentialKey) {
      authConfig.credentialKey = requestedAuthConfig.credentialKey;
    }
    await new AuthHandler(authConfig).parseAndStoreAuthResponse(state);
  }
  const toolsToResume = /* @__PURE__ */ new Set();
  for (const fcId of authFcIds) {
    const requestedAuthConfig = requestedAuthConfigById[fcId];
    if (!requestedAuthConfig) {
      continue;
    }
    for (const event of events) {
      const eventFunctionCalls = getFunctionCalls(event);
      for (const functionCall of eventFunctionCalls) {
        if (functionCall.id === fcId && functionCall.name === REQUEST_EUC_FUNCTION_CALL_NAME) {
          const args = functionCall.args;
          if (args && args.functionCallId) {
            if (args.functionCallId.startsWith(TOOLSET_AUTH_CREDENTIAL_ID_PREFIX)) {
              continue;
            }
            toolsToResume.add(args.functionCallId);
          }
        }
      }
    }
  }
  return toolsToResume;
}
class AuthPreprocessor extends BaseLlmRequestProcessor {
  runAsync(invocationContext) {
    return __asyncGenerator(this, null, function* () {
      const agent = invocationContext.agent;
      if (!isLlmAgent(agent)) {
        return;
      }
      const events = invocationContext.session.events;
      if (!events || events.length === 0) {
        return;
      }
      let lastEventWithContent = null;
      for (let i = events.length - 1; i >= 0; i--) {
        const event = events[i];
        if (event.content !== void 0) {
          lastEventWithContent = event;
          break;
        }
      }
      if (!lastEventWithContent || lastEventWithContent.author !== "user") {
        return;
      }
      const responses = getFunctionResponses(lastEventWithContent);
      if (!responses || responses.length === 0) {
        return;
      }
      const authFcIds = /* @__PURE__ */ new Set();
      const authResponses = {};
      for (const functionCallResponse of responses) {
        if (functionCallResponse.name !== REQUEST_EUC_FUNCTION_CALL_NAME) {
          continue;
        }
        if (functionCallResponse.id) {
          authFcIds.add(functionCallResponse.id);
          authResponses[functionCallResponse.id] = functionCallResponse.response;
        }
      }
      if (authFcIds.size === 0) {
        return;
      }
      const state = new State(invocationContext.session.state);
      const toolsToResume = yield new __await(storeAuthAndCollectResumeTargets(
        events,
        authFcIds,
        authResponses,
        state
      ));
      if (toolsToResume.size === 0) {
        return;
      }
      for (let i = events.length - 2; i >= 0; i--) {
        const event = events[i];
        const functionCalls = getFunctionCalls(event);
        if (!functionCalls || functionCalls.length === 0) {
          continue;
        }
        const hasMatchingCall = functionCalls.some(
          (call) => call.id ? toolsToResume.has(call.id) : false
        );
        if (hasMatchingCall) {
          const canonicalTools = yield new __await(agent.canonicalTools(
            new ReadonlyContext(invocationContext)
          ));
          const toolsDict = {};
          for (const tool of canonicalTools) {
            toolsDict[tool.name] = tool;
          }
          const functionResponseEvent = yield new __await(handleFunctionCallsAsync({
            invocationContext,
            functionCallEvent: event,
            toolsDict,
            beforeToolCallbacks: agent.canonicalBeforeToolCallbacks,
            afterToolCallbacks: agent.canonicalAfterToolCallbacks,
            filters: toolsToResume
          }));
          if (functionResponseEvent) {
            yield functionResponseEvent;
          }
          return;
        }
      }
    });
  }
}
const AUTH_PREPROCESSOR = new AuthPreprocessor();
export {
  AUTH_PREPROCESSOR,
  AuthPreprocessor
};
