// @google/adk-devtools/dist/esm/utils/agent_loader.js

const Module = await import('node:module'); 
const require = /* @__PURE__ */ (Module?.default?.createRequire ?? Module?.createRequire)(import.meta.url);
const __filename = /* @__PURE__ */ (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = /* @__PURE__ */ (await import("node:path")).dirname(__filename);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { isBaseAgent } from "@google/adk";
import esbuild from "esbuild";
import { shimPlugin } from "esbuild-shim-plugin";
import * as fsPromises from "node:fs/promises";
import * as path from "node:path";
import { pathToFileURL } from "node:url";
import {
  getTempDir,
  isFile,
  isFileExists,
  isFolderExists,
  loadFileData,
  removeFolder,
  tryToFindFileRecursively
} from "./file_utils.js";
const JS_FILES_EXTENSIONS = [".js", ".cjs", ".mjs", ".ts", ".mts", ".cts"];
var FileModuleType = /* @__PURE__ */ ((FileModuleType2) => {
  FileModuleType2["CJS"] = "cjs";
  FileModuleType2["ESM"] = "esm";
  return FileModuleType2;
})(FileModuleType || {});
const FILE_MODULE_TYPE_EXTENSION_MAP = {
  ["cjs" /* CJS */]: ".cjs",
  ["esm" /* ESM */]: ".mjs"
};
class AgentFileLoadingError extends Error {
}
const DEFAULT_AGENT_FILE_OPTIONS = {
  compile: true,
  bundle: true
};
function replaceDirnamePlugin(filePath, originalDir) {
  return {
    name: "replace-dirname",
    setup(build) {
      build.onLoad({ filter: /.*/ }, async (args) => {
        if (args.path === filePath) {
          const content = await fsPromises.readFile(args.path, "utf8");
          const fileUrl = pathToFileURL(filePath).href;
          const loader = [".ts", ".mts", ".cts"].includes(
            path.extname(filePath)
          ) ? "ts" : "js";
          const transformResult = await esbuild.transform(content, {
            loader,
            define: {
              "__dirname": JSON.stringify(originalDir),
              "__filename": JSON.stringify(filePath),
              "import.meta.url": JSON.stringify(fileUrl)
            }
          });
          return {
            contents: transformResult.code,
            loader: "js"
          };
        }
        return void 0;
      });
    }
  };
}
class AgentFile {
  constructor(filePath, options = DEFAULT_AGENT_FILE_OPTIONS) {
    this.filePath = filePath;
    this.options = options;
    this.disposed = false;
  }
  async load() {
    if (this.agent) {
      return this.agent;
    }
    try {
      await fsPromises.stat(this.filePath);
    } catch (e) {
      if (e.code === "ENOENT") {
        throw new AgentFileLoadingError(
          `Agent file ${this.filePath} does not exists`
        );
      }
    }
    let filePath = this.filePath;
    const shouldCompile = this.options.compile || this.options.bundle;
    if (shouldCompile) {
      const moduleType = this.options.moduleType || await getFileModuleType(filePath);
      const parsedPath = path.parse(filePath);
      const outputDir = getTempDir("adk_agent_loader");
      const compiledFilePath = path.join(
        outputDir,
        parsedPath.name + FILE_MODULE_TYPE_EXTENSION_MAP[moduleType]
      );
      const originalDir = path.dirname(filePath);
      await fsPromises.mkdir(outputDir, { recursive: true });
      await linkProjectNodeModules(outputDir, parsedPath.dir);
      await esbuild.build({
        entryPoints: [filePath],
        outfile: compiledFilePath,
        target: "node16",
        platform: "node",
        format: moduleType,
        packages: "bundle",
        bundle: this.options.bundle,
        minify: this.options.bundle,
        allowOverwrite: true,
        plugins: [replaceDirnamePlugin(filePath, originalDir), shimPlugin()],
        // See http://mikro-orm.io/docs/deployment#deploy-a-bundle-of-entities-and-dependencies-with-esbuild for more details
        external: [
          "sqlite3",
          "better-sqlite3",
          "mysql",
          "mysql2",
          // Native addons must remain external so Node can resolve their
          // platform-specific assets at runtime.
          "onnxruntime-node",
          "oracledb",
          "pg-native",
          "pg-query-stream",
          "tedious",
          "libsql"
        ]
      });
      this.cleanupDirPath = outputDir;
      this.cleanupFilePath = compiledFilePath;
      filePath = compiledFilePath;
    }
    const jsModule = await import(pathToFileURL(filePath).href);
    if (jsModule) {
      if (isBaseAgent(jsModule.rootAgent)) {
        return this.agent = jsModule.rootAgent;
      }
      if (isBaseAgent(jsModule.default)) {
        return this.agent = jsModule.default;
      }
      const rootAgents = Object.values(jsModule).filter(
        (exportValue) => isBaseAgent(exportValue)
      );
      if (rootAgents.length > 1) {
        console.warn(
          `Multiple agents found in ${filePath}. Using the ${rootAgents[0].name} as a root agent.`
        );
      }
      if (rootAgents.length > 0) {
        return this.agent = rootAgents[0];
      }
    }
    await this.dispose();
    throw new AgentFileLoadingError(
      `Failed to load agent ${filePath}: No @google/adk BaseAgent class instance found. Please check that file is not empty and it has export of @google/adk BaseAgent class (e.g. LlmAgent) instance.`
    );
  }
  getFilePath() {
    if (!this.agent) {
      throw new Error("Agent is not loaded yet");
    }
    if (this.disposed) {
      throw new Error("Agent is disposed and can not be used");
    }
    return this.cleanupFilePath || this.filePath;
  }
  async [Symbol.asyncDispose]() {
    return this.dispose();
  }
  async dispose() {
    if (this.disposed) {
      return;
    }
    if (this.cleanupFilePath) {
      this.disposed = true;
      await fsPromises.unlink(this.cleanupFilePath);
      if (this.cleanupDirPath) {
        await removeFolder(this.cleanupDirPath);
      }
    }
  }
}
class AgentLoader {
  constructor(agentsDirPath = process.cwd(), options = DEFAULT_AGENT_FILE_OPTIONS) {
    this.agentsDirPath = agentsDirPath;
    this.options = options;
    this.agentsAlreadyPreloaded = false;
    this.preloadedAgents = {};
    const exitHandler = async ({
      exit,
      cleanup
    }) => {
      if (cleanup) {
        await this.disposeAll();
      }
      if (exit) {
        process.exit();
      }
    };
    process.on("exit", () => exitHandler({ cleanup: true }));
    process.on("SIGINT", () => exitHandler({ exit: true }));
    process.on("SIGUSR1", () => exitHandler({ exit: true }));
    process.on("SIGUSR2", () => exitHandler({ exit: true }));
    process.on("uncaughtException", () => exitHandler({ exit: true }));
  }
  async listAgents() {
    await this.preloadAgents();
    return Object.keys(this.preloadedAgents).sort();
  }
  async getAgentFile(agentName) {
    await this.preloadAgents();
    return this.preloadedAgents[agentName];
  }
  async disposeAll() {
    await Promise.all(
      Object.values(this.preloadedAgents).map((f) => f.dispose())
    );
  }
  async preloadAgents() {
    if (this.agentsAlreadyPreloaded) {
      return;
    }
    const files = await isFile(this.agentsDirPath) ? [await getFileMetadata(this.agentsDirPath)] : await getDirFiles(this.agentsDirPath);
    await Promise.all(
      files.map(async (fileOrDir) => {
        if (fileOrDir.isFile && isJsFile(fileOrDir.ext)) {
          return this.loadAgentFromFile(fileOrDir);
        }
        if (fileOrDir.isDirectory) {
          return this.loadAgentFromDirectory(fileOrDir);
        }
      })
    );
    this.agentsAlreadyPreloaded = true;
    return;
  }
  async loadAgentFromFile(file) {
    try {
      const agentFile = new AgentFile(file.path, this.options);
      await agentFile.load();
      this.preloadedAgents[file.name] = agentFile;
    } catch (e) {
      if (e instanceof AgentFileLoadingError) {
        return;
      }
      throw e;
    }
  }
  async loadAgentFromDirectory(dir) {
    const subFiles = await getDirFiles(dir.path);
    const possibleAgentJsFile = subFiles.find(
      (f) => f.isFile && f.name === "agent" && isJsFile(f.ext)
    );
    if (!possibleAgentJsFile) {
      return;
    }
    try {
      const agentFile = new AgentFile(possibleAgentJsFile.path, this.options);
      await agentFile.load();
      this.preloadedAgents[dir.name] = agentFile;
    } catch (e) {
      if (e instanceof AgentFileLoadingError) {
        return;
      }
      throw e;
    }
  }
}
function isJsFile(fileExt) {
  return !!fileExt && JS_FILES_EXTENSIONS.includes(fileExt);
}
async function getDirFiles(dir) {
  const files = await fsPromises.readdir(dir);
  return await Promise.all(
    files.map((filePath) => getFileMetadata(path.join(dir, filePath)))
  );
}
async function getFileMetadata(filePath) {
  const fileStats = await fsPromises.stat(filePath);
  const isFile2 = fileStats.isFile();
  const baseName = path.basename(filePath);
  const ext = path.extname(filePath);
  return {
    path: filePath,
    name: isFile2 ? baseName.slice(0, baseName.length - ext.length) : baseName,
    ext: isFile2 ? path.extname(filePath) : void 0,
    isFile: isFile2,
    isDirectory: fileStats.isDirectory()
  };
}
async function getFileModuleType(filePath) {
  const { ext } = path.parse(filePath);
  if ([".cjs", ".cts"].includes(ext)) {
    return "cjs" /* CJS */;
  }
  if ([".mts", ".mjs"].includes(ext)) {
    return "esm" /* ESM */;
  }
  if ([".js", ".ts"].includes(ext)) {
    return getTypeFromPackageJson(path.dirname(filePath));
  }
  return "cjs" /* CJS */;
}
async function getTypeFromPackageJson(dir) {
  const packagePath = path.join(dir, "package.json");
  if (await isFileExists(packagePath)) {
    try {
      const packageJson = await loadFileData(packagePath);
      return packageJson.type === "module" ? "esm" /* ESM */ : "cjs" /* CJS */;
    } catch {
      return "cjs" /* CJS */;
    }
  }
  const parentDir = path.dirname(dir);
  if (parentDir === dir) {
    return "cjs" /* CJS */;
  }
  return getTypeFromPackageJson(parentDir);
}
async function linkProjectNodeModules(outputDir, sourceDir) {
  const nodeModulesDir = await getProjectNodeModulesDir(sourceDir);
  if (!nodeModulesDir) {
    return;
  }
  const linkPath = path.join(outputDir, "node_modules");
  if (await isFolderExists(linkPath)) {
    return;
  }
  try {
    await fsPromises.symlink(
      path.resolve(nodeModulesDir),
      linkPath,
      process.platform === "win32" ? "junction" : "dir"
    );
  } catch (error) {
    if (error.code !== "EEXIST") {
      throw error;
    }
  }
}
async function getProjectNodeModulesDir(sourceDir) {
  try {
    const packageJsonPath = await tryToFindFileRecursively(
      sourceDir,
      "package.json",
      10
    );
    const nodeModulesDir = path.join(
      path.dirname(packageJsonPath),
      "node_modules"
    );
    return await isFolderExists(nodeModulesDir) ? nodeModulesDir : void 0;
  } catch {
    return void 0;
  }
}
export {
  AgentFile,
  AgentLoader,
  FileModuleType,
  replaceDirnamePlugin
};
