// @google/adk-devtools/dist/esm/utils/file_utils.js

const Module = await import('node:module'); 
const require = /* @__PURE__ */ (Module?.default?.createRequire ?? Module?.createRequire)(import.meta.url);
const __filename = /* @__PURE__ */ (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = /* @__PURE__ */ (await import("node:path")).dirname(__filename);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
async function isFolderExists(folderPath) {
  try {
    await fs.access(folderPath);
    const stat = await fs.stat(folderPath);
    return stat.isDirectory();
  } catch (_e) {
    return false;
  }
}
async function isFileExists(folderPath) {
  try {
    const stat = await fs.stat(folderPath);
    return stat.isFile();
  } catch (_e) {
    return false;
  }
}
async function createFolder(folderPath) {
  try {
    await fs.mkdir(folderPath);
  } catch (e) {
    console.error(`Failed to create folder ${folderPath}`, e);
  }
}
async function removeFolder(folderPath) {
  try {
    await fs.rm(folderPath, { recursive: true });
  } catch (e) {
    console.error(`Failed to remove folder ${folderPath}`, e);
  }
}
async function listFiles(folderPath) {
  try {
    return await fs.readdir(folderPath);
  } catch (e) {
    console.error(`Failed to list files in folder ${folderPath}`, e);
    return [];
  }
}
async function isFile(filePath) {
  try {
    const stat = await fs.stat(filePath);
    return stat.isFile();
  } catch (_e) {
    return false;
  }
}
async function loadFileData(filePath) {
  try {
    return JSON.parse(await fs.readFile(filePath, { encoding: "utf-8" }));
  } catch (e) {
    console.error(`Failed to read or parse file ${filePath}:`, e);
    throw e;
  }
}
async function saveToFile(filePath, data) {
  try {
    await fs.writeFile(
      filePath,
      typeof data === "string" ? data : JSON.stringify(data, null, 2),
      { encoding: "utf-8" }
    );
  } catch (e) {
    console.error(`Failed to write file ${filePath}:`, e);
    throw e;
  }
}
function getTempDir(prefix) {
  const pathParts = [os.tmpdir()];
  if (prefix) {
    pathParts.push(prefix);
  }
  pathParts.push(Date.now().toString());
  return path.join(...pathParts);
}
async function tryToFindFileRecursively(sourceFolder, fileName, maxIterations) {
  let currentFolder = sourceFolder;
  for (let i = 0; i < maxIterations; i++) {
    const filePath = path.join(currentFolder, fileName);
    if (await isFileExists(filePath)) {
      return filePath;
    }
    currentFolder = path.dirname(currentFolder);
  }
  throw new Error(
    `No ${fileName} found in ${sourceFolder} or its parent folders up to ${maxIterations} levels.`
  );
}
export {
  createFolder,
  getTempDir,
  isFile,
  isFileExists,
  isFolderExists,
  listFiles,
  loadFileData,
  removeFolder,
  saveToFile,
  tryToFindFileRecursively
};
