// @google/adk-devtools/dist/esm/utils/logger.js

const Module = await import('node:module'); 
const require = /* @__PURE__ */ (Module?.default?.createRequire ?? Module?.createRequire)(import.meta.url);
const __filename = /* @__PURE__ */ (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = /* @__PURE__ */ (await import("node:path")).dirname(__filename);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { LogLevel } from "@google/adk";
import * as winston from "winston";
class AdkLogger {
  constructor(options) {
    this.logLevel = LogLevel.INFO;
    var _a;
    const formats = [
      winston.format.label({
        label: options.label,
        message: (_a = options.colorize) == null ? void 0 : _a.all
      }),
      winston.format((info) => {
        info.level = info.level.toUpperCase();
        return info;
      })()
    ];
    if (options.colorize) {
      formats.push(winston.format.colorize(options.colorize));
    }
    if (options.timestamp) {
      formats.push(winston.format.timestamp());
    }
    if (options.printFormat) {
      formats.push(winston.format.printf(options.printFormat));
    } else {
      formats.push(winston.format.printf((info) => info.message));
    }
    this.logger = winston.createLogger({
      levels: {
        "debug": LogLevel.DEBUG,
        "info": LogLevel.INFO,
        "warn": LogLevel.WARN,
        "error": LogLevel.ERROR
      },
      level: "error",
      format: winston.format.combine(...formats),
      transports: [new winston.transports.Console()]
    });
  }
  setLogLevel(level) {
    this.logLevel = level;
  }
  log(level, ...messages) {
    if (this.logLevel > level) {
      return;
    }
    this.logger.log(level.toString(), messages.join(" "));
  }
  debug(...messages) {
    if (this.logLevel > LogLevel.DEBUG) {
      return;
    }
    this.logger.debug(messages.join(" "));
  }
  info(...messages) {
    if (this.logLevel > LogLevel.INFO) {
      return;
    }
    this.logger.info(messages.join(" "));
  }
  warn(...messages) {
    if (this.logLevel > LogLevel.WARN) {
      return;
    }
    this.logger.warn(messages.join(" "));
  }
  error(...messages) {
    if (this.logLevel > LogLevel.ERROR) {
      return;
    }
    this.logger.error(messages.join(" "));
  }
}
export {
  AdkLogger
};
