// @google/adk-devtools/dist/esm/utils/telemetry_utils.js

const Module = await import('node:module'); 
const require = /* @__PURE__ */ (Module?.default?.createRequire ?? Module?.createRequire)(import.meta.url);
const __filename = /* @__PURE__ */ (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = /* @__PURE__ */ (await import("node:path")).dirname(__filename);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  getGcpExporters,
  getGcpResource,
  maybeSetOtelProviders
} from "@google/adk";
function hrTimeToNanoseconds(hrTime) {
  if (!hrTime || !Array.isArray(hrTime) || hrTime.length !== 2) {
    return 0;
  }
  const [seconds, nanoseconds] = hrTime;
  return seconds * 1e9 + nanoseconds;
}
class ApiServerSpanExporter {
  constructor(traceDict) {
    this.traceDict = traceDict;
  }
  export(spans, resultCallback) {
    for (const span of spans) {
      if (span.name === "call_llm" || span.name === "send_data" || span.name.startsWith("execute_tool")) {
        const attributes = { ...span.attributes };
        attributes["trace_id"] = span.spanContext().traceId;
        attributes["span_id"] = span.spanContext().spanId;
        const eventId = attributes["gcp.vertex.agent.event_id"];
        if (eventId) {
          this.traceDict[eventId] = attributes;
        }
      }
    }
    resultCallback({ code: 0 });
  }
  forceFlush() {
    return Promise.resolve();
  }
  shutdown() {
    return Promise.resolve();
  }
}
class InMemoryExporter {
  constructor(traceDict) {
    this.spans = [];
    this.traceDict = traceDict;
  }
  export(spans, resultCallback) {
    for (const span of spans) {
      const traceId = span.spanContext().traceId;
      if (span.name === "call_llm") {
        const attributes = { ...span.attributes };
        const sessionId = attributes["gcp.vertex.agent.session_id"];
        if (sessionId) {
          if (!this.traceDict[sessionId]) {
            this.traceDict[sessionId] = [traceId];
          } else {
            this.traceDict[sessionId].push(traceId);
          }
        }
      }
    }
    this.spans.push(...spans);
    resultCallback({ code: 0 });
  }
  forceFlush() {
    return Promise.resolve();
  }
  shutdown() {
    return Promise.resolve();
  }
  getFinishedSpans(sessionId) {
    const traceIds = this.traceDict[sessionId];
    if (!traceIds || traceIds.length === 0) {
      return [];
    }
    return this.spans.filter(
      (span) => traceIds.includes(span.spanContext().traceId)
    );
  }
  clear() {
    this.spans = [];
  }
}
function otelEnvVarsEnabled() {
  const endpointVars = [
    "OTEL_EXPORTER_OTLP_ENDPOINT",
    "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT",
    "OTEL_EXPORTER_OTLP_METRICS_ENDPOINT",
    "OTEL_EXPORTER_OTLP_LOGS_ENDPOINT"
  ];
  return endpointVars.some((varName) => process.env[varName]);
}
async function setupTelemetry(otelToCloud = false, internalExporters = []) {
  if (otelToCloud) {
    await setupGcpTelemetryExperimental(internalExporters);
  } else if (otelEnvVarsEnabled()) {
    await setupTelemetryFromEnvExperimental(internalExporters);
  } else {
    const otelHooks = {
      spanProcessors: internalExporters
    };
    maybeSetOtelProviders([otelHooks]);
  }
}
async function setupGcpTelemetryExperimental(internalExporters = []) {
  const otelHooksToAdd = [];
  if (internalExporters.length > 0) {
    otelHooksToAdd.push({
      spanProcessors: internalExporters
    });
  }
  const gcpExporters = await getGcpExporters({
    enableTracing: true,
    enableLogging: false,
    enableMetrics: true
  });
  otelHooksToAdd.push(gcpExporters);
  const otelResource = getGcpResource();
  maybeSetOtelProviders(otelHooksToAdd, otelResource);
}
async function setupTelemetryFromEnvExperimental(internalExporters = []) {
  const otelHooksToAdd = [];
  if (internalExporters.length > 0) {
    otelHooksToAdd.push({
      spanProcessors: internalExporters
    });
  }
  maybeSetOtelProviders(otelHooksToAdd);
}
export {
  ApiServerSpanExporter,
  InMemoryExporter,
  hrTimeToNanoseconds,
  setupTelemetry
};
