// @google/adk/dist/cjs/artifacts/gcs_artifact_service.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var gcs_artifact_service_exports = {};
__export(gcs_artifact_service_exports, {
  GcsArtifactService: () => GcsArtifactService
});
module.exports = __toCommonJS(gcs_artifact_service_exports);
var import_storage = require("@google-cloud/storage");
var import_genai = require("@google/genai");
var import_logger = require("../utils/logger.js");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class GcsArtifactService {
  constructor(bucket) {
    this.bucket = new import_storage.Storage().bucket(bucket);
  }
  async saveArtifact(request) {
    if (!request.artifact.inlineData && !request.artifact.text) {
      throw new Error("Artifact must have either inlineData or text content.");
    }
    const versions = await this.listVersions(request);
    const version = versions.length > 0 ? Math.max(...versions) + 1 : 0;
    const file = this.bucket.file(
      getFileName({
        ...request,
        version
      })
    );
    const metadata = request.customMetadata || {};
    if (request.artifact.inlineData) {
      await file.save(
        Buffer.from(request.artifact.inlineData.data || "", "base64"),
        {
          contentType: request.artifact.inlineData.mimeType,
          metadata
        }
      );
      return version;
    }
    await file.save(request.artifact.text, {
      contentType: "text/plain",
      metadata
    });
    return version;
  }
  async loadArtifact(request) {
    try {
      let version = request.version;
      if (version === void 0) {
        const versions = await this.listVersions(request);
        if (versions.length === 0) {
          return void 0;
        }
        version = Math.max(...versions);
      }
      const file = this.bucket.file(
        getFileName({
          ...request,
          version
        })
      );
      const [[metadata], [rawDataBuffer]] = await Promise.all([
        file.getMetadata(),
        file.download()
      ]);
      if (metadata.contentType === "text/plain") {
        return (0, import_genai.createPartFromText)(rawDataBuffer.toString("utf-8"));
      }
      return (0, import_genai.createPartFromBase64)(
        rawDataBuffer.toString("base64"),
        metadata.contentType
      );
    } catch (e) {
      import_logger.logger.warn(
        `[GcsArtifactService] loadArtifact: Failed to load artifact ${request.filename}`,
        e
      );
      return void 0;
    }
  }
  async listArtifactKeys(request) {
    const sessionPrefix = `${request.appName}/${request.userId}/${request.sessionId}/`;
    const usernamePrefix = `${request.appName}/${request.userId}/user/`;
    const [[sessionFiles], [userSessionFiles]] = await Promise.all([
      this.bucket.getFiles({ prefix: sessionPrefix }),
      this.bucket.getFiles({ prefix: usernamePrefix })
    ]);
    return [
      ...extractArtifactKeys(sessionFiles, sessionPrefix),
      ...extractArtifactKeys(userSessionFiles, usernamePrefix, "user:")
    ].sort((a, b) => a.localeCompare(b));
  }
  async deleteArtifact(request) {
    const versions = await this.listVersions(request);
    await Promise.all(
      versions.map((version) => {
        const file = this.bucket.file(
          getFileName({
            ...request,
            version
          })
        );
        return file.delete();
      })
    );
    return;
  }
  async listVersions(request) {
    const prefix = getFileName(request);
    const searchPrefix = prefix + "/";
    const [files] = await this.bucket.getFiles({ prefix: searchPrefix });
    const versions = [];
    for (const file of files) {
      const version = file.name.split("/").pop();
      const v = parseInt(version, 10);
      if (!isNaN(v)) {
        versions.push(v);
      }
    }
    return versions.sort((a, b) => a - b);
  }
  async listArtifactVersions(request) {
    const versions = await this.listVersions(request);
    const artifactVersions = [];
    for (const version of versions) {
      const artifactVersion = await this.getArtifactVersion({
        ...request,
        version
      });
      if (artifactVersion) {
        artifactVersions.push(artifactVersion);
      }
    }
    return artifactVersions;
  }
  async getArtifactVersion(request) {
    try {
      let version = request.version;
      if (version === void 0) {
        const versions = await this.listVersions(request);
        if (versions.length === 0) {
          return void 0;
        }
        version = Math.max(...versions);
      }
      const file = this.bucket.file(
        getFileName({
          ...request,
          version
        })
      );
      const [metadata] = await file.getMetadata();
      return {
        version,
        mimeType: metadata.contentType,
        customMetadata: metadata.metadata,
        canonicalUri: file.publicUrl()
      };
    } catch (e) {
      import_logger.logger.warn(
        `[GcsArtifactService] getArtifactVersion: Failed to get artifact version for userId: ${request.userId} sessionId: ${request.sessionId} filename: ${request.filename} version: ${request.version}`,
        e
      );
      return void 0;
    }
  }
}
function getFileName({
  appName,
  userId,
  sessionId,
  filename,
  version
}) {
  const isUser = filename.startsWith("user:");
  const cleanFilename = isUser ? filename.substring(5) : filename;
  const prefix = isUser ? `${appName}/${userId}/user/${cleanFilename}` : `${appName}/${userId}/${sessionId}/${cleanFilename}`;
  return version !== void 0 ? `${prefix}/${version}` : prefix;
}
function extractArtifactKeys(files, fileNamePrefix, keyPrefix = "") {
  const keys = /* @__PURE__ */ new Set();
  for (const file of files) {
    if (!file.name.startsWith(fileNamePrefix)) {
      continue;
    }
    const relative = file.name.substring(fileNamePrefix.length);
    const name = getFileNameFromPath(relative);
    keys.add(`${keyPrefix}${name}`);
  }
  return [...keys];
}
function getFileNameFromPath(filePath) {
  const parts = filePath.split("/");
  if (parts.length < 2) {
    return filePath;
  }
  return parts.slice(0, -1).join("/");
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  GcsArtifactService
});
