// @google/adk/dist/web/code_executors/unsafe_local_code_executor.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { spawn } from "child_process";
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import { getMimeTypeAndEncoding } from "../utils/file_extension_utils.js";
import { materializeFiles } from "../utils/file_utils.js";
import { logger } from "../utils/logger.js";
import { BaseCodeExecutor } from "./base_code_executor.js";
import {
  CodeExecutionLanguage
} from "./code_execution_utils.js";
const IS_WINDOWS = os.platform() === "win32";
async function createTempScriptFile(code, language, shellCommandPath) {
  const tempDir = path.join(
    os.tmpdir(),
    "adk_js_unsafe_code_executor",
    Date.now().toString() + "_" + Math.random().toString(36).slice(2)
  );
  await fs.mkdir(tempDir, { recursive: true });
  const ext = getExtensionForLanguage(language, shellCommandPath) || ".js";
  const filePath = path.join(tempDir, "script".concat(ext));
  await fs.writeFile(filePath, code);
  return { filePath, tempDir };
}
function getExtensionForLanguage(language, shellCommandPath) {
  if (language === CodeExecutionLanguage.JAVASCRIPT) {
    return ".js";
  }
  if (language === CodeExecutionLanguage.PYTHON) {
    return ".py";
  }
  if (language === CodeExecutionLanguage.POWERSHELL) {
    return ".ps1";
  }
  if (language === CodeExecutionLanguage.WINDOWS_CMD) {
    return ".bat";
  }
  if (language === CodeExecutionLanguage.SHELL) {
    if (IS_WINDOWS) {
      if (shellCommandPath && shellCommandPath.toLowerCase().includes("cmd")) {
        return ".bat";
      }
      return ".ps1";
    }
    return ".sh";
  }
  return void 0;
}
class UnsafeLocalCodeExecutor extends BaseCodeExecutor {
  constructor(options = {}) {
    var _a, _b, _c, _d;
    super();
    this.timeoutSeconds = (_a = options.timeoutSeconds) != null ? _a : 30;
    this.nodeCommandPath = (_b = options.commandPath) != null ? _b : process.execPath;
    this.pythonCommandPath = (_c = options.pythonCommandPath) != null ? _c : IS_WINDOWS ? "python" : "python3";
    this.shellCommandPath = (_d = options.shellCommandPath) != null ? _d : IS_WINDOWS ? "powershell" : "bash";
    this.stateful = false;
    this.optimizeDataFile = false;
  }
  async executeCode(params) {
    var _a;
    const { code, language } = params.codeExecutionInput;
    if (![
      CodeExecutionLanguage.JAVASCRIPT,
      CodeExecutionLanguage.PYTHON,
      CodeExecutionLanguage.SHELL,
      CodeExecutionLanguage.WINDOWS_CMD,
      CodeExecutionLanguage.POWERSHELL
    ].includes(language)) {
      return {
        stdout: "",
        stderr: "Unsupported language: ".concat(language),
        outputFiles: []
      };
    }
    logger.warn(
      "\n====================================================================================\n\u26A0\uFE0F DANGER: UnsafeLocalCodeExecutor is executing code locally on this host machine!\nThis component provides NO sandboxing or container isolation. Arbitrary shell/script\ncommands generated by AI or untrusted sources could lead to serious security risks.\n====================================================================================\n"
    );
    let tempDir;
    try {
      const res = await createTempScriptFile(
        code,
        language,
        this.shellCommandPath
      );
      const filePath = res.filePath;
      tempDir = res.tempDir;
      if (params.codeExecutionInput.inputFiles) {
        await materializeFiles(params.codeExecutionInput.inputFiles, tempDir);
      }
      let command = this.nodeCommandPath;
      let args = [filePath];
      if (language === CodeExecutionLanguage.PYTHON) {
        command = this.pythonCommandPath;
      } else if (language === CodeExecutionLanguage.SHELL) {
        command = this.shellCommandPath;
        if (this.shellCommandPath.toLowerCase().includes("powershell")) {
          args = ["-NoLogo", "-ExecutionPolicy", "Bypass", "-File", filePath];
        } else if (this.shellCommandPath.toLowerCase().includes("cmd")) {
          args = ["/c", filePath];
        }
      } else if (language === CodeExecutionLanguage.POWERSHELL) {
        command = IS_WINDOWS ? "powershell" : "pwsh";
        args = ["-NoLogo", "-ExecutionPolicy", "Bypass", "-File", filePath];
      } else if (language === CodeExecutionLanguage.WINDOWS_CMD) {
        command = "cmd.exe";
        args = ["/c", filePath];
      }
      if (params.codeExecutionInput.args) {
        if (Array.isArray(params.codeExecutionInput.args)) {
          args.push(...params.codeExecutionInput.args);
        } else {
          for (const [k, v] of Object.entries(params.codeExecutionInput.args)) {
            args.push("--".concat(k), String(v));
          }
        }
      }
      const executionResult = await new Promise((resolve) => {
        const child = spawn(command, args, {
          timeout: this.timeoutSeconds * 1e3,
          killSignal: "SIGKILL",
          cwd: tempDir
        });
        let stdout = "";
        let stderr = "";
        if (child.stdout) {
          child.stdout.on("data", (data) => {
            stdout += data.toString();
          });
        }
        if (child.stderr) {
          child.stderr.on("data", (data) => {
            stderr += data.toString();
          });
        }
        child.on("error", (err) => {
          stderr += "Process error: ".concat(err.message, "\n");
        });
        child.on("close", (exitCode, signal) => {
          if (signal === "SIGKILL" || signal === "SIGTERM") {
            stderr += "\nCode execution timed out after ".concat(this.timeoutSeconds, " seconds.");
          } else if (exitCode !== 0 && exitCode !== null) {
            if (!stderr) {
              stderr = "Exit code ".concat(exitCode);
            }
          }
          resolve({ stdout, stderr, exitCode });
        });
      });
      const outputFiles = [];
      try {
        const allFiles = await fs.readdir(tempDir, { recursive: true });
        for (const relativeFilePath of allFiles) {
          const fullPath = path.join(tempDir, relativeFilePath);
          const stat = await fs.lstat(fullPath);
          if (!stat.isFile()) {
            continue;
          }
          if (relativeFilePath === path.basename(filePath)) {
            continue;
          }
          const isInputFile = (_a = params.codeExecutionInput.inputFiles) == null ? void 0 : _a.some(
            (f) => f.name === relativeFilePath
          );
          if (isInputFile) {
            continue;
          }
          const fileContent = await fs.readFile(fullPath);
          const { mimeType, encoding } = getMimeTypeAndEncoding(
            path.extname(relativeFilePath)
          );
          outputFiles.push({
            name: relativeFilePath,
            content: fileContent.toString(encoding),
            contentEncoding: encoding,
            mimeType
          });
        }
      } catch (e) {
        logger.error("Error scanning output files: ".concat(e));
      }
      return {
        stdout: executionResult.stdout,
        stderr: executionResult.stderr,
        outputFiles
      };
    } finally {
      if (tempDir) {
        await fs.rm(tempDir, { recursive: true, force: true });
      }
    }
  }
}
export {
  UnsafeLocalCodeExecutor
};
