// @google/adk/dist/cjs/features/feature_registry.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var feature_registry_exports = {};
__export(feature_registry_exports, {
  FeatureName: () => FeatureName,
  FeatureStage: () => FeatureStage,
  getFeatureConfig: () => getFeatureConfig,
  isFeatureEnabled: () => isFeatureEnabled,
  overrideFeatureEnabled: () => overrideFeatureEnabled,
  registerFeature: () => registerFeature,
  withTemporaryFeatureOverride: () => withTemporaryFeatureOverride
});
module.exports = __toCommonJS(feature_registry_exports);
var import_env_aware_utils = require("../utils/env_aware_utils.js");
var import_logger = require("../utils/logger.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var FeatureName = /* @__PURE__ */ ((FeatureName2) => {
  FeatureName2["PROGRESSIVE_SSE_STREAMING"] = "PROGRESSIVE_SSE_STREAMING";
  return FeatureName2;
})(FeatureName || {});
var FeatureStage = /* @__PURE__ */ ((FeatureStage2) => {
  FeatureStage2["WIP"] = "wip";
  FeatureStage2["EXPERIMENTAL"] = "experimental";
  FeatureStage2["STABLE"] = "stable";
  return FeatureStage2;
})(FeatureStage || {});
const FEATURE_REGISTRY = {
  ["PROGRESSIVE_SSE_STREAMING" /* PROGRESSIVE_SSE_STREAMING */]: {
    stage: "experimental" /* EXPERIMENTAL */,
    defaultOn: false
  }
};
const WARNED_FEATURES = /* @__PURE__ */ new Set();
const FEATURE_OVERRIDES = {};
function getFeatureConfig(featureName) {
  return FEATURE_REGISTRY[featureName];
}
function registerFeature(featureName, config) {
  FEATURE_REGISTRY[featureName] = config;
}
function overrideFeatureEnabled(featureName, enabled) {
  const config = getFeatureConfig(featureName);
  if (!config) {
    throw new Error(`Feature ${featureName} is not registered.`);
  }
  if (enabled === void 0) {
    delete FEATURE_OVERRIDES[featureName];
  } else {
    FEATURE_OVERRIDES[featureName] = enabled;
  }
}
function isFeatureEnabled(featureName) {
  const config = getFeatureConfig(featureName);
  if (!config) {
    throw new Error(`Feature ${featureName} is not registered.`);
  }
  if (featureName in FEATURE_OVERRIDES) {
    const enabled = FEATURE_OVERRIDES[featureName];
    if (enabled && config.stage !== "stable" /* STABLE */) {
      emitNonStableWarningOnce(featureName, config.stage);
    }
    return enabled;
  }
  const enableVar = `ADK_ENABLE_${featureName}`;
  const disableVar = `ADK_DISABLE_${featureName}`;
  if ((0, import_env_aware_utils.getBooleanEnvVar)(enableVar)) {
    if (config.stage !== "stable" /* STABLE */) {
      emitNonStableWarningOnce(featureName, config.stage);
    }
    return true;
  }
  if ((0, import_env_aware_utils.getBooleanEnvVar)(disableVar)) {
    return false;
  }
  if (config.stage !== "stable" /* STABLE */ && config.defaultOn) {
    emitNonStableWarningOnce(featureName, config.stage);
  }
  return config.defaultOn;
}
function emitNonStableWarningOnce(featureName, featureStage) {
  if (!WARNED_FEATURES.has(featureName)) {
    WARNED_FEATURES.add(featureName);
    import_logger.logger.warn(
      `[${featureStage.toUpperCase()}] feature ${featureName} is enabled.`
    );
  }
}
async function withTemporaryFeatureOverride(featureName, enabled, callback) {
  const config = getFeatureConfig(featureName);
  if (!config) {
    throw new Error(`Feature ${featureName} is not registered.`);
  }
  const hadOverride = featureName in FEATURE_OVERRIDES;
  const originalValue = FEATURE_OVERRIDES[featureName];
  FEATURE_OVERRIDES[featureName] = enabled;
  try {
    return await callback();
  } finally {
    if (hadOverride) {
      FEATURE_OVERRIDES[featureName] = originalValue;
    } else {
      delete FEATURE_OVERRIDES[featureName];
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FeatureName,
  FeatureStage,
  getFeatureConfig,
  isFeatureEnabled,
  overrideFeatureEnabled,
  registerFeature,
  withTemporaryFeatureOverride
});
