// @google/adk/dist/web/context/summarizers/llm_summarizer.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __forAwait = (obj, it, method) => (it = obj[__knownSymbol("asyncIterator")]) ? it.call(obj) : (obj = obj[__knownSymbol("iterator")](), it = {}, method = (key, fn) => (fn = obj[key]) && (it[key] = (arg) => new Promise((yes, no, done) => (arg = fn.call(obj, arg), done = arg.done, Promise.resolve(arg.value).then((value) => yes({ value, done }), no)))), method("next"), method("return"), it);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  createCompactedEvent
} from "../../events/compacted_event.js";
import { stringifyContent } from "../../events/event.js";
const DEFAULT_PROMPT = "The following is a conversation history between a user and an AI agent. Please summarize the conversation, focusing on key information and decisions made, as well as any unresolved questions or tasks. The summary should be concise and capture the essence of the interaction.";
class LlmSummarizer {
  constructor(options) {
    this.llm = options.llm;
    this.prompt = options.prompt || DEFAULT_PROMPT;
  }
  async summarize(events) {
    var _a, _b, _c, _d, _e, _f;
    if (events.length === 0) {
      throw new Error("Cannot summarize an empty list of events.");
    }
    const startTime = events[0].timestamp;
    const endTime = events[events.length - 1].timestamp;
    let formattedEvents = "";
    for (let i = 0; i < events.length; i++) {
      formattedEvents += "[Event ".concat(i + 1, " - Author: ").concat(events[i].author, "]\n");
      formattedEvents += "".concat(stringifyContent(events[i]), "\n\n");
    }
    const fullPrompt = "".concat(this.prompt, "\n\n").concat(formattedEvents);
    const request = {
      contents: [{ role: "user", parts: [{ text: fullPrompt }] }],
      toolsDict: {},
      liveConnectConfig: {}
    };
    const responseGen = this.llm.generateContentAsync(request, false);
    let compactedContent = "";
    const firstResponse = await responseGen.next();
    if (firstResponse.done || !((_c = (_b = (_a = firstResponse.value.content) == null ? void 0 : _a.parts) == null ? void 0 : _b[0]) == null ? void 0 : _c.text)) {
      throw new Error("LLM failed to return a valid summary.");
    }
    compactedContent += firstResponse.value.content.parts[0].text;
    try {
      for (var iter = __forAwait(responseGen), more, temp, error; more = !(temp = await iter.next()).done; more = false) {
        const chunk = temp.value;
        if ((_f = (_e = (_d = chunk.content) == null ? void 0 : _d.parts) == null ? void 0 : _e[0]) == null ? void 0 : _f.text) {
          compactedContent += chunk.content.parts[0].text;
        }
      }
    } catch (temp) {
      error = [temp];
    } finally {
      try {
        more && (temp = iter.return) && await temp.call(iter);
      } finally {
        if (error)
          throw error[0];
      }
    }
    return createCompactedEvent({
      author: "system",
      content: {
        role: "model",
        parts: [{ text: compactedContent }]
      },
      startTime,
      endTime,
      compactedContent
    });
  }
}
export {
  LlmSummarizer
};
