// @google/adk/dist/esm/utils/gemini_schema_util.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { Type } from "@google/genai";
import { z } from "zod";
const MCPToolSchemaObject = z.object({
  type: z.literal("object"),
  properties: z.record(z.string(), z.unknown()).optional(),
  required: z.string().array().optional()
});
function toGeminiType(mcpType) {
  if (!mcpType) return Type.TYPE_UNSPECIFIED;
  switch (mcpType.toLowerCase()) {
    case "text":
    case "string":
      return Type.STRING;
    case "number":
      return Type.NUMBER;
    case "boolean":
      return Type.BOOLEAN;
    case "integer":
      return Type.INTEGER;
    case "array":
      return Type.ARRAY;
    case "object":
      return Type.OBJECT;
    case "null":
      return Type.NULL;
    default:
      return Type.TYPE_UNSPECIFIED;
  }
}
const getTypeFromArrayItem = (mcpType) => {
  var _a, _b;
  if (typeof mcpType === "string") {
    return mcpType.toLowerCase();
  }
  return (_b = (_a = mcpType == null ? void 0 : mcpType.type) == null ? void 0 : _a.toLowerCase) == null ? void 0 : _b.call(_a);
};
function toGeminiSchema(mcpSchema) {
  if (!mcpSchema) {
    return void 0;
  }
  function recursiveConvert(mcp) {
    var _a;
    const sourceType = (_a = mcp.anyOf) != null ? _a : mcp.type;
    let isNullable = false;
    let nonNullTypes;
    if (Array.isArray(sourceType)) {
      nonNullTypes = sourceType.filter(
        (t) => getTypeFromArrayItem(t) !== "null"
      );
      isNullable = sourceType.some(
        (t) => getTypeFromArrayItem(t) === "null"
      );
      if (nonNullTypes.length === 1) {
        const nonNullType = nonNullTypes[0];
        if (typeof nonNullType === "object") {
          mcp = nonNullType;
        } else {
          const { type: _removed, anyOf: _removedAnyOf, ...rest } = mcp;
          mcp = { ...rest, type: nonNullType };
        }
      } else if (nonNullTypes.length === 0 && isNullable) {
        const { type: _removed, anyOf: _removedAnyOf, ...rest } = mcp;
        mcp = { ...rest, type: "null" };
      } else if (typeof mcp.anyOf === "undefined") {
        const anyOfItems = mcp.type.map((t) => ({ type: t }));
        const { type: _removed, ...rest } = mcp;
        mcp = { ...rest, anyOf: anyOfItems };
      }
    }
    if (!mcp.type) {
      if (mcp.properties || mcp.$ref) {
        mcp.type = "object";
      } else if (mcp.items) {
        mcp.type = "array";
      } else if (isNullable) {
        mcp.type = "null";
      }
    }
    const geminiType = toGeminiType(mcp.type);
    const geminiSchema = {};
    if (mcp.anyOf) {
      geminiSchema.anyOf = mcp.anyOf.map(
        (item) => recursiveConvert(item)
      );
    } else {
      geminiSchema.type = geminiType;
    }
    if (mcp.description) {
      geminiSchema.description = mcp.description;
    }
    if (isNullable && mcp.type !== "null") {
      geminiSchema.nullable = true;
    }
    if (geminiType === Type.OBJECT) {
      geminiSchema.properties = {};
      if (mcp.properties) {
        for (const name in mcp.properties) {
          geminiSchema.properties[name] = recursiveConvert(
            mcp.properties[name]
          );
        }
      }
      if (mcp.required) {
        geminiSchema.required = mcp.required;
      }
    } else if (geminiType === Type.ARRAY) {
      if (mcp.items) {
        geminiSchema.items = recursiveConvert(mcp.items);
      }
    }
    return geminiSchema;
  }
  return recursiveConvert(mcpSchema);
}
export {
  toGeminiSchema
};
