// @google/adk/dist/esm/utils/logger.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import * as winston from "winston";
var LogLevel = /* @__PURE__ */ ((LogLevel2) => {
  LogLevel2[LogLevel2["DEBUG"] = 0] = "DEBUG";
  LogLevel2[LogLevel2["INFO"] = 1] = "INFO";
  LogLevel2[LogLevel2["WARN"] = 2] = "WARN";
  LogLevel2[LogLevel2["ERROR"] = 3] = "ERROR";
  return LogLevel2;
})(LogLevel || {});
class SimpleLogger {
  constructor() {
    this.logLevel = 1 /* INFO */;
    this.logger = winston.createLogger({
      levels: {
        "debug": 0 /* DEBUG */,
        "info": 1 /* INFO */,
        "warn": 2 /* WARN */,
        "error": 3 /* ERROR */
      },
      level: "error",
      format: winston.format.combine(
        winston.format.label({ label: "ADK" }),
        winston.format((info) => {
          info.level = info.level.toUpperCase();
          return info;
        })(),
        winston.format.colorize(),
        winston.format.timestamp(),
        winston.format.printf((info) => {
          return `${info.level}: [${info.label}] ${info.timestamp} ${info.message}`;
        })
      ),
      transports: [new winston.transports.Console()]
    });
  }
  setLogLevel(level) {
    this.logLevel = level;
  }
  log(level, ...messages) {
    if (this.logLevel > level) {
      return;
    }
    this.logger.log(level.toString(), messages.join(" "));
  }
  debug(...messages) {
    if (this.logLevel > 0 /* DEBUG */) {
      return;
    }
    this.logger.debug(messages.join(" "));
  }
  info(...messages) {
    if (this.logLevel > 1 /* INFO */) {
      return;
    }
    this.logger.info(messages.join(" "));
  }
  warn(...messages) {
    if (this.logLevel > 2 /* WARN */) {
      return;
    }
    this.logger.warn(messages.join(" "));
  }
  error(...messages) {
    if (this.logLevel > 3 /* ERROR */) {
      return;
    }
    this.logger.error(messages.join(" "));
  }
}
class NoOpLogger {
  setLogLevel(_level) {
  }
  log(_level, ..._args) {
  }
  debug(..._args) {
  }
  info(..._args) {
  }
  warn(..._args) {
  }
  error(..._args) {
  }
}
let currentLogger = new SimpleLogger();
function setLogger(customLogger) {
  currentLogger = customLogger != null ? customLogger : new NoOpLogger();
}
function getLogger() {
  return currentLogger;
}
function resetLogger() {
  currentLogger = new SimpleLogger();
}
function setLogLevel(level) {
  logger.setLogLevel(level);
}
const logger = {
  setLogLevel(level) {
    currentLogger.setLogLevel(level);
  },
  log(level, ...args) {
    currentLogger.log(level, ...args);
  },
  debug(...args) {
    currentLogger.debug(...args);
  },
  info(...args) {
    currentLogger.info(...args);
  },
  warn(...args) {
    currentLogger.warn(...args);
  },
  error(...args) {
    currentLogger.error(...args);
  }
};
export {
  LogLevel,
  getLogger,
  logger,
  resetLogger,
  setLogLevel,
  setLogger
};
