// @google/adk/dist/esm/utils/streaming_utils.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  FinishReason
} from "@google/genai";
import { JSONPath } from "jsonpath-plus";
import { generateClientFunctionCallId } from "../agents/functions.js";
import { FeatureName, isFeatureEnabled } from "../features/feature_registry.js";
import { createLlmResponse } from "../models/llm_response.js";
class StreamingResponseAggregator {
  constructor(isProgressiveMode = isFeatureEnabled(
    FeatureName.PROGRESSIVE_SSE_STREAMING
  )) {
    this.isProgressiveMode = isProgressiveMode;
    // For non-progressive SSE streaming mode
    this.text = "";
    this.thoughtText = "";
    // For progressive SSE streaming mode: accumulate parts in order
    this.partsSequence = [];
    this.currentTextBuffer = "";
    this.currentFcArgs = {};
  }
  flushTextBufferToSequence() {
    if (!this.currentTextBuffer) {
      return;
    }
    if (this.currentTextIsThought) {
      this.partsSequence.push({
        text: this.currentTextBuffer,
        thought: true
      });
    } else {
      this.partsSequence.push({
        text: this.currentTextBuffer
      });
    }
    this.currentTextBuffer = "";
    this.currentTextIsThought = void 0;
  }
  getValueFromPartialArg(partialArg, jsonPath) {
    let value = null;
    let hasValue = false;
    const stringValue = partialArg.stringValue;
    const numberValue = partialArg.numberValue;
    const boolValue = partialArg.boolValue;
    const nullValue = partialArg.nullValue;
    if (stringValue !== void 0) {
      const stringChunk = stringValue;
      hasValue = true;
      const pathParts = JSONPath.toPathArray(jsonPath).filter(
        (p) => p !== "$" && p !== "$["
      );
      let existingValue = this.currentFcArgs;
      for (const part of pathParts) {
        if (existingValue && typeof existingValue === "object" && part in existingValue) {
          existingValue = existingValue[part];
        } else {
          existingValue = void 0;
          break;
        }
      }
      if (typeof existingValue === "string") {
        value = existingValue + stringChunk;
      } else {
        value = stringChunk;
      }
    } else if (numberValue !== void 0) {
      value = numberValue;
      hasValue = true;
    } else if (boolValue !== void 0) {
      value = boolValue;
      hasValue = true;
    } else if (nullValue !== void 0) {
      value = null;
      hasValue = true;
    }
    return [value, hasValue];
  }
  setValueByJsonPath(jsonPath, value) {
    const pathParts = JSONPath.toPathArray(jsonPath).filter(
      (p) => p !== "$" && p !== "$["
    );
    let current = this.currentFcArgs;
    for (let i = 0; i < pathParts.length - 1; i++) {
      const part = pathParts[i];
      if (!(part in current)) {
        current[part] = {};
      }
      current = current[part];
    }
    if (pathParts.length > 0) {
      current[pathParts[pathParts.length - 1]] = value;
    }
  }
  flushFunctionCallToSequence() {
    var _a;
    if (this.currentFcName) {
      const fcPart = {
        functionCall: {
          name: this.currentFcName,
          args: JSON.parse(JSON.stringify(this.currentFcArgs)),
          id: (_a = this.currentFcId) != null ? _a : generateClientFunctionCallId()
        }
      };
      if (this.currentThoughtSignature) {
        fcPart.thoughtSignature = this.currentThoughtSignature.toString();
      }
      this.partsSequence.push(fcPart);
      this.currentFcName = void 0;
      this.currentFcArgs = {};
      this.currentFcId = void 0;
      this.currentThoughtSignature = void 0;
    }
  }
  processStreamingFunctionCall(fc) {
    if (fc.name) {
      this.currentFcName = fc.name;
    }
    if (fc.id) {
      this.currentFcId = fc.id;
    }
    for (const partialArg of fc.partialArgs || []) {
      const jsonPath = partialArg.jsonPath;
      if (!jsonPath) {
        continue;
      }
      const [value, hasValue] = this.getValueFromPartialArg(
        partialArg,
        jsonPath
      );
      if (hasValue) {
        this.setValueByJsonPath(jsonPath, value);
      }
    }
    if (!fc.willContinue) {
      this.flushTextBufferToSequence();
      this.flushFunctionCallToSequence();
    }
  }
  processFunctionCallPart(part) {
    const fc = part.functionCall;
    if (!fc) {
      return;
    }
    if (part.thoughtSignature) {
      this.lastThoughtSignature = part.thoughtSignature;
    } else if (this.lastThoughtSignature) {
      part.thoughtSignature = this.lastThoughtSignature.toString();
    }
    if (fc.partialArgs || fc.willContinue) {
      if (!fc.id && !this.currentFcId) {
        fc.id = generateClientFunctionCallId();
      }
      if (part.thoughtSignature && !this.currentThoughtSignature) {
        this.currentThoughtSignature = part.thoughtSignature;
      }
      this.processStreamingFunctionCall(fc);
    } else {
      if (fc.name) {
        if (!fc.id) {
          fc.id = generateClientFunctionCallId();
        }
        this.flushTextBufferToSequence();
        this.partsSequence.push(part);
      }
    }
  }
  async *processResponse(response) {
    var _a, _b, _c;
    this.response = response;
    const llmResponse = createLlmResponse(response);
    this.usageMetadata = llmResponse.usageMetadata;
    if (llmResponse.groundingMetadata) {
      this.groundingMetadata = llmResponse.groundingMetadata;
    }
    if (llmResponse.citationMetadata) {
      this.citationMetadata = llmResponse.citationMetadata;
    }
    if (llmResponse.finishReason) {
      this.finishReason = llmResponse.finishReason;
    }
    if (llmResponse.content && llmResponse.content.parts) {
      for (const part of llmResponse.content.parts) {
        if (part.thoughtSignature) {
          this.lastThoughtSignature = part.thoughtSignature;
        } else if (part.functionCall && this.lastThoughtSignature) {
          part.thoughtSignature = this.lastThoughtSignature.toString();
        }
      }
    }
    if (this.isProgressiveMode) {
      if (llmResponse.content && llmResponse.content.parts) {
        for (const part of llmResponse.content.parts) {
          if (part.text) {
            const isThought = (_a = part.thought) != null ? _a : false;
            if (this.currentTextBuffer && isThought !== this.currentTextIsThought) {
              this.flushTextBufferToSequence();
            }
            if (!this.currentTextBuffer) {
              this.currentTextIsThought = isThought;
            }
            this.currentTextBuffer += part.text;
          } else if (part.functionCall) {
            this.processFunctionCallPart(part);
          } else {
            this.flushTextBufferToSequence();
            this.partsSequence.push(part);
          }
        }
      }
      llmResponse.partial = true;
      yield llmResponse;
      return;
    }
    if (llmResponse.content && llmResponse.content.parts && typeof ((_b = llmResponse.content.parts[0]) == null ? void 0 : _b.text) === "string") {
      const part0 = llmResponse.content.parts[0];
      const partText = part0.text || "";
      if (part0.thought) {
        this.thoughtText += partText;
      } else {
        this.text += partText;
      }
      llmResponse.partial = true;
    } else if ((this.thoughtText || this.text) && (!llmResponse.content || !llmResponse.content.parts || !((_c = llmResponse.content.parts[0]) == null ? void 0 : _c.inlineData))) {
      const parts = [];
      if (this.thoughtText) {
        parts.push({ text: this.thoughtText, thought: true });
      }
      if (this.text) {
        parts.push({ text: this.text });
      }
      yield {
        content: {
          role: "model",
          parts
        },
        usageMetadata: llmResponse.usageMetadata,
        partial: false
      };
      this.thoughtText = "";
      this.text = "";
    }
    yield llmResponse;
  }
  close() {
    var _a, _b;
    if (!((_a = this.response) == null ? void 0 : _a.candidates) || this.response.candidates.length === 0) {
      return;
    }
    if (this.isProgressiveMode) {
      this.flushTextBufferToSequence();
      this.flushFunctionCallToSequence();
      const finalParts = this.partsSequence;
      if (finalParts.length === 0) {
        return;
      }
      const candidate = this.response.candidates[0];
      const finishReason = (_b = this.finishReason) != null ? _b : candidate.finishReason;
      return {
        content: {
          role: "model",
          parts: finalParts
        },
        groundingMetadata: this.groundingMetadata,
        citationMetadata: this.citationMetadata,
        errorCode: finishReason === FinishReason.STOP ? void 0 : finishReason,
        errorMessage: finishReason === FinishReason.STOP ? void 0 : candidate.finishMessage,
        usageMetadata: this.usageMetadata,
        finishReason,
        partial: false
      };
    }
    if ((this.text || this.thoughtText) && this.response.candidates && this.response.candidates.length > 0) {
      const parts = [];
      if (this.thoughtText) {
        parts.push({ text: this.thoughtText, thought: true });
      }
      if (this.text) {
        parts.push({ text: this.text });
      }
      const candidate = this.response.candidates[0];
      const finishReason = candidate.finishReason;
      return {
        content: {
          role: "model",
          parts
        },
        groundingMetadata: this.groundingMetadata,
        citationMetadata: this.citationMetadata,
        errorCode: finishReason === FinishReason.STOP ? void 0 : finishReason,
        errorMessage: finishReason === FinishReason.STOP ? void 0 : candidate.finishMessage,
        usageMetadata: this.usageMetadata,
        finishReason,
        partial: false
      };
    }
    return void 0;
  }
}
export {
  StreamingResponseAggregator
};
