// @google/adk/dist/cjs/models/apigee_llm.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var apigee_llm_exports = {};
__export(apigee_llm_exports, {
  ApigeeLlm: () => ApigeeLlm
});
module.exports = __toCommonJS(apigee_llm_exports);
var import_env_aware_utils = require("../utils/env_aware_utils.js");
var import_logger = require("../utils/logger.js");
var import_google_llm = require("./google_llm.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const APIGEE_PROXY_URL_ENV_VARIABLE_NAME = "APIGEE_PROXY_URL";
class ApigeeLlm extends import_google_llm.Gemini {
  constructor({
    model,
    proxyUrl,
    apiKey,
    vertexai,
    location,
    project,
    headers
  }) {
    var _a;
    if (!validateModel(model)) {
      throw new Error(
        `Model ${model} is not a valid Apigee model, expected apigee/[<provider>/][<version>/]<model_id>`
      );
    }
    super({
      ...apigeeToGeminiInitParams({ model, vertexai, project, location, apiKey }),
      headers
    });
    this.proxyUrl = proxyUrl != null ? proxyUrl : "";
    if (!(0, import_env_aware_utils.isBrowser)() && !this.proxyUrl) {
      this.proxyUrl = (_a = process.env[APIGEE_PROXY_URL_ENV_VARIABLE_NAME]) != null ? _a : "";
    }
    if (!this.proxyUrl) {
      throw new Error(
        `Proxy URL must be provided via the constructor or ${APIGEE_PROXY_URL_ENV_VARIABLE_NAME} environment variable.`
      );
    }
  }
  getHttpOptions() {
    const opts = super.getHttpOptions();
    opts.baseUrl = this.proxyUrl;
    return opts;
  }
  getLiveHttpOptions() {
    const opts = super.getLiveHttpOptions();
    opts.baseUrl = this.proxyUrl;
    return opts;
  }
  identifyApiVersion() {
    const modelTrimmed = this.model.startsWith("apigee/") ? this.model.substring("apigee/".length) : this.model;
    const components = modelTrimmed.split("/");
    if (components.length === 3) {
      return components[1];
    }
    if (components.length === 2) {
      if (components[0] != "vertex_ai" && components[0] != "gemini" && components[0].startsWith("v")) {
        return components[0];
      }
    }
    return this.vertexai ? "v1beta1" : "v1alpha";
  }
  get liveApiVersion() {
    if (!this._apigeeLiveApiVersion) {
      this._apigeeLiveApiVersion = this.identifyApiVersion();
    }
    return this._apigeeLiveApiVersion;
  }
  async *generateContentAsync(llmRequest, stream = false, abortSignal) {
    var _a;
    const modelToUse = (_a = llmRequest.model) != null ? _a : this.model;
    llmRequest.model = getModelId(modelToUse);
    yield* super.generateContentAsync(llmRequest, stream, abortSignal);
  }
  async connect(llmRequest) {
    var _a;
    const modelToUse = (_a = llmRequest.model) != null ? _a : this.model;
    llmRequest.model = getModelId(modelToUse);
    return super.connect(llmRequest);
  }
}
/**
 * A list of model name patterns that are supported by this LLM.
 *
 * @returns A list of supported models.
 */
ApigeeLlm.supportedModels = [
  /apigee\/.*/
];
function apigeeToGeminiInitParams({
  model,
  vertexai,
  project,
  location,
  apiKey
}) {
  var _a;
  const params = (0, import_google_llm.geminiInitParams)({ model, vertexai, project, location, apiKey });
  params.vertexai = params.vertexai || ((_a = params.model) == null ? void 0 : _a.startsWith("apigee/vertex_ai/"));
  if (params.vertexai) {
    return params;
  }
  if (!params.apiKey) {
    import_logger.logger.warn(
      `No API key provided when using a Gemini model, using a fake key "-".`
    );
    params.apiKey = "-";
  }
  return params;
}
function getModelId(model) {
  if (!validateModel(model)) {
    throw new Error(
      `Model ${model} is not a valid Apigee model, expected apigee/[<provider>/][<version>/]<model_id>`
    );
  }
  const components = model.split("/");
  return components[components.length - 1];
}
function validateModel(model) {
  const validProviders = ["vertex_ai", "gemini"];
  if (!model.startsWith("apigee/")) {
    return false;
  }
  const modelPart = model.substring("apigee/".length);
  if (modelPart.length === 0) {
    return false;
  }
  const components = modelPart.split("/", -1);
  if (components[components.length - 1].length === 0) {
    return false;
  }
  if (components.length == 1) {
    return true;
  }
  if (components.length == 2) {
    if (validProviders.includes(components[0])) {
      return true;
    }
    return components[0].startsWith("v");
  }
  if (components.length == 3) {
    if (!validProviders.includes(components[0])) {
      return false;
    }
    return components[1].startsWith("v");
  }
  return false;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApigeeLlm
});
