// @google/adk/dist/esm/models/apigee_llm.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { isBrowser } from "../utils/env_aware_utils.js";
import { logger } from "../utils/logger.js";
import { Gemini, geminiInitParams } from "./google_llm.js";
const APIGEE_PROXY_URL_ENV_VARIABLE_NAME = "APIGEE_PROXY_URL";
class ApigeeLlm extends Gemini {
  constructor({
    model,
    proxyUrl,
    apiKey,
    vertexai,
    location,
    project,
    headers
  }) {
    var _a;
    if (!validateModel(model)) {
      throw new Error(
        `Model ${model} is not a valid Apigee model, expected apigee/[<provider>/][<version>/]<model_id>`
      );
    }
    super({
      ...apigeeToGeminiInitParams({ model, vertexai, project, location, apiKey }),
      headers
    });
    this.proxyUrl = proxyUrl != null ? proxyUrl : "";
    if (!isBrowser() && !this.proxyUrl) {
      this.proxyUrl = (_a = process.env[APIGEE_PROXY_URL_ENV_VARIABLE_NAME]) != null ? _a : "";
    }
    if (!this.proxyUrl) {
      throw new Error(
        `Proxy URL must be provided via the constructor or ${APIGEE_PROXY_URL_ENV_VARIABLE_NAME} environment variable.`
      );
    }
  }
  getHttpOptions() {
    const opts = super.getHttpOptions();
    opts.baseUrl = this.proxyUrl;
    return opts;
  }
  getLiveHttpOptions() {
    const opts = super.getLiveHttpOptions();
    opts.baseUrl = this.proxyUrl;
    return opts;
  }
  identifyApiVersion() {
    const modelTrimmed = this.model.startsWith("apigee/") ? this.model.substring("apigee/".length) : this.model;
    const components = modelTrimmed.split("/");
    if (components.length === 3) {
      return components[1];
    }
    if (components.length === 2) {
      if (components[0] != "vertex_ai" && components[0] != "gemini" && components[0].startsWith("v")) {
        return components[0];
      }
    }
    return this.vertexai ? "v1beta1" : "v1alpha";
  }
  get liveApiVersion() {
    if (!this._apigeeLiveApiVersion) {
      this._apigeeLiveApiVersion = this.identifyApiVersion();
    }
    return this._apigeeLiveApiVersion;
  }
  async *generateContentAsync(llmRequest, stream = false, abortSignal) {
    var _a;
    const modelToUse = (_a = llmRequest.model) != null ? _a : this.model;
    llmRequest.model = getModelId(modelToUse);
    yield* super.generateContentAsync(llmRequest, stream, abortSignal);
  }
  async connect(llmRequest) {
    var _a;
    const modelToUse = (_a = llmRequest.model) != null ? _a : this.model;
    llmRequest.model = getModelId(modelToUse);
    return super.connect(llmRequest);
  }
}
/**
 * A list of model name patterns that are supported by this LLM.
 *
 * @returns A list of supported models.
 */
ApigeeLlm.supportedModels = [
  /apigee\/.*/
];
function apigeeToGeminiInitParams({
  model,
  vertexai,
  project,
  location,
  apiKey
}) {
  var _a;
  const params = geminiInitParams({ model, vertexai, project, location, apiKey });
  params.vertexai = params.vertexai || ((_a = params.model) == null ? void 0 : _a.startsWith("apigee/vertex_ai/"));
  if (params.vertexai) {
    return params;
  }
  if (!params.apiKey) {
    logger.warn(
      `No API key provided when using a Gemini model, using a fake key "-".`
    );
    params.apiKey = "-";
  }
  return params;
}
function getModelId(model) {
  if (!validateModel(model)) {
    throw new Error(
      `Model ${model} is not a valid Apigee model, expected apigee/[<provider>/][<version>/]<model_id>`
    );
  }
  const components = model.split("/");
  return components[components.length - 1];
}
function validateModel(model) {
  const validProviders = ["vertex_ai", "gemini"];
  if (!model.startsWith("apigee/")) {
    return false;
  }
  const modelPart = model.substring("apigee/".length);
  if (modelPart.length === 0) {
    return false;
  }
  const components = modelPart.split("/", -1);
  if (components[components.length - 1].length === 0) {
    return false;
  }
  if (components.length == 1) {
    return true;
  }
  if (components.length == 2) {
    if (validProviders.includes(components[0])) {
      return true;
    }
    return components[0].startsWith("v");
  }
  if (components.length == 3) {
    if (!validProviders.includes(components[0])) {
      return false;
    }
    return components[1].startsWith("v");
  }
  return false;
}
export {
  ApigeeLlm
};
