// @google/adk/dist/cjs/utils/gemini_schema_util.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var gemini_schema_util_exports = {};
__export(gemini_schema_util_exports, {
  toGeminiSchema: () => toGeminiSchema
});
module.exports = __toCommonJS(gemini_schema_util_exports);
var import_genai = require("@google/genai");
var import_zod = require("zod");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const MCPToolSchemaObject = import_zod.z.object({
  type: import_zod.z.literal("object"),
  properties: import_zod.z.record(import_zod.z.string(), import_zod.z.unknown()).optional(),
  required: import_zod.z.string().array().optional()
});
function toGeminiType(mcpType) {
  if (!mcpType) return import_genai.Type.TYPE_UNSPECIFIED;
  switch (mcpType.toLowerCase()) {
    case "text":
    case "string":
      return import_genai.Type.STRING;
    case "number":
      return import_genai.Type.NUMBER;
    case "boolean":
      return import_genai.Type.BOOLEAN;
    case "integer":
      return import_genai.Type.INTEGER;
    case "array":
      return import_genai.Type.ARRAY;
    case "object":
      return import_genai.Type.OBJECT;
    case "null":
      return import_genai.Type.NULL;
    default:
      return import_genai.Type.TYPE_UNSPECIFIED;
  }
}
const getTypeFromArrayItem = (mcpType) => {
  var _a, _b;
  if (typeof mcpType === "string") {
    return mcpType.toLowerCase();
  }
  return (_b = (_a = mcpType == null ? void 0 : mcpType.type) == null ? void 0 : _a.toLowerCase) == null ? void 0 : _b.call(_a);
};
function toGeminiSchema(mcpSchema) {
  if (!mcpSchema) {
    return void 0;
  }
  function recursiveConvert(mcp) {
    var _a;
    const sourceType = (_a = mcp.anyOf) != null ? _a : mcp.type;
    let isNullable = false;
    let nonNullTypes;
    if (Array.isArray(sourceType)) {
      nonNullTypes = sourceType.filter(
        (t) => getTypeFromArrayItem(t) !== "null"
      );
      isNullable = sourceType.some(
        (t) => getTypeFromArrayItem(t) === "null"
      );
      if (nonNullTypes.length === 1) {
        const nonNullType = nonNullTypes[0];
        if (typeof nonNullType === "object") {
          mcp = nonNullType;
        } else {
          const { type: _removed, anyOf: _removedAnyOf, ...rest } = mcp;
          mcp = { ...rest, type: nonNullType };
        }
      } else if (nonNullTypes.length === 0 && isNullable) {
        const { type: _removed, anyOf: _removedAnyOf, ...rest } = mcp;
        mcp = { ...rest, type: "null" };
      } else if (typeof mcp.anyOf === "undefined") {
        const anyOfItems = mcp.type.map((t) => ({ type: t }));
        const { type: _removed, ...rest } = mcp;
        mcp = { ...rest, anyOf: anyOfItems };
      }
    }
    if (!mcp.type) {
      if (mcp.properties || mcp.$ref) {
        mcp.type = "object";
      } else if (mcp.items) {
        mcp.type = "array";
      } else if (isNullable) {
        mcp.type = "null";
      }
    }
    const geminiType = toGeminiType(mcp.type);
    const geminiSchema = {};
    if (mcp.anyOf) {
      geminiSchema.anyOf = mcp.anyOf.map(
        (item) => recursiveConvert(item)
      );
    } else {
      geminiSchema.type = geminiType;
    }
    if (mcp.description) {
      geminiSchema.description = mcp.description;
    }
    if (isNullable && mcp.type !== "null") {
      geminiSchema.nullable = true;
    }
    if (geminiType === import_genai.Type.OBJECT) {
      geminiSchema.properties = {};
      if (mcp.properties) {
        for (const name in mcp.properties) {
          geminiSchema.properties[name] = recursiveConvert(
            mcp.properties[name]
          );
        }
      }
      if (mcp.required) {
        geminiSchema.required = mcp.required;
      }
    } else if (geminiType === import_genai.Type.ARRAY) {
      if (mcp.items) {
        geminiSchema.items = recursiveConvert(mcp.items);
      }
    }
    return geminiSchema;
  }
  return recursiveConvert(mcpSchema);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  toGeminiSchema
});
