// @google/adk/dist/cjs/utils/logger.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var logger_exports = {};
__export(logger_exports, {
  LogLevel: () => LogLevel,
  getLogger: () => getLogger,
  logger: () => logger,
  resetLogger: () => resetLogger,
  setLogLevel: () => setLogLevel,
  setLogger: () => setLogger
});
module.exports = __toCommonJS(logger_exports);
var winston = __toESM(require("winston"), 1);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var LogLevel = /* @__PURE__ */ ((LogLevel2) => {
  LogLevel2[LogLevel2["DEBUG"] = 0] = "DEBUG";
  LogLevel2[LogLevel2["INFO"] = 1] = "INFO";
  LogLevel2[LogLevel2["WARN"] = 2] = "WARN";
  LogLevel2[LogLevel2["ERROR"] = 3] = "ERROR";
  return LogLevel2;
})(LogLevel || {});
class SimpleLogger {
  constructor() {
    this.logLevel = 1 /* INFO */;
    this.logger = winston.createLogger({
      levels: {
        "debug": 0 /* DEBUG */,
        "info": 1 /* INFO */,
        "warn": 2 /* WARN */,
        "error": 3 /* ERROR */
      },
      level: "error",
      format: winston.format.combine(
        winston.format.label({ label: "ADK" }),
        winston.format((info) => {
          info.level = info.level.toUpperCase();
          return info;
        })(),
        winston.format.colorize(),
        winston.format.timestamp(),
        winston.format.printf((info) => {
          return `${info.level}: [${info.label}] ${info.timestamp} ${info.message}`;
        })
      ),
      transports: [new winston.transports.Console()]
    });
  }
  setLogLevel(level) {
    this.logLevel = level;
  }
  log(level, ...messages) {
    if (this.logLevel > level) {
      return;
    }
    this.logger.log(level.toString(), messages.join(" "));
  }
  debug(...messages) {
    if (this.logLevel > 0 /* DEBUG */) {
      return;
    }
    this.logger.debug(messages.join(" "));
  }
  info(...messages) {
    if (this.logLevel > 1 /* INFO */) {
      return;
    }
    this.logger.info(messages.join(" "));
  }
  warn(...messages) {
    if (this.logLevel > 2 /* WARN */) {
      return;
    }
    this.logger.warn(messages.join(" "));
  }
  error(...messages) {
    if (this.logLevel > 3 /* ERROR */) {
      return;
    }
    this.logger.error(messages.join(" "));
  }
}
class NoOpLogger {
  setLogLevel(_level) {
  }
  log(_level, ..._args) {
  }
  debug(..._args) {
  }
  info(..._args) {
  }
  warn(..._args) {
  }
  error(..._args) {
  }
}
let currentLogger = new SimpleLogger();
function setLogger(customLogger) {
  currentLogger = customLogger != null ? customLogger : new NoOpLogger();
}
function getLogger() {
  return currentLogger;
}
function resetLogger() {
  currentLogger = new SimpleLogger();
}
function setLogLevel(level) {
  logger.setLogLevel(level);
}
const logger = {
  setLogLevel(level) {
    currentLogger.setLogLevel(level);
  },
  log(level, ...args) {
    currentLogger.log(level, ...args);
  },
  debug(...args) {
    currentLogger.debug(...args);
  },
  info(...args) {
    currentLogger.info(...args);
  },
  warn(...args) {
    currentLogger.warn(...args);
  },
  error(...args) {
    currentLogger.error(...args);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LogLevel,
  getLogger,
  logger,
  resetLogger,
  setLogLevel,
  setLogger
});
