// @google/adk/dist/web/features/feature_registry.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { getBooleanEnvVar } from "../utils/env_aware_utils.js";
import { logger } from "../utils/logger.js";
var FeatureName = /* @__PURE__ */ ((FeatureName2) => {
  FeatureName2["PROGRESSIVE_SSE_STREAMING"] = "PROGRESSIVE_SSE_STREAMING";
  return FeatureName2;
})(FeatureName || {});
var FeatureStage = /* @__PURE__ */ ((FeatureStage2) => {
  FeatureStage2["WIP"] = "wip";
  FeatureStage2["EXPERIMENTAL"] = "experimental";
  FeatureStage2["STABLE"] = "stable";
  return FeatureStage2;
})(FeatureStage || {});
const FEATURE_REGISTRY = {
  ["PROGRESSIVE_SSE_STREAMING" /* PROGRESSIVE_SSE_STREAMING */]: {
    stage: "experimental" /* EXPERIMENTAL */,
    defaultOn: false
  }
};
const WARNED_FEATURES = /* @__PURE__ */ new Set();
const FEATURE_OVERRIDES = {};
function getFeatureConfig(featureName) {
  return FEATURE_REGISTRY[featureName];
}
function registerFeature(featureName, config) {
  FEATURE_REGISTRY[featureName] = config;
}
function overrideFeatureEnabled(featureName, enabled) {
  const config = getFeatureConfig(featureName);
  if (!config) {
    throw new Error("Feature ".concat(featureName, " is not registered."));
  }
  if (enabled === void 0) {
    delete FEATURE_OVERRIDES[featureName];
  } else {
    FEATURE_OVERRIDES[featureName] = enabled;
  }
}
function isFeatureEnabled(featureName) {
  const config = getFeatureConfig(featureName);
  if (!config) {
    throw new Error("Feature ".concat(featureName, " is not registered."));
  }
  if (featureName in FEATURE_OVERRIDES) {
    const enabled = FEATURE_OVERRIDES[featureName];
    if (enabled && config.stage !== "stable" /* STABLE */) {
      emitNonStableWarningOnce(featureName, config.stage);
    }
    return enabled;
  }
  const enableVar = "ADK_ENABLE_".concat(featureName);
  const disableVar = "ADK_DISABLE_".concat(featureName);
  if (getBooleanEnvVar(enableVar)) {
    if (config.stage !== "stable" /* STABLE */) {
      emitNonStableWarningOnce(featureName, config.stage);
    }
    return true;
  }
  if (getBooleanEnvVar(disableVar)) {
    return false;
  }
  if (config.stage !== "stable" /* STABLE */ && config.defaultOn) {
    emitNonStableWarningOnce(featureName, config.stage);
  }
  return config.defaultOn;
}
function emitNonStableWarningOnce(featureName, featureStage) {
  if (!WARNED_FEATURES.has(featureName)) {
    WARNED_FEATURES.add(featureName);
    logger.warn(
      "[".concat(featureStage.toUpperCase(), "] feature ").concat(featureName, " is enabled.")
    );
  }
}
async function withTemporaryFeatureOverride(featureName, enabled, callback) {
  const config = getFeatureConfig(featureName);
  if (!config) {
    throw new Error("Feature ".concat(featureName, " is not registered."));
  }
  const hadOverride = featureName in FEATURE_OVERRIDES;
  const originalValue = FEATURE_OVERRIDES[featureName];
  FEATURE_OVERRIDES[featureName] = enabled;
  try {
    return await callback();
  } finally {
    if (hadOverride) {
      FEATURE_OVERRIDES[featureName] = originalValue;
    } else {
      delete FEATURE_OVERRIDES[featureName];
    }
  }
}
export {
  FeatureName,
  FeatureStage,
  getFeatureConfig,
  isFeatureEnabled,
  overrideFeatureEnabled,
  registerFeature,
  withTemporaryFeatureOverride
};
