// @google/adk/dist/cjs/auth/oauth2/oauth2_credential_refresher.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var oauth2_credential_refresher_exports = {};
__export(oauth2_credential_refresher_exports, {
  OAuth2CredentialRefresher: () => OAuth2CredentialRefresher
});
module.exports = __toCommonJS(oauth2_credential_refresher_exports);
var import_logger = require("../../utils/logger.js");
var import_oauth2_utils = require("./oauth2_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class OAuth2CredentialRefresher {
  /**
   * Check if the OAuth2 credential needs to be refreshed.
   *
   * @param authCredential The OAuth2 credential to check.
   * @param authScheme The OAuth2 authentication scheme (optional).
   * @returns True if the credential needs to be refreshed, False otherwise.
   */
  async isRefreshNeeded(authCredential) {
    if (!authCredential.oauth2) {
      return false;
    }
    if (authCredential.oauth2 && authCredential.oauth2.expiresAt) {
      return (0, import_oauth2_utils.isTokenExpired)(authCredential.oauth2);
    }
    return false;
  }
  /**
   * Refresh the OAuth2 credential.
   *
   * @param authCredential The OAuth2 credential to refresh.
   * @param authScheme The OAuth2 authentication scheme.
   * @returns The refreshed credential.
   */
  async refresh(authCredential, authScheme) {
    if (!authCredential.oauth2 || !authScheme) {
      return authCredential;
    }
    if (!authCredential.oauth2.refreshToken) {
      import_logger.logger.warn("No refresh token available to refresh credential");
      return authCredential;
    }
    const isNeeded = await this.isRefreshNeeded(authCredential);
    if (!isNeeded) {
      return authCredential;
    }
    const tokenEndpoint = (0, import_oauth2_utils.getTokenEndpoint)(authScheme);
    if (!tokenEndpoint) {
      import_logger.logger.warn("Token endpoint not found in auth scheme.");
      return authCredential;
    }
    if (!authCredential.oauth2.clientId || !authCredential.oauth2.clientSecret) {
      import_logger.logger.warn("clientId and clientSecret are required for token refresh.");
      return authCredential;
    }
    const body = new URLSearchParams();
    body.set("grant_type", "refresh_token");
    body.set("refresh_token", authCredential.oauth2.refreshToken);
    body.set("client_id", authCredential.oauth2.clientId);
    body.set("client_secret", authCredential.oauth2.clientSecret);
    try {
      const data = await (0, import_oauth2_utils.fetchOAuth2Tokens)(tokenEndpoint, body);
      const updatedOAuth2 = {
        ...authCredential.oauth2,
        accessToken: data.accessToken || authCredential.oauth2.accessToken,
        refreshToken: data.refreshToken || authCredential.oauth2.refreshToken,
        expiresIn: data.expiresIn,
        expiresAt: data.expiresAt || authCredential.oauth2.expiresAt
      };
      return {
        ...authCredential,
        oauth2: updatedOAuth2
      };
    } catch (error) {
      import_logger.logger.error("Failed to refresh tokens:", error);
      return authCredential;
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  OAuth2CredentialRefresher
});
