// @google/adk/dist/cjs/auth/oauth2/oauth2_discovery.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var oauth2_discovery_exports = {};
__export(oauth2_discovery_exports, {
  AuthorizationServerMetadataSchema: () => AuthorizationServerMetadataSchema,
  OAuth2DiscoveryManager: () => OAuth2DiscoveryManager,
  ProtectedResourceMetadataSchema: () => ProtectedResourceMetadataSchema
});
module.exports = __toCommonJS(oauth2_discovery_exports);
var import_zod = require("zod");
var import_logger = require("../../utils/logger.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const AuthorizationServerMetadataSchema = import_zod.z.object({
  issuer: import_zod.z.string(),
  authorization_endpoint: import_zod.z.string(),
  token_endpoint: import_zod.z.string(),
  scopes_supported: import_zod.z.array(import_zod.z.string()).optional(),
  registration_endpoint: import_zod.z.string().optional()
});
const ProtectedResourceMetadataSchema = import_zod.z.object({
  resource: import_zod.z.string(),
  authorization_servers: import_zod.z.array(import_zod.z.string()).default([])
});
class OAuth2DiscoveryManager {
  /**
   * Discovers the OAuth2 authorization server metadata.
   */
  async discoverAuthServerMetadata(issuerUrl) {
    if (!validateDiscoveryUrl(issuerUrl)) {
      return void 0;
    }
    let baseUrl;
    let path;
    try {
      const url = new URL(issuerUrl);
      baseUrl = `${url.protocol}//${url.host}`;
      path = url.pathname;
    } catch (e) {
      import_logger.logger.warn(`Failed to parse issuerUrl ${issuerUrl}: ${e}`);
      return void 0;
    }
    const endpointsToTry = [];
    if (path && path !== "/") {
      endpointsToTry.push(
        `${baseUrl}/.well-known/oauth-authorization-server${path}`,
        `${baseUrl}/.well-known/openid-configuration${path}`,
        `${baseUrl}${path}/.well-known/openid-configuration`
      );
    } else {
      endpointsToTry.push(
        `${baseUrl}/.well-known/oauth-authorization-server`,
        `${baseUrl}/.well-known/openid-configuration`
      );
    }
    for (const endpoint of endpointsToTry) {
      try {
        const response = await fetch(endpoint, {
          method: "GET",
          headers: {
            Accept: "application/json"
          }
        });
        if (!response.ok) {
          continue;
        }
        const data = await response.json();
        const metadata = AuthorizationServerMetadataSchema.parse(data);
        if (metadata.issuer.replace(/\/$/, "") === issuerUrl.replace(/\/$/, "")) {
          return metadata;
        } else {
          import_logger.logger.warn(
            `Issuer in metadata ${metadata.issuer} does not match issuerUrl ${issuerUrl}`
          );
        }
      } catch (e) {
        import_logger.logger.debug(`Failed to fetch metadata from ${endpoint}: ${e}`);
      }
    }
    return void 0;
  }
  /**
   * Discovers the OAuth2 protected resource metadata.
   */
  async discoverResourceMetadata(resourceUrl) {
    if (!validateDiscoveryUrl(resourceUrl)) {
      return void 0;
    }
    let baseUrl;
    let path;
    try {
      const url = new URL(resourceUrl);
      baseUrl = `${url.protocol}//${url.host}`;
      path = url.pathname;
    } catch (e) {
      import_logger.logger.warn(`Failed to parse resourceUrl ${resourceUrl}: ${e}`);
      return void 0;
    }
    let wellKnownEndpoint;
    if (path && path !== "/") {
      wellKnownEndpoint = `${baseUrl}/.well-known/oauth-protected-resource${path}`;
    } else {
      wellKnownEndpoint = `${baseUrl}/.well-known/oauth-protected-resource`;
    }
    try {
      const response = await fetch(wellKnownEndpoint, {
        method: "GET",
        headers: {
          Accept: "application/json"
        }
      });
      if (!response.ok) {
        return void 0;
      }
      const data = await response.json();
      const metadata = ProtectedResourceMetadataSchema.parse(data);
      if (metadata.resource.replace(/\/$/, "") === resourceUrl.replace(/\/$/, "")) {
        return metadata;
      } else {
        import_logger.logger.warn(
          `Resource in metadata ${metadata.resource} does not match resourceUrl ${resourceUrl}`
        );
      }
    } catch (e) {
      import_logger.logger.debug(`Failed to fetch metadata from ${wellKnownEndpoint}: ${e}`);
    }
    return void 0;
  }
}
function validateDiscoveryUrl(urlStr) {
  try {
    const url = new URL(urlStr);
    if (url.protocol !== "https:") {
      import_logger.logger.warn(`Unsafe protocol for discovery URL: ${url.protocol}`);
      return false;
    }
    const host = url.hostname.toLowerCase();
    if (host === "localhost" || host === "127.0.0.1" || host === "[::1]" || host.startsWith("10.") || host.startsWith("192.168.") || host.startsWith("169.254.")) {
      import_logger.logger.warn(`Unsafe host for discovery URL: ${host}`);
      return false;
    }
    const match = host.match(/^172\.(\d+)\./);
    if (match) {
      const secondOctet = parseInt(match[1], 10);
      if (secondOctet >= 16 && secondOctet <= 31) {
        import_logger.logger.warn(`Unsafe host for discovery URL: ${host}`);
        return false;
      }
    }
    return true;
  } catch (e) {
    import_logger.logger.warn(`Failed to parse URL for validation ${urlStr}: ${e}`);
    return false;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AuthorizationServerMetadataSchema,
  OAuth2DiscoveryManager,
  ProtectedResourceMetadataSchema
});
