// @google/adk/dist/cjs/auth/oauth2/oauth2_credential_exchanger.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var oauth2_credential_exchanger_exports = {};
__export(oauth2_credential_exchanger_exports, {
  OAuth2CredentialExchanger: () => OAuth2CredentialExchanger,
  determineGrantType: () => determineGrantType,
  exchangeAuthorizationCode: () => exchangeAuthorizationCode,
  exchangeClientCredentials: () => exchangeClientCredentials
});
module.exports = __toCommonJS(oauth2_credential_exchanger_exports);
var import_logger = require("../../utils/logger.js");
var import_auth_schemes = require("../auth_schemes.js");
var import_base_credential_exchanger = require("../exchanger/base_credential_exchanger.js");
var import_oauth2_utils = require("./oauth2_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class OAuth2CredentialExchanger {
  async exchange({
    authCredential,
    authScheme
  }) {
    var _a;
    if (!authScheme) {
      throw new import_base_credential_exchanger.CredentialExchangeError(
        "authScheme is required for OAuth2 credential exchange"
      );
    }
    if ((_a = authCredential.oauth2) == null ? void 0 : _a.accessToken) {
      return {
        credential: authCredential,
        wasExchanged: false
      };
    }
    const grantType = determineGrantType(authScheme);
    if (grantType === import_auth_schemes.OAuthGrantType.CLIENT_CREDENTIALS) {
      return exchangeClientCredentials({ authCredential, authScheme });
    }
    if (grantType === import_auth_schemes.OAuthGrantType.AUTHORIZATION_CODE) {
      return exchangeAuthorizationCode({ authCredential, authScheme });
    }
    import_logger.logger.warn(`Unsupported OAuth2 grant type: ${grantType}`);
    return {
      credential: authCredential,
      wasExchanged: false
    };
  }
}
function determineGrantType(authScheme) {
  var _a;
  if ("flows" in authScheme && authScheme.flows) {
    return (0, import_auth_schemes.getOAuthGrantTypeFromFlow)(authScheme.flows);
  }
  if (authScheme.grantTypesSupported) {
    const oidcScheme = authScheme;
    if ((_a = oidcScheme.grantTypesSupported) == null ? void 0 : _a.includes("client_credentials")) {
      return import_auth_schemes.OAuthGrantType.CLIENT_CREDENTIALS;
    }
    return import_auth_schemes.OAuthGrantType.AUTHORIZATION_CODE;
  }
  return void 0;
}
async function exchangeClientCredentials({
  authCredential,
  authScheme
}) {
  var _a, _b;
  const tokenEndpoint = (0, import_oauth2_utils.getTokenEndpoint)(authScheme);
  if (!tokenEndpoint) {
    throw new import_base_credential_exchanger.CredentialExchangeError(
      "Token endpoint not found in auth scheme."
    );
  }
  if (!((_a = authCredential.oauth2) == null ? void 0 : _a.clientId) || !((_b = authCredential.oauth2) == null ? void 0 : _b.clientSecret)) {
    throw new import_base_credential_exchanger.CredentialExchangeError(
      "clientId and clientSecret are required for client credentials exchange."
    );
  }
  const body = (0, import_oauth2_utils.createOAuth2TokenRequestBody)({
    grantType: "client_credentials",
    clientId: authCredential.oauth2.clientId,
    clientSecret: authCredential.oauth2.clientSecret
  });
  try {
    const oauth2Auth = await (0, import_oauth2_utils.fetchOAuth2Tokens)(tokenEndpoint, body);
    return {
      credential: {
        ...authCredential,
        oauth2: {
          ...authCredential.oauth2,
          ...oauth2Auth
        }
      },
      wasExchanged: true
    };
  } catch (error) {
    throw new import_base_credential_exchanger.CredentialExchangeError(
      `Failed to exchange tokens: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}
async function exchangeAuthorizationCode({
  authCredential,
  authScheme
}) {
  var _a, _b, _c, _d;
  const tokenEndpoint = (0, import_oauth2_utils.getTokenEndpoint)(authScheme);
  if (!tokenEndpoint) {
    throw new import_base_credential_exchanger.CredentialExchangeError(
      "Token endpoint not found in auth scheme."
    );
  }
  if (!((_a = authCredential.oauth2) == null ? void 0 : _a.clientId) || !((_b = authCredential.oauth2) == null ? void 0 : _b.clientSecret) || !((_c = authCredential.oauth2) == null ? void 0 : _c.authCode) && !((_d = authCredential.oauth2) == null ? void 0 : _d.authResponseUri)) {
    throw new import_base_credential_exchanger.CredentialExchangeError(
      "clientId, clientSecret, and either authCode or authResponseUri are required for authorization code exchange."
    );
  }
  let code = authCredential.oauth2.authCode;
  if (!code && authCredential.oauth2.authResponseUri) {
    code = (0, import_oauth2_utils.parseAuthorizationCode)(authCredential.oauth2.authResponseUri);
  }
  if (authCredential.oauth2.authResponseUri && authCredential.oauth2.state) {
    try {
      const url = new URL(authCredential.oauth2.authResponseUri);
      const receivedState = url.searchParams.get("state") || void 0;
      if (authCredential.oauth2.state !== receivedState) {
        throw new import_base_credential_exchanger.CredentialExchangeError(
          "State mismatch detected. Potential CSRF attack."
        );
      }
    } catch (e) {
      throw new import_base_credential_exchanger.CredentialExchangeError(
        `Failed to parse authResponseUri for state validation: ${e instanceof Error ? e.message : String(e)}`
      );
    }
  }
  if (!code) {
    throw new import_base_credential_exchanger.CredentialExchangeError(
      "Authorization code not found in auth response."
    );
  }
  const body = (0, import_oauth2_utils.createOAuth2TokenRequestBody)({
    grantType: "authorization_code",
    clientId: authCredential.oauth2.clientId,
    clientSecret: authCredential.oauth2.clientSecret,
    code,
    redirectUri: authCredential.oauth2.redirectUri,
    codeVerifier: authCredential.oauth2.codeVerifier
  });
  try {
    const oauth2Auth = await (0, import_oauth2_utils.fetchOAuth2Tokens)(tokenEndpoint, body);
    return {
      credential: {
        ...authCredential,
        oauth2: {
          ...authCredential.oauth2,
          ...oauth2Auth
        }
      },
      wasExchanged: true
    };
  } catch (error) {
    throw new import_base_credential_exchanger.CredentialExchangeError(
      `Failed to exchange tokens: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  OAuth2CredentialExchanger,
  determineGrantType,
  exchangeAuthorizationCode,
  exchangeClientCredentials
});
