// @google/adk/dist/cjs/auth/oauth2/oauth2_utils.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var oauth2_utils_exports = {};
__export(oauth2_utils_exports, {
  createOAuth2TokenRequestBody: () => createOAuth2TokenRequestBody,
  fetchOAuth2Tokens: () => fetchOAuth2Tokens,
  getTokenEndpoint: () => getTokenEndpoint,
  isTokenExpired: () => isTokenExpired,
  parseAuthorizationCode: () => parseAuthorizationCode
});
module.exports = __toCommonJS(oauth2_utils_exports);
var import_logger = require("../../utils/logger.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function getTokenEndpoint(authScheme) {
  if (authScheme.type === "openIdConnect" && authScheme.tokenEndpoint) {
    return authScheme.tokenEndpoint;
  }
  if (authScheme.type === "oauth2" && authScheme.flows) {
    const flows = authScheme.flows;
    const flow = flows.authorizationCode || flows.clientCredentials || flows.password || flows.implicit;
    if (flow && "tokenUrl" in flow) {
      return flow.tokenUrl;
    }
  }
  return void 0;
}
async function fetchOAuth2Tokens(endpoint, body) {
  try {
    const response = await fetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded"
      },
      body: body.toString()
    });
    if (!response.ok) {
      throw new Error(`Token request failed with status ${response.status}`);
    }
    const data = await response.json();
    return {
      accessToken: data.access_token,
      refreshToken: data.refresh_token,
      idToken: data.id_token,
      expiresIn: data.expires_in,
      expiresAt: data.expires_in ? Date.now() + data.expires_in * 1e3 : void 0
    };
  } catch (e) {
    import_logger.logger.error(`Failed to fetch OAuth2 tokens: ${e}`);
    throw e;
  }
}
function parseAuthorizationCode(uri) {
  try {
    const url = new URL(uri);
    return url.searchParams.get("code") || void 0;
  } catch (e) {
    import_logger.logger.warn(`Failed to parse authorization URI ${uri}: ${e}`);
    return void 0;
  }
}
function createOAuth2TokenRequestBody(params) {
  const body = new URLSearchParams();
  body.set("grant_type", params.grantType);
  body.set("client_id", params.clientId);
  body.set("client_secret", params.clientSecret);
  if (params.grantType === "authorization_code") {
    body.set("code", params.code);
    if (params.redirectUri) {
      body.set("redirect_uri", params.redirectUri);
    }
    if (params.codeVerifier) {
      body.set("code_verifier", params.codeVerifier);
    }
  } else if (params.grantType === "refresh_token") {
    body.set("refresh_token", params.refreshToken);
  }
  return body;
}
function isTokenExpired(token, leeway = 60) {
  if (typeof token.expiresAt !== "number") {
    return false;
  }
  const expirationThreshold = token.expiresAt - leeway * 1e3;
  return expirationThreshold < Date.now();
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createOAuth2TokenRequestBody,
  fetchOAuth2Tokens,
  getTokenEndpoint,
  isTokenExpired,
  parseAuthorizationCode
});
