// @google/adk/dist/esm/tools/mcp/mcp_tool.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { toGeminiSchema } from "../../utils/gemini_schema_util.js";
import { BaseTool } from "../base_tool.js";
class MCPTool extends BaseTool {
  constructor(mcpTool, mcpSessionManager, originalName) {
    super({ name: mcpTool.name, description: mcpTool.description || "" });
    this.mcpTool = mcpTool;
    this.mcpSessionManager = mcpSessionManager;
    this.originalName = originalName || mcpTool.name;
  }
  _getDeclaration() {
    return {
      name: this.mcpTool.name,
      description: this.mcpTool.description,
      parameters: toGeminiSchema(this.mcpTool.inputSchema),
      // TODO: need revisit, refer to this
      // https://modelcontextprotocol.io/specification/2025-06-18/server/tools#tool-result
      response: toGeminiSchema(this.mcpTool.outputSchema)
    };
  }
  async runAsync(request) {
    const session = await this.mcpSessionManager.createSession();
    const callRequest = {};
    callRequest.params = { name: this.originalName, arguments: request.args };
    const result = await session.callTool(callRequest.params, void 0, {
      signal: request.toolContext.abortSignal
    });
    return result;
  }
}
export {
  MCPTool
};
