// @mikro-orm/mssql/MsSqlQueryBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsSqlQueryBuilder = void 0;
const core_1 = require("@mikro-orm/core");
const knex_1 = require("@mikro-orm/knex");
class MsSqlQueryBuilder extends knex_1.QueryBuilder {
    insert(data) {
        this.checkIdentityInsert(data);
        return super.insert(data);
    }
    getKnex() {
        const qb = super.getKnex();
        if (this.flags.has(core_1.QueryFlag.IDENTITY_INSERT)) {
            this.appendIdentityInsert(qb);
        }
        if (!this.flags.has(core_1.QueryFlag.IDENTITY_INSERT) && this.type === knex_1.QueryType.INSERT && this.metadata.has(this.mainAlias.entityName)) {
            const meta = this.metadata.get(this.mainAlias.entityName);
            if (meta.hasTriggers) {
                this.appendOutputTable(qb);
            }
        }
        return qb;
    }
    getKnexQuery(processVirtualEntity = true) {
        if (this.type === knex_1.QueryType.TRUNCATE) {
            const tableName = this.driver.getTableName(this.mainAlias.metadata, { schema: this._schema }, false);
            const tableNameQuoted = this.platform.quoteIdentifier(tableName);
            const sql = `delete from ${tableNameQuoted}; declare @count int = case @@rowcount when 0 then 1 else 0 end; dbcc checkident ('${tableName}', reseed, @count)`;
            this._query = {};
            return this._query.qb = this.knex.raw(sql);
        }
        return super.getKnexQuery(processVirtualEntity);
    }
    appendIdentityInsert(qb) {
        const meta = this.metadata.get(this.mainAlias.entityName);
        const table = this.driver.getTableName(meta, { schema: this._schema });
        const originalToSQL = qb.toSQL;
        qb.toSQL = () => {
            const res = originalToSQL.apply(qb);
            return {
                ...res,
                sql: `set identity_insert ${table} on; ${res.sql}; set identity_insert ${table} off;`,
                toNative: () => res.toNative(),
            };
        };
    }
    checkIdentityInsert(data) {
        const meta = this.metadata.find(this.mainAlias.entityName);
        if (!meta) {
            return;
        }
        const dataKeys = core_1.Utils.unique(core_1.Utils.asArray(data).flatMap(Object.keys));
        const hasAutoincrement = dataKeys.some(x => meta.properties[x]?.autoincrement);
        if (hasAutoincrement) {
            this.setFlag(core_1.QueryFlag.IDENTITY_INSERT);
        }
    }
    appendOutputTable(qb) {
        const meta = this.metadata.find(this.mainAlias.entityName);
        const returningProps = meta.props.filter(prop => prop.primary || prop.defaultRaw);
        const returningFields = core_1.Utils.flatten(returningProps.map(prop => prop.fieldNames));
        const table = this.driver.getTableName(meta, { schema: this._schema });
        const selections = returningFields
            .map((field) => `[t].${this.platform.quoteIdentifier(field)}`)
            .join(',');
        const originalToSQL = qb.toSQL;
        qb.toSQL = () => {
            const res = originalToSQL.apply(qb);
            const position = res.sql.indexOf(' values ');
            const sqlBeforeValues = res.sql.substring(0, position);
            const sqlAfterValues = res.sql.substring(position + 1);
            let outputSql = `select top(0) ${selections} into #out from ${table} as t left join ${table} on 0 = 1; `;
            outputSql += `${sqlBeforeValues} into #out ${sqlAfterValues}; `;
            outputSql += `select ${selections} from #out as t; `;
            outputSql += `drop table #out`;
            return {
                ...res,
                sql: outputSql,
                toNative: () => res.toNative(),
            };
        };
    }
}
exports.MsSqlQueryBuilder = MsSqlQueryBuilder;
