// @google/adk/dist/esm/context/token_based_context_compactor.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { isCompactedEvent } from "../events/compacted_event.js";
import { stringifyContent } from "../events/event.js";
class TokenBasedContextCompactor {
  constructor(options) {
    this.tokenThreshold = options.tokenThreshold;
    this.eventRetentionSize = options.eventRetentionSize;
    this.summarizer = options.summarizer;
  }
  getActiveEvents(events) {
    let latestCompactedEvent = void 0;
    for (let i = events.length - 1; i >= 0; i--) {
      const e = events[i];
      if (isCompactedEvent(e)) {
        if (!latestCompactedEvent || e.endTime > latestCompactedEvent.endTime) {
          latestCompactedEvent = e;
        }
      }
    }
    if (!latestCompactedEvent) {
      return events;
    }
    const activeRawEvents = events.filter(
      (e) => !isCompactedEvent(e) && e.timestamp > latestCompactedEvent.endTime
    );
    return [latestCompactedEvent, ...activeRawEvents];
  }
  shouldCompact(invocationContext) {
    const events = invocationContext.session.events;
    const activeEvents = this.getActiveEvents(events);
    const rawEvents = activeEvents.filter((e) => !isCompactedEvent(e));
    if (rawEvents.length <= this.eventRetentionSize) {
      return false;
    }
    let totalTokens = 0;
    for (const event of activeEvents) {
      totalTokens += getEventTokens(event);
    }
    return totalTokens > this.tokenThreshold;
  }
  async compact(invocationContext) {
    const events = invocationContext.session.events;
    const activeEvents = this.getActiveEvents(events);
    const rawEvents = activeEvents.filter((e) => !isCompactedEvent(e));
    if (rawEvents.length <= this.eventRetentionSize) {
      return;
    }
    let retainStartIndex = Math.max(
      0,
      rawEvents.length - this.eventRetentionSize
    );
    while (retainStartIndex > 0) {
      const eventToRetain = rawEvents[retainStartIndex];
      const previousEvent = rawEvents[retainStartIndex - 1];
      if (hasFunctionResponse(eventToRetain) && hasFunctionCall(previousEvent)) {
        retainStartIndex--;
      } else {
        break;
      }
    }
    if (retainStartIndex === 0) {
      return;
    }
    const rawEventsToCompact = rawEvents.slice(0, retainStartIndex);
    const compactedEventPresent = activeEvents.find(isCompactedEvent);
    const eventsToCompact = compactedEventPresent ? [compactedEventPresent, ...rawEventsToCompact] : rawEventsToCompact;
    const compactedEvent = await this.summarizer.summarize(eventsToCompact);
    if (!compactedEvent.actions) {
      compactedEvent.actions = {
        stateDelta: {},
        artifactDelta: {},
        requestedAuthConfigs: [],
        requestedToolConfirmations: {}
      };
    }
    invocationContext.session.events.push(compactedEvent);
  }
}
function getEventTokens(event) {
  var _a;
  if (((_a = event.usageMetadata) == null ? void 0 : _a.promptTokenCount) !== void 0) {
    return event.usageMetadata.promptTokenCount;
  }
  const contentStr = stringifyContent(event);
  return Math.ceil(contentStr.length / 4);
}
function hasFunctionCall(event) {
  var _a, _b;
  return !!((_b = (_a = event.content) == null ? void 0 : _a.parts) == null ? void 0 : _b.some(
    (part) => part.functionCall !== void 0
  ));
}
function hasFunctionResponse(event) {
  var _a, _b;
  return !!((_b = (_a = event.content) == null ? void 0 : _a.parts) == null ? void 0 : _b.some(
    (part) => part.functionResponse !== void 0
  ));
}
export {
  TokenBasedContextCompactor
};
