// @google/adk/dist/cjs/index.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  A2AAgentExecutor: () => import_agent_executor.A2AAgentExecutor,
  AGENT_CARD_PATH: () => import_a2a_remote_agent.AGENT_CARD_PATH,
  DatabaseSessionService: () => import_database_session_service.DatabaseSessionService,
  FileArtifactService: () => import_file_artifact_service.FileArtifactService,
  GcsArtifactService: () => import_gcs_artifact_service.GcsArtifactService,
  RemoteA2AAgent: () => import_a2a_remote_agent.RemoteA2AAgent,
  RunSkillInlineScriptTool: () => import_run_skill_inline_script_tool.RunSkillInlineScriptTool,
  RunSkillScriptTool: () => import_run_skill_script_tool.RunSkillScriptTool,
  UnsafeLocalCodeExecutor: () => import_unsafe_local_code_executor.UnsafeLocalCodeExecutor,
  getA2AAgentCard: () => import_agent_card.getA2AAgentCard,
  getArtifactServiceFromUri: () => import_registry.getArtifactServiceFromUri,
  getSessionServiceFromUri: () => import_registry2.getSessionServiceFromUri,
  loadAllSkillsInDir: () => import_loader.loadAllSkillsInDir,
  loadSkillFromDir: () => import_loader.loadSkillFromDir,
  toA2a: () => import_agent_to_a2a.toA2a,
  validateSkillDir: () => import_loader.validateSkillDir
});
module.exports = __toCommonJS(index_exports);
var import_a2a_remote_agent = require("./a2a/a2a_remote_agent.js");
var import_agent_card = require("./a2a/agent_card.js");
var import_agent_executor = require("./a2a/agent_executor.js");
var import_agent_to_a2a = require("./a2a/agent_to_a2a.js");
var import_file_artifact_service = require("./artifacts/file_artifact_service.js");
var import_gcs_artifact_service = require("./artifacts/gcs_artifact_service.js");
var import_registry = require("./artifacts/registry.js");
var import_unsafe_local_code_executor = require("./code_executors/unsafe_local_code_executor.js");
__reExport(index_exports, require("./common.js"), module.exports);
var import_database_session_service = require("./sessions/database_session_service.js");
var import_registry2 = require("./sessions/registry.js");
var import_loader = require("./skills/loader.js");
var import_run_skill_inline_script_tool = require("./tools/skill/run_skill_inline_script_tool.js");
var import_run_skill_script_tool = require("./tools/skill/run_skill_script_tool.js");
__reExport(index_exports, require("./telemetry/google_cloud.js"), module.exports);
__reExport(index_exports, require("./telemetry/setup.js"), module.exports);
__reExport(index_exports, require("./tools/mcp/mcp_session_manager.js"), module.exports);
__reExport(index_exports, require("./tools/mcp/mcp_tool.js"), module.exports);
__reExport(index_exports, require("./tools/mcp/mcp_toolset.js"), module.exports);
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  A2AAgentExecutor,
  AGENT_CARD_PATH,
  DatabaseSessionService,
  FileArtifactService,
  GcsArtifactService,
  RemoteA2AAgent,
  RunSkillInlineScriptTool,
  RunSkillScriptTool,
  UnsafeLocalCodeExecutor,
  getA2AAgentCard,
  getArtifactServiceFromUri,
  getSessionServiceFromUri,
  loadAllSkillsInDir,
  loadSkillFromDir,
  toA2a,
  validateSkillDir,
  ...require("./common.js"),
  ...require("./telemetry/google_cloud.js"),
  ...require("./telemetry/setup.js"),
  ...require("./tools/mcp/mcp_session_manager.js"),
  ...require("./tools/mcp/mcp_tool.js"),
  ...require("./tools/mcp/mcp_toolset.js")
});
