// @google/adk/dist/esm/auth/oauth2/oauth2_utils.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { logger } from "../../utils/logger.js";
function getTokenEndpoint(authScheme) {
  if (authScheme.type === "openIdConnect" && authScheme.tokenEndpoint) {
    return authScheme.tokenEndpoint;
  }
  if (authScheme.type === "oauth2" && authScheme.flows) {
    const flows = authScheme.flows;
    const flow = flows.authorizationCode || flows.clientCredentials || flows.password || flows.implicit;
    if (flow && "tokenUrl" in flow) {
      return flow.tokenUrl;
    }
  }
  return void 0;
}
async function fetchOAuth2Tokens(endpoint, body) {
  try {
    const response = await fetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded"
      },
      body: body.toString()
    });
    if (!response.ok) {
      throw new Error(`Token request failed with status ${response.status}`);
    }
    const data = await response.json();
    return {
      accessToken: data.access_token,
      refreshToken: data.refresh_token,
      idToken: data.id_token,
      expiresIn: data.expires_in,
      expiresAt: data.expires_in ? Date.now() + data.expires_in * 1e3 : void 0
    };
  } catch (e) {
    logger.error(`Failed to fetch OAuth2 tokens: ${e}`);
    throw e;
  }
}
function parseAuthorizationCode(uri) {
  try {
    const url = new URL(uri);
    return url.searchParams.get("code") || void 0;
  } catch (e) {
    logger.warn(`Failed to parse authorization URI ${uri}: ${e}`);
    return void 0;
  }
}
function createOAuth2TokenRequestBody(params) {
  const body = new URLSearchParams();
  body.set("grant_type", params.grantType);
  body.set("client_id", params.clientId);
  body.set("client_secret", params.clientSecret);
  if (params.grantType === "authorization_code") {
    body.set("code", params.code);
    if (params.redirectUri) {
      body.set("redirect_uri", params.redirectUri);
    }
    if (params.codeVerifier) {
      body.set("code_verifier", params.codeVerifier);
    }
  } else if (params.grantType === "refresh_token") {
    body.set("refresh_token", params.refreshToken);
  }
  return body;
}
function isTokenExpired(token, leeway = 60) {
  if (typeof token.expiresAt !== "number") {
    return false;
  }
  const expirationThreshold = token.expiresAt - leeway * 1e3;
  return expirationThreshold < Date.now();
}
export {
  createOAuth2TokenRequestBody,
  fetchOAuth2Tokens,
  getTokenEndpoint,
  isTokenExpired,
  parseAuthorizationCode
};
