// @google/adk/dist/esm/auth/oauth2/oauth2_credential_exchanger.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { logger } from "../../utils/logger.js";
import {
  getOAuthGrantTypeFromFlow,
  OAuthGrantType
} from "../auth_schemes.js";
import {
  CredentialExchangeError
} from "../exchanger/base_credential_exchanger.js";
import {
  createOAuth2TokenRequestBody,
  fetchOAuth2Tokens,
  getTokenEndpoint,
  parseAuthorizationCode
} from "./oauth2_utils.js";
class OAuth2CredentialExchanger {
  async exchange({
    authCredential,
    authScheme
  }) {
    var _a;
    if (!authScheme) {
      throw new CredentialExchangeError(
        "authScheme is required for OAuth2 credential exchange"
      );
    }
    if ((_a = authCredential.oauth2) == null ? void 0 : _a.accessToken) {
      return {
        credential: authCredential,
        wasExchanged: false
      };
    }
    const grantType = determineGrantType(authScheme);
    if (grantType === OAuthGrantType.CLIENT_CREDENTIALS) {
      return exchangeClientCredentials({ authCredential, authScheme });
    }
    if (grantType === OAuthGrantType.AUTHORIZATION_CODE) {
      return exchangeAuthorizationCode({ authCredential, authScheme });
    }
    logger.warn(`Unsupported OAuth2 grant type: ${grantType}`);
    return {
      credential: authCredential,
      wasExchanged: false
    };
  }
}
function determineGrantType(authScheme) {
  var _a;
  if ("flows" in authScheme && authScheme.flows) {
    return getOAuthGrantTypeFromFlow(authScheme.flows);
  }
  if (authScheme.grantTypesSupported) {
    const oidcScheme = authScheme;
    if ((_a = oidcScheme.grantTypesSupported) == null ? void 0 : _a.includes("client_credentials")) {
      return OAuthGrantType.CLIENT_CREDENTIALS;
    }
    return OAuthGrantType.AUTHORIZATION_CODE;
  }
  return void 0;
}
async function exchangeClientCredentials({
  authCredential,
  authScheme
}) {
  var _a, _b;
  const tokenEndpoint = getTokenEndpoint(authScheme);
  if (!tokenEndpoint) {
    throw new CredentialExchangeError(
      "Token endpoint not found in auth scheme."
    );
  }
  if (!((_a = authCredential.oauth2) == null ? void 0 : _a.clientId) || !((_b = authCredential.oauth2) == null ? void 0 : _b.clientSecret)) {
    throw new CredentialExchangeError(
      "clientId and clientSecret are required for client credentials exchange."
    );
  }
  const body = createOAuth2TokenRequestBody({
    grantType: "client_credentials",
    clientId: authCredential.oauth2.clientId,
    clientSecret: authCredential.oauth2.clientSecret
  });
  try {
    const oauth2Auth = await fetchOAuth2Tokens(tokenEndpoint, body);
    return {
      credential: {
        ...authCredential,
        oauth2: {
          ...authCredential.oauth2,
          ...oauth2Auth
        }
      },
      wasExchanged: true
    };
  } catch (error) {
    throw new CredentialExchangeError(
      `Failed to exchange tokens: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}
async function exchangeAuthorizationCode({
  authCredential,
  authScheme
}) {
  var _a, _b, _c, _d;
  const tokenEndpoint = getTokenEndpoint(authScheme);
  if (!tokenEndpoint) {
    throw new CredentialExchangeError(
      "Token endpoint not found in auth scheme."
    );
  }
  if (!((_a = authCredential.oauth2) == null ? void 0 : _a.clientId) || !((_b = authCredential.oauth2) == null ? void 0 : _b.clientSecret) || !((_c = authCredential.oauth2) == null ? void 0 : _c.authCode) && !((_d = authCredential.oauth2) == null ? void 0 : _d.authResponseUri)) {
    throw new CredentialExchangeError(
      "clientId, clientSecret, and either authCode or authResponseUri are required for authorization code exchange."
    );
  }
  let code = authCredential.oauth2.authCode;
  if (!code && authCredential.oauth2.authResponseUri) {
    code = parseAuthorizationCode(authCredential.oauth2.authResponseUri);
  }
  if (authCredential.oauth2.authResponseUri && authCredential.oauth2.state) {
    try {
      const url = new URL(authCredential.oauth2.authResponseUri);
      const receivedState = url.searchParams.get("state") || void 0;
      if (authCredential.oauth2.state !== receivedState) {
        throw new CredentialExchangeError(
          "State mismatch detected. Potential CSRF attack."
        );
      }
    } catch (e) {
      throw new CredentialExchangeError(
        `Failed to parse authResponseUri for state validation: ${e instanceof Error ? e.message : String(e)}`
      );
    }
  }
  if (!code) {
    throw new CredentialExchangeError(
      "Authorization code not found in auth response."
    );
  }
  const body = createOAuth2TokenRequestBody({
    grantType: "authorization_code",
    clientId: authCredential.oauth2.clientId,
    clientSecret: authCredential.oauth2.clientSecret,
    code,
    redirectUri: authCredential.oauth2.redirectUri,
    codeVerifier: authCredential.oauth2.codeVerifier
  });
  try {
    const oauth2Auth = await fetchOAuth2Tokens(tokenEndpoint, body);
    return {
      credential: {
        ...authCredential,
        oauth2: {
          ...authCredential.oauth2,
          ...oauth2Auth
        }
      },
      wasExchanged: true
    };
  } catch (error) {
    throw new CredentialExchangeError(
      `Failed to exchange tokens: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}
export {
  OAuth2CredentialExchanger,
  determineGrantType,
  exchangeAuthorizationCode,
  exchangeClientCredentials
};
