// @google/adk/dist/cjs/a2a/a2a_remote_agent_run_processor.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var a2a_remote_agent_run_processor_exports = {};
__export(a2a_remote_agent_run_processor_exports, {
  A2ARemoteAgentRunProcessor: () => A2ARemoteAgentRunProcessor
});
module.exports = __toCommonJS(a2a_remote_agent_run_processor_exports);
var import_genai = require("@google/genai");
var import_event = require("../events/event.js");
var import_a2a_event = require("./a2a_event.js");
var import_metadata_converter_utils = require("./metadata_converter_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class A2ARemoteAgentRunProcessor {
  constructor(request) {
    this.request = request;
    this.aggregations = /* @__PURE__ */ new Map();
    this.aggregationOrder = [];
  }
  /**
   * aggregatePartial stores contents of partial events to emit them with the terminal event.
   * It can return multiple events to emit instead of just the provided one.
   */
  aggregatePartial(context, a2aEvent, adkEvent) {
    const metadata = (0, import_a2a_event.getEventMetadata)(a2aEvent);
    if (metadata[import_metadata_converter_utils.A2AMetadataKeys.PARTIAL]) {
      return [adkEvent];
    }
    if ((0, import_a2a_event.isTaskStatusUpdateEvent)(a2aEvent) && a2aEvent.final) {
      const events = [];
      for (const aid of this.aggregationOrder) {
        const agg = this.aggregations.get(aid);
        if (agg) {
          events.push(this.buildNonPartialAggregation(context, agg));
        }
      }
      this.aggregations.clear();
      this.aggregationOrder = [];
      return [...events, adkEvent];
    }
    if ((0, import_a2a_event.isTask)(a2aEvent)) {
      this.aggregations.clear();
      this.aggregationOrder = [];
      return [adkEvent];
    }
    if (!(0, import_a2a_event.isTaskArtifactUpdateEvent)(a2aEvent)) {
      return [adkEvent];
    }
    const artifactId = a2aEvent.artifact.artifactId;
    if (!a2aEvent.append) {
      this.removeAggregation(artifactId);
      if (a2aEvent.lastChunk) {
        adkEvent.partial = false;
        return [adkEvent];
      }
    }
    let aggregation = this.aggregations.get(artifactId);
    if (!aggregation) {
      aggregation = {
        aggregatedText: "",
        aggregatedThoughts: "",
        parts: []
      };
      this.aggregations.set(artifactId, aggregation);
      this.aggregationOrder.push(artifactId);
    } else {
      this.aggregationOrder = this.aggregationOrder.filter(
        (id) => id !== artifactId
      );
      this.aggregationOrder.push(artifactId);
    }
    this.updateAggregation(aggregation, adkEvent);
    if (!a2aEvent.lastChunk) {
      return [adkEvent];
    }
    this.removeAggregation(artifactId);
    return [adkEvent, this.buildNonPartialAggregation(context, aggregation)];
  }
  removeAggregation(artifactId) {
    this.aggregations.delete(artifactId);
    this.aggregationOrder = this.aggregationOrder.filter(
      (id) => id !== artifactId
    );
  }
  updateAggregation(agg, event) {
    var _a;
    const parts = ((_a = event.content) == null ? void 0 : _a.parts) || [];
    for (const part of parts) {
      if (part.text && part.text !== "") {
        if (part.thought) {
          agg.aggregatedThoughts += part.text;
        } else {
          agg.aggregatedText += part.text;
        }
      } else {
        this.promoteTextBlocksToParts(agg);
        agg.parts.push(part);
      }
    }
    if (event.citationMetadata) {
      if (!agg.citations) {
        agg.citations = { citations: [] };
      }
      if (!agg.citations.citations) {
        agg.citations.citations = [];
      }
      agg.citations.citations.push(...event.citationMetadata.citations || []);
    }
    if (event.customMetadata) {
      if (!agg.customMeta) {
        agg.customMeta = {};
      }
      Object.assign(agg.customMeta, event.customMetadata);
    }
    if (event.groundingMetadata) {
      agg.grounding = event.groundingMetadata;
    }
    if (event.usageMetadata) {
      agg.usage = event.usageMetadata;
    }
  }
  buildNonPartialAggregation(context, agg) {
    this.promoteTextBlocksToParts(agg);
    const result = (0, import_event.createEvent)({
      author: context.agent.name,
      invocationId: context.invocationId,
      content: agg.parts.length > 0 ? (0, import_genai.createModelContent)([...agg.parts]) : void 0,
      customMetadata: agg.customMeta,
      groundingMetadata: agg.grounding,
      citationMetadata: agg.citations,
      usageMetadata: agg.usage,
      turnComplete: false,
      partial: false
    });
    return result;
  }
  promoteTextBlocksToParts(agg) {
    if (agg.aggregatedThoughts !== "") {
      agg.parts.push({ thought: true, text: agg.aggregatedThoughts });
      agg.aggregatedThoughts = "";
    }
    if (agg.aggregatedText !== "") {
      agg.parts.push({ text: agg.aggregatedText });
      agg.aggregatedText = "";
    }
  }
  /**
   * Adds request and response metadata to the event.
   */
  updateCustomMetadata(event, response) {
    const toAdd = {};
    if (this.request && event.turnComplete) {
      toAdd["request"] = this.request;
    }
    if (response) {
      toAdd["response"] = response;
      if ((0, import_a2a_event.isTask)(response)) {
        if (response.id) toAdd["task_id"] = response.id;
        if (response.contextId) toAdd["context_id"] = response.contextId;
      } else if (response.taskId) {
        toAdd["task_id"] = response.taskId;
        if (response.contextId) toAdd["context_id"] = response.contextId;
      }
    }
    if (Object.keys(toAdd).length === 0) {
      return;
    }
    if (!event.customMetadata) {
      event.customMetadata = {};
    }
    for (const [k, v] of Object.entries(toAdd)) {
      if (v === void 0 || v === null) continue;
      event.customMetadata[`a2a:${k}`] = v;
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  A2ARemoteAgentRunProcessor
});
