// @google/adk/dist/cjs/a2a/a2a_remote_agent.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var a2a_remote_agent_exports = {};
__export(a2a_remote_agent_exports, {
  AGENT_CARD_PATH: () => import_sdk.AGENT_CARD_PATH,
  RemoteA2AAgent: () => RemoteA2AAgent
});
module.exports = __toCommonJS(a2a_remote_agent_exports);
var import_sdk = require("@a2a-js/sdk");
var import_client = require("@a2a-js/sdk/client");
var import_base_agent = require("../agents/base_agent.js");
var import_event = require("../events/event.js");
var import_env_aware_utils = require("../utils/env_aware_utils.js");
var import_logger = require("../utils/logger.js");
var import_a2a_event = require("./a2a_event.js");
var import_a2a_remote_agent_run_processor = require("./a2a_remote_agent_run_processor.js");
var import_a2a_remote_agent_utils = require("./a2a_remote_agent_utils.js");
var import_agent_card = require("./agent_card.js");
var import_event_converter_utils = require("./event_converter_utils.js");
var import_metadata_converter_utils = require("./metadata_converter_utils.js");
var import_part_converter_utils = require("./part_converter_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class RemoteA2AAgent extends import_base_agent.BaseAgent {
  constructor(a2aConfig) {
    super(a2aConfig);
    this.a2aConfig = a2aConfig;
    this.isInitialized = false;
    if (!a2aConfig.agentCard && !a2aConfig.client) {
      throw new Error("Either AgentCard or Client must be provided");
    }
  }
  async init() {
    if (this.isInitialized) {
      return;
    }
    if (this.a2aConfig.client) {
      this.client = this.a2aConfig.client;
    }
    if (this.a2aConfig.agentCard) {
      this.card = await (0, import_agent_card.resolveAgentCard)(this.a2aConfig.agentCard);
      if (!this.client) {
        const factory = this.a2aConfig.clientFactory || new import_client.ClientFactory();
        this.client = await factory.createFromAgentCard(this.card);
      }
    }
    this.isInitialized = true;
  }
  async *runAsyncImpl(context) {
    var _a, _b;
    await this.init();
    try {
      const events = context.session.events;
      if (events.length === 0) {
        throw new Error("No events in session to send");
      }
      const userFnCall = (0, import_a2a_remote_agent_utils.getUserFunctionCallAt)(
        context.session,
        events.length - 1
      );
      let parts;
      let taskId = void 0;
      let contextId = void 0;
      if (userFnCall) {
        const event = userFnCall.response;
        parts = (0, import_part_converter_utils.toA2AParts)(
          ((_a = event.content) == null ? void 0 : _a.parts) || [],
          event.longRunningToolIds
        );
        taskId = userFnCall.taskId;
        contextId = userFnCall.contextId;
      } else {
        const missing = (0, import_a2a_remote_agent_utils.toMissingRemoteSessionParts)(context, context.session);
        parts = missing.parts;
        contextId = missing.contextId;
      }
      const message = {
        kind: "message",
        messageId: (0, import_env_aware_utils.randomUUID)(),
        role: import_a2a_event.MessageRole.USER,
        parts,
        metadata: (0, import_metadata_converter_utils.getA2ASessionMetadata)({
          appName: context.session.appName,
          userId: context.session.userId,
          sessionId: context.session.id
        })
      };
      if (taskId) message.taskId = taskId;
      if (contextId) message.contextId = contextId;
      const params = {
        message,
        configuration: this.a2aConfig.messageSendConfig
      };
      const processor = new import_a2a_remote_agent_run_processor.A2ARemoteAgentRunProcessor(params);
      if (this.a2aConfig.beforeRequestCallbacks) {
        for (const callback of this.a2aConfig.beforeRequestCallbacks) {
          await callback(context, params);
        }
      }
      const useStreaming = this.card ? ((_b = this.card.capabilities) == null ? void 0 : _b.streaming) !== false : true;
      if (useStreaming) {
        for await (const chunk of this.client.sendMessageStream(params)) {
          if (this.a2aConfig.afterRequestCallbacks) {
            for (const callback of this.a2aConfig.afterRequestCallbacks) {
              await callback(context, chunk);
            }
          }
          const adkEvent = (0, import_event_converter_utils.toAdkEvent)(chunk, context.invocationId, this.name);
          if (!adkEvent) {
            continue;
          }
          processor.updateCustomMetadata(adkEvent, chunk);
          const eventsToEmit = processor.aggregatePartial(
            context,
            chunk,
            adkEvent
          );
          for (const ev of eventsToEmit) {
            yield ev;
          }
        }
      } else {
        const result = await this.client.sendMessage(params);
        if (this.a2aConfig.afterRequestCallbacks) {
          for (const callback of this.a2aConfig.afterRequestCallbacks) {
            await callback(context, result);
          }
        }
        const adkEvent = (0, import_event_converter_utils.toAdkEvent)(result, context.invocationId, this.name);
        if (adkEvent) {
          processor.updateCustomMetadata(adkEvent, result);
          yield adkEvent;
        }
      }
    } catch (e) {
      const error = e;
      import_logger.logger.error(`A2ARemoteAgent ${this.name} failed:`, error);
      yield (0, import_event.createEvent)({
        author: this.name,
        invocationId: context.invocationId,
        errorMessage: error.message,
        turnComplete: true
      });
    }
  }
  runLiveImpl(_context) {
    throw new Error("Live mode is not supported in A2ARemoteAgent yet.");
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AGENT_CARD_PATH,
  RemoteA2AAgent
});
