// @google/adk/dist/cjs/a2a/a2a_remote_agent_utils.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var a2a_remote_agent_utils_exports = {};
__export(a2a_remote_agent_utils_exports, {
  getFunctionResponseCallId: () => getFunctionResponseCallId,
  getUserFunctionCallAt: () => getUserFunctionCallAt,
  isFunctionCallEvent: () => isFunctionCallEvent,
  presentAsUserMessage: () => presentAsUserMessage,
  toMissingRemoteSessionParts: () => toMissingRemoteSessionParts
});
module.exports = __toCommonJS(a2a_remote_agent_utils_exports);
var import_event = require("../events/event.js");
var import_metadata_converter_utils = require("./metadata_converter_utils.js");
var import_part_converter_utils = require("./part_converter_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function getUserFunctionCallAt(session, index) {
  const events = session.events;
  if (index < 0 || index >= events.length) {
    return void 0;
  }
  const candidate = events[index];
  if (candidate.author !== "user") {
    return void 0;
  }
  const fnCallId = getFunctionResponseCallId(candidate);
  if (!fnCallId) {
    return void 0;
  }
  for (let i = index - 1; i >= 0; i--) {
    const request = events[i];
    if (!isFunctionCallEvent(request, fnCallId)) {
      continue;
    }
    const metadata = request.customMetadata || {};
    const taskId = metadata[import_metadata_converter_utils.AdkMetadataKeys.TASK_ID] || "";
    const contextId = metadata[import_metadata_converter_utils.AdkMetadataKeys.CONTEXT_ID] || "";
    return {
      response: candidate,
      taskId,
      contextId
    };
  }
  return void 0;
}
function isFunctionCallEvent(event, callId) {
  if (!event || !event.content || !event.content.parts) {
    return false;
  }
  return event.content.parts.some(
    (part) => part.functionCall && part.functionCall.id === callId
  );
}
function getFunctionResponseCallId(event) {
  var _a;
  if (!event || !event.content || !event.content.parts) {
    return void 0;
  }
  const responsePart = event.content.parts.find(
    (part) => part.functionResponse
  );
  return (_a = responsePart == null ? void 0 : responsePart.functionResponse) == null ? void 0 : _a.id;
}
function toMissingRemoteSessionParts(ctx, session) {
  const events = session.events;
  let contextId = void 0;
  let lastRemoteResponseIndex = -1;
  for (let i = events.length - 1; i >= 0; i--) {
    const event = events[i];
    if (event.author === ctx.agent.name) {
      lastRemoteResponseIndex = i;
      const metadata = event.customMetadata || {};
      contextId = metadata[import_metadata_converter_utils.AdkMetadataKeys.CONTEXT_ID];
      break;
    }
  }
  const missingParts = [];
  for (let i = lastRemoteResponseIndex + 1; i < events.length; i++) {
    let event = events[i];
    if (event.author !== "user" && event.author !== ctx.agent.name) {
      event = presentAsUserMessage(ctx, event);
    }
    if (!event.content || !event.content.parts || event.content.parts.length === 0) {
      continue;
    }
    const parts = (0, import_part_converter_utils.toA2AParts)(event.content.parts, event.longRunningToolIds);
    missingParts.push(...parts);
  }
  return {
    parts: missingParts,
    contextId
  };
}
function presentAsUserMessage(ctx, agentEvent) {
  const event = (0, import_event.createEvent)({
    author: "user",
    invocationId: ctx.invocationId
  });
  if (!agentEvent.content || !agentEvent.content.parts) {
    return event;
  }
  const parts = [{ text: "For context:" }];
  for (const part of agentEvent.content.parts) {
    if (part.thought) {
      continue;
    }
    if (part.text) {
      parts.push({
        text: `[${agentEvent.author}] said: ${part.text}`
      });
    } else if (part.functionCall) {
      const call = part.functionCall;
      parts.push({
        text: `[${agentEvent.author}] called tool ${call.name} with parameters: ${JSON.stringify(call.args)}`
      });
    } else if (part.functionResponse) {
      const resp = part.functionResponse;
      parts.push({
        text: `[${agentEvent.author}] ${resp.name} tool returned result: ${JSON.stringify(resp.response)}`
      });
    } else {
      parts.push(part);
    }
  }
  if (parts.length > 1) {
    event.content = {
      role: "user",
      parts
    };
  }
  return event;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getFunctionResponseCallId,
  getUserFunctionCallAt,
  isFunctionCallEvent,
  presentAsUserMessage,
  toMissingRemoteSessionParts
});
