// @google/adk/dist/cjs/a2a/a2a_event.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var a2a_event_exports = {};
__export(a2a_event_exports, {
  MessageRole: () => MessageRole,
  TaskState: () => TaskState,
  createInputMissingErrorEvent: () => createInputMissingErrorEvent,
  createTask: () => createTask,
  createTaskArtifactUpdateEvent: () => createTaskArtifactUpdateEvent,
  createTaskCompletedEvent: () => createTaskCompletedEvent,
  createTaskFailedEvent: () => createTaskFailedEvent,
  createTaskInputRequiredEvent: () => createTaskInputRequiredEvent,
  createTaskSubmittedEvent: () => createTaskSubmittedEvent,
  createTaskWorkingEvent: () => createTaskWorkingEvent,
  getEventMetadata: () => getEventMetadata,
  getFailedTaskStatusUpdateEventError: () => getFailedTaskStatusUpdateEventError,
  isFailedTaskStatusUpdateEvent: () => isFailedTaskStatusUpdateEvent,
  isInputRequiredTaskStatusUpdateEvent: () => isInputRequiredTaskStatusUpdateEvent,
  isMessage: () => isMessage,
  isTask: () => isTask,
  isTaskArtifactUpdateEvent: () => isTaskArtifactUpdateEvent,
  isTaskStatusUpdateEvent: () => isTaskStatusUpdateEvent,
  isTerminalTaskStatusUpdateEvent: () => isTerminalTaskStatusUpdateEvent
});
module.exports = __toCommonJS(a2a_event_exports);
var import_env_aware_utils = require("../utils/env_aware_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var MessageRole = /* @__PURE__ */ ((MessageRole2) => {
  MessageRole2["USER"] = "user";
  MessageRole2["AGENT"] = "agent";
  return MessageRole2;
})(MessageRole || {});
var TaskState = /* @__PURE__ */ ((TaskState2) => {
  TaskState2["SUBMITTED"] = "submitted";
  TaskState2["WORKING"] = "working";
  TaskState2["COMPLETED"] = "completed";
  TaskState2["FAILED"] = "failed";
  TaskState2["CANCELED"] = "canceled";
  TaskState2["REJECTED"] = "rejected";
  TaskState2["INPUT_REQUIRED"] = "input-required";
  return TaskState2;
})(TaskState || {});
function isTaskStatusUpdateEvent(event) {
  return (event == null ? void 0 : event.kind) === "status-update";
}
function isTaskArtifactUpdateEvent(event) {
  return (event == null ? void 0 : event.kind) === "artifact-update";
}
function isMessage(event) {
  return (event == null ? void 0 : event.kind) === "message";
}
function isTask(event) {
  return (event == null ? void 0 : event.kind) === "task";
}
function getEventMetadata(event) {
  if (isTaskArtifactUpdateEvent(event)) {
    return event.artifact.metadata || {};
  }
  if (isTaskStatusUpdateEvent(event) || isTask(event) || isMessage(event)) {
    return event.metadata || {};
  }
  return {};
}
function isFailedTaskStatusUpdateEvent(event) {
  return (isTaskStatusUpdateEvent(event) || isTask(event)) && event.status.state === "failed" /* FAILED */;
}
function isTerminalTaskStatusUpdateEvent(event) {
  return (isTaskStatusUpdateEvent(event) || isTask(event)) && [
    "completed" /* COMPLETED */,
    "failed" /* FAILED */,
    "canceled" /* CANCELED */,
    "rejected" /* REJECTED */
  ].includes(event.status.state);
}
function isInputRequiredTaskStatusUpdateEvent(event) {
  return (isTaskStatusUpdateEvent(event) || isTask(event)) && event.status.state === "input-required" /* INPUT_REQUIRED */;
}
function getFailedTaskStatusUpdateEventError(event) {
  var _a;
  if (!isFailedTaskStatusUpdateEvent(event)) {
    return void 0;
  }
  const parts = ((_a = event.status.message) == null ? void 0 : _a.parts) || [];
  if (parts.length === 0) {
    return void 0;
  }
  if (parts[0].kind !== "text") {
    return void 0;
  }
  return parts[0].text;
}
function createTaskSubmittedEvent({
  taskId,
  contextId,
  message,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: false,
    status: {
      state: "submitted" /* SUBMITTED */,
      message,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTask({
  contextId,
  message,
  taskId,
  metadata
}) {
  return {
    kind: "task",
    id: taskId || (0, import_env_aware_utils.randomUUID)(),
    contextId,
    history: [message],
    status: {
      state: "submitted" /* SUBMITTED */,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTaskWorkingEvent({
  taskId,
  contextId,
  message,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: false,
    status: {
      state: "working" /* WORKING */,
      message,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTaskCompletedEvent({
  taskId,
  contextId,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: true,
    status: {
      state: "completed" /* COMPLETED */,
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTaskArtifactUpdateEvent({
  taskId,
  contextId,
  artifactId,
  parts = [],
  metadata,
  append,
  lastChunk
}) {
  return {
    kind: "artifact-update",
    taskId,
    contextId,
    append,
    lastChunk,
    artifact: {
      artifactId: artifactId || (0, import_env_aware_utils.randomUUID)(),
      parts
    },
    metadata
  };
}
function createTaskFailedEvent({
  taskId,
  contextId,
  error,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: true,
    status: {
      state: "failed" /* FAILED */,
      message: {
        kind: "message",
        messageId: (0, import_env_aware_utils.randomUUID)(),
        role: "agent",
        taskId,
        contextId,
        parts: [
          {
            kind: "text",
            text: error.message
          }
        ]
      },
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createTaskInputRequiredEvent({
  taskId,
  contextId,
  parts,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: true,
    status: {
      state: "input-required" /* INPUT_REQUIRED */,
      message: {
        kind: "message",
        messageId: (0, import_env_aware_utils.randomUUID)(),
        role: "agent",
        taskId,
        contextId,
        parts
      },
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
function createInputMissingErrorEvent({
  taskId,
  contextId,
  parts,
  metadata
}) {
  return {
    kind: "status-update",
    taskId,
    contextId,
    final: true,
    status: {
      state: "input-required" /* INPUT_REQUIRED */,
      message: {
        kind: "message",
        messageId: (0, import_env_aware_utils.randomUUID)(),
        role: "agent",
        taskId,
        contextId,
        parts
      },
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    },
    metadata
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MessageRole,
  TaskState,
  createInputMissingErrorEvent,
  createTask,
  createTaskArtifactUpdateEvent,
  createTaskCompletedEvent,
  createTaskFailedEvent,
  createTaskInputRequiredEvent,
  createTaskSubmittedEvent,
  createTaskWorkingEvent,
  getEventMetadata,
  getFailedTaskStatusUpdateEventError,
  isFailedTaskStatusUpdateEvent,
  isInputRequiredTaskStatusUpdateEvent,
  isMessage,
  isTask,
  isTaskArtifactUpdateEvent,
  isTaskStatusUpdateEvent,
  isTerminalTaskStatusUpdateEvent
});
