// @google/adk/dist/cjs/a2a/agent_to_a2a.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var agent_to_a2a_exports = {};
__export(agent_to_a2a_exports, {
  toA2a: () => toA2a
});
module.exports = __toCommonJS(agent_to_a2a_exports);
var import_sdk = require("@a2a-js/sdk");
var import_server = require("@a2a-js/sdk/server");
var import_express = require("@a2a-js/sdk/server/express");
var import_express2 = __toESM(require("express"), 1);
var import_run_config = require("../agents/run_config.js");
var import_in_memory_session_service = require("../sessions/in_memory_session_service.js");
var import_agent_card = require("./agent_card.js");
var import_agent_executor = require("./agent_executor.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
async function toA2a(agent, options = {}) {
  var _a, _b, _c, _d;
  const host = (_a = options.host) != null ? _a : "localhost";
  const port = (_b = options.port) != null ? _b : 8e3;
  const protocol = (_c = options.protocol) != null ? _c : "http";
  const basePath = options.basePath || "";
  const rpcUrl = `${protocol}://${host}:${port}${basePath}`;
  const agentCard = options.agentCard ? await (0, import_agent_card.resolveAgentCard)(options.agentCard) : await (0, import_agent_card.getA2AAgentCard)(agent, [
    {
      url: `${rpcUrl}/jsonrpc`,
      transport: "JSONRPC"
    },
    {
      url: `${rpcUrl}/rest`,
      transport: "HTTP+JSON"
    }
  ]);
  const agentExecutor = new import_agent_executor.A2AAgentExecutor({
    runner: options.runner || {
      agent,
      appName: agent.name,
      sessionService: options.sessionService || new import_in_memory_session_service.InMemorySessionService(),
      memoryService: options.memoryService,
      artifactService: options.artifactService
    },
    runConfig: {
      streamingMode: import_run_config.StreamingMode.SSE
    }
  });
  const requestHandler = new import_server.DefaultRequestHandler(
    agentCard,
    new import_server.InMemoryTaskStore(),
    agentExecutor
  );
  const app = (_d = options.app) != null ? _d : (0, import_express2.default)();
  if (!options.app) {
    app.use(import_express2.default.urlencoded({ limit: "50mb", extended: true }));
    app.use(import_express2.default.json({ limit: "50mb" }));
  }
  app.use(
    `${basePath}/${import_sdk.AGENT_CARD_PATH}`,
    (0, import_express.agentCardHandler)({ agentCardProvider: requestHandler })
  );
  app.use(
    `${basePath}/rest`,
    (0, import_express.restHandler)({
      requestHandler,
      userBuilder: import_express.UserBuilder.noAuthentication
    })
  );
  app.use(
    `${basePath}/jsonrpc`,
    (0, import_express.jsonRpcHandler)({
      requestHandler,
      userBuilder: import_express.UserBuilder.noAuthentication
    })
  );
  return app;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  toA2a
});
