// @google/adk/dist/cjs/a2a/agent_executor.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var agent_executor_exports = {};
__export(agent_executor_exports, {
  A2AAgentExecutor: () => A2AAgentExecutor
});
module.exports = __toCommonJS(agent_executor_exports);
var import_runner = require("../runner/runner.js");
var import_env_aware_utils = require("../utils/env_aware_utils.js");
var import_logger = require("../utils/logger.js");
var import_a2a_event = require("./a2a_event.js");
var import_event_processor_utils = require("./event_processor_utils.js");
var import_executor_context = require("./executor_context.js");
var import_metadata_converter_utils = require("./metadata_converter_utils.js");
var import_part_converter_utils = require("./part_converter_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class A2AAgentExecutor {
  constructor(config) {
    this.config = config;
    this.agentPartialArtifactIdsMap = {};
  }
  async execute(ctx, eventBus) {
    var _a, _b;
    const a2aUserMessage = ctx.userMessage;
    if (!a2aUserMessage) {
      throw new Error("message not provided");
    }
    const userId = `A2A_USER_${ctx.contextId}`;
    const sessionId = ctx.contextId;
    const genAIUserMessage = (0, import_part_converter_utils.toGenAIContent)(a2aUserMessage);
    const adkRunner = await getAdkRunner(this.config.runner);
    const session = await getAdkSession(
      userId,
      sessionId,
      adkRunner.sessionService,
      adkRunner.appName
    );
    const executorContext = (0, import_executor_context.createExecutorContext)({
      session,
      userContent: genAIUserMessage,
      requestContext: ctx
    });
    try {
      if (this.config.beforeExecuteCallback) {
        await this.config.beforeExecuteCallback(ctx);
      }
      if (ctx.task) {
        const inputRequiredEvent = (0, import_event_processor_utils.getTaskInputRequiredEvent)(
          ctx.task,
          genAIUserMessage
        );
        if (inputRequiredEvent) {
          await this.publishFinalTaskStatus({
            executorContext,
            eventBus,
            event: inputRequiredEvent
          });
          return;
        }
      }
      if (!ctx.task) {
        eventBus.publish(
          (0, import_a2a_event.createTask)({
            taskId: ctx.taskId,
            contextId: ctx.contextId,
            message: a2aUserMessage
          })
        );
      }
      eventBus.publish(
        (0, import_a2a_event.createTaskWorkingEvent)({
          taskId: ctx.taskId,
          contextId: ctx.contextId
        })
      );
      const adkEvents = [];
      for await (const adkEvent of adkRunner.runAsync({
        userId,
        sessionId,
        newMessage: genAIUserMessage,
        runConfig: this.config.runConfig
      })) {
        adkEvents.push(adkEvent);
        const a2aEvent = this.convertAdkEventToA2AEvent(
          adkEvent,
          executorContext
        );
        if (!a2aEvent) {
          continue;
        }
        await ((_b = (_a = this.config).afterEventCallback) == null ? void 0 : _b.call(
          _a,
          executorContext,
          adkEvent,
          a2aEvent
        ));
        eventBus.publish(a2aEvent);
      }
      await this.publishFinalTaskStatus({
        executorContext,
        eventBus,
        event: (0, import_event_processor_utils.getFinalTaskStatusUpdate)(adkEvents, executorContext)
      });
    } catch (e) {
      const error = e;
      await this.publishFinalTaskStatus({
        executorContext,
        eventBus,
        error,
        event: (0, import_a2a_event.createTaskFailedEvent)({
          taskId: ctx.taskId,
          contextId: ctx.contextId,
          error: new Error(`Agent run failed: ${error.message}`),
          metadata: (0, import_metadata_converter_utils.getA2ASessionMetadata)(executorContext)
        })
      });
    }
  }
  // Task cancellation is not supported in this implementation yet.
  async cancelTask(_taskId) {
    throw new Error("Task cancellation is not supported yet.");
  }
  convertAdkEventToA2AEvent(adkEvent, executorContext) {
    var _a;
    const a2aParts = (0, import_part_converter_utils.toA2AParts)(
      (_a = adkEvent.content) == null ? void 0 : _a.parts,
      adkEvent.longRunningToolIds
    );
    if (a2aParts.length === 0) {
      return void 0;
    }
    const artifactId = this.agentPartialArtifactIdsMap[adkEvent.author] || (0, import_env_aware_utils.randomUUID)();
    const a2aEvent = (0, import_a2a_event.createTaskArtifactUpdateEvent)({
      taskId: executorContext.requestContext.taskId,
      contextId: executorContext.requestContext.contextId,
      artifactId,
      parts: a2aParts,
      metadata: (0, import_metadata_converter_utils.getA2AEventMetadata)(adkEvent, executorContext),
      append: adkEvent.partial,
      lastChunk: !adkEvent.partial
    });
    if (adkEvent.partial) {
      this.agentPartialArtifactIdsMap[adkEvent.author] = artifactId;
    } else {
      delete this.agentPartialArtifactIdsMap[adkEvent.author];
    }
    return a2aEvent;
  }
  /**
   * Writes the final status event to the queue.
   */
  async publishFinalTaskStatus({
    executorContext,
    eventBus,
    event,
    error
  }) {
    var _a, _b;
    try {
      await ((_b = (_a = this.config).afterExecuteCallback) == null ? void 0 : _b.call(_a, executorContext, event, error));
    } catch (e) {
      import_logger.logger.error("Error in afterExecuteCallback:", e);
    }
    eventBus.publish(event);
  }
}
async function getAdkSession(userId, sessionId, sessionService, appName) {
  const session = await sessionService.getSession({
    appName,
    userId,
    sessionId
  });
  if (session) {
    return session;
  }
  return sessionService.createSession({
    appName,
    userId,
    sessionId
  });
}
async function getAdkRunner(runnerOrConfig) {
  if (typeof runnerOrConfig === "function") {
    const result = await runnerOrConfig();
    return getAdkRunner(result);
  }
  if ((0, import_runner.isRunner)(runnerOrConfig)) {
    return runnerOrConfig;
  }
  return new import_runner.Runner(runnerOrConfig);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  A2AAgentExecutor
});
