// @google/adk/dist/cjs/a2a/event_converter_utils.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var event_converter_utils_exports = {};
__export(event_converter_utils_exports, {
  toA2AMessage: () => toA2AMessage,
  toAdkEvent: () => toAdkEvent
});
module.exports = __toCommonJS(event_converter_utils_exports);
var import_genai = require("@google/genai");
var import_event = require("../events/event.js");
var import_event_actions = require("../events/event_actions.js");
var import_env_aware_utils = require("../utils/env_aware_utils.js");
var import_a2a_event = require("./a2a_event.js");
var import_metadata_converter_utils = require("./metadata_converter_utils.js");
var import_part_converter_utils = require("./part_converter_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function toA2AMessage(event, {
  appName,
  userId,
  sessionId
}) {
  var _a;
  return {
    kind: "message",
    messageId: (0, import_env_aware_utils.randomUUID)(),
    role: event.author === import_a2a_event.MessageRole.USER ? import_a2a_event.MessageRole.USER : import_a2a_event.MessageRole.AGENT,
    parts: (0, import_part_converter_utils.toA2AParts)(((_a = event.content) == null ? void 0 : _a.parts) || [], event.longRunningToolIds),
    metadata: (0, import_metadata_converter_utils.getA2AEventMetadata)(event, { appName, userId, sessionId })
  };
}
function toAdkEvent(event, invocationId, agentName) {
  if ((0, import_a2a_event.isMessage)(event)) {
    return messageToAdkEvent(event, invocationId, agentName);
  }
  if ((0, import_a2a_event.isTask)(event)) {
    return taskToAdkEvent(event, invocationId, agentName);
  }
  if ((0, import_a2a_event.isTaskArtifactUpdateEvent)(event)) {
    return artifactUpdateToAdkEvent(event, invocationId, agentName);
  }
  if ((0, import_a2a_event.isTaskStatusUpdateEvent)(event)) {
    return event.final ? finalTaskStatusUpdateToAdkEvent(event, invocationId, agentName) : taskStatusUpdateToAdkEvent(event, invocationId, agentName);
  }
  return void 0;
}
function messageToAdkEvent(msg, invocationId, agentName, parentEvent) {
  const parts = (0, import_part_converter_utils.toGenAIParts)(msg.parts);
  return {
    ...createAdkEventFromMetadata(parentEvent || msg),
    invocationId,
    author: msg.role === import_a2a_event.MessageRole.USER ? import_a2a_event.MessageRole.USER : agentName,
    content: msg.role === import_a2a_event.MessageRole.USER ? (0, import_genai.createUserContent)(parts) : (0, import_genai.createModelContent)(parts),
    turnComplete: true,
    partial: false
  };
}
function artifactUpdateToAdkEvent(a2aEvent, invocationId, agentName) {
  var _a;
  const partsToConvert = ((_a = a2aEvent.artifact) == null ? void 0 : _a.parts) || [];
  if (partsToConvert.length === 0) {
    return void 0;
  }
  const partial = !!(0, import_a2a_event.getEventMetadata)(a2aEvent)[import_metadata_converter_utils.A2AMetadataKeys.PARTIAL] || a2aEvent.append || !a2aEvent.lastChunk;
  return {
    ...createAdkEventFromMetadata(a2aEvent),
    invocationId,
    author: agentName,
    content: (0, import_genai.createModelContent)((0, import_part_converter_utils.toGenAIParts)(partsToConvert)),
    longRunningToolIds: getLongRunningToolIDs(partsToConvert),
    partial
  };
}
function finalTaskStatusUpdateToAdkEvent(a2aEvent, invocationId, agentName) {
  var _a;
  const partsToConvert = ((_a = a2aEvent.status.message) == null ? void 0 : _a.parts) || [];
  if (partsToConvert.length === 0) {
    return void 0;
  }
  const parts = (0, import_part_converter_utils.toGenAIParts)(partsToConvert);
  const isFailedTask = (0, import_a2a_event.isFailedTaskStatusUpdateEvent)(a2aEvent);
  const hasContent = !isFailedTask && parts.length > 0;
  return {
    ...createAdkEventFromMetadata(a2aEvent),
    invocationId,
    author: agentName,
    errorMessage: isFailedTask ? (0, import_a2a_event.getFailedTaskStatusUpdateEventError)(a2aEvent) : void 0,
    content: hasContent ? (0, import_genai.createModelContent)(parts) : void 0,
    longRunningToolIds: getLongRunningToolIDs(partsToConvert),
    turnComplete: true
  };
}
function taskStatusUpdateToAdkEvent(a2aEvent, invocationId, agentName) {
  const msg = a2aEvent.status.message;
  if (!msg) {
    return void 0;
  }
  const parts = (0, import_part_converter_utils.toGenAIParts)(msg.parts);
  return {
    ...createAdkEventFromMetadata(a2aEvent),
    invocationId,
    author: agentName,
    content: (0, import_genai.createModelContent)(parts),
    turnComplete: false,
    partial: true
  };
}
function taskToAdkEvent(a2aTask, invocationId, agentName) {
  var _a, _b;
  const parts = [];
  const longRunningToolIds = [];
  if (a2aTask.artifacts) {
    for (const artifact of a2aTask.artifacts) {
      if (((_a = artifact.parts) == null ? void 0 : _a.length) > 0) {
        const artifactParts = (0, import_part_converter_utils.toGenAIParts)(artifact.parts);
        parts.push(...artifactParts);
        longRunningToolIds.push(...getLongRunningToolIDs(artifact.parts));
      }
    }
  }
  if ((_b = a2aTask.status) == null ? void 0 : _b.message) {
    const a2aParts = a2aTask.status.message.parts;
    const genAIParts = (0, import_part_converter_utils.toGenAIParts)(a2aParts);
    parts.push(...genAIParts);
    longRunningToolIds.push(...getLongRunningToolIDs(a2aParts));
  }
  const isTerminal = (0, import_a2a_event.isTerminalTaskStatusUpdateEvent)(a2aTask) || (0, import_a2a_event.isInputRequiredTaskStatusUpdateEvent)(a2aTask);
  const isFailed = (0, import_a2a_event.isFailedTaskStatusUpdateEvent)(a2aTask);
  if (parts.length === 0 && !isTerminal) {
    return void 0;
  }
  return {
    ...createAdkEventFromMetadata(a2aTask),
    invocationId,
    author: agentName,
    content: isFailed ? void 0 : (0, import_genai.createModelContent)(parts),
    errorMessage: isFailed ? (0, import_a2a_event.getFailedTaskStatusUpdateEventError)(a2aTask) : void 0,
    longRunningToolIds,
    turnComplete: isTerminal
  };
}
function createAdkEventFromMetadata(a2aEvent) {
  const metadata = a2aEvent.metadata || {};
  return (0, import_event.createEvent)({
    branch: metadata[import_metadata_converter_utils.A2AMetadataKeys.BRANCH],
    author: metadata[import_metadata_converter_utils.A2AMetadataKeys.AUTHOR],
    partial: metadata[import_metadata_converter_utils.A2AMetadataKeys.PARTIAL],
    errorCode: metadata[import_metadata_converter_utils.A2AMetadataKeys.ERROR_CODE],
    errorMessage: metadata[import_metadata_converter_utils.A2AMetadataKeys.ERROR_MESSAGE],
    citationMetadata: metadata[import_metadata_converter_utils.A2AMetadataKeys.CITATION_METADATA],
    groundingMetadata: metadata[import_metadata_converter_utils.A2AMetadataKeys.GROUNDING_METADATA],
    usageMetadata: metadata[import_metadata_converter_utils.A2AMetadataKeys.USAGE_METADATA],
    customMetadata: metadata[import_metadata_converter_utils.A2AMetadataKeys.CUSTOM_METADATA],
    actions: (0, import_event_actions.createEventActions)({
      escalate: !!metadata[import_metadata_converter_utils.A2AMetadataKeys.ESCALATE],
      transferToAgent: metadata[import_metadata_converter_utils.A2AMetadataKeys.TRANSFER_TO_AGENT]
    })
  });
}
function getLongRunningToolIDs(parts) {
  const ids = [];
  for (const a2aPart of parts) {
    if (a2aPart.metadata && a2aPart.metadata[import_metadata_converter_utils.A2AMetadataKeys.IS_LONG_RUNNING]) {
      const genAIPart = (0, import_part_converter_utils.toGenAIPart)(a2aPart);
      if (genAIPart.functionCall && genAIPart.functionCall.id) {
        ids.push(genAIPart.functionCall.id);
      }
    }
  }
  return ids;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  toA2AMessage,
  toAdkEvent
});
