// @google/adk/dist/cjs/a2a/event_processor_utils.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var event_processor_utils_exports = {};
__export(event_processor_utils_exports, {
  getFinalTaskStatusUpdate: () => getFinalTaskStatusUpdate,
  getTaskInputRequiredEvent: () => getTaskInputRequiredEvent
});
module.exports = __toCommonJS(event_processor_utils_exports);
var import_event_actions = require("../events/event_actions.js");
var import_a2a_event = require("./a2a_event.js");
var import_metadata_converter_utils = require("./metadata_converter_utils.js");
var import_part_converter_utils = require("./part_converter_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function getFinalTaskStatusUpdate(adkEvents, context) {
  var _a, _b;
  const finalEventActions = (0, import_event_actions.createEventActions)();
  for (const adkEvent of adkEvents) {
    if (adkEvent.errorCode || adkEvent.errorMessage) {
      return (0, import_a2a_event.createTaskFailedEvent)({
        taskId: context.requestContext.taskId,
        contextId: context.requestContext.contextId,
        error: new Error(adkEvent.errorMessage || adkEvent.errorCode),
        metadata: {
          ...(0, import_metadata_converter_utils.getA2AEventMetadata)(adkEvent, context),
          ...(0, import_metadata_converter_utils.getA2AEventMetadataFromActions)(finalEventActions)
        }
      });
    }
    finalEventActions.escalate = finalEventActions.escalate || ((_a = adkEvent.actions) == null ? void 0 : _a.escalate);
    if ((_b = adkEvent.actions) == null ? void 0 : _b.transferToAgent) {
      finalEventActions.transferToAgent = adkEvent.actions.transferToAgent;
    }
  }
  const inputRequiredEvent = scanForInputRequiredEvents(adkEvents, context);
  if (inputRequiredEvent) {
    return {
      ...inputRequiredEvent,
      metadata: {
        ...inputRequiredEvent.metadata,
        ...(0, import_metadata_converter_utils.getA2AEventMetadataFromActions)(finalEventActions)
      }
    };
  }
  return (0, import_a2a_event.createTaskCompletedEvent)({
    taskId: context.requestContext.taskId,
    contextId: context.requestContext.contextId,
    metadata: {
      ...(0, import_metadata_converter_utils.getA2ASessionMetadata)(context),
      ...(0, import_metadata_converter_utils.getA2AEventMetadataFromActions)(finalEventActions)
    }
  });
}
function scanForInputRequiredEvents(adkEvents, context) {
  var _a, _b;
  const inputRequiredParts = [];
  const inputRequiredFunctionCallIds = /* @__PURE__ */ new Set();
  for (const adkEvent of adkEvents) {
    if (!((_b = (_a = adkEvent.content) == null ? void 0 : _a.parts) == null ? void 0 : _b.length)) {
      continue;
    }
    for (const genAIPart of adkEvent.content.parts) {
      const longRunningFunctionCallId = getLongRunnningFunctionCallId(
        genAIPart,
        adkEvent.longRunningToolIds,
        inputRequiredParts
      );
      if (!longRunningFunctionCallId) {
        continue;
      }
      const isAlreadyAdded = inputRequiredFunctionCallIds.has(
        longRunningFunctionCallId
      );
      if (isAlreadyAdded) {
        continue;
      }
      inputRequiredParts.push(genAIPart);
      inputRequiredFunctionCallIds.add(longRunningFunctionCallId);
    }
  }
  if (inputRequiredParts.length > 0) {
    return (0, import_a2a_event.createTaskInputRequiredEvent)({
      taskId: context.requestContext.taskId,
      contextId: context.requestContext.contextId,
      parts: (0, import_part_converter_utils.toA2AParts)(inputRequiredParts, [...inputRequiredFunctionCallIds]),
      metadata: (0, import_metadata_converter_utils.getA2ASessionMetadata)(context)
    });
  }
  return void 0;
}
function getLongRunnningFunctionCallId(genAIPart, longRunningToolIds = [], inputRequiredParts = []) {
  var _a, _b, _c;
  const functionCallId = (_a = genAIPart.functionCall) == null ? void 0 : _a.id;
  const functionResponseId = (_b = genAIPart.functionResponse) == null ? void 0 : _b.id;
  if (!functionCallId && !functionResponseId) {
    return;
  }
  if (functionCallId && longRunningToolIds.includes(functionCallId)) {
    return functionCallId;
  }
  if (functionResponseId && longRunningToolIds.includes(functionResponseId)) {
    return functionResponseId;
  }
  for (const part of inputRequiredParts) {
    if (((_c = part.functionCall) == null ? void 0 : _c.id) === functionResponseId) {
      return functionResponseId;
    }
  }
  return;
}
function getTaskInputRequiredEvent(task, genAIContent) {
  var _a;
  if (!task || !(0, import_a2a_event.isInputRequiredTaskStatusUpdateEvent)(task) || !task.status.message) {
    return void 0;
  }
  const statusMsg = task.status.message;
  const taskParts = (0, import_part_converter_utils.toGenAIParts)(statusMsg.parts);
  for (const taskPart of taskParts) {
    const functionCallId = (_a = taskPart.functionCall) == null ? void 0 : _a.id;
    if (!functionCallId) {
      continue;
    }
    const hasMatchingResponse = ((genAIContent == null ? void 0 : genAIContent.parts) || []).some(
      (p) => {
        var _a2;
        return ((_a2 = p.functionResponse) == null ? void 0 : _a2.id) === functionCallId;
      }
    );
    if (!hasMatchingResponse) {
      return (0, import_a2a_event.createInputMissingErrorEvent)({
        taskId: task.id,
        contextId: task.contextId,
        parts: [
          ...statusMsg.parts.filter((p) => {
            var _a2;
            return !((_a2 = p.metadata) == null ? void 0 : _a2.validation_error);
          }),
          {
            kind: "text",
            text: `No input provided for function call id ${functionCallId}`,
            metadata: {
              validation_error: true
            }
          }
        ]
      });
    }
  }
  return void 0;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getFinalTaskStatusUpdate,
  getTaskInputRequiredEvent
});
