// @google/adk/dist/cjs/a2a/metadata_converter_utils.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var metadata_converter_utils_exports = {};
__export(metadata_converter_utils_exports, {
  A2AMetadataKeys: () => A2AMetadataKeys,
  AdkMetadataKeys: () => AdkMetadataKeys,
  getA2AEventMetadata: () => getA2AEventMetadata,
  getA2AEventMetadataFromActions: () => getA2AEventMetadataFromActions,
  getA2ASessionMetadata: () => getA2ASessionMetadata,
  getAdkEventMetadata: () => getAdkEventMetadata
});
module.exports = __toCommonJS(metadata_converter_utils_exports);
var import_a2a_event = require("./a2a_event.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const ADK_METADATA_KEY_PREFIX = "adk_";
const A2A_METADATA_KEY_PREFIX = "a2a:";
var A2AMetadataKeys = ((A2AMetadataKeys2) => {
  A2AMetadataKeys2["APP_NAME"] = `${ADK_METADATA_KEY_PREFIX}app_name`;
  A2AMetadataKeys2["USER_ID"] = `${ADK_METADATA_KEY_PREFIX}user_id`;
  A2AMetadataKeys2["SESSION_ID"] = `${ADK_METADATA_KEY_PREFIX}session_id`;
  A2AMetadataKeys2["INVOCATION_ID"] = `${ADK_METADATA_KEY_PREFIX}invocation_id`;
  A2AMetadataKeys2["AUTHOR"] = `${ADK_METADATA_KEY_PREFIX}author`;
  A2AMetadataKeys2["BRANCH"] = `${ADK_METADATA_KEY_PREFIX}branch`;
  A2AMetadataKeys2["DATA_PART_TYPE"] = `${ADK_METADATA_KEY_PREFIX}type`;
  A2AMetadataKeys2["PARTIAL"] = `${ADK_METADATA_KEY_PREFIX}partial`;
  A2AMetadataKeys2["ESCALATE"] = `${ADK_METADATA_KEY_PREFIX}escalate`;
  A2AMetadataKeys2["TRANSFER_TO_AGENT"] = `${ADK_METADATA_KEY_PREFIX}transfer_to_agent`;
  A2AMetadataKeys2["IS_LONG_RUNNING"] = `${ADK_METADATA_KEY_PREFIX}is_long_running`;
  A2AMetadataKeys2["THOUGHT"] = `${ADK_METADATA_KEY_PREFIX}thought`;
  A2AMetadataKeys2["ERROR_CODE"] = `${ADK_METADATA_KEY_PREFIX}error_code`;
  A2AMetadataKeys2["ERROR_MESSAGE"] = `${ADK_METADATA_KEY_PREFIX}error_message`;
  A2AMetadataKeys2["CITATION_METADATA"] = `${ADK_METADATA_KEY_PREFIX}citation_metadata`;
  A2AMetadataKeys2["GROUNDING_METADATA"] = `${ADK_METADATA_KEY_PREFIX}grounding_metadata`;
  A2AMetadataKeys2["USAGE_METADATA"] = `${ADK_METADATA_KEY_PREFIX}usage_metadata`;
  A2AMetadataKeys2["CUSTOM_METADATA"] = `${ADK_METADATA_KEY_PREFIX}custom_metadata`;
  A2AMetadataKeys2["VIDEO_METADATA"] = `${ADK_METADATA_KEY_PREFIX}video_metadata`;
  return A2AMetadataKeys2;
})(A2AMetadataKeys || {});
var AdkMetadataKeys = ((AdkMetadataKeys2) => {
  AdkMetadataKeys2["TASK_ID"] = `${A2A_METADATA_KEY_PREFIX}task_id`;
  AdkMetadataKeys2["CONTEXT_ID"] = `${A2A_METADATA_KEY_PREFIX}context_id`;
  return AdkMetadataKeys2;
})(AdkMetadataKeys || {});
function getAdkEventMetadata(a2aEvent) {
  return {
    [AdkMetadataKeys.TASK_ID]: (0, import_a2a_event.isTask)(a2aEvent) ? a2aEvent.id : a2aEvent.taskId,
    [AdkMetadataKeys.CONTEXT_ID]: a2aEvent.contextId
  };
}
function getA2AEventMetadata(adkEvent, {
  appName,
  userId,
  sessionId
}) {
  return {
    ...getA2AEventMetadataFromActions(adkEvent.actions),
    ...getA2ASessionMetadata({
      appName,
      userId,
      sessionId
    }),
    [A2AMetadataKeys.INVOCATION_ID]: adkEvent.invocationId,
    [A2AMetadataKeys.AUTHOR]: adkEvent.author,
    [A2AMetadataKeys.BRANCH]: adkEvent.branch,
    [A2AMetadataKeys.ERROR_CODE]: adkEvent.errorMessage,
    [A2AMetadataKeys.ERROR_MESSAGE]: adkEvent.errorMessage,
    [A2AMetadataKeys.CITATION_METADATA]: adkEvent.citationMetadata,
    [A2AMetadataKeys.GROUNDING_METADATA]: adkEvent.groundingMetadata,
    [A2AMetadataKeys.USAGE_METADATA]: adkEvent.usageMetadata,
    [A2AMetadataKeys.CUSTOM_METADATA]: adkEvent.customMetadata,
    [A2AMetadataKeys.PARTIAL]: adkEvent.partial,
    [A2AMetadataKeys.IS_LONG_RUNNING]: (adkEvent.longRunningToolIds || []).length > 0
  };
}
function getA2ASessionMetadata({
  appName,
  userId,
  sessionId
}) {
  return {
    [A2AMetadataKeys.APP_NAME]: appName,
    [A2AMetadataKeys.USER_ID]: userId,
    [A2AMetadataKeys.SESSION_ID]: sessionId
  };
}
function getA2AEventMetadataFromActions(actions) {
  return {
    [A2AMetadataKeys.ESCALATE]: actions.escalate,
    [A2AMetadataKeys.TRANSFER_TO_AGENT]: actions.transferToAgent
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  A2AMetadataKeys,
  AdkMetadataKeys,
  getA2AEventMetadata,
  getA2AEventMetadataFromActions,
  getA2ASessionMetadata,
  getAdkEventMetadata
});
