// @google/adk/dist/cjs/a2a/part_converter_utils.js
/**
  * @license
  * Copyright 2026 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var part_converter_utils_exports = {};
__export(part_converter_utils_exports, {
  toA2ADataPart: () => toA2ADataPart,
  toA2AFilePart: () => toA2AFilePart,
  toA2APart: () => toA2APart,
  toA2AParts: () => toA2AParts,
  toA2ATextPart: () => toA2ATextPart,
  toGenAIContent: () => toGenAIContent,
  toGenAIPart: () => toGenAIPart,
  toGenAIPartData: () => toGenAIPartData,
  toGenAIPartFile: () => toGenAIPartFile,
  toGenAIPartText: () => toGenAIPartText,
  toGenAIParts: () => toGenAIParts
});
module.exports = __toCommonJS(part_converter_utils_exports);
var import_genai = require("@google/genai");
var import_metadata_converter_utils = require("./metadata_converter_utils.js");
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var DataPartType = /* @__PURE__ */ ((DataPartType2) => {
  DataPartType2["FUNCTION_CALL"] = "function_call";
  DataPartType2["FUNCTION_RESPONSE"] = "function_response";
  DataPartType2["CODE_EXEC_RESULT"] = "code_execution_result";
  DataPartType2["CODE_EXECUTABLE_CODE"] = "executable_code";
  return DataPartType2;
})(DataPartType || {});
function toA2AParts(parts = [], longRunningToolIDs = []) {
  return parts.map((part) => toA2APart(part, longRunningToolIDs));
}
function toA2APart(part, longRunningToolIDs) {
  if (part.text !== void 0 && part.text !== null) {
    return toA2ATextPart(part);
  }
  if (part.inlineData || part.fileData) {
    return toA2AFilePart(part);
  }
  return toA2ADataPart(part, longRunningToolIDs);
}
function toA2ATextPart(part) {
  const a2aPart = { kind: "text", text: part.text || "" };
  if (part.thought) {
    a2aPart.metadata = {
      [import_metadata_converter_utils.A2AMetadataKeys.THOUGHT]: true
    };
  }
  return a2aPart;
}
function toA2AFilePart(part) {
  const metadata = {};
  if (part.videoMetadata) {
    metadata[import_metadata_converter_utils.A2AMetadataKeys.VIDEO_METADATA] = part.videoMetadata;
  }
  if (part.fileData) {
    return {
      kind: "file",
      file: {
        uri: part.fileData.fileUri || "",
        mimeType: part.fileData.mimeType
      },
      metadata
    };
  }
  if (part.inlineData) {
    return {
      kind: "file",
      file: {
        bytes: part.inlineData.data || "",
        mimeType: part.inlineData.mimeType
      },
      metadata
    };
  }
  throw new Error(`Not a file part: ${JSON.stringify(part)}`);
}
function toA2ADataPart(part, longRunningToolIDs = []) {
  let dataPartType;
  let data;
  if (part.functionCall) {
    dataPartType = "function_call" /* FUNCTION_CALL */;
    data = part.functionCall;
  } else if (part.functionResponse) {
    dataPartType = "function_response" /* FUNCTION_RESPONSE */;
    data = part.functionResponse;
  } else if (part.executableCode) {
    dataPartType = "executable_code" /* CODE_EXECUTABLE_CODE */;
    data = part.executableCode;
  } else if (part.codeExecutionResult) {
    dataPartType = "code_execution_result" /* CODE_EXEC_RESULT */;
    data = part.codeExecutionResult;
  } else {
    return {
      kind: "data",
      data: {},
      metadata: {}
    };
  }
  const metadata = {
    [import_metadata_converter_utils.A2AMetadataKeys.DATA_PART_TYPE]: dataPartType
  };
  if (part.functionCall && part.functionCall.id && longRunningToolIDs.includes(part.functionCall.id)) {
    metadata[import_metadata_converter_utils.A2AMetadataKeys.IS_LONG_RUNNING] = true;
  }
  if (part.functionResponse && part.functionResponse.id && longRunningToolIDs.includes(part.functionResponse.id)) {
    metadata[import_metadata_converter_utils.A2AMetadataKeys.IS_LONG_RUNNING] = true;
  }
  return {
    kind: "data",
    data,
    metadata
  };
}
function toGenAIContent(a2aMessage) {
  const parts = toGenAIParts(a2aMessage.parts);
  return a2aMessage.role === "user" ? (0, import_genai.createUserContent)(parts) : (0, import_genai.createModelContent)(parts);
}
function toGenAIParts(a2aParts) {
  return a2aParts.map((a2aPart) => toGenAIPart(a2aPart));
}
function toGenAIPart(a2aPart) {
  if (a2aPart.kind === "text") {
    return toGenAIPartText(a2aPart);
  }
  if (a2aPart.kind === "file") {
    return toGenAIPartFile(a2aPart);
  }
  if (a2aPart.kind === "data") {
    return toGenAIPartData(a2aPart);
  }
  throw new Error(`Unknown part kind: ${JSON.stringify(a2aPart)}`);
}
function toGenAIPartText(a2aPart) {
  var _a;
  return {
    text: a2aPart.text,
    thought: !!((_a = a2aPart.metadata) == null ? void 0 : _a[import_metadata_converter_utils.A2AMetadataKeys.THOUGHT])
  };
}
function toGenAIPartFile(a2aPart) {
  var _a;
  const part = {};
  if ((_a = a2aPart.metadata) == null ? void 0 : _a[import_metadata_converter_utils.A2AMetadataKeys.VIDEO_METADATA]) {
    part.videoMetadata = a2aPart.metadata[import_metadata_converter_utils.A2AMetadataKeys.VIDEO_METADATA];
  }
  if ("bytes" in a2aPart.file) {
    part.inlineData = {
      data: a2aPart.file.bytes,
      mimeType: a2aPart.file.mimeType || ""
    };
    return part;
  }
  if ("uri" in a2aPart.file) {
    part.fileData = {
      fileUri: a2aPart.file.uri,
      mimeType: a2aPart.file.mimeType || ""
    };
    return part;
  }
  throw new Error(`Not a file part: ${JSON.stringify(a2aPart)}`);
}
function toGenAIPartData(a2aPart) {
  var _a;
  if (!a2aPart.data) {
    throw new Error(`No data in part: ${JSON.stringify(a2aPart)}`);
  }
  const data = a2aPart.data;
  const type = (_a = a2aPart.metadata) == null ? void 0 : _a[import_metadata_converter_utils.A2AMetadataKeys.DATA_PART_TYPE];
  if (type === "function_call" /* FUNCTION_CALL */) {
    return { functionCall: data };
  }
  if (type === "function_response" /* FUNCTION_RESPONSE */) {
    return { functionResponse: data };
  }
  if (type === "executable_code" /* CODE_EXECUTABLE_CODE */) {
    return { executableCode: data };
  }
  if (type === "code_execution_result" /* CODE_EXEC_RESULT */) {
    return { codeExecutionResult: data };
  }
  return {
    text: JSON.stringify(a2aPart.data)
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  toA2ADataPart,
  toA2AFilePart,
  toA2APart,
  toA2AParts,
  toA2ATextPart,
  toGenAIContent,
  toGenAIPart,
  toGenAIPartData,
  toGenAIPartFile,
  toGenAIPartText,
  toGenAIParts
});
