// @google/adk/dist/web/a2a/a2a_remote_agent_run_processor.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  createModelContent
} from "@google/genai";
import { createEvent } from "../events/event.js";
import {
  getEventMetadata,
  isTask,
  isTaskArtifactUpdateEvent,
  isTaskStatusUpdateEvent
} from "./a2a_event.js";
import { A2AMetadataKeys } from "./metadata_converter_utils.js";
class A2ARemoteAgentRunProcessor {
  constructor(request) {
    this.request = request;
    this.aggregations = /* @__PURE__ */ new Map();
    this.aggregationOrder = [];
  }
  /**
   * aggregatePartial stores contents of partial events to emit them with the terminal event.
   * It can return multiple events to emit instead of just the provided one.
   */
  aggregatePartial(context, a2aEvent, adkEvent) {
    const metadata = getEventMetadata(a2aEvent);
    if (metadata[A2AMetadataKeys.PARTIAL]) {
      return [adkEvent];
    }
    if (isTaskStatusUpdateEvent(a2aEvent) && a2aEvent.final) {
      const events = [];
      for (const aid of this.aggregationOrder) {
        const agg = this.aggregations.get(aid);
        if (agg) {
          events.push(this.buildNonPartialAggregation(context, agg));
        }
      }
      this.aggregations.clear();
      this.aggregationOrder = [];
      return [...events, adkEvent];
    }
    if (isTask(a2aEvent)) {
      this.aggregations.clear();
      this.aggregationOrder = [];
      return [adkEvent];
    }
    if (!isTaskArtifactUpdateEvent(a2aEvent)) {
      return [adkEvent];
    }
    const artifactId = a2aEvent.artifact.artifactId;
    if (!a2aEvent.append) {
      this.removeAggregation(artifactId);
      if (a2aEvent.lastChunk) {
        adkEvent.partial = false;
        return [adkEvent];
      }
    }
    let aggregation = this.aggregations.get(artifactId);
    if (!aggregation) {
      aggregation = {
        aggregatedText: "",
        aggregatedThoughts: "",
        parts: []
      };
      this.aggregations.set(artifactId, aggregation);
      this.aggregationOrder.push(artifactId);
    } else {
      this.aggregationOrder = this.aggregationOrder.filter(
        (id) => id !== artifactId
      );
      this.aggregationOrder.push(artifactId);
    }
    this.updateAggregation(aggregation, adkEvent);
    if (!a2aEvent.lastChunk) {
      return [adkEvent];
    }
    this.removeAggregation(artifactId);
    return [adkEvent, this.buildNonPartialAggregation(context, aggregation)];
  }
  removeAggregation(artifactId) {
    this.aggregations.delete(artifactId);
    this.aggregationOrder = this.aggregationOrder.filter(
      (id) => id !== artifactId
    );
  }
  updateAggregation(agg, event) {
    var _a;
    const parts = ((_a = event.content) == null ? void 0 : _a.parts) || [];
    for (const part of parts) {
      if (part.text && part.text !== "") {
        if (part.thought) {
          agg.aggregatedThoughts += part.text;
        } else {
          agg.aggregatedText += part.text;
        }
      } else {
        this.promoteTextBlocksToParts(agg);
        agg.parts.push(part);
      }
    }
    if (event.citationMetadata) {
      if (!agg.citations) {
        agg.citations = { citations: [] };
      }
      if (!agg.citations.citations) {
        agg.citations.citations = [];
      }
      agg.citations.citations.push(...event.citationMetadata.citations || []);
    }
    if (event.customMetadata) {
      if (!agg.customMeta) {
        agg.customMeta = {};
      }
      Object.assign(agg.customMeta, event.customMetadata);
    }
    if (event.groundingMetadata) {
      agg.grounding = event.groundingMetadata;
    }
    if (event.usageMetadata) {
      agg.usage = event.usageMetadata;
    }
  }
  buildNonPartialAggregation(context, agg) {
    this.promoteTextBlocksToParts(agg);
    const result = createEvent({
      author: context.agent.name,
      invocationId: context.invocationId,
      content: agg.parts.length > 0 ? createModelContent([...agg.parts]) : void 0,
      customMetadata: agg.customMeta,
      groundingMetadata: agg.grounding,
      citationMetadata: agg.citations,
      usageMetadata: agg.usage,
      turnComplete: false,
      partial: false
    });
    return result;
  }
  promoteTextBlocksToParts(agg) {
    if (agg.aggregatedThoughts !== "") {
      agg.parts.push({ thought: true, text: agg.aggregatedThoughts });
      agg.aggregatedThoughts = "";
    }
    if (agg.aggregatedText !== "") {
      agg.parts.push({ text: agg.aggregatedText });
      agg.aggregatedText = "";
    }
  }
  /**
   * Adds request and response metadata to the event.
   */
  updateCustomMetadata(event, response) {
    const toAdd = {};
    if (this.request && event.turnComplete) {
      toAdd["request"] = this.request;
    }
    if (response) {
      toAdd["response"] = response;
      if (isTask(response)) {
        if (response.id) toAdd["task_id"] = response.id;
        if (response.contextId) toAdd["context_id"] = response.contextId;
      } else if (response.taskId) {
        toAdd["task_id"] = response.taskId;
        if (response.contextId) toAdd["context_id"] = response.contextId;
      }
    }
    if (Object.keys(toAdd).length === 0) {
      return;
    }
    if (!event.customMetadata) {
      event.customMetadata = {};
    }
    for (const [k, v] of Object.entries(toAdd)) {
      if (v === void 0 || v === null) continue;
      event.customMetadata["a2a:".concat(k)] = v;
    }
  }
}
export {
  A2ARemoteAgentRunProcessor
};
