// @google/adk/dist/web/a2a/a2a_remote_agent_utils.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { createEvent } from "../events/event.js";
import { AdkMetadataKeys } from "./metadata_converter_utils.js";
import { toA2AParts } from "./part_converter_utils.js";
function getUserFunctionCallAt(session, index) {
  const events = session.events;
  if (index < 0 || index >= events.length) {
    return void 0;
  }
  const candidate = events[index];
  if (candidate.author !== "user") {
    return void 0;
  }
  const fnCallId = getFunctionResponseCallId(candidate);
  if (!fnCallId) {
    return void 0;
  }
  for (let i = index - 1; i >= 0; i--) {
    const request = events[i];
    if (!isFunctionCallEvent(request, fnCallId)) {
      continue;
    }
    const metadata = request.customMetadata || {};
    const taskId = metadata[AdkMetadataKeys.TASK_ID] || "";
    const contextId = metadata[AdkMetadataKeys.CONTEXT_ID] || "";
    return {
      response: candidate,
      taskId,
      contextId
    };
  }
  return void 0;
}
function isFunctionCallEvent(event, callId) {
  if (!event || !event.content || !event.content.parts) {
    return false;
  }
  return event.content.parts.some(
    (part) => part.functionCall && part.functionCall.id === callId
  );
}
function getFunctionResponseCallId(event) {
  var _a;
  if (!event || !event.content || !event.content.parts) {
    return void 0;
  }
  const responsePart = event.content.parts.find(
    (part) => part.functionResponse
  );
  return (_a = responsePart == null ? void 0 : responsePart.functionResponse) == null ? void 0 : _a.id;
}
function toMissingRemoteSessionParts(ctx, session) {
  const events = session.events;
  let contextId = void 0;
  let lastRemoteResponseIndex = -1;
  for (let i = events.length - 1; i >= 0; i--) {
    const event = events[i];
    if (event.author === ctx.agent.name) {
      lastRemoteResponseIndex = i;
      const metadata = event.customMetadata || {};
      contextId = metadata[AdkMetadataKeys.CONTEXT_ID];
      break;
    }
  }
  const missingParts = [];
  for (let i = lastRemoteResponseIndex + 1; i < events.length; i++) {
    let event = events[i];
    if (event.author !== "user" && event.author !== ctx.agent.name) {
      event = presentAsUserMessage(ctx, event);
    }
    if (!event.content || !event.content.parts || event.content.parts.length === 0) {
      continue;
    }
    const parts = toA2AParts(event.content.parts, event.longRunningToolIds);
    missingParts.push(...parts);
  }
  return {
    parts: missingParts,
    contextId
  };
}
function presentAsUserMessage(ctx, agentEvent) {
  const event = createEvent({
    author: "user",
    invocationId: ctx.invocationId
  });
  if (!agentEvent.content || !agentEvent.content.parts) {
    return event;
  }
  const parts = [{ text: "For context:" }];
  for (const part of agentEvent.content.parts) {
    if (part.thought) {
      continue;
    }
    if (part.text) {
      parts.push({
        text: "[".concat(agentEvent.author, "] said: ").concat(part.text)
      });
    } else if (part.functionCall) {
      const call = part.functionCall;
      parts.push({
        text: "[".concat(agentEvent.author, "] called tool ").concat(call.name, " with parameters: ").concat(JSON.stringify(call.args))
      });
    } else if (part.functionResponse) {
      const resp = part.functionResponse;
      parts.push({
        text: "[".concat(agentEvent.author, "] ").concat(resp.name, " tool returned result: ").concat(JSON.stringify(resp.response))
      });
    } else {
      parts.push(part);
    }
  }
  if (parts.length > 1) {
    event.content = {
      role: "user",
      parts
    };
  }
  return event;
}
export {
  getFunctionResponseCallId,
  getUserFunctionCallAt,
  isFunctionCallEvent,
  presentAsUserMessage,
  toMissingRemoteSessionParts
};
