// @google/adk/dist/web/a2a/a2a_remote_agent.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __await = function(promise, isYieldStar) {
  this[0] = promise;
  this[1] = isYieldStar;
};
var __asyncGenerator = (__this, __arguments, generator) => {
  var resume = (k, v, yes, no) => {
    try {
      var x = generator[k](v), isAwait = (v = x.value) instanceof __await, done = x.done;
      Promise.resolve(isAwait ? v[0] : v).then((y) => isAwait ? resume(k === "return" ? k : "next", v[1] ? { done: y.done, value: y.value } : y, yes, no) : yes({ value: y, done })).catch((e) => resume("throw", e, yes, no));
    } catch (e) {
      no(e);
    }
  }, method = (k) => it[k] = (x) => new Promise((yes, no) => resume(k, x, yes, no)), it = {};
  return generator = generator.apply(__this, __arguments), it[__knownSymbol("asyncIterator")] = () => it, method("next"), method("throw"), method("return"), it;
};
var __forAwait = (obj, it, method) => (it = obj[__knownSymbol("asyncIterator")]) ? it.call(obj) : (obj = obj[__knownSymbol("iterator")](), it = {}, method = (key, fn) => (fn = obj[key]) && (it[key] = (arg) => new Promise((yes, no, done) => (arg = fn.call(obj, arg), done = arg.done, Promise.resolve(arg.value).then((value) => yes({ value, done }), no)))), method("next"), method("return"), it);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  AGENT_CARD_PATH
} from "@a2a-js/sdk";
import { ClientFactory } from "@a2a-js/sdk/client";
import { BaseAgent } from "../agents/base_agent.js";
import { createEvent } from "../events/event.js";
import { randomUUID } from "../utils/env_aware_utils.js";
import { logger } from "../utils/logger.js";
import { MessageRole } from "./a2a_event.js";
import { A2ARemoteAgentRunProcessor } from "./a2a_remote_agent_run_processor.js";
import {
  getUserFunctionCallAt,
  toMissingRemoteSessionParts
} from "./a2a_remote_agent_utils.js";
import { resolveAgentCard } from "./agent_card.js";
import { toAdkEvent } from "./event_converter_utils.js";
import { getA2ASessionMetadata } from "./metadata_converter_utils.js";
import { toA2AParts } from "./part_converter_utils.js";
class RemoteA2AAgent extends BaseAgent {
  constructor(a2aConfig) {
    super(a2aConfig);
    this.a2aConfig = a2aConfig;
    this.isInitialized = false;
    if (!a2aConfig.agentCard && !a2aConfig.client) {
      throw new Error("Either AgentCard or Client must be provided");
    }
  }
  async init() {
    if (this.isInitialized) {
      return;
    }
    if (this.a2aConfig.client) {
      this.client = this.a2aConfig.client;
    }
    if (this.a2aConfig.agentCard) {
      this.card = await resolveAgentCard(this.a2aConfig.agentCard);
      if (!this.client) {
        const factory = this.a2aConfig.clientFactory || new ClientFactory();
        this.client = await factory.createFromAgentCard(this.card);
      }
    }
    this.isInitialized = true;
  }
  runAsyncImpl(context) {
    return __asyncGenerator(this, null, function* () {
      var _a, _b;
      yield new __await(this.init());
      try {
        const events = context.session.events;
        if (events.length === 0) {
          throw new Error("No events in session to send");
        }
        const userFnCall = getUserFunctionCallAt(
          context.session,
          events.length - 1
        );
        let parts;
        let taskId = void 0;
        let contextId = void 0;
        if (userFnCall) {
          const event = userFnCall.response;
          parts = toA2AParts(
            ((_a = event.content) == null ? void 0 : _a.parts) || [],
            event.longRunningToolIds
          );
          taskId = userFnCall.taskId;
          contextId = userFnCall.contextId;
        } else {
          const missing = toMissingRemoteSessionParts(context, context.session);
          parts = missing.parts;
          contextId = missing.contextId;
        }
        const message = {
          kind: "message",
          messageId: randomUUID(),
          role: MessageRole.USER,
          parts,
          metadata: getA2ASessionMetadata({
            appName: context.session.appName,
            userId: context.session.userId,
            sessionId: context.session.id
          })
        };
        if (taskId) message.taskId = taskId;
        if (contextId) message.contextId = contextId;
        const params = {
          message,
          configuration: this.a2aConfig.messageSendConfig
        };
        const processor = new A2ARemoteAgentRunProcessor(params);
        if (this.a2aConfig.beforeRequestCallbacks) {
          for (const callback of this.a2aConfig.beforeRequestCallbacks) {
            yield new __await(callback(context, params));
          }
        }
        const useStreaming = this.card ? ((_b = this.card.capabilities) == null ? void 0 : _b.streaming) !== false : true;
        if (useStreaming) {
          try {
            for (var iter = __forAwait(this.client.sendMessageStream(params)), more, temp, error; more = !(temp = yield new __await(iter.next())).done; more = false) {
              const chunk = temp.value;
              if (this.a2aConfig.afterRequestCallbacks) {
                for (const callback of this.a2aConfig.afterRequestCallbacks) {
                  yield new __await(callback(context, chunk));
                }
              }
              const adkEvent = toAdkEvent(chunk, context.invocationId, this.name);
              if (!adkEvent) {
                continue;
              }
              processor.updateCustomMetadata(adkEvent, chunk);
              const eventsToEmit = processor.aggregatePartial(
                context,
                chunk,
                adkEvent
              );
              for (const ev of eventsToEmit) {
                yield ev;
              }
            }
          } catch (temp) {
            error = [temp];
          } finally {
            try {
              more && (temp = iter.return) && (yield new __await(temp.call(iter)));
            } finally {
              if (error)
                throw error[0];
            }
          }
        } else {
          const result = yield new __await(this.client.sendMessage(params));
          if (this.a2aConfig.afterRequestCallbacks) {
            for (const callback of this.a2aConfig.afterRequestCallbacks) {
              yield new __await(callback(context, result));
            }
          }
          const adkEvent = toAdkEvent(result, context.invocationId, this.name);
          if (adkEvent) {
            processor.updateCustomMetadata(adkEvent, result);
            yield adkEvent;
          }
        }
      } catch (e) {
        const error2 = e;
        logger.error("A2ARemoteAgent ".concat(this.name, " failed:"), error2);
        yield createEvent({
          author: this.name,
          invocationId: context.invocationId,
          errorMessage: error2.message,
          turnComplete: true
        });
      }
    });
  }
  runLiveImpl(_context) {
    throw new Error("Live mode is not supported in A2ARemoteAgent yet.");
  }
}
export {
  AGENT_CARD_PATH,
  RemoteA2AAgent
};
