// @google/adk/dist/web/a2a/agent_card.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { DefaultAgentCardResolver } from "@a2a-js/sdk/client";
import * as fs from "node:fs/promises";
import {
  InvocationContext
} from "../agents/invocation_context.js";
import { isLlmAgent } from "../agents/llm_agent.js";
import { isLoopAgent } from "../agents/loop_agent.js";
import { isParallelAgent } from "../agents/parallel_agent.js";
import { ReadonlyContext } from "../agents/readonly_context.js";
import { isSequentialAgent } from "../agents/sequential_agent.js";
import { isBaseTool } from "../tools/base_tool.js";
import { isBaseToolset } from "../tools/base_toolset.js";
import { logger } from "../utils/logger.js";
async function resolveAgentCard(agentCard) {
  if (typeof agentCard === "object") {
    return agentCard;
  }
  const source = agentCard;
  if (source.startsWith("http://") || source.startsWith("https://")) {
    const resolver = new DefaultAgentCardResolver();
    return await resolver.resolve(source);
  }
  try {
    const content = await fs.readFile(source, "utf-8");
    return JSON.parse(content);
  } catch (err) {
    throw new Error(
      "Failed to read agent card from file ".concat(source, ": ").concat(err.message)
    );
  }
}
async function getA2AAgentCard(agent, transports) {
  return {
    name: agent.name,
    description: agent.description || "",
    protocolVersion: "0.3.0",
    version: "1.0.0",
    skills: await buildAgentSkills(agent),
    url: transports[0].url,
    preferredTransport: transports[0].transport,
    capabilities: {
      extensions: [],
      stateTransitionHistory: false,
      pushNotifications: false,
      streaming: true
    },
    defaultInputModes: ["text"],
    defaultOutputModes: ["text"],
    additionalInterfaces: transports
  };
}
async function buildAgentSkills(agent) {
  const [primarySkills, subAgentSkills] = await Promise.all([
    buildPrimarySkills(agent),
    buildSubAgentSkills(agent)
  ]);
  return [...primarySkills, ...subAgentSkills];
}
async function buildPrimarySkills(agent) {
  if (isLlmAgent(agent)) {
    return buildLLMAgentSkills(agent);
  }
  return buildNonLLMAgentSkills(agent);
}
async function buildSubAgentSkills(agent) {
  const subAgents = agent.subAgents;
  const result = [];
  for (const sub of subAgents) {
    const skills = await buildPrimarySkills(sub);
    for (const subSkill of skills) {
      const skill = {
        id: "".concat(sub.name, "_").concat(subSkill.id),
        name: "".concat(sub.name, ": ").concat(subSkill.name),
        description: subSkill.description,
        tags: ["sub_agent:".concat(sub.name), ...subSkill.tags]
      };
      result.push(skill);
    }
  }
  return result;
}
async function buildLLMAgentSkills(agent) {
  const skills = [
    {
      id: agent.name,
      name: "model",
      description: await buildDescriptionFromInstructions(agent),
      tags: ["llm"]
    }
  ];
  if (agent.tools && agent.tools.length > 0) {
    for (const toolUnion of agent.tools) {
      if (isBaseTool(toolUnion)) {
        skills.push(toolToSkill(agent.name, toolUnion));
      } else if (isBaseToolset(toolUnion)) {
        const tools = await toolUnion.getTools();
        for (const tool of tools) {
          skills.push(toolToSkill(agent.name, tool));
        }
      }
    }
  }
  return skills;
}
function toolToSkill(prefix, tool) {
  let description = tool.description;
  if (!description) {
    description = "Tool: ".concat(tool.name);
  }
  return {
    id: "".concat(prefix, "-").concat(tool.name),
    name: tool.name,
    description,
    tags: ["llm", "tools"]
  };
}
function buildNonLLMAgentSkills(agent) {
  const skills = [
    {
      id: agent.name,
      name: getAgentSkillName(agent),
      description: buildAgentDescription(agent),
      tags: [getAgentTypeTag(agent)]
    }
  ];
  const subAgents = agent.subAgents;
  if (subAgents.length > 0) {
    const descriptions = subAgents.map(
      (sub) => sub.description || "No description"
    );
    skills.push({
      id: "".concat(agent.name, "-sub-agents"),
      name: "sub-agents",
      description: "Orchestrates: ".concat(descriptions.join("; ")),
      tags: [getAgentTypeTag(agent), "orchestration"]
    });
  }
  return skills;
}
function buildAgentDescription(agent) {
  const descriptionParts = [];
  if (agent.description) {
    descriptionParts.push(agent.description);
  }
  if (agent.subAgents.length > 0) {
    if (isLoopAgent(agent)) {
      descriptionParts.push(buildLoopAgentDescription(agent));
    } else if (isParallelAgent(agent)) {
      descriptionParts.push(buildParallelAgentDescription(agent));
    } else if (isSequentialAgent(agent)) {
      descriptionParts.push(buildSequentialAgentDescription(agent));
    }
  }
  if (descriptionParts.length > 0) {
    return descriptionParts.join(" ");
  } else {
    return getDefaultAgentDescription(agent);
  }
}
function buildSequentialAgentDescription(agent) {
  const subAgents = agent.subAgents;
  const descriptions = [];
  subAgents.forEach((sub, i) => {
    let subDescription = sub.description;
    if (!subDescription) {
      subDescription = "execute the ".concat(sub.name, " agent");
    }
    if (i === 0) {
      descriptions.push("First, this agent will ".concat(subDescription, "."));
    } else if (i === subAgents.length - 1) {
      descriptions.push("Finally, this agent will ".concat(subDescription, "."));
    } else {
      descriptions.push("Then, this agent will ".concat(subDescription, "."));
    }
  });
  return descriptions.join(" ");
}
function buildParallelAgentDescription(agent) {
  const subAgents = agent.subAgents;
  const descriptions = [];
  subAgents.forEach((sub, i) => {
    let subDescription = sub.description;
    if (!subDescription) {
      subDescription = "execute the ".concat(sub.name, " agent");
    }
    if (i === 0) {
      descriptions.push("This agent will ".concat(subDescription));
    } else if (i === subAgents.length - 1) {
      descriptions.push("and ".concat(subDescription));
    } else {
      descriptions.push(", ".concat(subDescription));
    }
  });
  return "".concat(descriptions.join(" "), " simultaneously.");
}
function buildLoopAgentDescription(agent) {
  const maxIterationsVal = agent.maxIterations;
  let maxIterations = "unlimited";
  if (typeof maxIterationsVal === "number" && maxIterationsVal < Number.MAX_SAFE_INTEGER) {
    maxIterations = maxIterationsVal.toString();
  }
  const subAgents = agent.subAgents;
  const descriptions = [];
  subAgents.forEach((sub, i) => {
    let subDescription = sub.description;
    if (!subDescription) {
      subDescription = "execute the ".concat(sub.name, " agent");
    }
    if (i === 0) {
      descriptions.push("This agent will ".concat(subDescription));
    } else if (i === subAgents.length - 1) {
      descriptions.push("and ".concat(subDescription));
    } else {
      descriptions.push(", ".concat(subDescription));
    }
  });
  return "".concat(descriptions.join(" "), " in a loop (max ").concat(maxIterations, " iterations).");
}
async function buildDescriptionFromInstructions(agent) {
  const descriptionParts = [];
  if (agent.description) {
    descriptionParts.push(agent.description);
  }
  if (agent.instruction) {
    let instructionStr;
    if (typeof agent.instruction === "function") {
      const dummyContext = new ReadonlyContext(
        new InvocationContext({
          agent
        })
      );
      try {
        instructionStr = await agent.instruction(dummyContext);
      } catch (e) {
        logger.warn("Failed to resolve dynamic instruction for AgentCard", e);
        instructionStr = "";
      }
    } else {
      instructionStr = agent.instruction;
    }
    if (instructionStr) {
      descriptionParts.push(replacePronouns(instructionStr));
    }
  }
  const root = agent.rootAgent;
  if (isLlmAgent(root) && root.globalInstruction) {
    let globalInstructionStr;
    if (typeof root.globalInstruction === "function") {
      const dummyContext = new ReadonlyContext(
        new InvocationContext({
          agent
        })
      );
      try {
        globalInstructionStr = await root.globalInstruction(dummyContext);
      } catch (e) {
        logger.warn(
          "Failed to resolve dynamic global instruction for AgentCard",
          e
        );
        globalInstructionStr = "";
      }
    } else {
      globalInstructionStr = root.globalInstruction;
    }
    if (globalInstructionStr) {
      descriptionParts.push(replacePronouns(globalInstructionStr));
    }
  }
  if (descriptionParts.length > 0) {
    return descriptionParts.join(" ");
  } else {
    return getDefaultAgentDescription(agent);
  }
}
function replacePronouns(instruction) {
  const substitutions = [
    { original: "you were", target: "I was" },
    { original: "you are", target: "I am" },
    { original: "you're", target: "I am" },
    { original: "you've", target: "I have" },
    { original: "yours", target: "mine" },
    { original: "your", target: "my" },
    { original: "you", target: "I" }
  ];
  let result = instruction;
  for (const sub of substitutions) {
    const pattern = new RegExp("\\b".concat(sub.original, "\\b"), "gi");
    result = result.replace(pattern, sub.target);
  }
  return result;
}
function getDefaultAgentDescription(agent) {
  if (isLoopAgent(agent)) {
    return "A loop workflow agent";
  } else if (isSequentialAgent(agent)) {
    return "A sequential workflow agent";
  } else if (isParallelAgent(agent)) {
    return "A parallel workflow agent";
  } else if (isLlmAgent(agent)) {
    return "An LLM-based agent";
  } else {
    return "A custom agent";
  }
}
function getAgentTypeTag(agent) {
  if (isLoopAgent(agent)) {
    return "loop_workflow";
  } else if (isSequentialAgent(agent)) {
    return "sequential_workflow";
  } else if (isParallelAgent(agent)) {
    return "parallel_workflow";
  } else if (isLlmAgent(agent)) {
    return "llm_agent";
  } else {
    return "custom_agent";
  }
}
function getAgentSkillName(agent) {
  if (isLlmAgent(agent)) {
    return "model";
  }
  if (isWorkflowAgent(agent)) {
    return "workflow";
  }
  return "custom";
}
function isWorkflowAgent(agent) {
  return isLoopAgent(agent) || isSequentialAgent(agent) || isParallelAgent(agent);
}
export {
  buildAgentSkills,
  getA2AAgentCard,
  resolveAgentCard
};
