// @google/adk/dist/web/a2a/agent_executor.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __forAwait = (obj, it, method) => (it = obj[__knownSymbol("asyncIterator")]) ? it.call(obj) : (obj = obj[__knownSymbol("iterator")](), it = {}, method = (key, fn) => (fn = obj[key]) && (it[key] = (arg) => new Promise((yes, no, done) => (arg = fn.call(obj, arg), done = arg.done, Promise.resolve(arg.value).then((value) => yes({ value, done }), no)))), method("next"), method("return"), it);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { isRunner, Runner } from "../runner/runner.js";
import { randomUUID } from "../utils/env_aware_utils.js";
import { logger } from "../utils/logger.js";
import {
  createTask,
  createTaskArtifactUpdateEvent,
  createTaskFailedEvent,
  createTaskWorkingEvent
} from "./a2a_event.js";
import {
  getFinalTaskStatusUpdate,
  getTaskInputRequiredEvent
} from "./event_processor_utils.js";
import { createExecutorContext } from "./executor_context.js";
import {
  getA2AEventMetadata,
  getA2ASessionMetadata
} from "./metadata_converter_utils.js";
import { toA2AParts, toGenAIContent } from "./part_converter_utils.js";
class A2AAgentExecutor {
  constructor(config) {
    this.config = config;
    this.agentPartialArtifactIdsMap = {};
  }
  async execute(ctx, eventBus) {
    var _a, _b;
    const a2aUserMessage = ctx.userMessage;
    if (!a2aUserMessage) {
      throw new Error("message not provided");
    }
    const userId = "A2A_USER_".concat(ctx.contextId);
    const sessionId = ctx.contextId;
    const genAIUserMessage = toGenAIContent(a2aUserMessage);
    const adkRunner = await getAdkRunner(this.config.runner);
    const session = await getAdkSession(
      userId,
      sessionId,
      adkRunner.sessionService,
      adkRunner.appName
    );
    const executorContext = createExecutorContext({
      session,
      userContent: genAIUserMessage,
      requestContext: ctx
    });
    try {
      if (this.config.beforeExecuteCallback) {
        await this.config.beforeExecuteCallback(ctx);
      }
      if (ctx.task) {
        const inputRequiredEvent = getTaskInputRequiredEvent(
          ctx.task,
          genAIUserMessage
        );
        if (inputRequiredEvent) {
          await this.publishFinalTaskStatus({
            executorContext,
            eventBus,
            event: inputRequiredEvent
          });
          return;
        }
      }
      if (!ctx.task) {
        eventBus.publish(
          createTask({
            taskId: ctx.taskId,
            contextId: ctx.contextId,
            message: a2aUserMessage
          })
        );
      }
      eventBus.publish(
        createTaskWorkingEvent({
          taskId: ctx.taskId,
          contextId: ctx.contextId
        })
      );
      const adkEvents = [];
      try {
        for (var iter = __forAwait(adkRunner.runAsync({
          userId,
          sessionId,
          newMessage: genAIUserMessage,
          runConfig: this.config.runConfig
        })), more, temp, error; more = !(temp = await iter.next()).done; more = false) {
          const adkEvent = temp.value;
          adkEvents.push(adkEvent);
          const a2aEvent = this.convertAdkEventToA2AEvent(
            adkEvent,
            executorContext
          );
          if (!a2aEvent) {
            continue;
          }
          await ((_b = (_a = this.config).afterEventCallback) == null ? void 0 : _b.call(
            _a,
            executorContext,
            adkEvent,
            a2aEvent
          ));
          eventBus.publish(a2aEvent);
        }
      } catch (temp) {
        error = [temp];
      } finally {
        try {
          more && (temp = iter.return) && await temp.call(iter);
        } finally {
          if (error)
            throw error[0];
        }
      }
      await this.publishFinalTaskStatus({
        executorContext,
        eventBus,
        event: getFinalTaskStatusUpdate(adkEvents, executorContext)
      });
    } catch (e) {
      const error2 = e;
      await this.publishFinalTaskStatus({
        executorContext,
        eventBus,
        error: error2,
        event: createTaskFailedEvent({
          taskId: ctx.taskId,
          contextId: ctx.contextId,
          error: new Error("Agent run failed: ".concat(error2.message)),
          metadata: getA2ASessionMetadata(executorContext)
        })
      });
    }
  }
  // Task cancellation is not supported in this implementation yet.
  async cancelTask(_taskId) {
    throw new Error("Task cancellation is not supported yet.");
  }
  convertAdkEventToA2AEvent(adkEvent, executorContext) {
    var _a;
    const a2aParts = toA2AParts(
      (_a = adkEvent.content) == null ? void 0 : _a.parts,
      adkEvent.longRunningToolIds
    );
    if (a2aParts.length === 0) {
      return void 0;
    }
    const artifactId = this.agentPartialArtifactIdsMap[adkEvent.author] || randomUUID();
    const a2aEvent = createTaskArtifactUpdateEvent({
      taskId: executorContext.requestContext.taskId,
      contextId: executorContext.requestContext.contextId,
      artifactId,
      parts: a2aParts,
      metadata: getA2AEventMetadata(adkEvent, executorContext),
      append: adkEvent.partial,
      lastChunk: !adkEvent.partial
    });
    if (adkEvent.partial) {
      this.agentPartialArtifactIdsMap[adkEvent.author] = artifactId;
    } else {
      delete this.agentPartialArtifactIdsMap[adkEvent.author];
    }
    return a2aEvent;
  }
  /**
   * Writes the final status event to the queue.
   */
  async publishFinalTaskStatus({
    executorContext,
    eventBus,
    event,
    error
  }) {
    var _a, _b;
    try {
      await ((_b = (_a = this.config).afterExecuteCallback) == null ? void 0 : _b.call(_a, executorContext, event, error));
    } catch (e) {
      logger.error("Error in afterExecuteCallback:", e);
    }
    eventBus.publish(event);
  }
}
async function getAdkSession(userId, sessionId, sessionService, appName) {
  const session = await sessionService.getSession({
    appName,
    userId,
    sessionId
  });
  if (session) {
    return session;
  }
  return sessionService.createSession({
    appName,
    userId,
    sessionId
  });
}
async function getAdkRunner(runnerOrConfig) {
  if (typeof runnerOrConfig === "function") {
    const result = await runnerOrConfig();
    return getAdkRunner(result);
  }
  if (isRunner(runnerOrConfig)) {
    return runnerOrConfig;
  }
  return new Runner(runnerOrConfig);
}
export {
  A2AAgentExecutor
};
