// @google/adk/dist/web/a2a/agent_to_a2a.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { AGENT_CARD_PATH } from "@a2a-js/sdk";
import { DefaultRequestHandler, InMemoryTaskStore } from "@a2a-js/sdk/server";
import {
  agentCardHandler,
  jsonRpcHandler,
  restHandler,
  UserBuilder
} from "@a2a-js/sdk/server/express";
import express from "express";
import { StreamingMode } from "../agents/run_config.js";
import { InMemorySessionService } from "../sessions/in_memory_session_service.js";
import { getA2AAgentCard, resolveAgentCard } from "./agent_card.js";
import { A2AAgentExecutor } from "./agent_executor.js";
async function toA2a(agent, options = {}) {
  var _a, _b, _c, _d;
  const host = (_a = options.host) != null ? _a : "localhost";
  const port = (_b = options.port) != null ? _b : 8e3;
  const protocol = (_c = options.protocol) != null ? _c : "http";
  const basePath = options.basePath || "";
  const rpcUrl = "".concat(protocol, "://").concat(host, ":").concat(port).concat(basePath);
  const agentCard = options.agentCard ? await resolveAgentCard(options.agentCard) : await getA2AAgentCard(agent, [
    {
      url: "".concat(rpcUrl, "/jsonrpc"),
      transport: "JSONRPC"
    },
    {
      url: "".concat(rpcUrl, "/rest"),
      transport: "HTTP+JSON"
    }
  ]);
  const agentExecutor = new A2AAgentExecutor({
    runner: options.runner || {
      agent,
      appName: agent.name,
      sessionService: options.sessionService || new InMemorySessionService(),
      memoryService: options.memoryService,
      artifactService: options.artifactService
    },
    runConfig: {
      streamingMode: StreamingMode.SSE
    }
  });
  const requestHandler = new DefaultRequestHandler(
    agentCard,
    new InMemoryTaskStore(),
    agentExecutor
  );
  const app = (_d = options.app) != null ? _d : express();
  if (!options.app) {
    app.use(express.urlencoded({ limit: "50mb", extended: true }));
    app.use(express.json({ limit: "50mb" }));
  }
  app.use(
    "".concat(basePath, "/").concat(AGENT_CARD_PATH),
    agentCardHandler({ agentCardProvider: requestHandler })
  );
  app.use(
    "".concat(basePath, "/rest"),
    restHandler({
      requestHandler,
      userBuilder: UserBuilder.noAuthentication
    })
  );
  app.use(
    "".concat(basePath, "/jsonrpc"),
    jsonRpcHandler({
      requestHandler,
      userBuilder: UserBuilder.noAuthentication
    })
  );
  return app;
}
export {
  toA2a
};
