// @google/adk/dist/web/a2a/event_processor_utils.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { createEventActions } from "../events/event_actions.js";
import {
  createInputMissingErrorEvent,
  createTaskCompletedEvent,
  createTaskFailedEvent,
  createTaskInputRequiredEvent,
  isInputRequiredTaskStatusUpdateEvent
} from "./a2a_event.js";
import {
  getA2AEventMetadata,
  getA2AEventMetadataFromActions,
  getA2ASessionMetadata
} from "./metadata_converter_utils.js";
import { toA2AParts, toGenAIParts } from "./part_converter_utils.js";
function getFinalTaskStatusUpdate(adkEvents, context) {
  var _a, _b;
  const finalEventActions = createEventActions();
  for (const adkEvent of adkEvents) {
    if (adkEvent.errorCode || adkEvent.errorMessage) {
      return createTaskFailedEvent({
        taskId: context.requestContext.taskId,
        contextId: context.requestContext.contextId,
        error: new Error(adkEvent.errorMessage || adkEvent.errorCode),
        metadata: __spreadValues(__spreadValues({}, getA2AEventMetadata(adkEvent, context)), getA2AEventMetadataFromActions(finalEventActions))
      });
    }
    finalEventActions.escalate = finalEventActions.escalate || ((_a = adkEvent.actions) == null ? void 0 : _a.escalate);
    if ((_b = adkEvent.actions) == null ? void 0 : _b.transferToAgent) {
      finalEventActions.transferToAgent = adkEvent.actions.transferToAgent;
    }
  }
  const inputRequiredEvent = scanForInputRequiredEvents(adkEvents, context);
  if (inputRequiredEvent) {
    return __spreadProps(__spreadValues({}, inputRequiredEvent), {
      metadata: __spreadValues(__spreadValues({}, inputRequiredEvent.metadata), getA2AEventMetadataFromActions(finalEventActions))
    });
  }
  return createTaskCompletedEvent({
    taskId: context.requestContext.taskId,
    contextId: context.requestContext.contextId,
    metadata: __spreadValues(__spreadValues({}, getA2ASessionMetadata(context)), getA2AEventMetadataFromActions(finalEventActions))
  });
}
function scanForInputRequiredEvents(adkEvents, context) {
  var _a, _b;
  const inputRequiredParts = [];
  const inputRequiredFunctionCallIds = /* @__PURE__ */ new Set();
  for (const adkEvent of adkEvents) {
    if (!((_b = (_a = adkEvent.content) == null ? void 0 : _a.parts) == null ? void 0 : _b.length)) {
      continue;
    }
    for (const genAIPart of adkEvent.content.parts) {
      const longRunningFunctionCallId = getLongRunnningFunctionCallId(
        genAIPart,
        adkEvent.longRunningToolIds,
        inputRequiredParts
      );
      if (!longRunningFunctionCallId) {
        continue;
      }
      const isAlreadyAdded = inputRequiredFunctionCallIds.has(
        longRunningFunctionCallId
      );
      if (isAlreadyAdded) {
        continue;
      }
      inputRequiredParts.push(genAIPart);
      inputRequiredFunctionCallIds.add(longRunningFunctionCallId);
    }
  }
  if (inputRequiredParts.length > 0) {
    return createTaskInputRequiredEvent({
      taskId: context.requestContext.taskId,
      contextId: context.requestContext.contextId,
      parts: toA2AParts(inputRequiredParts, [...inputRequiredFunctionCallIds]),
      metadata: getA2ASessionMetadata(context)
    });
  }
  return void 0;
}
function getLongRunnningFunctionCallId(genAIPart, longRunningToolIds = [], inputRequiredParts = []) {
  var _a, _b, _c;
  const functionCallId = (_a = genAIPart.functionCall) == null ? void 0 : _a.id;
  const functionResponseId = (_b = genAIPart.functionResponse) == null ? void 0 : _b.id;
  if (!functionCallId && !functionResponseId) {
    return;
  }
  if (functionCallId && longRunningToolIds.includes(functionCallId)) {
    return functionCallId;
  }
  if (functionResponseId && longRunningToolIds.includes(functionResponseId)) {
    return functionResponseId;
  }
  for (const part of inputRequiredParts) {
    if (((_c = part.functionCall) == null ? void 0 : _c.id) === functionResponseId) {
      return functionResponseId;
    }
  }
  return;
}
function getTaskInputRequiredEvent(task, genAIContent) {
  var _a;
  if (!task || !isInputRequiredTaskStatusUpdateEvent(task) || !task.status.message) {
    return void 0;
  }
  const statusMsg = task.status.message;
  const taskParts = toGenAIParts(statusMsg.parts);
  for (const taskPart of taskParts) {
    const functionCallId = (_a = taskPart.functionCall) == null ? void 0 : _a.id;
    if (!functionCallId) {
      continue;
    }
    const hasMatchingResponse = ((genAIContent == null ? void 0 : genAIContent.parts) || []).some(
      (p) => {
        var _a2;
        return ((_a2 = p.functionResponse) == null ? void 0 : _a2.id) === functionCallId;
      }
    );
    if (!hasMatchingResponse) {
      return createInputMissingErrorEvent({
        taskId: task.id,
        contextId: task.contextId,
        parts: [
          ...statusMsg.parts.filter((p) => {
            var _a2;
            return !((_a2 = p.metadata) == null ? void 0 : _a2.validation_error);
          }),
          {
            kind: "text",
            text: "No input provided for function call id ".concat(functionCallId),
            metadata: {
              validation_error: true
            }
          }
        ]
      });
    }
  }
  return void 0;
}
export {
  getFinalTaskStatusUpdate,
  getTaskInputRequiredEvent
};
