// @google/adk/dist/web/a2a/metadata_converter_utils.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { isTask } from "./a2a_event.js";
const ADK_METADATA_KEY_PREFIX = "adk_";
const A2A_METADATA_KEY_PREFIX = "a2a:";
var A2AMetadataKeys = ((A2AMetadataKeys2) => {
  A2AMetadataKeys2[A2AMetadataKeys2["APP_NAME"] = "".concat(ADK_METADATA_KEY_PREFIX, "app_name")] = "APP_NAME";
  A2AMetadataKeys2[A2AMetadataKeys2["USER_ID"] = "".concat(ADK_METADATA_KEY_PREFIX, "user_id")] = "USER_ID";
  A2AMetadataKeys2[A2AMetadataKeys2["SESSION_ID"] = "".concat(ADK_METADATA_KEY_PREFIX, "session_id")] = "SESSION_ID";
  A2AMetadataKeys2[A2AMetadataKeys2["INVOCATION_ID"] = "".concat(ADK_METADATA_KEY_PREFIX, "invocation_id")] = "INVOCATION_ID";
  A2AMetadataKeys2[A2AMetadataKeys2["AUTHOR"] = "".concat(ADK_METADATA_KEY_PREFIX, "author")] = "AUTHOR";
  A2AMetadataKeys2[A2AMetadataKeys2["BRANCH"] = "".concat(ADK_METADATA_KEY_PREFIX, "branch")] = "BRANCH";
  A2AMetadataKeys2[A2AMetadataKeys2["DATA_PART_TYPE"] = "".concat(ADK_METADATA_KEY_PREFIX, "type")] = "DATA_PART_TYPE";
  A2AMetadataKeys2[A2AMetadataKeys2["PARTIAL"] = "".concat(ADK_METADATA_KEY_PREFIX, "partial")] = "PARTIAL";
  A2AMetadataKeys2[A2AMetadataKeys2["ESCALATE"] = "".concat(ADK_METADATA_KEY_PREFIX, "escalate")] = "ESCALATE";
  A2AMetadataKeys2[A2AMetadataKeys2["TRANSFER_TO_AGENT"] = "".concat(ADK_METADATA_KEY_PREFIX, "transfer_to_agent")] = "TRANSFER_TO_AGENT";
  A2AMetadataKeys2[A2AMetadataKeys2["IS_LONG_RUNNING"] = "".concat(ADK_METADATA_KEY_PREFIX, "is_long_running")] = "IS_LONG_RUNNING";
  A2AMetadataKeys2[A2AMetadataKeys2["THOUGHT"] = "".concat(ADK_METADATA_KEY_PREFIX, "thought")] = "THOUGHT";
  A2AMetadataKeys2[A2AMetadataKeys2["ERROR_CODE"] = "".concat(ADK_METADATA_KEY_PREFIX, "error_code")] = "ERROR_CODE";
  A2AMetadataKeys2[A2AMetadataKeys2["ERROR_MESSAGE"] = "".concat(ADK_METADATA_KEY_PREFIX, "error_message")] = "ERROR_MESSAGE";
  A2AMetadataKeys2[A2AMetadataKeys2["CITATION_METADATA"] = "".concat(ADK_METADATA_KEY_PREFIX, "citation_metadata")] = "CITATION_METADATA";
  A2AMetadataKeys2[A2AMetadataKeys2["GROUNDING_METADATA"] = "".concat(ADK_METADATA_KEY_PREFIX, "grounding_metadata")] = "GROUNDING_METADATA";
  A2AMetadataKeys2[A2AMetadataKeys2["USAGE_METADATA"] = "".concat(ADK_METADATA_KEY_PREFIX, "usage_metadata")] = "USAGE_METADATA";
  A2AMetadataKeys2[A2AMetadataKeys2["CUSTOM_METADATA"] = "".concat(ADK_METADATA_KEY_PREFIX, "custom_metadata")] = "CUSTOM_METADATA";
  A2AMetadataKeys2[A2AMetadataKeys2["VIDEO_METADATA"] = "".concat(ADK_METADATA_KEY_PREFIX, "video_metadata")] = "VIDEO_METADATA";
  return A2AMetadataKeys2;
})(A2AMetadataKeys || {});
var AdkMetadataKeys = ((AdkMetadataKeys2) => {
  AdkMetadataKeys2[AdkMetadataKeys2["TASK_ID"] = "".concat(A2A_METADATA_KEY_PREFIX, "task_id")] = "TASK_ID";
  AdkMetadataKeys2[AdkMetadataKeys2["CONTEXT_ID"] = "".concat(A2A_METADATA_KEY_PREFIX, "context_id")] = "CONTEXT_ID";
  return AdkMetadataKeys2;
})(AdkMetadataKeys || {});
function getAdkEventMetadata(a2aEvent) {
  return {
    [AdkMetadataKeys.TASK_ID]: isTask(a2aEvent) ? a2aEvent.id : a2aEvent.taskId,
    [AdkMetadataKeys.CONTEXT_ID]: a2aEvent.contextId
  };
}
function getA2AEventMetadata(adkEvent, {
  appName,
  userId,
  sessionId
}) {
  return __spreadProps(__spreadValues(__spreadValues({}, getA2AEventMetadataFromActions(adkEvent.actions)), getA2ASessionMetadata({
    appName,
    userId,
    sessionId
  })), {
    [A2AMetadataKeys.INVOCATION_ID]: adkEvent.invocationId,
    [A2AMetadataKeys.AUTHOR]: adkEvent.author,
    [A2AMetadataKeys.BRANCH]: adkEvent.branch,
    [A2AMetadataKeys.ERROR_CODE]: adkEvent.errorMessage,
    [A2AMetadataKeys.ERROR_MESSAGE]: adkEvent.errorMessage,
    [A2AMetadataKeys.CITATION_METADATA]: adkEvent.citationMetadata,
    [A2AMetadataKeys.GROUNDING_METADATA]: adkEvent.groundingMetadata,
    [A2AMetadataKeys.USAGE_METADATA]: adkEvent.usageMetadata,
    [A2AMetadataKeys.CUSTOM_METADATA]: adkEvent.customMetadata,
    [A2AMetadataKeys.PARTIAL]: adkEvent.partial,
    [A2AMetadataKeys.IS_LONG_RUNNING]: (adkEvent.longRunningToolIds || []).length > 0
  });
}
function getA2ASessionMetadata({
  appName,
  userId,
  sessionId
}) {
  return {
    [A2AMetadataKeys.APP_NAME]: appName,
    [A2AMetadataKeys.USER_ID]: userId,
    [A2AMetadataKeys.SESSION_ID]: sessionId
  };
}
function getA2AEventMetadataFromActions(actions) {
  return {
    [A2AMetadataKeys.ESCALATE]: actions.escalate,
    [A2AMetadataKeys.TRANSFER_TO_AGENT]: actions.transferToAgent
  };
}
export {
  A2AMetadataKeys,
  AdkMetadataKeys,
  getA2AEventMetadata,
  getA2AEventMetadataFromActions,
  getA2ASessionMetadata,
  getAdkEventMetadata
};
