// @google/adk/dist/web/a2a/part_converter_utils.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  createModelContent,
  createUserContent
} from "@google/genai";
import { A2AMetadataKeys } from "./metadata_converter_utils.js";
var DataPartType = /* @__PURE__ */ ((DataPartType2) => {
  DataPartType2["FUNCTION_CALL"] = "function_call";
  DataPartType2["FUNCTION_RESPONSE"] = "function_response";
  DataPartType2["CODE_EXEC_RESULT"] = "code_execution_result";
  DataPartType2["CODE_EXECUTABLE_CODE"] = "executable_code";
  return DataPartType2;
})(DataPartType || {});
function toA2AParts(parts = [], longRunningToolIDs = []) {
  return parts.map((part) => toA2APart(part, longRunningToolIDs));
}
function toA2APart(part, longRunningToolIDs) {
  if (part.text !== void 0 && part.text !== null) {
    return toA2ATextPart(part);
  }
  if (part.inlineData || part.fileData) {
    return toA2AFilePart(part);
  }
  return toA2ADataPart(part, longRunningToolIDs);
}
function toA2ATextPart(part) {
  const a2aPart = { kind: "text", text: part.text || "" };
  if (part.thought) {
    a2aPart.metadata = {
      [A2AMetadataKeys.THOUGHT]: true
    };
  }
  return a2aPart;
}
function toA2AFilePart(part) {
  const metadata = {};
  if (part.videoMetadata) {
    metadata[A2AMetadataKeys.VIDEO_METADATA] = part.videoMetadata;
  }
  if (part.fileData) {
    return {
      kind: "file",
      file: {
        uri: part.fileData.fileUri || "",
        mimeType: part.fileData.mimeType
      },
      metadata
    };
  }
  if (part.inlineData) {
    return {
      kind: "file",
      file: {
        bytes: part.inlineData.data || "",
        mimeType: part.inlineData.mimeType
      },
      metadata
    };
  }
  throw new Error("Not a file part: ".concat(JSON.stringify(part)));
}
function toA2ADataPart(part, longRunningToolIDs = []) {
  let dataPartType;
  let data;
  if (part.functionCall) {
    dataPartType = "function_call" /* FUNCTION_CALL */;
    data = part.functionCall;
  } else if (part.functionResponse) {
    dataPartType = "function_response" /* FUNCTION_RESPONSE */;
    data = part.functionResponse;
  } else if (part.executableCode) {
    dataPartType = "executable_code" /* CODE_EXECUTABLE_CODE */;
    data = part.executableCode;
  } else if (part.codeExecutionResult) {
    dataPartType = "code_execution_result" /* CODE_EXEC_RESULT */;
    data = part.codeExecutionResult;
  } else {
    return {
      kind: "data",
      data: {},
      metadata: {}
    };
  }
  const metadata = {
    [A2AMetadataKeys.DATA_PART_TYPE]: dataPartType
  };
  if (part.functionCall && part.functionCall.id && longRunningToolIDs.includes(part.functionCall.id)) {
    metadata[A2AMetadataKeys.IS_LONG_RUNNING] = true;
  }
  if (part.functionResponse && part.functionResponse.id && longRunningToolIDs.includes(part.functionResponse.id)) {
    metadata[A2AMetadataKeys.IS_LONG_RUNNING] = true;
  }
  return {
    kind: "data",
    data,
    metadata
  };
}
function toGenAIContent(a2aMessage) {
  const parts = toGenAIParts(a2aMessage.parts);
  return a2aMessage.role === "user" ? createUserContent(parts) : createModelContent(parts);
}
function toGenAIParts(a2aParts) {
  return a2aParts.map((a2aPart) => toGenAIPart(a2aPart));
}
function toGenAIPart(a2aPart) {
  if (a2aPart.kind === "text") {
    return toGenAIPartText(a2aPart);
  }
  if (a2aPart.kind === "file") {
    return toGenAIPartFile(a2aPart);
  }
  if (a2aPart.kind === "data") {
    return toGenAIPartData(a2aPart);
  }
  throw new Error("Unknown part kind: ".concat(JSON.stringify(a2aPart)));
}
function toGenAIPartText(a2aPart) {
  var _a;
  return {
    text: a2aPart.text,
    thought: !!((_a = a2aPart.metadata) == null ? void 0 : _a[A2AMetadataKeys.THOUGHT])
  };
}
function toGenAIPartFile(a2aPart) {
  var _a;
  const part = {};
  if ((_a = a2aPart.metadata) == null ? void 0 : _a[A2AMetadataKeys.VIDEO_METADATA]) {
    part.videoMetadata = a2aPart.metadata[A2AMetadataKeys.VIDEO_METADATA];
  }
  if ("bytes" in a2aPart.file) {
    part.inlineData = {
      data: a2aPart.file.bytes,
      mimeType: a2aPart.file.mimeType || ""
    };
    return part;
  }
  if ("uri" in a2aPart.file) {
    part.fileData = {
      fileUri: a2aPart.file.uri,
      mimeType: a2aPart.file.mimeType || ""
    };
    return part;
  }
  throw new Error("Not a file part: ".concat(JSON.stringify(a2aPart)));
}
function toGenAIPartData(a2aPart) {
  var _a;
  if (!a2aPart.data) {
    throw new Error("No data in part: ".concat(JSON.stringify(a2aPart)));
  }
  const data = a2aPart.data;
  const type = (_a = a2aPart.metadata) == null ? void 0 : _a[A2AMetadataKeys.DATA_PART_TYPE];
  if (type === "function_call" /* FUNCTION_CALL */) {
    return { functionCall: data };
  }
  if (type === "function_response" /* FUNCTION_RESPONSE */) {
    return { functionResponse: data };
  }
  if (type === "executable_code" /* CODE_EXECUTABLE_CODE */) {
    return { executableCode: data };
  }
  if (type === "code_execution_result" /* CODE_EXEC_RESULT */) {
    return { codeExecutionResult: data };
  }
  return {
    text: JSON.stringify(a2aPart.data)
  };
}
export {
  toA2ADataPart,
  toA2AFilePart,
  toA2APart,
  toA2AParts,
  toA2ATextPart,
  toGenAIContent,
  toGenAIPart,
  toGenAIPartData,
  toGenAIPartFile,
  toGenAIPartText,
  toGenAIParts
};
