// @google/adk-devtools/dist/esm/integration/agent_registry.js

const Module = await import('node:module'); 
const require = /* @__PURE__ */ (Module?.default?.createRequire ?? Module?.createRequire)(import.meta.url);
const __filename = /* @__PURE__ */ (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = /* @__PURE__ */ (await import("node:path")).dirname(__filename);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  AgentTool,
  LlmAgent,
  LoopAgent,
  MCPToolset,
  ParallelAgent,
  SequentialAgent
} from "@google/adk";
import * as path from "node:path";
const BUILTIN_TOOLS = [
  "exit_loop",
  "google_search",
  "url_context",
  "google_maps_grounding"
];
class AgentRegistry {
  constructor(integrationRegistry) {
    this.agents = /* @__PURE__ */ new Map();
    this.configs = /* @__PURE__ */ new Map();
    this.instantiating = /* @__PURE__ */ new Set();
    this.integrationRegistry = integrationRegistry;
  }
  registerAgent(name, agent) {
    this.agents.set(name, agent);
  }
  getAgent(name) {
    if (this.agents.has(name)) {
      return this.agents.get(name);
    }
    if (this.configs.has(name)) {
      return this.instantiateAgent(name, this.configs.get(name));
    }
    return void 0;
  }
  // Returns the root agent based on the last directory name.
  // if there are somehow more than one root agent in a directory, an arbitrary one is returned
  getRootAgentByShortName(shortName) {
    for (const [name, config] of this.configs) {
      if (path.basename(path.dirname(name)) === shortName && config.isRootAgent) {
        return this.getAgent(name);
      }
    }
    return void 0;
  }
  registerAgentConfig(name, config) {
    this.configs.set(name, config);
  }
  findToolOrThrow(name) {
    const tool = this.integrationRegistry.getTool(name);
    if (!tool) {
      console.log("Tool not found in registry", name);
      throw new Error(`Tool ${name} not found in registry`);
    }
    return tool;
  }
  instantiateAgent(name, config) {
    var _a, _b, _c, _d, _e;
    console.log(
      "Instantiating",
      name,
      "of class",
      config.agentClass ?? "LlmAgent"
    );
    if (this.instantiating.has(name)) {
      throw new Error(`Circular dependency detected for agent ${name}`);
    }
    this.instantiating.add(name);
    try {
      const beforeAgentCallbacks = (_a = config.beforeAgentCallbacks) == null ? void 0 : _a.map(
        (callbackInfo) => {
          const callback = this.integrationRegistry.getBeforeAgentCallback(
            callbackInfo.name
          );
          if (!callback) {
            throw new Error(
              `BeforeAgentCallback ${callbackInfo.name} not found in registry`
            );
          }
          return callback;
        }
      );
      const afterAgentCallbacks = (_b = config.afterAgentCallbacks) == null ? void 0 : _b.map(
        (callbackInfo) => {
          const callback = this.integrationRegistry.getAfterAgentCallback(
            callbackInfo.name
          );
          if (!callback) {
            throw new Error(
              `AfterAgentCallback ${callbackInfo.name} not found in registry`
            );
          }
          return callback;
        }
      );
      const subAgents = (_c = config.subAgents) == null ? void 0 : _c.map((ref) => {
        const subAgent = this.getAgent(ref.configPath);
        if (!subAgent) {
          throw new Error(
            `SubAgent ${ref.configPath} not found in registry (referenced by ${name})`
          );
        }
        return subAgent;
      });
      const tools = (_d = config.tools) == null ? void 0 : _d.map((toolConfig) => {
        if (BUILTIN_TOOLS.includes(toolConfig.name)) {
          return void 0;
        }
        if (toolConfig.name == "LongRunningFunctionTool") {
          const args = toolConfig.args;
          const subTool = this.findToolOrThrow(args.func);
          return subTool;
        }
        if (toolConfig.name === "AgentTool") {
          const args = toolConfig.args;
          const subAgent = this.getAgent(args.agent.configPath);
          if (!subAgent) {
            throw new Error(
              `Agent ${args.agent.configPath} not found in registry (referenced by AgentTool in ${name})`
            );
          }
          return new AgentTool({ agent: subAgent });
        }
        if (toolConfig.name === "McpToolset") {
          const args = toolConfig.args;
          if (args.stdioConnectionParams.type !== "StdioConnectionParams") {
            args.stdioConnectionParams.type = "StdioConnectionParams";
          }
          const toolset = new MCPToolset(
            args.stdioConnectionParams,
            args.toolFilter
          );
          return toolset;
        }
        return this.findToolOrThrow(toolConfig.name);
      }).filter((tool) => tool !== void 0);
      const options = {
        name: config.name,
        model: config.model,
        description: config.description,
        instruction: config.instruction,
        beforeAgentCallback: beforeAgentCallbacks ?? [],
        afterAgentCallback: afterAgentCallbacks ?? [],
        subAgents: subAgents ?? [],
        tools: tools ?? [],
        disallowTransferToParent: config.disallowTransferToParent === "true",
        disallowTransferToPeers: config.disallowTransferToPeers === "true",
        temperature: (_e = config.generateContentConfig) == null ? void 0 : _e.temperature
      };
      let agent;
      switch (config.agentClass) {
        case "LoopAgent":
          agent = new LoopAgent({
            ...options,
            maxIterations: config.maxIterations ? parseInt(config.maxIterations, 10) : void 0
          });
          break;
        case "ParallelAgent":
          agent = new ParallelAgent(options);
          break;
        case "SequentialAgent":
          agent = new SequentialAgent(options);
          break;
        case "LlmAgent":
        default:
          agent = new LlmAgent(options);
          break;
      }
      this.registerAgent(name, agent);
      return agent;
    } finally {
      this.instantiating.delete(name);
    }
  }
  summary() {
    return `${this.configs.size} configs, ${this.agents.size} instantiated agents`;
  }
}
export {
  AgentRegistry
};
