// @google/adk-devtools/dist/esm/integration/integration_registry.js

const Module = await import('node:module'); 
const require = /* @__PURE__ */ (Module?.default?.createRequire ?? Module?.createRequire)(import.meta.url);
const __filename = /* @__PURE__ */ (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = /* @__PURE__ */ (await import("node:path")).dirname(__filename);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
class IntegrationRegistry {
  constructor() {
    this.tools = /* @__PURE__ */ new Map();
    this.beforeAgentCallbacks = /* @__PURE__ */ new Map();
    this.afterAgentCallbacks = /* @__PURE__ */ new Map();
    this.plugins = /* @__PURE__ */ new Map();
  }
  summary() {
    return `${this.tools.size} tools, ${this.beforeAgentCallbacks.size} before agent callbacks, ${this.afterAgentCallbacks.size} after agent callbacks, and ${this.plugins.size} plugins.`;
  }
  registerTool(name, tool) {
    this.tools.set(name, tool);
  }
  getTool(name) {
    return this.tools.get(name);
  }
  registerBeforeAgentCallback(name, callback) {
    this.beforeAgentCallbacks.set(name, callback);
  }
  getBeforeAgentCallback(name) {
    return this.beforeAgentCallbacks.get(name);
  }
  registerAfterAgentCallback(name, callback) {
    this.afterAgentCallbacks.set(name, callback);
  }
  getAfterAgentCallback(name) {
    return this.afterAgentCallbacks.get(name);
  }
  registerPlugin(name, plugin) {
    this.plugins.set(name, plugin);
  }
  getPlugin(name) {
    return this.plugins.get(name);
  }
}
export {
  IntegrationRegistry
};
