// @google/adk-devtools/dist/esm/integration/test_runner.js

const Module = await import('node:module'); 
const require = /* @__PURE__ */ (Module?.default?.createRequire ?? Module?.createRequire)(import.meta.url);
const __filename = /* @__PURE__ */ (await import("node:url")).fileURLToPath(import.meta.url);
const __dirname = /* @__PURE__ */ (await import("node:path")).dirname(__filename);
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  InMemorySessionService,
  isLlmAgent,
  Runner
} from "@google/adk";
import { cloneDeep } from "lodash-es";
import * as assert from "node:assert";
import { DummyLlm } from "./dummy_llm.js";
import { ReplayPlugin } from "./replay_plugin.js";
const SKIPPED_TESTS = [
  {
    name: "tool/example_tool_001",
    reason: "ExampleTool is not implemented yet."
  },
  { name: "workflow/loop_001", reason: "ExitLoopTool is not implemented yet." },
  {
    name: "core/multi_005",
    reason: "Suspected broken test. Need to re-evaluate."
  },
  {
    name: "tool/long_running_tool_001",
    reason: "Suspected broken test. Need to re-evaluate."
  }
];
class TestRunner {
  constructor(agentRegistry) {
    this.agentRegistry = agentRegistry;
  }
  async run(testInfo, force) {
    if (!force) {
      for (const skip of SKIPPED_TESTS) {
        if (skip.name == testInfo.name) {
          console.log("Skipping test", testInfo.name, "because:", skip.reason);
          return true;
        }
      }
    }
    const agentName = testInfo.spec.agent;
    const agent = this.agentRegistry.getRootAgentByShortName(agentName);
    if (!agent) {
      throw new Error(`Agent ${agentName} not found in registry`);
    }
    const recordings = cloneDeep(testInfo.recordings.recordings);
    const context = { userMessageIndex: 0 };
    injectDummyLlm(agent);
    const replayPlugin = new ReplayPlugin(recordings, context);
    const sessionService = new InMemorySessionService();
    const runner = new Runner({
      agent,
      sessionService,
      plugins: [replayPlugin],
      appName: "test-runner"
    });
    const userId = "test-user";
    const sessionId = "test-session";
    await sessionService.createSession({
      appName: "test-runner",
      userId,
      sessionId
    });
    const userMessages = testInfo.spec.userMessages;
    for (let i = 0; i < userMessages.length; i++) {
      context.userMessageIndex = i;
      const userMsg = userMessages[i];
      const content = userMessageToContent(userMsg);
      const iterator = runner.runAsync({
        userId,
        sessionId,
        newMessage: content,
        stateDelta: i === 0 ? testInfo.spec.initialState : void 0
      });
      for await (const _ of iterator) {
      }
    }
    const session = await sessionService.getSession({
      appName: "test-runner",
      userId,
      sessionId
    });
    if (!session) {
      throw new Error("Session not found after execution");
    }
    validateSession(session, testInfo.session);
    return false;
  }
}
function injectDummyLlm(agent) {
  if (isLlmAgent(agent)) {
    agent.model = new DummyLlm();
  }
  const subAgents = agent.subAgents;
  if (subAgents && Array.isArray(subAgents)) {
    for (const sub of subAgents) {
      injectDummyLlm(sub);
    }
  }
}
function userMessageToContent(msg) {
  if (msg.content) {
    const content = msg.content;
    content.role = "user";
    return content;
  }
  if (msg.text) {
    return { role: "user", parts: [{ text: msg.text }] };
  }
  throw new Error("Either Content text or content field is required");
}
function validateSession(actual, expected) {
  const actualEvents = actual.events.map(normalizeEvent);
  const expectedEvents = expected.events.map(normalizeEvent);
  assert.deepStrictEqual(actualEvents, expectedEvents);
}
function normalizeEvent(event) {
  const filteredEvent = event;
  filterEventFields(filteredEvent);
  removeEmptyAndUndefinedFields(
    filteredEvent
  );
  return filteredEvent;
}
function removeEmptyAndUndefinedFields(obj) {
  for (const key in obj) {
    if (Object.hasOwn(obj, key)) {
      if (obj[key] === void 0 || obj[key] === null) {
        delete obj[key];
      } else if (Array.isArray(obj[key])) {
        for (let i = 0; i < obj[key].length; i++) {
          removeEmptyAndUndefinedFields(obj[key][i]);
        }
        if (obj[key].length === 0) {
          delete obj[key];
          continue;
        }
      } else if (typeof obj[key] === "object") {
        removeEmptyAndUndefinedFields(obj[key]);
        if (Object.keys(obj[key]).length === 0) {
          delete obj[key];
          continue;
        }
      }
    }
  }
}
function filterEventActionsStateDelta(actions) {
  if (!(actions == null ? void 0 : actions.stateDelta)) {
    return;
  }
  delete actions.stateDelta["_adk_recordings_config"];
  delete actions.stateDelta["_adk_replay_config"];
}
function filterPartFields(part) {
  delete part.thoughtSignature;
  delete part.functionCall;
  delete part.functionResponse;
}
function filterEventFields(event) {
  var _a;
  delete event.id;
  delete event.timestamp;
  delete event.invocationId;
  delete event.longRunningToolIds;
  filterEventActionsStateDelta(event.actions);
  if (event.content) {
    (_a = event.content.parts) == null ? void 0 : _a.forEach(filterPartFields);
  }
}
export {
  TestRunner
};
