// @google/adk-devtools/dist/cjs/cli/cli.js
#! /usr/bin/env node

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var cli_exports = {};
__export(cli_exports, {
  createProgram: () => createProgram
});
module.exports = __toCommonJS(cli_exports);
var import_adk = require("@google/adk");
var import_commander = require("commander");
var import_dotenv = __toESM(require("dotenv"), 1);
var path = __toESM(require("path"), 1);
var import_run_integration_tests = require("../integration/run_integration_tests.js");
var import_adk_api_server = require("../server/adk_api_server.js");
var import_agent_loader = require("../utils/agent_loader.js");
var import_file_utils = require("../utils/file_utils.js");
var import_logger = require("../utils/logger.js");
var import_version = require("../version.js");
var import_cli_create = require("./cli_create.js");
var import_cli_deploy = require("./cli_deploy.js");
var import_cli_run = require("./cli_run.js");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import_dotenv.default.config({ quiet: true });
const LOG_LEVEL_MAP = {
  "debug": import_adk.LogLevel.DEBUG,
  "info": import_adk.LogLevel.INFO,
  "warn": import_adk.LogLevel.WARN,
  "error": import_adk.LogLevel.ERROR
};
function getLogLevelFromOptions(options) {
  if (options.verbose) {
    return import_adk.LogLevel.DEBUG;
  }
  if (typeof options.log_level === "string") {
    return LOG_LEVEL_MAP[options.log_level.toLowerCase()] || import_adk.LogLevel.INFO;
  }
  return import_adk.LogLevel.INFO;
}
function getAbsolutePath(p) {
  return path.isAbsolute(p) ? p : path.join(process.cwd(), p);
}
function getSessionServiceFromOptions(options) {
  return (0, import_adk.getSessionServiceFromUri)(
    options["session_service_uri"] || process.env.DATABASE_URL || "memory://"
  );
}
function getArtifactServiceFromOptions(options) {
  return (0, import_adk.getArtifactServiceFromUri)(
    options["artifact_service_uri"] || "memory://"
  );
}
function getAgentFileOptions(options) {
  return {
    compile: getBoolean(options["compile"]),
    bundle: getBoolean(options["bundle"]),
    moduleType: options["file_type"]
  };
}
function getBoolean(option) {
  if (typeof option === "boolean") {
    return option;
  }
  if (typeof option === "string") {
    return option === "true" || option === "1";
  }
  return false;
}
const AGENT_DIR_ARGUMENT = new import_commander.Argument(
  "[agents_dir]",
  "Agent file or directory of agents to serve. For directory the internal structure should be agents_dir/{agentName}.js or agents_dir/{agentName}/agent.js. Agent file should has export of the rootAgent as instance of BaseAgent (e.g LlmAgent)"
).default(process.cwd());
const HOST_OPTION = new import_commander.Option(
  "-h, --host <string>",
  "Optional. The binding host of the server"
).default("localhost");
const PORT_OPTION = new import_commander.Option(
  "-p, --port <number>",
  "Optional. The port of the server"
).default("8000");
const ORIGINS_OPTION = new import_commander.Option(
  "--allow_origins <string>",
  "Optional. The allow origins of the server"
).default("");
const VERBOSE_OPTION = new import_commander.Option(
  "-v, --verbose [boolean]",
  "Optional. The verbose level of the server"
).default(false);
const LOG_LEVEL_OPTION = new import_commander.Option(
  "--log_level <string>",
  "Optional. The log level of the server"
).default("info");
const SESSION_SERVICE_URI_OPTION = new import_commander.Option(
  "--session_service_uri <string>",
  "Optional. The URI of the session service. Supported URIs: memory:// for in-memory session service."
);
const ARTIFACT_SERVICE_URI_OPTION = new import_commander.Option(
  "--artifact_service_uri <string>",
  "Optional. The URI of the artifact service. Supported URIs: gs://<bucket name> for GCS artifact service."
);
const OTEL_TO_CLOUD_OPTION = new import_commander.Option(
  "--otel_to_cloud [boolean]",
  "Optional. Whether to send otel traces to cloud."
).default(false);
const COMPILE_AGENT_FILE = new import_commander.Option(
  "--compile [boolean]",
  "Optional. Whether to compile ts agent file to js before execution"
).default(true);
const BUNDLE_AGENT_FILE = new import_commander.Option(
  "--bundle [boolean]",
  "Optional. Whether to compile ts agent file to js before execution"
).default(true);
const A2A_OPTION = new import_commander.Option(
  "--a2a [boolean]",
  "Optional. Whether to enable A2A for web/api server. Default: false"
).default(false);
const AGENT_FILE_MODULE_TYPE = new import_commander.Option("--file_type <string>", "Optional. ");
AGENT_FILE_MODULE_TYPE.argChoices = [import_agent_loader.FileModuleType.CJS, import_agent_loader.FileModuleType.ESM];
function createProgram() {
  const logger = new import_logger.AdkLogger({
    label: "ADK CLI",
    colorize: { all: true }
  });
  const program = new import_commander.Command("ADK CLI");
  program.addOption(new import_commander.Option("-v, --version", "Get ADK CLI version")).action(() => {
    console.log(import_version.version);
  });
  program.command("web").description("Start ADK web server").addArgument(AGENT_DIR_ARGUMENT).addOption(HOST_OPTION).addOption(PORT_OPTION).addOption(ORIGINS_OPTION).addOption(VERBOSE_OPTION).addOption(LOG_LEVEL_OPTION).addOption(SESSION_SERVICE_URI_OPTION).addOption(ARTIFACT_SERVICE_URI_OPTION).addOption(OTEL_TO_CLOUD_OPTION).addOption(COMPILE_AGENT_FILE).addOption(BUNDLE_AGENT_FILE).addOption(AGENT_FILE_MODULE_TYPE).addOption(A2A_OPTION).action(async (agentsDir, options) => {
    const logLevel = getLogLevelFromOptions(options);
    (0, import_adk.setLogLevel)(logLevel);
    try {
      const server = new import_adk_api_server.AdkApiServer({
        logLevel,
        agentsDir: getAbsolutePath(agentsDir),
        host: options["host"],
        port: parseInt(options["port"], 10),
        serveDebugUI: true,
        allowOrigins: options["allow_origins"],
        sessionService: getSessionServiceFromOptions(options),
        artifactService: getArtifactServiceFromOptions(options),
        otelToCloud: options["otel_to_cloud"] ? true : false,
        agentFileLoadOptions: getAgentFileOptions(options),
        a2a: getBoolean(options["a2a"])
      });
      await server.start();
    } catch (error) {
      logger.error("Error starting web server:", error.message);
    }
  });
  program.command("api_server").description("Start ADK API server").addArgument(AGENT_DIR_ARGUMENT).addOption(HOST_OPTION).addOption(PORT_OPTION).addOption(ORIGINS_OPTION).addOption(VERBOSE_OPTION).addOption(LOG_LEVEL_OPTION).addOption(SESSION_SERVICE_URI_OPTION).addOption(ARTIFACT_SERVICE_URI_OPTION).addOption(OTEL_TO_CLOUD_OPTION).addOption(COMPILE_AGENT_FILE).addOption(BUNDLE_AGENT_FILE).addOption(AGENT_FILE_MODULE_TYPE).addOption(A2A_OPTION).action(async (agentsDir, options) => {
    const logLevel = getLogLevelFromOptions(options);
    (0, import_adk.setLogLevel)(logLevel);
    try {
      const server = new import_adk_api_server.AdkApiServer({
        logLevel,
        agentsDir: getAbsolutePath(agentsDir),
        host: options["host"],
        port: parseInt(options["port"], 10),
        serveDebugUI: false,
        allowOrigins: options["allow_origins"],
        sessionService: getSessionServiceFromOptions(options),
        artifactService: getArtifactServiceFromOptions(options),
        otelToCloud: options["otel_to_cloud"] ? true : false,
        agentFileLoadOptions: getAgentFileOptions(options),
        a2a: getBoolean(options["a2a"])
      });
      await server.start();
    } catch (error) {
      logger.error("Error starting API server:", error.message);
    }
  });
  program.command("create").description("Creates a new agent").argument("[agent]", "Name to give the new agent", "adk_agent").option("-y, --yes", "Optional. Skip confirmation prompts.").option("--model <string>", "Optional. THe model used for the root_agent").option(
    "--api_key <string>",
    "Optional. The API Key needed to access the model, e.g. Google AI API Key."
  ).option(
    "--project <string>",
    "Optional. The Google Cloud Project for using VertexAI as backend."
  ).option(
    "--region <string>",
    "Optional. The Google Cloud Region for using VertexAI as backend."
  ).option(
    "--language <string>",
    "Optional. Either ts or js as the language to output."
  ).action(async (agentName, options) => {
    try {
      await (0, import_cli_create.createAgent)({
        agentName,
        forceYes: !!options["yes"],
        model: options["model"],
        apiKey: options["api_key"],
        project: options["project"],
        region: options["region"],
        language: options["language"]
      });
    } catch (error) {
      logger.error("Error creating agent:", error.message);
    }
  });
  program.command("run").description("Runs agent").argument("<agent>", "Agent file path (.js or .ts)").option(
    "--save_session [boolean]",
    "Optional. Whether to save the session to a json file on exit.",
    false
  ).option(
    "--session_id <string>",
    "Optional. The session ID to save the session to on exit when --save_session is set to true. User will be prompted to enter a session ID if not set."
  ).option(
    "--replay <string>",
    "The json file that contains the initial state of the session and user queries. A new session will be created using this state. And user queries are run against the newly created session. Users cannot continue to interact with the agent."
  ).option(
    "--resume <string>",
    "The json file that contains a previously saved session (by --save_session option). The previous session will be re-displayed. And user can continue to interact with the agent."
  ).addOption(VERBOSE_OPTION).addOption(LOG_LEVEL_OPTION).addOption(SESSION_SERVICE_URI_OPTION).addOption(ARTIFACT_SERVICE_URI_OPTION).addOption(OTEL_TO_CLOUD_OPTION).addOption(COMPILE_AGENT_FILE).addOption(BUNDLE_AGENT_FILE).addOption(AGENT_FILE_MODULE_TYPE).action(async (agentPath, options) => {
    (0, import_adk.setLogLevel)(getLogLevelFromOptions(options));
    try {
      await (0, import_cli_run.runAgent)({
        agentPath,
        inputFile: options["replay"],
        savedSessionFile: options["resume"],
        saveSession: getBoolean(options["save_session"]),
        sessionId: options["session_id"],
        sessionService: getSessionServiceFromOptions(options),
        artifactService: getArtifactServiceFromOptions(options),
        otelToCloud: options["otel_to_cloud"] ? true : false,
        agentFileLoadOptions: getAgentFileOptions(options)
      });
    } catch (error) {
      logger.error("Error running agent:", error.message);
    }
  });
  const DEPLOY_COMMAND = program.command("deploy").description("Deploy agent").allowUnknownOption().allowExcessArguments();
  DEPLOY_COMMAND.command("cloud_run").addArgument(AGENT_DIR_ARGUMENT).allowUnknownOption().allowExcessArguments().addOption(PORT_OPTION).option(
    "--project [string]",
    "Optional. Google Cloud project to deploy the agent. If not set, default project from gcloud config is used"
  ).option(
    "--region [string]",
    "Optional. Google Cloud region to deploy the agent. If not set, default run/region from gcloud config is used"
  ).option(
    "--service_name [string]",
    'Optional. The service name to use in Cloud Run. Default: "adk-default-service-name"',
    "adk-default-service-name"
  ).option(
    "--temp_folder [string]",
    "Optional. Temp folder for the generated Cloud Run source files (default: a timestamped folder in the system temp directory).",
    (0, import_file_utils.getTempDir)("cloud_run_deploy_src")
  ).option(
    "--adk_version [string]",
    "Optional. ADK version to use in the Cloud Run service. If not set, default to the latest version available on npm",
    "latest"
  ).option(
    "--with_ui [boolean]",
    "Optional. Deploy ADK Web UI if set. (default: deploy ADK API server only)",
    false
  ).addOption(ORIGINS_OPTION).addOption(VERBOSE_OPTION).addOption(LOG_LEVEL_OPTION).addOption(SESSION_SERVICE_URI_OPTION).addOption(ARTIFACT_SERVICE_URI_OPTION).addOption(COMPILE_AGENT_FILE).addOption(BUNDLE_AGENT_FILE).addOption(AGENT_FILE_MODULE_TYPE).addOption(A2A_OPTION).action(async (agentPath, options) => {
    const extraGcloudArgs = [];
    for (const arg of process.argv.slice(5)) {
      let argName = arg.replace(/^-+/, "");
      if (argName.includes("=")) {
        argName = argName.split("=")[0];
      }
      if (argName in options) {
        continue;
      }
      extraGcloudArgs.push(arg);
    }
    try {
      await (0, import_cli_deploy.deployToCloudRun)({
        agentPath: getAbsolutePath(agentPath),
        project: options["project"],
        region: options["region"],
        serviceName: options["service_name"],
        tempFolder: options["temp_folder"],
        port: parseInt(options["port"], 10),
        withUi: getBoolean(options["with_ui"]),
        logLevel: options["log_level"],
        adkVersion: options["adk_version"],
        allowOrigins: options["allow_origins"],
        sessionServiceUri: options["session_service_uri"],
        artifactServiceUri: options["artifact_service_uri"],
        agentFileLoadOptions: getAgentFileOptions(options),
        a2a: getBoolean(options["a2a"]),
        extraGcloudArgs
      });
    } catch (error) {
      logger.error("Error deploying agent:", error.message);
    }
  });
  const CONFORMANCE_COMMAND = program.command("integration").description("Run ADK integration and conformance tests");
  CONFORMANCE_COMMAND.command("conformance").description("Run ADK conformance tests").addOption(VERBOSE_OPTION).addOption(LOG_LEVEL_OPTION).option(
    "--agents_dir [dir]",
    "Directory of conformance test agent definitions. Recursively searched for .yaml files with agent definitions.",
    process.cwd()
  ).option(
    "--tests_dir [dir]",
    "Directory of conformance test definitions. Recursively searched for .yaml files with test definitions.",
    process.cwd()
  ).option("--force", "Force run skipped tests.").action(async (options) => {
    (0, import_run_integration_tests.runIntegrationTests)({
      agentsDir: options["agents_dir"],
      testsDir: options["tests_dir"],
      forceRunAll: getBoolean(options["force"])
    });
  });
  return program;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createProgram
});
