// @google/adk-devtools/dist/cjs/cli/cli_run.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function") __typeError("Object expected");
    var dispose, inner;
    if (async) dispose = value[__knownSymbol("asyncDispose")];
    if (dispose === void 0) {
      dispose = value[__knownSymbol("dispose")];
      if (async) inner = dispose;
    }
    if (typeof dispose !== "function") __typeError("Object not disposable");
    if (inner) dispose = function() {
      try {
        inner.call(this);
      } catch (e) {
        return Promise.reject(e);
      }
    };
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  };
  var fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e);
  var next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0]) return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError) throw error;
  };
  return next();
};
var cli_run_exports = {};
__export(cli_run_exports, {
  runAgent: () => runAgent
});
module.exports = __toCommonJS(cli_run_exports);
var import_adk = require("@google/adk");
var path = __toESM(require("node:path"), 1);
var readline = __toESM(require("node:readline"), 1);
var import_agent_loader = require("../utils/agent_loader.js");
var import_file_utils = require("../utils/file_utils.js");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const dirname = process.cwd();
async function getUserInput(prompt) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  return new Promise((resolve) => {
    rl.question(prompt, (answer) => {
      rl.close();
      resolve(answer);
    });
  });
}
async function runFromInputFile(options) {
  const fileContent = await (0, import_file_utils.loadFileData)(
    path.join(dirname, options.filePath)
  );
  if (!fileContent) {
    return;
  }
  fileContent.state["_time"] = (/* @__PURE__ */ new Date()).toISOString();
  const session = await options.sessionService.createSession({
    appName: options.appName,
    userId: options.userId,
    state: fileContent.state
  });
  const runner = new import_adk.Runner(options);
  for (const query of fileContent.queries) {
    console.log(`[user]: ${query}`);
    const runOptions = {
      userId: session.userId,
      sessionId: session.id,
      newMessage: { role: "user", parts: [{ text: query }] }
    };
    for await (const event of runner.runAsync(runOptions)) {
      if (event.content && event.content.parts) {
        const text = event.content.parts.map((part) => part.text || "").join("");
        if (text) {
          console.log(`[${event.author}]: ${text}`);
        }
      }
    }
  }
  return session;
}
async function runInteractively(options) {
  const runner = new import_adk.Runner({
    appName: options.rootAgent.name,
    agent: options.rootAgent,
    artifactService: options.artifactService,
    sessionService: options.sessionService,
    memoryService: options.memoryService
  });
  while (true) {
    const query = await getUserInput("[user]: ");
    if (!query || !query.trim()) {
      continue;
    }
    if (query === "exit") {
      break;
    }
    for await (const event of runner.runAsync({
      userId: options.session.userId,
      sessionId: options.session.id,
      newMessage: { role: "user", parts: [{ text: query }] }
    })) {
      if (event.content && event.content.parts) {
        const text = event.content.parts.map((part) => part.text || "").join("");
        if (text) {
          console.log(`[${event.author}]: ${text}`);
        }
      }
    }
  }
}
async function runAgent(options) {
  var _a;
  try {
    var _stack = [];
    try {
      const userId = "test_user";
      const artifactService = options.artifactService || new import_adk.InMemoryArtifactService();
      const sessionService = options.sessionService || new import_adk.InMemorySessionService();
      const memoryService = options.memoryService || new import_adk.InMemoryMemoryService();
      const agentFile = __using(_stack, new import_agent_loader.AgentFile(
        path.join(dirname, options.agentPath),
        options.agentFileLoadOptions
      ), true);
      const rootAgent = await agentFile.load();
      let session = await sessionService.createSession({
        appName: rootAgent.name,
        userId
      });
      if (options.inputFile) {
        session = await runFromInputFile({
          appName: rootAgent.name,
          userId,
          agent: rootAgent,
          artifactService,
          sessionService,
          memoryService,
          filePath: options.inputFile
        }) || session;
      } else if (options.savedSessionFile) {
        const loadedSession = await (0, import_file_utils.loadFileData)(
          options.savedSessionFile
        );
        if (loadedSession) {
          for (const event of loadedSession.events) {
            await sessionService.appendEvent({ session, event });
            const content = event.content;
            if (content && ((_a = content.parts) == null ? void 0 : _a.length)) {
              const text = content.parts.map((part) => part.text || "").join("");
              if (text) {
                console.log(`[${event.author}]: ${text}`);
              }
            }
          }
        }
        await runInteractively({
          rootAgent,
          artifactService,
          sessionService,
          memoryService,
          session
        });
      } else {
        console.log(`Running agent ${rootAgent.name}, type exit to exit.`);
        await runInteractively({
          rootAgent,
          artifactService,
          sessionService,
          memoryService,
          session
        });
      }
      if (options.saveSession) {
        const sessionId = options.sessionId || await getUserInput("Session ID to save: ");
        const sessionPath = path.join(
          options.agentPath,
          `${sessionId}.session.json`
        );
        const sessionToStore = await sessionService.getSession({
          appName: session.appName,
          userId: session.userId,
          sessionId: session.id
        });
        await (0, import_file_utils.saveToFile)(path.join(dirname, sessionPath), sessionToStore);
        console.log("Session saved to", sessionPath);
      }
    } catch (_) {
      var _error = _, _hasError = true;
    } finally {
      var _promise = __callDispose(_stack, _error, _hasError);
      _promise && await _promise;
    }
  } catch (e) {
    console.log(e);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  runAgent
});
