// @google/adk-devtools/dist/cjs/cli/cli_create.js

const importMetaUrl = /* @__PURE__ */ require("node:url").pathToFileURL(__filename).toString();
const importMetaFilename = /* @__PURE__ */ __filename;
const importMetaDirname = /* @__PURE__ */ __dirname;
const importMetaResolve = /* @__PURE__ */ require.resolve;
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var cli_create_exports = {};
__export(cli_create_exports, {
  createAgent: () => createAgent
});
module.exports = __toCommonJS(cli_create_exports);
var import_prompts = require("@clack/prompts");
var import_node_child_process = require("node:child_process");
var path = __toESM(require("node:path"), 1);
var import_node_util = require("node:util");
var import_file_utils = require("../utils/file_utils.js");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const execPromise = (0, import_node_util.promisify)(import_node_child_process.exec);
const dirname = process.cwd();
const TS_CONFIG = `{
  "compilerOptions": {
    "target": "esnext",
    "module": "nodenext",
    "rootDir": "./",
    "outDir": "dist",
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "declaration": true,
    "declarationMap": true,
    "esModuleInterop": true,
    "exactOptionalPropertyTypes": true,
    "noEmitOnError": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitReturns": true,
    "noUncheckedIndexedAccess": true,
    "pretty": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true
  }
}
`.trim();
const PACKAGE_JSON = (agentName, language) => `
{
  "name": "${agentName}",
  "version": "1.0.0",
  "description": "",
  "main": "agent.${language}",
  "scripts": {
    "web": "npx @google/adk-devtools web",
    "cli": "npx @google/adk-devtools run agent.${language}"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}
`.trim();
const AGENT_TEMPLATE = (model) => `
import {FunctionTool, LlmAgent} from '@google/adk';
import {z} from 'zod';
import dotenv from 'dotenv';

dotenv.config();

/* Mock tool implementation */
const getCurrentTime = new FunctionTool({
  name: 'get_current_time',
  description: 'Returns the current time in a specified city.',
  parameters: z.object({
    city: z.string().describe("The name of the city for which to retrieve the current time."),
  }),
  execute: ({city}) => {
    return {status: 'success', report: \`The current time in \${city} is 10:30 AM\`};
  },
});

export const rootAgent = new LlmAgent({
  name: 'hello_time_agent',
  model: '${model}',
  description: 'Tells the current time in a specified city.',
  instruction: \`You are a helpful assistant that tells the current time in a city.
                Use the 'getCurrentTime' tool for this purpose.\`,
  tools: [getCurrentTime],
});
`.trim();
async function getGcpProject() {
  if (process.env.GOOGLE_CLOUD_PROJECT) {
    return process.env.GOOGLE_CLOUD_PROJECT;
  }
  try {
    const stdout = (0, import_node_child_process.execSync)("gcloud config get-value project", {
      encoding: "utf-8",
      stdio: "pipe"
    });
    return stdout.trim();
  } catch (_e) {
    return "";
  }
}
async function getGcpRegion() {
  if (process.env.GOOGLE_CLOUD_LOCATION) {
    return process.env.GOOGLE_CLOUD_LOCATION;
  }
  try {
    const stdout = (0, import_node_child_process.execSync)("gcloud config get-value compute/region", {
      encoding: "utf-8",
      stdio: "pipe"
    });
    return stdout.trim();
  } catch (_e) {
    return "";
  }
}
async function generateAgentFolder(agentDir, forceYes) {
  if (!await (0, import_file_utils.isFolderExists)(agentDir)) {
    return await (0, import_file_utils.createFolder)(agentDir);
  }
  const overwriteFolderResponse = forceYes ? true : await (0, import_prompts.select)({
    message: `Folder ${agentDir} already exists. Would you like to overwrite existing folder?`,
    options: [
      { label: "Yes", value: true },
      { label: "No", value: false }
    ]
  });
  if ((0, import_prompts.isCancel)(overwriteFolderResponse)) {
    process.exit(0);
  }
  if (!overwriteFolderResponse) {
    console.error(`Agent directory ${agentDir} already exists.`);
    process.exit(0);
  }
  await (0, import_file_utils.removeFolder)(agentDir);
  await (0, import_file_utils.createFolder)(agentDir);
}
function generateEnvFile(options) {
  const lines = [];
  if (options.apiKey) {
    lines.push(`GOOGLE_API_KEY=${options.apiKey}`);
    lines.push(`GOOGLE_GENAI_USE_VERTEXAI=0`);
  }
  if (options.project) {
    lines.push(`GOOGLE_CLOUD_PROJECT=${options.project}`);
  }
  if (options.region) {
    lines.push(`GOOGLE_CLOUD_LOCATION=${options.region}`);
  }
  if (options.region && options.project) {
    lines.push(`GOOGLE_GENAI_USE_VERTEXAI=1`);
  }
  return lines.join("\n");
}
async function generateFiles(options) {
  const agentDir = path.join(dirname, options.agentName);
  await (0, import_file_utils.saveToFile)(
    path.join(agentDir, `agent.${options.language}`),
    AGENT_TEMPLATE(options.model || "gemini-2.5-flash")
  );
  await (0, import_file_utils.saveToFile)(path.join(agentDir, ".env"), generateEnvFile(options));
  await (0, import_file_utils.saveToFile)(
    path.join(agentDir, "package.json"),
    PACKAGE_JSON(options.agentName, options.language)
  );
  if (options.language === "ts") {
    await (0, import_file_utils.saveToFile)(path.join(agentDir, "tsconfig.json"), TS_CONFIG);
  }
}
async function createAgent(options) {
  const agentDir = path.join(dirname, options.agentName);
  await generateAgentFolder(agentDir, options.forceYes);
  if (!options.model) {
    const model = options.forceYes ? "gemini-2.5-flash" : await (0, import_prompts.select)({
      message: "Choose a model for the root agent",
      options: [
        { label: "gemini-2.5-flash", value: "gemini-2.5-flash" },
        { label: "gemini-2.5-pro", value: "gemini-2.5-pro" },
        {
          label: "gemini-3-flash-preview",
          value: "gemini-3-flash-preview"
        },
        { label: "gemini-3-pro-preview", value: "gemini-3-pro-preview" }
      ]
    });
    if ((0, import_prompts.isCancel)(model)) {
      process.exit(0);
    }
    options.model = model;
  }
  if (options.language !== "js" && options.language !== "ts") {
    const language = options.forceYes ? "ts" : await (0, import_prompts.select)({
      message: "Choose a language for the agent",
      options: [
        { label: "TypeScript", value: "ts" },
        { label: "JavaScript", value: "js" }
      ]
    });
    if ((0, import_prompts.isCancel)(language)) {
      process.exit(0);
    }
    options.language = language;
  }
  if (!options.apiKey && !options.project) {
    const backend = options.forceYes ? "googleai" : await (0, import_prompts.select)({
      message: "Choose a backend",
      options: [
        { label: "Google AI", value: "googleai" },
        { label: "Vertex AI", value: "vertex" }
      ]
    });
    if ((0, import_prompts.isCancel)(backend)) {
      process.exit(0);
    }
    if (backend === "vertex") {
      const defaultProject = await getGcpProject();
      const defaultRegion = await getGcpRegion();
      const projectResponse = options.forceYes ? defaultProject : (await (0, import_prompts.text)({
        message: "Enter the Google Cloud Project ID",
        initialValue: defaultProject
      })).toString();
      if ((0, import_prompts.isCancel)(projectResponse)) {
        process.exit(0);
      }
      options.project = projectResponse;
      const regionResponse = options.forceYes ? defaultRegion : await (0, import_prompts.text)({
        message: "Enter the Google Cloud Region",
        initialValue: defaultRegion
      });
      if ((0, import_prompts.isCancel)(regionResponse)) {
        process.exit(0);
      }
      options.region = regionResponse;
    } else {
      const apiKeyResponse = options.forceYes ? "" : await (0, import_prompts.text)({
        message: "Enter the Google API Key"
      });
      if ((0, import_prompts.isCancel)(apiKeyResponse)) {
        process.exit(0);
      }
      options.apiKey = apiKeyResponse;
    }
  }
  await generateFiles(options);
  if (options.language === "ts") {
    await execPromise(`npm install typescript --save-dev`, { cwd: agentDir });
  }
  await execPromise(`npm install @google/adk @google/adk-devtools zod dotenv`, {
    cwd: agentDir
  });
  const files = await (0, import_file_utils.listFiles)(agentDir);
  console.log(`
Created the following files in ${agentDir}:`);
  files.forEach((file) => {
    console.log(`  - ${file}`);
  });
  console.log(
    `Run 'cd ${options.agentName} && npm run web' to start the agent in a web interface`
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createAgent
});
