// @google/adk/dist/web/utils/gemini_schema_util.js
import {createRequire as topLevelCreateRequire} from 'module';
const require = topLevelCreateRequire(import.meta.url);
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { Type } from "@google/genai";
import { z } from "zod";
const MCPToolSchemaObject = z.object({
  type: z.literal("object"),
  properties: z.record(z.string(), z.unknown()).optional(),
  required: z.string().array().optional()
});
function toGeminiType(mcpType) {
  if (!mcpType) return Type.TYPE_UNSPECIFIED;
  switch (mcpType.toLowerCase()) {
    case "text":
    case "string":
      return Type.STRING;
    case "number":
      return Type.NUMBER;
    case "boolean":
      return Type.BOOLEAN;
    case "integer":
      return Type.INTEGER;
    case "array":
      return Type.ARRAY;
    case "object":
      return Type.OBJECT;
    case "null":
      return Type.NULL;
    default:
      return Type.TYPE_UNSPECIFIED;
  }
}
const getTypeFromArrayItem = (mcpType) => {
  var _a, _b;
  if (typeof mcpType === "string") {
    return mcpType.toLowerCase();
  }
  return (_b = (_a = mcpType == null ? void 0 : mcpType.type) == null ? void 0 : _a.toLowerCase) == null ? void 0 : _b.call(_a);
};
function toGeminiSchema(mcpSchema) {
  if (!mcpSchema) {
    return void 0;
  }
  function recursiveConvert(mcp) {
    var _a;
    const sourceType = (_a = mcp.anyOf) != null ? _a : mcp.type;
    let isNullable = false;
    let nonNullTypes;
    if (Array.isArray(sourceType)) {
      nonNullTypes = sourceType.filter(
        (t) => getTypeFromArrayItem(t) !== "null"
      );
      isNullable = sourceType.some(
        (t) => getTypeFromArrayItem(t) === "null"
      );
      if (nonNullTypes.length === 1) {
        const nonNullType = nonNullTypes[0];
        if (typeof nonNullType === "object") {
          mcp = nonNullType;
        } else {
          const _b = mcp, { type: _removed, anyOf: _removedAnyOf } = _b, rest = __objRest(_b, ["type", "anyOf"]);
          mcp = __spreadProps(__spreadValues({}, rest), { type: nonNullType });
        }
      } else if (nonNullTypes.length === 0 && isNullable) {
        const _c = mcp, { type: _removed, anyOf: _removedAnyOf } = _c, rest = __objRest(_c, ["type", "anyOf"]);
        mcp = __spreadProps(__spreadValues({}, rest), { type: "null" });
      } else if (typeof mcp.anyOf === "undefined") {
        const anyOfItems = mcp.type.map((t) => ({ type: t }));
        const _d = mcp, { type: _removed } = _d, rest = __objRest(_d, ["type"]);
        mcp = __spreadProps(__spreadValues({}, rest), { anyOf: anyOfItems });
      }
    }
    if (!mcp.type) {
      if (mcp.properties || mcp.$ref) {
        mcp.type = "object";
      } else if (mcp.items) {
        mcp.type = "array";
      } else if (isNullable) {
        mcp.type = "null";
      }
    }
    const geminiType = toGeminiType(mcp.type);
    const geminiSchema = {};
    if (mcp.anyOf) {
      geminiSchema.anyOf = mcp.anyOf.map(
        (item) => recursiveConvert(item)
      );
    } else {
      geminiSchema.type = geminiType;
    }
    if (mcp.description) {
      geminiSchema.description = mcp.description;
    }
    if (isNullable && mcp.type !== "null") {
      geminiSchema.nullable = true;
    }
    if (geminiType === Type.OBJECT) {
      geminiSchema.properties = {};
      if (mcp.properties) {
        for (const name in mcp.properties) {
          geminiSchema.properties[name] = recursiveConvert(
            mcp.properties[name]
          );
        }
      }
      if (mcp.required) {
        geminiSchema.required = mcp.required;
      }
    } else if (geminiType === Type.ARRAY) {
      if (mcp.items) {
        geminiSchema.items = recursiveConvert(mcp.items);
      }
    }
    return geminiSchema;
  }
  return recursiveConvert(mcpSchema);
}
export {
  toGeminiSchema
};
